<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Lecture 01: Introduction and Foundations – Statistical Genetics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../course_logo.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-8991b0faf62225856095ff2fe360da1f.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN" && texText && texText.data) {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-sidebar docked quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Lecture 01: Introduction and Foundations</li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
      <a href="../index.html" class="sidebar-logo-link">
      <img src="../course_logo.png" alt="" class="sidebar-logo light-content py-0 d-lg-inline d-none">
      <img src="../course_logo.png" alt="" class="sidebar-logo dark-content py-0 d-lg-inline d-none">
      </a>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link"><i class="bi bi-calendar" role="img">
</i> 
 <span class="menu-text">Schedule</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../syllabus.html" class="sidebar-item-text sidebar-link"><i class="bi bi-book" role="img">
</i> 
 <span class="menu-text">Syllabus</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../faq.html" class="sidebar-item-text sidebar-link"><i class="bi bi-question" role="img">
</i> 
 <span class="menu-text">FAQ</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../links.html" class="sidebar-item-text sidebar-link"><i class="bi bi-link" role="img">
</i> 
 <span class="menu-text">Useful Links</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://github.com/gwcbi/StatGen" class="sidebar-item-text sidebar-link" target="_blank"><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text">GitHub</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://www.zotero.org/groups/6113424/pubh-8878/library" class="sidebar-item-text sidebar-link" target="_blank"><i class="bi bi-book" role="img">
</i> 
 <span class="menu-text">Zotero Library</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#about-the-instructors" id="toc-about-the-instructors" class="nav-link active" data-scroll-target="#about-the-instructors">About the Instructors</a></li>
  <li><a href="#introductions" id="toc-introductions" class="nav-link" data-scroll-target="#introductions">Introductions</a></li>
  <li><a href="#course-site" id="toc-course-site" class="nav-link" data-scroll-target="#course-site">Course Site</a></li>
  <li><a href="#zotero" id="toc-zotero" class="nav-link" data-scroll-target="#zotero">Zotero</a></li>
  <li><a href="#course-logistics-and-expectations" id="toc-course-logistics-and-expectations" class="nav-link" data-scroll-target="#course-logistics-and-expectations">Course Logistics and Expectations</a>
  <ul class="collapse">
  <li><a href="#grading-breakdown" id="toc-grading-breakdown" class="nav-link" data-scroll-target="#grading-breakdown">Grading Breakdown</a></li>
  </ul></li>
  <li><a href="#problem-sets-60-of-grade" id="toc-problem-sets-60-of-grade" class="nav-link" data-scroll-target="#problem-sets-60-of-grade">Problem Sets (60% of grade)</a></li>
  <li><a href="#class-participation-20-of-grade" id="toc-class-participation-20-of-grade" class="nav-link" data-scroll-target="#class-participation-20-of-grade">Class Participation (20% of grade)</a></li>
  <li><a href="#research-project-20-of-grade" id="toc-research-project-20-of-grade" class="nav-link" data-scroll-target="#research-project-20-of-grade">Research Project (20% of grade)</a></li>
  <li><a href="#course-structure-overview" id="toc-course-structure-overview" class="nav-link" data-scroll-target="#course-structure-overview">Course Structure Overview</a></li>
  <li><a href="#technology-setup-requirements" id="toc-technology-setup-requirements" class="nav-link" data-scroll-target="#technology-setup-requirements">Technology Setup Requirements</a></li>
  <li><a href="#course-textbook" id="toc-course-textbook" class="nav-link" data-scroll-target="#course-textbook">Course Textbook</a></li>
  <li><a href="#what-is-statistical-genetics" id="toc-what-is-statistical-genetics" class="nav-link" data-scroll-target="#what-is-statistical-genetics">What is statistical genetics?</a></li>
  <li><a href="#a-statistical-geneticist-may-want-to-know" id="toc-a-statistical-geneticist-may-want-to-know" class="nav-link" data-scroll-target="#a-statistical-geneticist-may-want-to-know">A statistical geneticist may want to know</a></li>
  <li><a href="#a-statistical-geneticist-may-want-to-know-1" id="toc-a-statistical-geneticist-may-want-to-know-1" class="nav-link" data-scroll-target="#a-statistical-geneticist-may-want-to-know-1">A statistical geneticist may want to know</a></li>
  <li><a href="#what-does-the-data-look-like" id="toc-what-does-the-data-look-like" class="nav-link" data-scroll-target="#what-does-the-data-look-like">What does the data look like?</a></li>
  <li><a href="#what-does-the-data-look-like-1" id="toc-what-does-the-data-look-like-1" class="nav-link" data-scroll-target="#what-does-the-data-look-like-1">What does the data look like?</a></li>
  <li><a href="#core-abstractions-for-quantitative-biology" id="toc-core-abstractions-for-quantitative-biology" class="nav-link" data-scroll-target="#core-abstractions-for-quantitative-biology">Core abstractions for quantitative biology</a></li>
  <li><a href="#core-abstractions-for-quantitative-biology-1" id="toc-core-abstractions-for-quantitative-biology-1" class="nav-link" data-scroll-target="#core-abstractions-for-quantitative-biology-1">Core abstractions for quantitative biology</a></li>
  <li><a href="#core-abstractions-for-quantitative-biology-2" id="toc-core-abstractions-for-quantitative-biology-2" class="nav-link" data-scroll-target="#core-abstractions-for-quantitative-biology-2">Core abstractions for quantitative biology</a></li>
  <li><a href="#core-abstractions-for-quantitative-biology-3" id="toc-core-abstractions-for-quantitative-biology-3" class="nav-link" data-scroll-target="#core-abstractions-for-quantitative-biology-3">Core abstractions for quantitative biology</a></li>
  <li><a href="#early-work-in-the-field" id="toc-early-work-in-the-field" class="nav-link" data-scroll-target="#early-work-in-the-field">Early work in the field</a></li>
  <li><a href="#genetics-statistics-and-eugenics-have-always-been-closely-linked" id="toc-genetics-statistics-and-eugenics-have-always-been-closely-linked" class="nav-link" data-scroll-target="#genetics-statistics-and-eugenics-have-always-been-closely-linked">Genetics, statistics, and eugenics, have always been closely linked</a></li>
  <li><a href="#the-ethics-of-genetics-and-genomics-iswas-of-great-importance" id="toc-the-ethics-of-genetics-and-genomics-iswas-of-great-importance" class="nav-link" data-scroll-target="#the-ethics-of-genetics-and-genomics-iswas-of-great-importance">The ethics of genetics and genomics is/(was) of great importance!</a></li>
  <li><a href="#some-vocabulary" id="toc-some-vocabulary" class="nav-link" data-scroll-target="#some-vocabulary">Some vocabulary</a></li>
  <li><a href="#some-vocabulary-1" id="toc-some-vocabulary-1" class="nav-link" data-scroll-target="#some-vocabulary-1">Some vocabulary</a></li>
  <li><a href="#genetic-variant-example" id="toc-genetic-variant-example" class="nav-link" data-scroll-target="#genetic-variant-example">Genetic Variant Example</a></li>
  <li><a href="#probability-refresher" id="toc-probability-refresher" class="nav-link" data-scroll-target="#probability-refresher">Probability Refresher</a>
  <ul class="collapse">
  <li><a href="#random-variables-and-distributions" id="toc-random-variables-and-distributions" class="nav-link" data-scroll-target="#random-variables-and-distributions">Random Variables and Distributions</a></li>
  </ul></li>
  <li><a href="#probability-distributions-vs-sampling-distributions" id="toc-probability-distributions-vs-sampling-distributions" class="nav-link" data-scroll-target="#probability-distributions-vs-sampling-distributions">Probability Distributions vs Sampling Distributions</a></li>
  <li><a href="#probability-refresher-1" id="toc-probability-refresher-1" class="nav-link" data-scroll-target="#probability-refresher-1">Probability Refresher</a>
  <ul class="collapse">
  <li><a href="#key-distributions-in-statistical-genetics" id="toc-key-distributions-in-statistical-genetics" class="nav-link" data-scroll-target="#key-distributions-in-statistical-genetics">Key Distributions in Statistical Genetics</a></li>
  </ul></li>
  <li><a href="#conditional-probability-and-independence" id="toc-conditional-probability-and-independence" class="nav-link" data-scroll-target="#conditional-probability-and-independence">Conditional Probability and Independence</a></li>
  <li><a href="#parameter-estimation" id="toc-parameter-estimation" class="nav-link" data-scroll-target="#parameter-estimation">Parameter Estimation</a></li>
  <li><a href="#genetic-models" id="toc-genetic-models" class="nav-link" data-scroll-target="#genetic-models">Genetic Models</a></li>
  <li><a href="#modeling" id="toc-modeling" class="nav-link" data-scroll-target="#modeling">Modeling</a></li>
  <li><a href="#likelihood-functions" id="toc-likelihood-functions" class="nav-link" data-scroll-target="#likelihood-functions">Likelihood Functions</a></li>
  <li><a href="#maximum-likelihood-estimator" id="toc-maximum-likelihood-estimator" class="nav-link" data-scroll-target="#maximum-likelihood-estimator">Maximum Likelihood Estimator</a></li>
  <li><a href="#modeling-but-make-it-bayesian" id="toc-modeling-but-make-it-bayesian" class="nav-link" data-scroll-target="#modeling-but-make-it-bayesian">Modeling, but make it bayesian</a></li>
  <li><a href="#modeling-but-make-it-bayesian-1" id="toc-modeling-but-make-it-bayesian-1" class="nav-link" data-scroll-target="#modeling-but-make-it-bayesian-1">Modeling, but make it bayesian</a></li>
  <li><a href="#mendels-laws" id="toc-mendels-laws" class="nav-link" data-scroll-target="#mendels-laws">Mendel’s Laws</a></li>
  <li><a href="#worked-example-segregation-31-ratio" id="toc-worked-example-segregation-31-ratio" class="nav-link" data-scroll-target="#worked-example-segregation-31-ratio">Worked Example: Segregation (3:1 Ratio)</a>
  <ul class="collapse">
  <li><a href="#cross-aa-aa" id="toc-cross-aa-aa" class="nav-link" data-scroll-target="#cross-aa-aa">Cross: Aa × Aa</a></li>
  </ul></li>
  <li><a href="#test-cross-example" id="toc-test-cross-example" class="nav-link" data-scroll-target="#test-cross-example">Test Cross Example</a></li>
  <li><a href="#test-cross-example-1" id="toc-test-cross-example-1" class="nav-link" data-scroll-target="#test-cross-example-1">Test Cross Example</a></li>
  <li><a href="#testing-via-simulation" id="toc-testing-via-simulation" class="nav-link" data-scroll-target="#testing-via-simulation">Testing via Simulation</a></li>
  <li><a href="#on-simulations" id="toc-on-simulations" class="nav-link" data-scroll-target="#on-simulations">On Simulations</a></li>
  <li><a href="#estimation-of-allele-frequencies" id="toc-estimation-of-allele-frequencies" class="nav-link" data-scroll-target="#estimation-of-allele-frequencies">Estimation of Allele Frequencies</a></li>
  <li><a href="#hardyweinberg-equilibrium-hwe-assumptions" id="toc-hardyweinberg-equilibrium-hwe-assumptions" class="nav-link" data-scroll-target="#hardyweinberg-equilibrium-hwe-assumptions">Hardy–Weinberg Equilibrium (HWE): Assumptions</a></li>
  <li><a href="#testing-for-hwe" id="toc-testing-for-hwe" class="nav-link" data-scroll-target="#testing-for-hwe">Testing for HWE</a></li>
  <li><a href="#hwe-worked-example" id="toc-hwe-worked-example" class="nav-link" data-scroll-target="#hwe-worked-example">HWE: Worked Example</a></li>
  <li><a href="#hwe-in-practice" id="toc-hwe-in-practice" class="nav-link" data-scroll-target="#hwe-in-practice">HWE in Practice</a></li>
  <li><a href="#sampling-distributions-intuition-via-simulation" id="toc-sampling-distributions-intuition-via-simulation" class="nav-link" data-scroll-target="#sampling-distributions-intuition-via-simulation">Sampling distributions: intuition via simulation</a></li>
  <li><a href="#sampling-distributions-intuition-via-simulation-1" id="toc-sampling-distributions-intuition-via-simulation-1" class="nav-link" data-scroll-target="#sampling-distributions-intuition-via-simulation-1">Sampling distributions: intuition via simulation</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="lecture-01.html"><i class="bi bi-file-slides"></i>RevealJS</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Lecture 01: Introduction and Foundations</h1>
<p class="subtitle lead">PUBH 8878, Statistical Genetics</p>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="about-the-instructors" class="level2">
<h2 class="anchored" data-anchor-id="about-the-instructors">About the Instructors</h2>
<div class="quarto-layout-panel" data-layout="[100, 70, 66.5]">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 42.3%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="https://publichealth.gwu.edu/sites/g/files/zaxdzs4586/files/styles/d07/public/2023-02/keith-crandall-up-2014-jmc_3352lowres.jpeg" class="img-fluid figure-img"></p>
<figcaption>Dr.&nbsp;Keith Crandall</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 29.6%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="https://publichealth.gwu.edu/sites/g/files/zaxdzs4586/files/styles/d07/public/2024-11/ali_rahnavard_102024.jpg" class="img-fluid figure-img"></p>
<figcaption>Dr.&nbsp;Ali Rahnavard</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 28.1%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="https://chiraaggohel.com/favicon.jpeg" class="img-fluid figure-img"></p>
<figcaption>Chiraag Gohel</figcaption>
</figure>
</div>
</div>
</div>
</div>
</section>
<section id="introductions" class="level2">
<h2 class="anchored" data-anchor-id="introductions">Introductions</h2>
<ol type="1">
<li>Name, Degree, any other ways you like to define yourself (favorite hot dog brand, etc.)</li>
<li>Current research focus or research interests.</li>
<li>What you hope to get out of this course.</li>
</ol>
</section>
<section id="course-site" class="level2">
<h2 class="anchored" data-anchor-id="course-site">Course Site</h2>
<p><a href="https://gwcbi.github.io/StatGen/" class="uri">https://gwcbi.github.io/StatGen/</a></p>
<iframe src="https://gwcbi.github.io/StatGen/" width="100%" height="600px">
</iframe>
</section>
<section id="zotero" class="level2">
<h2 class="anchored" data-anchor-id="zotero">Zotero</h2>
<p><img src="images/paste-1.png" class="img-fluid"></p>
</section>
<section id="course-logistics-and-expectations" class="level2">
<h2 class="anchored" data-anchor-id="course-logistics-and-expectations">Course Logistics and Expectations</h2>
<section id="grading-breakdown" class="level3">
<h3 class="anchored" data-anchor-id="grading-breakdown">Grading Breakdown</h3>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Assignment Type</th>
<th>% of Grade</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Problem Sets</td>
<td>60%</td>
</tr>
<tr class="even">
<td>Class Participation</td>
<td>20%</td>
</tr>
<tr class="odd">
<td>Research Project</td>
<td>20%</td>
</tr>
</tbody>
</table>
<p><strong>Total Workload</strong>: 112.5 hours (5+ hrs/week independent + 6 hrs/week class/async)</p>
</section>
</section>
<section id="problem-sets-60-of-grade" class="level2">
<h2 class="anchored" data-anchor-id="problem-sets-60-of-grade">Problem Sets (60% of grade)</h2>
<ul>
<li><strong>Weekly assignments</strong> blending mathematical derivations with coding</li>
<li><strong>Collaborative time</strong> provided in class</li>
<li><strong>Individual work</strong> required – substantial time outside class</li>
<li><strong>Late penalty</strong>: 1% per hour (first 5 hours), then 5% per day</li>
<li><strong>Format</strong>: Mix of theory problems and R/Bioconductor analysis</li>
</ul>
</section>
<section id="class-participation-20-of-grade" class="level2">
<h2 class="anchored" data-anchor-id="class-participation-20-of-grade">Class Participation (20% of grade)</h2>
<p><strong>More than just showing up!</strong></p>
<ul>
<li>Come <strong>prepared</strong> with questions from async materials</li>
<li><strong>Engage thoughtfully</strong> in discussions and problem-solving</li>
<li><strong>Demonstrate mastery</strong> through insightful contributions</li>
<li><strong>Collaborate respectfully</strong> - help others understand concepts</li>
</ul>
<p><strong>Quantitative + Qualitative Assessment</strong>: Async responses + live session engagement</p>
</section>
<section id="research-project-20-of-grade" class="level2">
<h2 class="anchored" data-anchor-id="research-project-20-of-grade">Research Project (20% of grade)</h2>
<p><strong>Final third of semester</strong> - Choose your own adventure!</p>
<div class="quarto-layout-panel" data-layout-ncol="2">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p><strong>Project Types</strong></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<ul>
<li>Computational methods</li>
<li>Applied analysis</li>
<li>Methodological development</li>
<li>All based on primary literature</li>
</ul>
</div>
</div>
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p><strong>Deliverables</strong></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<ul>
<li>Scientific paper (Genetics journal format)</li>
<li>Conference-style presentation</li>
<li>Groups of up to 3 encouraged</li>
<li>Individual contributions must be clear</li>
</ul>
</div>
</div>
</div>
</section>
<section id="course-structure-overview" class="level2">
<h2 class="anchored" data-anchor-id="course-structure-overview">Course Structure Overview</h2>
<p><strong>Foundation → Scale → Specialization</strong> (10 weeks)</p>
<ul>
<li><strong>Weeks 1-4</strong>: Foundations (Mendelian genetics, heritability, likelihood algorithms, population structure &amp; Bayesian methods)</li>
<li><strong>Week 5</strong>: GWAS at scale (design, linear mixed models, meta-analysis)</li>
<li><strong>Weeks 6-7</strong>: Prediction models &amp; multiple testing control</li>
<li><strong>Weeks 8-9</strong>: Binary traits &amp; causal inference (Mendelian randomization)</li>
<li><strong>Week 10</strong>: Advanced AI topics in statistical genetics</li>
</ul>
<p><strong>Lab Component</strong>: Hands-on R/Bioconductor workflows parallel to lectures</p>
</section>
<section id="technology-setup-requirements" class="level2">
<h2 class="anchored" data-anchor-id="technology-setup-requirements">Technology Setup Requirements</h2>
<div class="incremental">
<ul class="incremental">
<li>R (4.5+) and IDE installation</li>
<li>Git setup and basic commands</li>
<li>Zotero for reference management</li>
<li>Zoom</li>
<li>Optional: HPC access</li>
</ul>
</div>
</section>
<section id="course-textbook" class="level2">
<h2 class="anchored" data-anchor-id="course-textbook">Course Textbook</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/paste-6.png" class="img-fluid figure-img"></p>
<figcaption><span class="citation" data-cites="sorensenStatisticalLearningGenetics2025">Sorensen (<a href="#ref-sorensenStatisticalLearningGenetics2025" role="doc-biblioref">2025</a>)</span></figcaption>
</figure>
</div>
</section>
<section id="what-is-statistical-genetics" class="level2">
<h2 class="anchored" data-anchor-id="what-is-statistical-genetics">What is statistical genetics?</h2>
<div class="quarto-layout-panel" data-layout-ncol="3">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 33.3%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/a/ae/Darwin%27s_finches_by_Gould.jpg/1590px-Darwin%27s_finches_by_Gould.jpg" class="img-fluid figure-img"></p>
<figcaption>Population Genetics</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 33.3%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41576-022-00478-5/MediaObjects/41576_2022_478_Fig2_HTML.png" class="img-fluid figure-img"></p>
<figcaption>Genetic Epidemiology</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 33.3%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="https://brightspotcdn.byu.edu/dims4/default/6658c6f/2147483647/strip/true/crop/1602x2400+0+0/resize/534x800!/quality/90/?url=https%3A%2F%2Fbrigham-young-brightspot-us-east-2.s3.us-east-2.amazonaws.com%2F9b%2Faf%2F9f5a135aaaf234b6420456b7312c%2Fbradley-measure.jpg" class="img-fluid figure-img"></p>
<figcaption>Quantitative Genetics</figcaption>
</figure>
</div>
</div>
</div>
</div>
<div class="notes">
<ul>
<li>Population genetics is the study of evolutionary processes affecting genetic variation between organisms. A classic example is Darwin’s finches, where beak shape and size vary between islands in the Galapagos, driven by natural selection.</li>
<li>Genetic Epidemiology is the study of how genetic factors contribute to health and disease in populations. A classic example is the study of genetic variants associated with diseases like diabetes or heart disease.</li>
<li>Quantitative genetics is the study of continuous traits influenced by multiple genes and environmental factors. A classic example is human height, which is influenced by many genetic variants and environmental factors like nutrition. Shawn Bradley was a basketball player who as 7’6 tall, The study found Bradley had 198 more height-associated genetic variants than the average person in the sample</li>
</ul>
</div>
</section>
<section id="a-statistical-geneticist-may-want-to-know" class="level2">
<h2 class="anchored" data-anchor-id="a-statistical-geneticist-may-want-to-know">A statistical geneticist may want to know</h2>
<ol type="1">
<li>Is there a genetic component contributing to the total variance of these traits?</li>
<li>Is the genetic component of the traits driven by a few genes located on a particular chromosome, or are there many genes scattered across many chromosomes? How many genes are involved and is this a scientifically sensible question?</li>
<li>Are the genes detected protein-coding genes, or are there also noncoding genes involved in gene regulation?</li>
<li>How is the strength of the signals captured in a statistical analysis related to the two types of genes? What fraction of the total genetic variation is allocated to both types of genes?</li>
<li>What are the frequencies of the genes in the sample? Are the frequencies associated with the magnitude of their effects on the traits?</li>
<li>What is the mode of action of the genes?</li>
</ol>
</section>
<section id="a-statistical-geneticist-may-want-to-know-1" class="level2">
<h2 class="anchored" data-anchor-id="a-statistical-geneticist-may-want-to-know-1">A statistical geneticist may want to know</h2>
<ol start="7" type="1">
<li>What proportion of the genetic variance estimated in 1 can be explained by the discovered genes?</li>
<li>Given the information on the set of genes carried by an individual, will a genetic score constructed before observing the trait help with early diagnosis and prevention?</li>
<li>How should the predictive ability of the score be measured?</li>
<li>Are there other non-genetic factors that affect the traits, such as smoking behavior, alcohol consumption, blood pressure measurements, body mass index and level of physical exercise?</li>
<li>Could the predictive ability of the genetic score be improved by incorporation of these non-genetic sources of information, either additively or considering interactions? What is the relative contribution from the different sources of information?</li>
</ol>
</section>
<section id="what-does-the-data-look-like" class="level2">
<h2 class="anchored" data-anchor-id="what-does-the-data-look-like">What does the data look like?</h2>
<p><strong>Family/Pedigree Studies</strong></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="https://www.genome.gov/sites/default/files/media/images/tg/Pedigree.jpg" class="img-fluid figure-img"></p>
<figcaption>Family/Pedigree Studies</figcaption>
</figure>
</div>
<div class="notes">
<ul>
<li>Family studies are used to understand the inheritance patterns of traits and diseases within families. They can help identify genetic factors contributing to diseases by analyzing the occurrence of traits in related individuals.</li>
<li>Pedigree charts visually represent family relationships and the transmission of genetic traits across generations. They are useful for tracking the inheritance of specific traits or diseases within a family.</li>
</ul>
</div>
</section>
<section id="what-does-the-data-look-like-1" class="level2">
<h2 class="anchored" data-anchor-id="what-does-the-data-look-like-1">What does the data look like?</h2>
<p><strong>Genome-Wide Association Studies (GWAS)</strong></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="https://upload.wikimedia.org/wikipedia/commons/4/4f/Manhattan_plot_from_a_GWAS_of_kidney_stone_disease.png" class="img-fluid figure-img" style="width:60.0%"></p>
<figcaption>Genome-Wide Association Studies</figcaption>
</figure>
</div>
<div class="notes">
<ul>
<li>GWAS are studies that look for associations between genetic variants across the genome and specific traits or diseases in a large population. They can help identify genetic factors contributing to complex diseases by analyzing the frequency of genetic variants in affected and unaffected individuals.</li>
<li>Manhattan plots are visual representations of GWAS results, where each point represents a genetic variant and its association with the trait of interest. The x-axis represents the genomic position, while the y-axis represents the significance of the association (usually -log10(p-value)). Peaks in the plot indicate regions of the genome that may be associated with the trait.</li>
</ul>
</div>
</section>
<section id="core-abstractions-for-quantitative-biology" class="level2">
<h2 class="anchored" data-anchor-id="core-abstractions-for-quantitative-biology">Core abstractions for quantitative biology</h2>
<ol type="1">
<li><strong>Model-Building</strong>
<ul>
<li>Formulate generative models linking genotype, environment, and phenotype (e.g., linear mixed, Bayesian hierarchical, non-parametric kernels).</li>
<li>Encode biological structure: linkage &amp; LD, population stratification, dominance/epistasis, multi-omics priors.</li>
<li>Balance realism and tractability to enable scalable computation on genome-scale data.</li>
</ul></li>
</ol>
<p><img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41467-021-25756-4/MediaObjects/41467_2021_25756_Fig1_HTML.png" class="img-fluid"></p>
</section>
<section id="core-abstractions-for-quantitative-biology-1" class="level2">
<h2 class="anchored" data-anchor-id="core-abstractions-for-quantitative-biology-1">Core abstractions for quantitative biology</h2>
<ol start="2" type="1">
<li><strong>Inference</strong>
<ul>
<li>Estimate unknown parameters and latent effects via likelihood maximisation, EM, MCMC, or SGD</li>
<li>Quantify uncertainty with standard errors, posterior intervals, and credible sets</li>
<li>Control false discoveries across millions of tests with FDR/Q-value, permutation, and empirical-Bayes shrinkage.</li>
</ul></li>
</ol>
<p><img src="https://www.cs.auckland.ac.nz/~georgy/research/texture/thesis-html/img133.png" class="img-fluid"></p>
</section>
<section id="core-abstractions-for-quantitative-biology-2" class="level2">
<h2 class="anchored" data-anchor-id="core-abstractions-for-quantitative-biology-2">Core abstractions for quantitative biology</h2>
<ol start="3" type="1">
<li><strong>Prediction</strong>
<ul>
<li>Use fitted models for out-of-sample trait prediction: BLUP/GBLUP, ridge/lasso/elastic-net, Bayesian whole-genome regressions, random forests, neural nets.</li>
<li>Evaluate accuracy (MSE, AUC), bias–variance trade-off, calibration, and portability across ancestries or cell types.</li>
<li>Translate genomic predictions into actionable scores for breeding, risk stratification, and drug-target prioritisation.</li>
</ul></li>
</ol>
<p><img src="https://www.cell.com/cms/10.1016/j.xgen.2024.100523/asset/29efca68-0928-4eed-94b9-9799e7cedda2/main.assets/gr5_lrg.jpg" class="img-fluid"></p>
</section>
<section id="core-abstractions-for-quantitative-biology-3" class="level2">
<h2 class="anchored" data-anchor-id="core-abstractions-for-quantitative-biology-3">Core abstractions for quantitative biology</h2>
<ol start="4" type="1">
<li><strong>Interpretation &amp; Validation</strong>
<ul>
<li>Integrate functional annotations, eQTL, and single-cell data to refine biological mechanisms.</li>
<li>Perform replication, cross-cohort meta-analysis, and sensitivity analyses to population assumptions.</li>
<li>Communicate findings with clear visualisations and reproducible workflows (R/Bioconductor, Git, notebooks).</li>
</ul></li>
</ol>
<div class="quarto-layout-panel" data-layout="[1.5, -.2 ,1, -.2 ,2]">
<div class="quarto-layout-row quarto-layout-valign-center">
<div class="quarto-layout-cell" style="flex-basis: 30.6%;justify-content: center;">
<p><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/d/db/Nature_journal_logo.svg/1200px-Nature_journal_logo.svg.png" class="img-fluid"></p>
</div>
<div class="quarto-figure-spacer quarto-layout-cell" style="flex-basis: 4.1%;justify-content: flex-start;">
<p>&nbsp;</p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 20.4%;justify-content: center;">
<p><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/3/38/Jupyter_logo.svg/1200px-Jupyter_logo.svg.png" class="img-fluid"></p>
</div>
<div class="quarto-figure-spacer quarto-layout-cell" style="flex-basis: 4.1%;justify-content: flex-start;">
<p>&nbsp;</p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 40.8%;justify-content: center;">
<p><img src="https://www.bioconductor.org/images/logo/png/bioconductor_logo_rgb.png" class="img-fluid"></p>
</div>
</div>
</div>
</section>
<section id="early-work-in-the-field" class="level2">
<h2 class="anchored" data-anchor-id="early-work-in-the-field">Early work in the field</h2>
<div class="quarto-layout-panel" data-layout="[70, 30]">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 70.0%;justify-content: center;">
<p><img src="https://cdn.kastatic.org/ka-perseus-images/b966bcd0e55933e24836c4afe4a2a9889247579c.png" class="img-fluid"></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 30.0%;justify-content: center;">
<p><img src="https://cdn.kastatic.org/ka-perseus-images/97ef4b490d977c359786b040ead67b9cdaeac1aa.png" class="img-fluid"></p>
</div>
</div>
</div>
<div class="notes">
<ul>
<li>Gregor Mendel (1822-1884) was an Austrian monk and scientist who is considered the father of modern genetics. He conducted experiments on pea plants and discovered the fundamental laws of inheritance, including the concepts of dominant and recessive traits, segregation, and independent assortment.</li>
<li>Mendel’s work laid the foundation for the field of genetics, but it was largely ignored during his lifetime. It wasn’t until the early 20th century that his findings were rediscovered and integrated into the broader understanding of biology.</li>
<li>It is hypothesized that Charles Darwin would have begun his work on evolution much earlier if he had access to Mendel’s work on inheritance.</li>
</ul>
</div>
</section>
<section id="genetics-statistics-and-eugenics-have-always-been-closely-linked" class="level2">
<h2 class="anchored" data-anchor-id="genetics-statistics-and-eugenics-have-always-been-closely-linked">Genetics, statistics, and eugenics, have always been closely linked</h2>
<div class="quarto-layout-panel" data-layout="[35, 65]">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 35.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/a/a7/PSM_V60_D232_Standard_eugenics_scheme_of_descent.png/960px-PSM_V60_D232_Standard_eugenics_scheme_of_descent.png" class="img-fluid figure-img"></p>
<figcaption>Standard eugenics scheme of descent</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 65.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/galton.png" class="img-fluid figure-img"></p>
<figcaption>Francis Galton and his work, “Inquires into Human Faculty and its Development”</figcaption>
</figure>
</div>
</div>
</div>
</div>
<div class="notes">
<ul>
<li>Francis Galton (1822-1911) was an English polymath who made significant contributions to various fields, including statistics, psychology, and anthropology. He is best known for his work on eugenics, a now-discredited movement that aimed to improve the genetic quality of the human population through selective breeding.</li>
<li>Galton was a cousin of Charles Darwin and was influenced by Darwin’s theory of evolution. He believed that intelligence and other traits were hereditary and could be measured and quantified. He developed statistical methods, such as correlation and regression, to study the inheritance of traits.</li>
<li>Galton’s work on eugenics has been widely criticized for its ethical implications and its association with discriminatory practices, such as forced sterilizations and racial segregation</li>
</ul>
</div>
</section>
<section id="the-ethics-of-genetics-and-genomics-iswas-of-great-importance" class="level2">
<h2 class="anchored" data-anchor-id="the-ethics-of-genetics-and-genomics-iswas-of-great-importance">The ethics of genetics and genomics is/(was) of great importance!</h2>
<p><img src="images/nhgri_ethics.png" class="img-fluid"></p>
</section>
<section id="some-vocabulary" class="level2">
<h2 class="anchored" data-anchor-id="some-vocabulary">Some vocabulary</h2>
<ul>
<li><strong>Trait/Phenotype</strong>: A measurable characteristic of an organism, such as height, weight, or disease status.</li>
<li><strong>Gene</strong>: Unit of inheritance</li>
<li><strong>Genotype</strong>: The genetic constitution of an individual, often represented by specific alleles at particular loci.</li>
<li><strong>Allele</strong>: A variant form of a gene that can exist at a specific locus on a chromosome.</li>
<li><strong>Locus</strong>: A specific, fixed position on a chromosome where a particular gene or genetic marker is located.</li>
<li><strong>Diploid</strong>: Having two sets of chromosomes (one from each parent)</li>
<li><strong>Penetrance</strong>: Likelihood of expressing a phenotype given a genotype</li>
<li><strong>Polymorphism</strong>: The occurrence of two or more genetically determined forms in a population, such as single nucleotide polymorphisms (SNPs) or copy number variations (CNVs).</li>
<li><strong>Genetic Marker</strong>: A specific DNA sequence with a known location on a chromosome that can be used to identify individuals or species, often used in genetic mapping or association studies.</li>
</ul>
</section>
<section id="some-vocabulary-1" class="level2">
<h2 class="anchored" data-anchor-id="some-vocabulary-1">Some vocabulary</h2>
<ul>
<li><strong>Mendelian Disease</strong>: A disease caused by a mutation in a single gene, following Mendelian inheritance patterns (dominant, recessive, X-linked).</li>
<li><strong>Linkage Disequilibrium (LD)</strong>: The non-random association of alleles at different loci, indicating that certain allele combinations occur together more frequently than expected by chance.</li>
<li><strong>Heritability</strong>: The proportion of phenotypic variance in a trait that can be attributed to genetic variance, often estimated through twin or family studies.</li>
<li><strong>Genome-Wide Association Study (GWAS)</strong>: A study that looks for associations between genetic variants across the genome and specific traits or diseases in a large population.</li>
<li><strong>Polygenic Score (PGS)</strong>: A score that aggregates the effects of multiple genetic variants to predict an individual’s genetic predisposition to a trait or disease.</li>
<li><strong>Quantitative Trait Locus (QTL)</strong>: A region of the genome that is associated with a quantitative trait, often identified through linkage or association mapping.</li>
<li><strong>Epistasis</strong>: The interaction between genes where the effect of one gene is modified by one or more other genes, influencing the expression of a trait.</li>
</ul>
</section>
<section id="genetic-variant-example" class="level2">
<h2 class="anchored" data-anchor-id="genetic-variant-example">Genetic Variant Example</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/paste-3.png" class="img-fluid figure-img"></p>
<figcaption><span class="citation" data-cites="laird2011">Laird and Lange (<a href="#ref-laird2011" role="doc-biblioref">2011</a>)</span></figcaption>
</figure>
</div>
</section>
<section id="probability-refresher" class="level2">
<h2 class="anchored" data-anchor-id="probability-refresher">Probability Refresher</h2>
<section id="random-variables-and-distributions" class="level3">
<h3 class="anchored" data-anchor-id="random-variables-and-distributions">Random Variables and Distributions</h3>
<div class="incremental">
<ul class="incremental">
<li>A <strong>random variable</strong> is a numerical summary of randomness (e.g., the count of (A) alleles in a sample). We will denote random variables with uppercase letters (e.g., <span class="math inline">\(X\)</span>, <span class="math inline">\(Y\)</span>).</li>
<li>We use a <strong>model</strong> (e.g., binomial/multinomial) to describe how data vary from sample to sample.
<ul class="incremental">
<li>This is otherwise known as a probability mass function (pmf) for discrete variables, or a probability density function (pdf) for continuous variables</li>
</ul></li>
<li>Key ideas:
<ul class="incremental">
<li><strong>Expectation (mean)</strong> and <strong>variance</strong>: long‑run average and spread.
<ul class="incremental">
<li><span class="math inline">\(E[X] = \sum x \cdot P(X = x)\)</span> (discrete) or <span class="math inline">\(E[X] = \int x f(x) dx\)</span> (continuous)</li>
<li><span class="math inline">\(\textsf{Var}(X) = E[(X - E[X])^2] = E[X^2] - (E[X])^2\)</span></li>
</ul></li>
<li><strong>Standard error (SE)</strong>: typical sampling variability of an estimator.
<ul class="incremental">
<li><span class="math inline">\(\textsf{SE} = \frac{\sigma}{\sqrt{n}}\)</span> (uncertainty in sample estimates)</li>
</ul></li>
<li><strong>Central Limit Theorem (CLT)</strong>: averages and proportions are roughly normal for large (n).
<ul class="incremental">
<li><span class="math inline">\(\bar{X} \sim N\left(\mu, \frac{\sigma^2}{n}\right)\)</span> as <span class="math inline">\(n \to \infty\)</span></li>
</ul></li>
</ul></li>
</ul>
</div>
</section>
</section>
<section id="probability-distributions-vs-sampling-distributions" class="level2">
<h2 class="anchored" data-anchor-id="probability-distributions-vs-sampling-distributions">Probability Distributions vs Sampling Distributions</h2>
<ul>
<li>A <strong>probability distribution</strong> describes how a random variable behaves in the population.
<ul>
<li>Example: <span class="math inline">\(X \sim \textsf{Binomial}(n, p)\)</span></li>
</ul></li>
<li>A <strong>sampling distribution</strong> describes how a statistic (e.g., sample mean, sample proportion) behaves across repeated samples from the population.
<ul>
<li>Example: <span class="math inline">\(\hat{p} \sim N\left(p, \frac{(p(1-p))}{n}\right)\)</span> as <span class="math inline">\(n \to \infty\)</span></li>
</ul></li>
</ul>
</section>
<section id="probability-refresher-1" class="level2">
<h2 class="anchored" data-anchor-id="probability-refresher-1">Probability Refresher</h2>
<section id="key-distributions-in-statistical-genetics" class="level3">
<h3 class="anchored" data-anchor-id="key-distributions-in-statistical-genetics">Key Distributions in Statistical Genetics</h3>
<div class="quarto-layout-panel" data-layout-ncol="2">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p><strong>Binomial Distribution</strong> <span class="math display">\[X \sim \textsf{Binomial}(n, p)\]</span> <span class="math display">\[P(X = k) = \binom{n}{k} p^k (1-p)^{n-k}\]</span></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p>Used for: Allele counts</p>
</div>
</div>
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p><strong>Normal Distribution</strong> <span class="math display">\[X \sim N(\mu, \sigma^2)\]</span> <span class="math display">\[f(x) = \frac{1}{\sqrt{2\pi\sigma^2}} e^{-\frac{(x-\mu)^2}{2\sigma^2}}\]</span></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p>Used for: Quantitative traits, effect sizes</p>
</div>
</div>
</div>
</section>
</section>
<section id="conditional-probability-and-independence" class="level2">
<h2 class="anchored" data-anchor-id="conditional-probability-and-independence">Conditional Probability and Independence</h2>
<ul>
<li><strong>Conditional Probability</strong>: <span class="math inline">\(P(A|B) = \frac{P(A \cap B)}{P(B)}\)</span></li>
<li><strong>Bayes’ Theorem</strong>: <span class="math inline">\(P(A|B) = \frac{P(B|A)P(A)}{P(B)}\)</span></li>
<li><strong>Independence</strong>: <span class="math inline">\(P(A \cap B) = P(A) \cdot P(B)\)</span> <span class="math inline">\(\iff\)</span> <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span> are independent</li>
</ul>
<p>Conditional probability and Bayes’ Theorem are crucial for understanding rare disease genetics! For example, consider a rare disease with a prevalence of 1 in 10,000. If a genetic test has a sensitivity of 99% and a specificity of 99%, we can use Bayes’ Theorem to calculate the probability that an individual has the disease given a positive test result.</p>
<div class="fragment">
<p><span class="math display">\[\begin{align*}
P(Disease|Positive) &amp;= \frac{P(Positive|Disease)P(Disease)}{P(Positive)} \\
&amp;= \frac{(0.99)(0.0001)}{(0.99)(0.0001) + (0.01)(0.9999)} \\
&amp;\approx 0.0098
\end{align*}\]</span></p>
</div>
</section>
<section id="parameter-estimation" class="level2">
<h2 class="anchored" data-anchor-id="parameter-estimation">Parameter Estimation</h2>
<ul>
<li>Imagine a geneticist is studying <strong>allele ages</strong> (i.e., how many generations ago an allele arose via mutation) under a simplified model where new mutations arise randomly and uniformly over a fixed window — say, the last <span class="math inline">\(\theta\)</span> generations.</li>
<li>We could model <span class="math inline">\(X \sim \textsf{Uniform}(0, \theta)\)</span>.</li>
<li>We have our observed sample <span class="math inline">\(\{x_1,x_2, \ldots, x_n\}\)</span>, ages in generations of <span class="math inline">\(n\)</span> observed alleles.
<ul>
<li>For concreteness, let our sample be <span class="math inline">\(\{0, 1, 8, 4, 3, 7, 7, 6\}\)</span>.</li>
</ul></li>
<li>How should we estimate <span class="math inline">\(\theta\)</span> ?</li>
</ul>
<div class="fragment">
<div class="cell quarto-layout-panel" data-layout-ncol="2">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="lecture-01_files/figure-html/unnamed-chunk-2-1.png" class="img-fluid" width="672"></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="lecture-01_files/figure-html/unnamed-chunk-2-2.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
</div>
</section>
<section id="genetic-models" class="level2">
<h2 class="anchored" data-anchor-id="genetic-models">Genetic Models</h2>
<ul>
<li>Understanding random variables, sampling distributions, and bias/variance of estimators helps motivate biological questions, and understand how they can be answered</li>
<li>Let y be the expression of a trait, G be the additive contribution of genetic variables, and an environmental value E;</li>
<li>What assumptions do we make via the following models?</li>
</ul>
<div class="quarto-layout-panel" data-layout="[44, -12, 44]">
<div class="quarto-layout-row">
<div class="column quarto-layout-cell" style="flex-basis: 44.0%;justify-content: flex-start;">
<p><span class="math display">\[\begin{align}

y &amp;= G + E \\
y &amp;= \beta_0 + \beta_1 G + \beta_2 E + \epsilon \\
y &amp;= \beta_0 + \beta_1 G + \beta_2 E + \epsilon, \quad \epsilon \sim N(0, \sigma^2) \\
y &amp;= \beta_0 + \beta_1 G + \beta_2 E + \beta_3 \left(G \times E\right) + \epsilon, \quad \epsilon \sim N(0, \sigma^2)

\end{align}\]</span></p>
</div>
<div class="quarto-figure-spacer quarto-layout-cell" style="flex-basis: 12.0%;justify-content: flex-start;">
<p>&nbsp;</p>
</div>
<div class="column quarto-layout-cell" style="flex-basis: 44.0%;justify-content: center;">
<p><img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fsj.tpj.6500085/MediaObjects/41397_2002_Article_BF6500085_Fig1_HTML.jpg" class="img-fluid"></p>
</div>
</div>
</div>
</section>
<section id="modeling" class="level2">
<h2 class="anchored" data-anchor-id="modeling">Modeling</h2>
<ul>
<li>Imagine that we have a sample size of <span class="math inline">\(n\)</span> unrelated haploid individuals from some population</li>
<li>We want to estimate allele frequencies for a biallelic SNP, say A/a</li>
<li>In our sample, we observe <span class="math inline">\(x\)</span> individuals with allele A and <span class="math inline">\(n - x\)</span> individuals with allele a</li>
<li>Let <span class="math inline">\(p\)</span> be the frequency of allele A and <span class="math inline">\(q = 1 - p\)</span> be the frequency of allele a</li>
</ul>
<p><span class="math display">\[Pr(X = x| n, p) = \binom{n}{x} p^x q^{n - x}\]</span></p>
<ul>
<li>Lets say we observe <span class="math inline">\(n = 27\)</span> and <span class="math inline">\(x = 11\)</span>. How do we estimate <span class="math inline">\(p\)</span>?</li>
</ul>
</section>
<section id="likelihood-functions" class="level2">
<h2 class="anchored" data-anchor-id="likelihood-functions">Likelihood Functions</h2>
<ul>
<li><p>The <strong>likelihood function</strong> is a function of the parameters of a statistical model, given specific observed data.</p></li>
<li><p>It represents the plausibility of different parameter values based on the observed data.</p></li>
<li><p>For a given set of data, the likelihood function is defined as: <span class="math display">\[L(\theta | data) = P(data | \theta)\]</span></p></li>
<li><p>In our case, the likelihood function for <span class="math inline">\(p\)</span> is: <span class="math display">\[L(p | x = 11, n = 27) = P(X = 11 | n = 27, p) = \binom{27}{11} p^{11} (1 - p)^{16}\]</span></p></li>
<li><p>We are asking, “how likely is it that we would observe this data for different values of <span class="math inline">\(p\)</span>?”</p></li>
</ul>
</section>
<section id="maximum-likelihood-estimator" class="level2">
<h2 class="anchored" data-anchor-id="maximum-likelihood-estimator">Maximum Likelihood Estimator</h2>
<p>Our likelihood function for <span class="math inline">\(p\)</span> is</p>
<p><span class="math display">\[L(P|x = 11, n = 27) = {27 \choose 11} p^{11} q^{27-11}\]</span></p>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">27</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">11</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>p_seq <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">length.out =</span> <span class="dv">200</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>likelihood <span class="ot">&lt;-</span> <span class="fu">dbinom</span>(x, n, p_seq)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>(<span class="at">p =</span> p_seq, <span class="at">likelihood =</span> likelihood),</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">x =</span> p, <span class="at">y =</span> likelihood)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>(<span class="at">linewidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">x =</span> <span class="st">"Probability of Success (p)"</span>,</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">y =</span> <span class="st">"Likelihood"</span>,</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">title =</span> <span class="st">"Binomial Likelihood Function"</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">xintercept =</span> x <span class="sc">/</span> n,</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">color =</span> <span class="st">"red"</span>,</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">linetype =</span> <span class="st">"dashed"</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_light</span>(<span class="at">base_size =</span> <span class="dv">18</span>) <span class="sc">+</span></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>())</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lecture-01_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="modeling-but-make-it-bayesian" class="level2">
<h2 class="anchored" data-anchor-id="modeling-but-make-it-bayesian">Modeling, but make it bayesian</h2>
<ul>
<li>In Bayesian statistics, we treat parameters as random variables and use probability distributions to represent our uncertainty about them.</li>
<li>We start with a <strong>prior distribution</strong> that reflects our beliefs about the parameter before seeing the data.</li>
<li>We then use the observed data to update our beliefs and obtain a <strong>posterior distribution</strong>.</li>
<li>The posterior distribution combines the information from the prior and the likelihood of the observed data using Bayes’ theorem: <span class="math display">\[P(p | data) = \frac{P(data | p) P(p)}{P(data)}\]</span></li>
<li>Let’s say we have run a previous experiment and observed <span class="math inline">\(n = 20\)</span> individuals and <span class="math inline">\(x = 3\)</span> individuals with allele A (again, bilelic SNP).</li>
<li>We can use this information to inform our prior distribution for <span class="math inline">\(p\)</span></li>
</ul>
</section>
<section id="modeling-but-make-it-bayesian-1" class="level2">
<h2 class="anchored" data-anchor-id="modeling-but-make-it-bayesian-1">Modeling, but make it bayesian</h2>
<div class="r-stack">
<div class="fragment fade-in-then-out">
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">length.out =</span> <span class="dv">1000</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Prior (Beta distribution from previous experiment)</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>prior <span class="ot">&lt;-</span> <span class="fu">dbeta</span>(p, <span class="dv">3</span> <span class="sc">+</span> <span class="dv">1</span>, <span class="dv">17</span> <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Likelihood (Beta distribution from current data)</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>likelihood <span class="ot">&lt;-</span> <span class="fu">dbeta</span>(p, <span class="dv">11</span> <span class="sc">+</span> <span class="dv">1</span>, <span class="dv">16</span> <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Posterior (Beta distribution combining both)</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>posterior <span class="ot">&lt;-</span> <span class="fu">dbeta</span>(p, <span class="dv">14</span> <span class="sc">+</span> <span class="dv">1</span>, <span class="dv">33</span> <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine into a data frame</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">p =</span> <span class="fu">rep</span>(p, <span class="dv">3</span>),</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">density =</span> <span class="fu">c</span>(prior, likelihood, posterior),</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">type =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"Prior"</span>, <span class="st">"Likelihood"</span>, <span class="st">"Posterior"</span>), <span class="at">each =</span> <span class="fu">length</span>(p))</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Add flat prior distribution</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>flat_prior <span class="ot">&lt;-</span> <span class="fu">dbeta</span>(p, <span class="dv">1</span>, <span class="dv">1</span>)</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>flat_posterior <span class="ot">&lt;-</span> <span class="fu">dbeta</span>(p, <span class="dv">12</span>, <span class="dv">17</span>)</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span> <span class="fu">rbind</span>(</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>    plot_data,</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>(</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">p =</span> <span class="fu">rep</span>(p, <span class="dv">3</span>),</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>        <span class="at">density =</span> <span class="fu">c</span>(flat_prior, likelihood, flat_posterior),</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>        <span class="at">type =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"Prior"</span>, <span class="st">"Likelihood"</span>, <span class="st">"Posterior"</span>), <span class="at">each =</span> <span class="fu">length</span>(p))</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Add a column to distinguish between informative and flat scenarios</span></span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>plot_data<span class="sc">$</span>scenario <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"Informative"</span>, <span class="dv">3000</span>), <span class="fu">rep</span>(<span class="st">"Flat"</span>, <span class="dv">3000</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</div>
<div class="fragment fade-in-then-out">
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>    plot_data <span class="sc">%&gt;%</span> <span class="fu">filter</span>(scenario <span class="sc">==</span> <span class="st">"Informative"</span>),</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">x =</span> p, <span class="at">y =</span> density, <span class="at">color =</span> type)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>(<span class="at">linewidth =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">11</span> <span class="sc">/</span> <span class="dv">27</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">x =</span> <span class="st">"Allele Frequency (p)"</span>,</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">y =</span> <span class="st">"Density"</span>,</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">title =</span> <span class="st">"Bayesian Update with Informative Prior"</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">15</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  plot_data <span class="sc">%&gt;%</span> <span class="fu">filter</span>(scenario <span class="sc">==</span> <span class="st">"Flat"</span>),</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aes</span>(<span class="at">x =</span> p, <span class="at">y =</span> density, <span class="at">color =</span> type, <span class="at">linetype =</span> type)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">linewidth =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_linetype_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"Prior"</span> <span class="ot">=</span> <span class="st">"dotted"</span>, <span class="st">"Likelihood"</span> <span class="ot">=</span> <span class="st">"solid"</span>, <span class="st">"Posterior"</span> <span class="ot">=</span> <span class="st">"dashed"</span>)) <span class="sc">+</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">11</span> <span class="sc">/</span> <span class="dv">27</span>, <span class="at">linetype =</span> <span class="st">"dotdash"</span>, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"Allele Frequency (p)"</span>,</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">"Density"</span>,</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"Bayesian Update with Flat Prior"</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">15</span>) <span class="sc">+</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">"text"</span>,</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="fl">0.7</span>, <span class="at">y =</span> <span class="fu">max</span>(flat_posterior) <span class="sc">*</span> <span class="fl">0.8</span>,</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">label =</span> <span class="st">"Likelihood ≈ Posterior"</span>, <span class="at">color =</span> <span class="st">"gray30"</span></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</div>
</div>
</div>
<div class="fragment">
<div class="cell quarto-layout-panel" data-layout-ncol="2">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="lecture-01_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="672"></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="lecture-01_files/figure-html/unnamed-chunk-7-2.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="mendels-laws" class="level2">
<h2 class="anchored" data-anchor-id="mendels-laws">Mendel’s Laws</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><strong>Mendel’s First Law (Segregation)</strong><br>
One allele of each parent is randomly transmitted, with probability 1/2, to the offspring; the alleles unite randomly to form the offspring’s genotype.</p>
<p><strong>Mendel’s Second Law (Independent Assortment)</strong><br>
At different loci, alleles are transmitted independently (when loci are unlinked).</p>
</div><div class="column" style="width:50%;">
<p><img src="https://upload.wikimedia.org/wikipedia/commons/8/8e/Intermediate_inheritance_P_-_F1_-_F2.png" class="img-fluid" style="width:100.0%" alt="Punnett-style diagram illustrating Mendelian segregation and independent assortment"></p>
</div>
</div>
<div class="notes">
<ul>
<li>Mendel’s First Law, the Law of Segregation, states that during the formation of gametes (sperm and egg cells), the two alleles for a given gene separate, so that each gamete receives only one allele. This means that each parent contributes one allele to their offspring, and the combination of these alleles determines the offspring’s genotype.</li>
<li>Mendel’s Second Law, the Law of Independent Assortment, states that alleles for different genes are transmitted independently of one another, provided the genes are located on different chromosomes or are far apart on the same chromosome. This means that the inheritance of one trait does not influence the inheritance of another trait, allowing for a variety of genetic combinations in the offspring.</li>
</ul>
</div>
</section>
<section id="worked-example-segregation-31-ratio" class="level2">
<h2 class="anchored" data-anchor-id="worked-example-segregation-31-ratio">Worked Example: Segregation (3:1 Ratio)</h2>
<section id="cross-aa-aa" class="level3">
<h3 class="anchored" data-anchor-id="cross-aa-aa">Cross: Aa × Aa</h3>
<div class="quarto-layout-panel" data-layout-ncol="2">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p><strong>Punnett Square</strong></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;"><strong>A</strong></th>
<th style="text-align: left;"><strong>a</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><strong>A</strong></td>
<td style="text-align: left;">AA</td>
<td style="text-align: left;">Aa</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>a</strong></td>
<td style="text-align: left;">Aa</td>
<td style="text-align: left;">aa</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p><strong>Expected Ratios</strong></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<ul>
<li><strong>Genotypic</strong>: 1 AA : 2 Aa : 1 aa</li>
<li><strong>Phenotypic</strong>: 3 dominant : 1 recessive</li>
<li><strong>Probabilities</strong>: <span class="math inline">\(P(AA)=\frac14, P(Aa)=\frac12, P(aa)=\frac14\)</span></li>
</ul>
</div>
</div>
</div>
</section>
</section>
<section id="test-cross-example" class="level2">
<h2 class="anchored" data-anchor-id="test-cross-example">Test Cross Example</h2>
<p>Proving law of segregation with a test cross</p>
<p>Example data:</p>
<ul>
<li>Cross: Aa × aa</li>
<li>Observed: 21 dominant, 28 recessive</li>
<li>Null hypothesis: Probability of transmission is 1:1.</li>
<li>Alternative hypothesis: Probability of transmission is not 1:1.</li>
<li>How to test this?</li>
</ul>
<p>Our question is related to observation of <strong>proportions</strong> –&gt; use a chi-squared goodness-of-fit test or a binomial test.</p>
</section>
<section id="test-cross-example-1" class="level2">
<h2 class="anchored" data-anchor-id="test-cross-example-1">Test Cross Example</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>observed_testcross <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">21</span>, <span class="dv">28</span>)     <span class="co"># Dominant, Recessive</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>chisq_test <span class="ot">&lt;-</span> <span class="fu">chisq.test</span>(observed_testcross, <span class="at">p =</span> <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">0.5</span>))</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>chisq_test</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Chi-squared test for given probabilities

data:  observed_testcross
X-squared = 1, df = 1, p-value = 0.3173</code></pre>
</div>
</div>
</section>
<section id="testing-via-simulation" class="level2">
<h2 class="anchored" data-anchor-id="testing-via-simulation">Testing via Simulation</h2>
<p>We could also simulate our idea of the null hypothesis to get a sense of how extreme our observed data is.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>n_sim <span class="ot">&lt;-</span> <span class="dv">10000</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="dv">49</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>simulated <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(n_sim, N, <span class="fl">0.5</span>)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>p_value_sim <span class="ot">&lt;-</span> <span class="fu">mean</span>(simulated <span class="sc">&lt;=</span> <span class="dv">21</span> <span class="sc">|</span> simulated <span class="sc">&gt;=</span> <span class="dv">28</span>)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>p_value_sim</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.3868</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(simulated, <span class="at">breaks =</span> <span class="dv">30</span>, <span class="at">main =</span> <span class="st">"Simulated Test Cross Outcomes"</span>, <span class="at">xlab =</span> <span class="st">"Number of Dominant Offspring"</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> <span class="fu">c</span>(<span class="dv">21</span>, <span class="dv">28</span>), <span class="at">col =</span> <span class="st">"red"</span>, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lecture-01_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="on-simulations" class="level2">
<h2 class="anchored" data-anchor-id="on-simulations">On Simulations</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/paste-5.png" class="img-fluid figure-img"></p>
<figcaption><span class="citation" data-cites="downey2016">Downey (<a href="#ref-downey2016" role="doc-biblioref">2015</a>)</span></figcaption>
</figure>
</div>
</section>
<section id="estimation-of-allele-frequencies" class="level2">
<h2 class="anchored" data-anchor-id="estimation-of-allele-frequencies">Estimation of Allele Frequencies</h2>
<p><span class="math inline">\(n_{AA}\)</span> = Number of individuals with genotype AA <span class="math inline">\(n_{Aa}\)</span> = Number of individuals with genotype Aa <span class="math inline">\(n_{aa}\)</span> = Number of individuals with genotype aa</p>
<p>where, <span class="math inline">\(n = n_{AA} + n_{Aa} + n_{aa} = N\)</span>.</p>
<p>The sample proportions of allele <span class="math inline">\(A\)</span> is:</p>
<p><span class="math display">\[
\hat{p} = \frac{2n_{AA} + n_{Aa}}{2n}
\]</span></p>
<p>The usual standard error for a proportion is <span class="math inline">\(\sqrt{\hat{p}(1 - \hat{p})/2n}\)</span>, but this assumes independence of the <span class="math inline">\(2n\)</span> sampled chromosomes.</p>
</section>
<section id="hardyweinberg-equilibrium-hwe-assumptions" class="level2">
<h2 class="anchored" data-anchor-id="hardyweinberg-equilibrium-hwe-assumptions">Hardy–Weinberg Equilibrium (HWE): Assumptions</h2>
<p>Let’s reframe the problem in terms of <strong>genotype frequencies</strong>. First, some assumptions:</p>
<ul>
<li><strong>Random mating</strong></li>
<li><strong>Large population</strong> (no strong drift in one generation)</li>
<li><strong>No selection</strong>, <strong>no migration</strong>, <strong>no mutation</strong> (in the generation under study)</li>
<li><strong>Accurate genotypes</strong> (no systematic genotyping error)</li>
</ul>
<p><strong>If these hold</strong>, and the allele frequency of <span class="math inline">\(A\)</span> is <span class="math inline">\(p\)</span> (so <span class="math inline">\(q = 1-p\)</span>), we say that the population is in <strong>Hardy-Weinberg Equilibrium (HWE)</strong>. The <strong>population genotype probabilities</strong> are:</p>
<p><span class="math display">\[
P(AA)=p^2 \quad P(Aa)=2pq \quad P(aa)=q^2
\]</span></p>
<p>As a consequence, the <strong>genotype frequencies</strong> in a random sample of <span class="math inline">\(n\)</span> individuals are approximately multinomially distributed with cell probabilities <span class="math inline">\((p^2, 2pq, q^2)\)</span>. Thus, the sampling error of <span class="math inline">\(\hat{p}\)</span> is approximately <span class="math inline">\(\sqrt{p(1-p)/2n}\)</span>.</p>
<p>This confidence interval is known as a <strong>Wald confidence interval</strong>.</p>
</section>
<section id="testing-for-hwe" class="level2">
<h2 class="anchored" data-anchor-id="testing-for-hwe">Testing for HWE</h2>
<ul>
<li>With large samples, we can test for HWE using a <strong>chi-squared goodness-of-fit test</strong>.</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/paste-4.png" class="img-fluid figure-img"></p>
<figcaption><span class="citation" data-cites="laird2011">Laird and Lange (<a href="#ref-laird2011" role="doc-biblioref">2011</a>)</span></figcaption>
</figure>
</div>
</section>
<section id="hwe-worked-example" class="level2">
<h2 class="anchored" data-anchor-id="hwe-worked-example">HWE: Worked Example</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">212</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>n_AA <span class="ot">&lt;-</span> <span class="dv">175</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>n_Aa <span class="ot">&lt;-</span> <span class="dv">33</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>n_aa <span class="ot">&lt;-</span> <span class="dv">4</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Estimate allele frequency</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>p_hat <span class="ot">&lt;-</span> (<span class="dv">2</span><span class="sc">*</span>n_AA <span class="sc">+</span> n_Aa) <span class="sc">/</span> (<span class="dv">2</span><span class="sc">*</span>n)</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>q_hat <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> p_hat</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>p_hat</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.9033019</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Expected counts</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>expected <span class="ot">&lt;-</span> <span class="fu">c</span>(n <span class="sc">*</span> p_hat<span class="sc">^</span><span class="dv">2</span>, <span class="dv">2</span><span class="sc">*</span>n <span class="sc">*</span> p_hat<span class="sc">*</span>q_hat, n <span class="sc">*</span> q_hat<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>observed <span class="ot">&lt;-</span> <span class="fu">c</span>(n_AA, n_Aa, n_aa)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>expected</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 172.982311  37.035377   1.982311</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Pearson chi-squared</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>chisq_val <span class="ot">&lt;-</span> <span class="fu">sum</span>((observed <span class="sc">-</span> expected)<span class="sc">^</span><span class="dv">2</span> <span class="sc">/</span> expected)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="dv">3</span> <span class="sc">-</span> <span class="dv">1</span> <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>p_val <span class="ot">&lt;-</span> <span class="fu">pchisq</span>(chisq_val, <span class="at">df =</span> df, <span class="at">lower.tail =</span> <span class="cn">FALSE</span>)</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>p_val</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1126299</code></pre>
</div>
</div>
</section>
<section id="hwe-in-practice" class="level2">
<h2 class="anchored" data-anchor-id="hwe-in-practice">HWE in Practice</h2>
<ul>
<li>HWE is a useful null model for quality control of genotype data.</li>
<li>Deviations from HWE can indicate genotyping errors, population stratification, or selection</li>
</ul>
</section>
<section id="sampling-distributions-intuition-via-simulation" class="level2">
<h2 class="anchored" data-anchor-id="sampling-distributions-intuition-via-simulation">Sampling distributions: intuition via simulation</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>true_p <span class="ot">&lt;-</span> <span class="fl">0.3</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>sample_sizes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">10</span>, <span class="dv">100</span>, <span class="dv">500</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>n_sims <span class="ot">&lt;-</span> <span class="dv">1000</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>sampling_results <span class="ot">&lt;-</span> <span class="fu">map_dfr</span>(sample_sizes, <span class="cf">function</span>(n) {</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>  p_hats <span class="ot">&lt;-</span> <span class="fu">replicate</span>(n_sims, {</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Sample n individuals (2n alleles total)</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>    alleles <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(n, <span class="dv">2</span>, true_p)  <span class="co"># each individual contributes 0,1,2 A alleles</span></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">sum</span>(alleles) <span class="sc">/</span> (<span class="dv">2</span> <span class="sc">*</span> n)           <span class="co"># sample allele frequency</span></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tibble</span>(<span class="at">sample_size =</span> n, <span class="at">p_hat =</span> p_hats,</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>         <span class="at">theoretical_se =</span> <span class="fu">sqrt</span>(true_p <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> true_p) <span class="sc">/</span> (<span class="dv">2</span> <span class="sc">*</span> n)))</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>})</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="sampling-distributions-intuition-via-simulation-1" class="level2">
<h2 class="anchored" data-anchor-id="sampling-distributions-intuition-via-simulation-1">Sampling distributions: intuition via simulation</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot sampling distributions</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(sampling_results, <span class="fu">aes</span>(<span class="at">x =</span> p_hat)) <span class="sc">+</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">y =</span> <span class="fu">after_stat</span>(density)), <span class="at">bins =</span> <span class="dv">30</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> true_p, <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">linewidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>sample_size_name, <span class="at">scales =</span> <span class="st">"free_y"</span>) <span class="sc">+</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Sampling Distribution of p̂"</span>, <span class="at">x =</span> <span class="st">"Sample Allele Frequency (p̂)"</span>, <span class="at">y =</span> <span class="st">"Density"</span>) <span class="sc">+</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lecture-01_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="references" class="level2">




</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-downey2016" class="csl-entry" role="listitem">
Downey,A.B. (2015) <a href="https://greenteapress.com/thinkstats2/">Think stats: Exploratory data analysis in python</a> 2nd ed. O’Reilly Media, Sebastopol, CA.
</div>
<div id="ref-laird2011" class="csl-entry" role="listitem">
Laird,N.M. and Lange,C. (2011) <a href="https://doi.org/10.1007/978-1-4419-7338-2">The fundamentals of modern statistical genetics</a> Springer, New York, NY.
</div>
<div id="ref-sorensenStatisticalLearningGenetics2025" class="csl-entry" role="listitem">
Sorensen,D. (2025) <a href="https://doi.org/10.1007/978-3-031-86274-8">Statistical learning in genetics: An introduction using r</a> 2nd ed. Springer, Cham.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>