<!DOCTYPE html>
<html lang="en"><head>
<link href="../course_logo.png" rel="icon" type="image/png">
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-html/tabby.min.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.8.25">

  <title>Statistical Genetics – Lecture 04: Population Structure and Bayesian Methods in Statistical Genetics</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    html { -webkit-text-size-adjust: 100%; }
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; font-weight: bold; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="../site_libs/revealjs/dist/theme/quarto-b35a3bfecdac5fc8de409df89ff20628.css">
  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.css">
  <link href="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Lecture 04: Population Structure and Bayesian Methods in Statistical Genetics</h1>
  <p class="subtitle">PUBH 8878, Statistical Genetics</p>

<div class="quarto-title-authors">
</div>

</section>
<section>
<section id="intro-to-bayesian-methods" class="title-slide slide level1 center">
<h1>Intro to Bayesian Methods</h1>

</section>
<section id="the-bayesian-paradigm" class="slide level2">
<h2>The Bayesian Paradigm</h2>
<style>
    figcaption, p.caption {
        text-align: center;
    }
</style>
<ul>
<li>Let <span class="math inline">\theta</span> be unknown parameters, <span class="math inline">y</span> observed data</li>
<li>While frequentist methods treat <span class="math inline">\theta</span> as fixed but unknown, Bayesian methods treat <span class="math inline">\theta</span> as <strong>random variables</strong></li>
</ul>

<img data-src="images/thomas-bayes.png" class="r-stretch quarto-figure-center"><p class="caption">Sir Thomas Bayes</p></section>
<section id="bayes-theorem" class="slide level2">
<h2>Bayes’ Theorem</h2>
<ul>
<li>Uncertainty in <span class="math inline">\theta</span> is modeled via a <strong>prior distribution</strong> <span class="math inline">p(\theta)</span></li>
<li>Observed data <span class="math inline">y</span> are modeled via a <strong>likelihood</strong> <span class="math inline">p(y|\theta)</span></li>
<li>Bayes’ theorem combines these to yield the <strong>posterior distribution</strong> <span class="math inline">p(\theta|y) \propto p(y|\theta)p(\theta)</span></li>
</ul>

<img data-src="images/bayes-dist.png" class="r-stretch"></section>
<section id="frequentist-vs-bayesian-terminology" class="slide level2">
<h2>Frequentist vs Bayesian Terminology</h2>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top" style="width:100%;">
<colgroup>
<col style="width: 22%">
<col style="width: 36%">
<col style="width: 41%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Concept</th>
<th style="text-align: left;">Frequentist</th>
<th style="text-align: left;">Bayesian</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Parameter</td>
<td style="text-align: left;">Fixed</td>
<td style="text-align: left;">Random</td>
</tr>
<tr class="even">
<td style="text-align: left;">Data</td>
<td style="text-align: left;">Random</td>
<td style="text-align: left;">Fixed</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Estimate</td>
<td style="text-align: left;">Point estimate</td>
<td style="text-align: left;">Posterior distribution</td>
</tr>
<tr class="even">
<td style="text-align: left;">Confidence Interval</td>
<td style="text-align: left;">Interval with coverage</td>
<td style="text-align: left;">Credible interval</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Hypothesis Test</td>
<td style="text-align: left;">Reject/Fail to reject</td>
<td style="text-align: left;">Posterior predictive check</td>
</tr>
<tr class="even">
<td style="text-align: left;">P-value</td>
<td style="text-align: left;">Probability of data under null</td>
<td style="text-align: left;">Probability of hypothesis given data</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Likelihood</td>
<td style="text-align: left;">Function of parameter given data</td>
<td style="text-align: left;">Function of data given parameter</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="interpretation-of-intervals" class="slide level2">
<h2>Interpretation of Intervals</h2>
<p>Compare statements on intervals</p>
<ul>
<li>Frequentist: “Over many repeated samples, 95% of such intervals will contain the true parameter.”</li>
<li>Bayesian: “Given the observed data and prior, there is a 95% probability that the parameter lies within this interval.”</li>
</ul>
</section>
<section id="bayesian-pros-and-cons" class="slide level2">
<h2>Bayesian Pros and Cons</h2>
<h3 id="pros">Pros</h3>
<ul>
<li>Intuitive: direct probability statements about parameters</li>
<li>Flexibility: complex models, hierarchical structures, small-n</li>
<li>Incorporate prior knowledge or expert opinion</li>
<li>Full uncertainty quantification via posterior distributions</li>
</ul>
<h3 id="cons">Cons</h3>
<ul>
<li>Computationally intensive: MCMC, variational inference</li>
<li>Sensitivity to prior choice</li>
<li>Not the status quo in many fields</li>
</ul>
</section>
<section id="further-resources" class="slide level2">
<h2>Further Resources</h2>
<div class="quarto-layout-panel" data-layout="[60, 40]">
<div class="quarto-layout-row quarto-layout-valign-bottom">
<div class="quarto-layout-cell" style="flex-basis: 60.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/statistical-rethinking.png"></p>
<figcaption><span class="citation" data-cites="mcelreathStatisticalRethinkingBayesian2020">McElreath (<a href="#/reference" role="doc-biblioref" onclick="">2020</a>)</span></figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 40.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://sites.stat.columbia.edu/gelman/book/bda_cover.png"></p>
<figcaption><span class="citation" data-cites="gelmanBayesianDataAnalysis2013">Gelman <em>et al.</em> (<a href="#/reference" role="doc-biblioref" onclick="">2013</a>)</span></figcaption>
</figure>
</div>
</div>
</div>
</div>
</section>
<section id="conjugacy" class="slide level2">
<h2>Conjugacy</h2>
<ul>
<li>A prior is <strong>conjugate</strong> to a likelihood if the posterior is in the same family as the prior.</li>
<li>Example: Beta prior + Binomial likelihood <span class="math inline">to</span> Beta posterior</li>
</ul>
</section>
<section id="conjugacy-example" class="slide level2">
<h2>Conjugacy Example</h2>
<p>Consider the allele-frequency estimation setup introduced in Lecture 1.</p>
<ul>
<li>Previous experiment: observed <span class="math inline">x_{\text{prev}} = 3</span> successes (allele A) out of <span class="math inline">n_{\text{prev}} = 20</span> trials.</li>
<li>Current data: <span class="math inline">x = 11</span> successes out of <span class="math inline">n = 27</span> trials.</li>
<li>Parameter of interest: allele (“success”) frequency <span class="math inline">p</span>.</li>
</ul>
</section>
<section id="step-1-prior" class="slide level2">
<h2>Step 1: Prior</h2>
<p>From the previous experiment (3 successes out of 20 trials), we can form a Beta prior:</p>
<p>Interpret <span class="math inline">\alpha</span> and <span class="math inline">\beta</span> as pseudo-counts of successes and failures, respectively. (Note that <span class="math inline">\text{Beta}(1,1)</span> is uniform on <span class="math inline">0,1</span>.)</p>
<p><span class="math display">\begin{gather*}
p \sim \text{Beta}(\alpha, \beta) \\
\alpha = x_{\text{prev}} + 1 = 4 \\
\beta = n_{\text{prev}} - x_{\text{prev}} + 1 = 18.
\end{gather*}</span></p>
</section>
<section id="step-2-likelihood" class="slide level2">
<h2>Step 2: Likelihood</h2>
<p>Current data likelihood (up to proportionality in <span class="math inline">p</span>): <span class="math display">L(p; x,n) \propto p^{x} (1-p)^{n-x} = p^{11}(1-p)^{16}.</span></p>
<p>Recognize this kernel as</p>
<p><span class="math display">
\text{Beta}(x+1, n-x+1)=\text{Beta}(12,17)
</span></p>
</section>
<section id="step-3-posterior-conjugacy" class="slide level2">
<h2>Step 3: Posterior (Conjugacy)</h2>
<p>Multiply prior and likelihood kernels: <span class="math display">p^{\alpha-1}(1-p)^{\beta-1}\times p^{x}(1-p)^{n-x} = p^{(\alpha + x) - 1}(1-p)^{(\beta + n - x) - 1}.</span></p>
<p>Thus</p>
<p><span class="math display">\begin{align*}
p | x \sim \text{Beta}(\alpha + x,\ \beta + n - x)
&amp;= \text{Beta}(4+11,\ 18+16) \\
&amp;= \text{Beta}(15,34).
\end{align*}</span></p>
</section>
<section id="posterior-summaries" class="slide level2">
<h2>Posterior Summaries</h2>
<ul>
<li><p>Posterior mean: <span class="math inline">E[p|x] = \frac{15}{15+34} = 0.306</span> (shrinks slightly toward prior mean <span class="math inline">\frac{4}{22}=0.182</span> relative to MLE <span class="math inline">\hat p = 11/27 \approx 0.407</span>).</p></li>
<li><p>Effective sample size intuition: prior contributes <span class="math inline">(\alpha+\beta-2)=20</span> pseudo-trials; data contribute 27 real trials.</p></li>
</ul>
</section>
<section id="visualization" class="slide level2">
<h2>Visualization</h2>

<img data-src="lecture-04_files/figure-revealjs/viz-1.png" width="960" class="r-stretch"></section>
<section id="non-conjugate-models" class="slide level2">
<h2>Non-Conjugate Models</h2>
<ul>
<li>Conjugacy is convenient but limited to simple models</li>
<li>Many realistic models (e.g., multinomial/Dirichlet with nonlinear transforms) are non-conjugate</li>
<li>Bayesian inference in these models typically requires computational methods like MCMC or variational inference</li>
<li>Richard McElreath has a great <a href="https://www.youtube.com/watch?v=rZk2FqX2XnY">youtube video</a> on MCMC</li>
</ul>
</section>
<section id="bayesian-revisit-abo-allele-frequencies" class="slide level2">
<h2>Bayesian Revisit: ABO Allele Frequencies</h2>
<ul>
<li>Goal: Infer allele frequencies <code>(p_A, p_B, p_O)</code> given phenotype counts <code>(n_A, n_AB, n_B, n_O)</code>.</li>
<li>Frequentist (Lecture 03): EM treats latent genotypes for A and B phenotypes.</li>
<li>Bayesian: Place a prior on allele frequencies; integrate (average) over uncertainty rather than impute expected counts.</li>
</ul>
</section>
<section id="data-sufficient-statistics" class="slide level2">
<h2>Data &amp; Sufficient Statistics</h2>
<ul>
<li>From Lecture 03, we have:</li>
<li><span class="math inline">n_A = 725</span> individuals with A phenotype</li>
<li><span class="math inline">n_{AB} = 72</span> individuals with AB phenotype</li>
<li><span class="math inline">n_{B} = 258</span> individuals with B phenotype</li>
<li><span class="math inline">n_{O} = 1073</span> individuals with O phenotype</li>
<li>Total sample size: <span class="math inline">N = 2128</span> individuals</li>
</ul>
</section>
<section id="model-specification" class="slide level2">
<h2>Model Specification</h2>
<ul>
<li>Allele frequency vector: <span class="math inline">\boldsymbol p=(p_A,p_B,p_O)</span>, <span class="math inline">\boldsymbol p\sim\text{Dirichlet}(\boldsymbol\alpha)</span>.</li>
<li>Under HWE, genotype frequencies: <span class="math inline">p_A^2, 2p_A p_O, p_B^2, 2 p_B p_O, 2 p_A p_B, p_O^2</span>.</li>
<li>Phenotype probabilities (aggregating ambiguous genotypes):</li>
<li><span class="math inline">P(\text{A}) = p_A^2 + 2 p_A p_O</span></li>
<li><span class="math inline">P(\text{AB}) = 2 p_A p_B</span></li>
<li><span class="math inline">P(\text{B}) = p_B^2 + 2 p_B p_O</span></li>
<li><span class="math inline">P(\text{O}) = p_O^2</span></li>
<li>Likelihood: <span class="math inline">(n_A,n_{AB},n_B,n_O) \sim \text{Multinomial}(N, \boldsymbol q)</span> with <span class="math inline">\boldsymbol q</span> above.</li>
</ul>
</section>
<section id="prior-families" class="slide level2">
<h2>Prior Families</h2>
<ul>
<li>Weak Dirichlet(1,1,1) (uniform over allele simplex)</li>
<li>Mild Dirichlet(2,2,2) (light shrink toward center)</li>
</ul>
</section>
<section id="using-historical-data" class="slide level2">
<h2>Using Historical Data</h2>
<ul>
<li>Consider global survey means <span class="math inline">(p_A = 0.26, p_B = 0.09, p_O = 0.65)</span> <span class="citation" data-cites="mourant1976 yamamoto2012abo">(<a href="#/reference" role="doc-biblioref" onclick="">Mourant <em>et al.</em>, 1976</a>; <a href="#/reference" role="doc-biblioref" onclick="">Yamamoto <em>et al.</em>, 2012</a>)</span>.</li>
<li>Dirichlet prior: <span class="math inline">\mathbf{\alpha} = k (0.26, 0.09, 0.65)</span></li>
<li>Effective sample size idea: acts like observing k allele draws before current data (2N alleles in sample).</li>
<li>Relative weight vs data (here total alleles = 2N = 4256):</li>
<li><span class="math inline">k = 200 \to</span> prior weight <span class="math inline">\approx 4.5\%</span> of total information.</li>
<li><span class="math inline">k = 1000 \to</span> prior weight <span class="math inline">\approx 19.0\%</span> of total information.</li>
</ul>
</section>
<section id="prior-shapes-marginal-densities" class="slide level2">
<h2>Prior Shapes (Marginal Densities)</h2>

<img data-src="lecture-04_files/figure-revealjs/unnamed-chunk-3-1.png" width="960" class="r-stretch"></section>
<section id="how-to-fit-these-models" class="slide level2">
<h2>How to fit these models?</h2>
<div class="quarto-layout-panel" data-layout="[80, -1, 20]">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="quarto-layout-cell" style="flex-basis: 79.2%;justify-content: flex-start;">
<ul>
<li>So we have our data, likelihood, and priors</li>
<li>We can use MCMC to sample from the posterior distribution of allele frequencies</li>
<li>There exist many software packages to do this! We will use Stan <span class="citation" data-cites="carpenterStanProbabilisticProgramming2017">(<a href="#/reference" role="doc-biblioref" onclick="">Carpenter <em>et al.</em>, 2017</a>)</span> via the <code>cmdstanr</code> R package</li>
</ul>
</div>
<div class="quarto-figure-spacer quarto-layout-cell" style="flex-basis: 1.0%;justify-content: flex-start;">
<p>&nbsp;</p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 19.8%;justify-content: center;">
<p><img data-src="https://mc-stan.org/img/logo_tm.png"></p>
</div>
</div>
</div>
</section>
<section id="stan-model" class="slide level2">
<h2>Stan Model</h2>
<h3 id="step-1-phenotype-probabilities">Step 1: Phenotype Probabilities</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="annotated-cell-1"><pre class="sourceCode numberSource stan code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode stan"><a class="code-annotation-anchor" data-target-cell="annotated-cell-1" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-1-1" class="code-annotation-target"><a href="#annotated-cell-1-1"></a><span class="kw">functions</span> {</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-1" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-1-2" class="code-annotation-target"><a href="#annotated-cell-1-2"></a>    <span class="dt">vector</span> abo_pheno_probs(<span class="dt">vector</span> p) {</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-1" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-1-3" class="code-annotation-target"><a href="#annotated-cell-1-3"></a>        <span class="dt">real</span> pA = p[<span class="dv">1</span>];</span>
<span id="annotated-cell-1-4"><a href="#annotated-cell-1-4"></a>        <span class="dt">real</span> pB = p[<span class="dv">2</span>];</span>
<span id="annotated-cell-1-5"><a href="#annotated-cell-1-5"></a>        <span class="dt">real</span> pO = p[<span class="dv">3</span>];</span>
<span id="annotated-cell-1-6"><a href="#annotated-cell-1-6"></a>        <span class="dt">vector</span>[<span class="dv">4</span>] q;</span>
<span id="annotated-cell-1-7"><a href="#annotated-cell-1-7"></a>        q[<span class="dv">1</span>] = pA * pA + <span class="dv">2</span> * pA * pO; <span class="co">// A phenotype</span></span>
<span id="annotated-cell-1-8"><a href="#annotated-cell-1-8"></a>        q[<span class="dv">2</span>] = <span class="dv">2</span> * pA * pB;            <span class="co">// AB phenotype</span></span>
<span id="annotated-cell-1-9"><a href="#annotated-cell-1-9"></a>        q[<span class="dv">3</span>] = pB * pB + <span class="dv">2</span> * pB * pO; <span class="co">// B phenotype</span></span>
<span id="annotated-cell-1-10"><a href="#annotated-cell-1-10"></a>        q[<span class="dv">4</span>] = pO * pO;               <span class="co">// O phenotype</span></span>
<span id="annotated-cell-1-11"><a href="#annotated-cell-1-11"></a>        <span class="cf">return</span> q;</span>
<span id="annotated-cell-1-12"><a href="#annotated-cell-1-12"></a>        }                                            </span>
<span id="annotated-cell-1-13"><a href="#annotated-cell-1-13"></a>}</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-1" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-1" data-code-lines="1,13" data-code-annotation="1">Declare the <code>functions</code> block (optional in Stan, but lets us encapsulate logic)</span>
</dd>
<dt data-target-cell="annotated-cell-1" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-1" data-code-lines="2" data-code-annotation="2">Define a helper that maps allele frequencies to phenotype probabilities</span>
</dd>
<dt data-target-cell="annotated-cell-1" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-1" data-code-lines="3,4,5" data-code-annotation="3">Extract components (for readability in later expressions)</span>
</dd>
</dl>
</section>
<section id="stan-model-1" class="slide level2">
<h2>Stan Model</h2>
<h3 id="step-1-phenotype-probabilities-1">Step 1: Phenotype Probabilities</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="annotated-cell-2"><pre class="sourceCode numberSource stan code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode stan"><span id="annotated-cell-2-1"><a href="#annotated-cell-2-1"></a><span class="kw">functions</span> {</span>
<span id="annotated-cell-2-2"><a href="#annotated-cell-2-2"></a>    <span class="dt">vector</span> abo_pheno_probs(<span class="dt">vector</span> p) {</span>
<span id="annotated-cell-2-3"><a href="#annotated-cell-2-3"></a>        <span class="dt">real</span> pA = p[<span class="dv">1</span>];</span>
<span id="annotated-cell-2-4"><a href="#annotated-cell-2-4"></a>        <span class="dt">real</span> pB = p[<span class="dv">2</span>];</span>
<span id="annotated-cell-2-5"><a href="#annotated-cell-2-5"></a>        <span class="dt">real</span> pO = p[<span class="dv">3</span>];</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-2-6" class="code-annotation-target"><a href="#annotated-cell-2-6"></a>        <span class="dt">vector</span>[<span class="dv">4</span>] q;</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="5" onclick="event.preventDefault();">5</a><span id="annotated-cell-2-7" class="code-annotation-target"><a href="#annotated-cell-2-7"></a>        q[<span class="dv">1</span>] = pA * pA + <span class="dv">2</span> * pA * pO; <span class="co">// A phenotype</span></span>
<span id="annotated-cell-2-8"><a href="#annotated-cell-2-8"></a>        q[<span class="dv">2</span>] = <span class="dv">2</span> * pA * pB;            <span class="co">// AB phenotype</span></span>
<span id="annotated-cell-2-9"><a href="#annotated-cell-2-9"></a>        q[<span class="dv">3</span>] = pB * pB + <span class="dv">2</span> * pB * pO; <span class="co">// B phenotype</span></span>
<span id="annotated-cell-2-10"><a href="#annotated-cell-2-10"></a>        q[<span class="dv">4</span>] = pO * pO;               <span class="co">// O phenotype</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="6" onclick="event.preventDefault();">6</a><span id="annotated-cell-2-11" class="code-annotation-target"><a href="#annotated-cell-2-11"></a>        <span class="cf">return</span> q;</span>
<span id="annotated-cell-2-12"><a href="#annotated-cell-2-12"></a>        }                                            </span>
<span id="annotated-cell-2-13"><a href="#annotated-cell-2-13"></a>}</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-2" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="6" data-code-annotation="4">Allocate a length-4 vector <code>q</code> (A, AB, B, O) for phenotype probabilities</span>
</dd>
<dt data-target-cell="annotated-cell-2" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="7,8,9,10" data-code-annotation="5">Hardy–Weinberg genotype algebra aggregated into phenotype probabilities</span>
</dd>
<dt data-target-cell="annotated-cell-2" data-target-annotation="6">6</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="11" data-code-annotation="6">Return the vector</span>
</dd>
</dl>
</section>
<section id="stan-model-2" class="slide level2">
<h2>Stan Model</h2>
<h3 id="step-2-data-transforms">Step 2: Data &amp; Transforms</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="annotated-cell-3"><pre class="sourceCode numberSource stan code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode stan"><a class="code-annotation-anchor" data-target-cell="annotated-cell-3" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-3-1" class="code-annotation-target"><a href="#annotated-cell-3-1"></a><span class="kw">data</span> {</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-3" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-3-2" class="code-annotation-target"><a href="#annotated-cell-3-2"></a>    <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; nA;</span>
<span id="annotated-cell-3-3"><a href="#annotated-cell-3-3"></a>    <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; nAB;</span>
<span id="annotated-cell-3-4"><a href="#annotated-cell-3-4"></a>    <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; nB;</span>
<span id="annotated-cell-3-5"><a href="#annotated-cell-3-5"></a>    <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; nO;</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-3" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-3-6" class="code-annotation-target"><a href="#annotated-cell-3-6"></a>    <span class="dt">vector</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt;[<span class="dv">3</span>] alpha; <span class="co"># Dirichlet hyperparameters</span></span>
<span id="annotated-cell-3-7"><a href="#annotated-cell-3-7"></a>}</span>
<span id="annotated-cell-3-8"><a href="#annotated-cell-3-8"></a></span>
<span id="annotated-cell-3-9"><a href="#annotated-cell-3-9"></a><span class="kw">transformed data</span> {</span>
<span id="annotated-cell-3-10"><a href="#annotated-cell-3-10"></a>    <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; N = nA + nAB + nB + nO;</span>
<span id="annotated-cell-3-11"><a href="#annotated-cell-3-11"></a>    <span class="dt">array</span>[<span class="dv">4</span>] <span class="dt">int</span> y = { nA, nAB, nB, nO };</span>
<span id="annotated-cell-3-12"><a href="#annotated-cell-3-12"></a>}</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-3" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-3" data-code-lines="1,7" data-code-annotation="1">Raw observed counts and prior hyperparameters enter in the <code>data</code> block.</span>
</dd>
<dt data-target-cell="annotated-cell-3" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-3" data-code-lines="2,3,4,5" data-code-annotation="2">ABO phenotype counts (non-negative integers) per category.</span>
</dd>
<dt data-target-cell="annotated-cell-3" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-3" data-code-lines="6" data-code-annotation="3">Dirichlet prior parameters supplied from R (allow different priors via <code>alpha</code>).</span>
</dd>
</dl>
</section>
<section id="stan-model-3" class="slide level2">
<h2>Stan Model</h2>
<h3 id="step-2-data-transforms-1">Step 2: Data &amp; Transforms</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="annotated-cell-4"><pre class="sourceCode numberSource stan code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode stan"><span id="annotated-cell-4-1"><a href="#annotated-cell-4-1"></a><span class="kw">data</span> {</span>
<span id="annotated-cell-4-2"><a href="#annotated-cell-4-2"></a>    <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; nA;</span>
<span id="annotated-cell-4-3"><a href="#annotated-cell-4-3"></a>    <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; nAB;</span>
<span id="annotated-cell-4-4"><a href="#annotated-cell-4-4"></a>    <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; nB;</span>
<span id="annotated-cell-4-5"><a href="#annotated-cell-4-5"></a>    <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; nO;</span>
<span id="annotated-cell-4-6"><a href="#annotated-cell-4-6"></a>    <span class="dt">vector</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt;[<span class="dv">3</span>] alpha; <span class="co"># Dirichlet hyperparameters</span></span>
<span id="annotated-cell-4-7"><a href="#annotated-cell-4-7"></a>}</span>
<span id="annotated-cell-4-8"><a href="#annotated-cell-4-8"></a></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-4-9" class="code-annotation-target"><a href="#annotated-cell-4-9"></a><span class="kw">transformed data</span> {</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="5" onclick="event.preventDefault();">5</a><span id="annotated-cell-4-10" class="code-annotation-target"><a href="#annotated-cell-4-10"></a>    <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; N = nA + nAB + nB + nO;</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="6" onclick="event.preventDefault();">6</a><span id="annotated-cell-4-11" class="code-annotation-target"><a href="#annotated-cell-4-11"></a>    <span class="dt">array</span>[<span class="dv">4</span>] <span class="dt">int</span> y = { nA, nAB, nB, nO };</span>
<span id="annotated-cell-4-12"><a href="#annotated-cell-4-12"></a>}</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-4" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="9,12" data-code-annotation="4"><code>transformed data</code> pre-computes deterministic quantities once</span>
</dd>
<dt data-target-cell="annotated-cell-4" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="10" data-code-annotation="5">Total sample size N used for reference or diagnostics</span>
</dd>
<dt data-target-cell="annotated-cell-4" data-target-annotation="6">6</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="11" data-code-annotation="6">Assemble counts into an array to pass to the multinomial</span>
</dd>
</dl>
</section>
<section id="stan-model-4" class="slide level2">
<h2>Stan Model</h2>
<h3 id="step-3-parameters-derived-q">Step 3: Parameters &amp; Derived <code>q</code></h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="annotated-cell-5"><pre class="sourceCode numberSource stan code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode stan"><a class="code-annotation-anchor" data-target-cell="annotated-cell-5" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-5-1" class="code-annotation-target"><a href="#annotated-cell-5-1"></a><span class="kw">parameters</span> {</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-5" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-5-2" class="code-annotation-target"><a href="#annotated-cell-5-2"></a>    <span class="dt">simplex</span>[<span class="dv">3</span>] p;  <span class="co"># (pA, pB, pO) on 2-simplex</span></span>
<span id="annotated-cell-5-3"><a href="#annotated-cell-5-3"></a>}</span>
<span id="annotated-cell-5-4"><a href="#annotated-cell-5-4"></a></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-5" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-5-5" class="code-annotation-target"><a href="#annotated-cell-5-5"></a><span class="kw">transformed parameters</span> {</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-5" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-5-6" class="code-annotation-target"><a href="#annotated-cell-5-6"></a>    <span class="dt">vector</span>[<span class="dv">4</span>] q = abo_pheno_probs(p);</span>
<span id="annotated-cell-5-7"><a href="#annotated-cell-5-7"></a>}</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-5" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-5" data-code-lines="1,3" data-code-annotation="1">Declare unknown quantities to infer in <code>parameters</code></span>
</dd>
<dt data-target-cell="annotated-cell-5" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-5" data-code-lines="2" data-code-annotation="2"><code>simplex[3]</code> enforces positivity and sum-to-one constraints automatically</span>
</dd>
<dt data-target-cell="annotated-cell-5" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-5" data-code-lines="5,7" data-code-annotation="3"><code>transformed parameters</code> recomputes per-draw derived values</span>
</dd>
<dt data-target-cell="annotated-cell-5" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-5" data-code-lines="6" data-code-annotation="4">Reuse helper to obtain phenotype probabilities from allele frequencies</span>
</dd>
</dl>
</section>
<section id="stan-model-step-4-prior-likelihood" class="slide level2">
<h2>Stan Model (Step 4: Prior &amp; Likelihood)</h2>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="annotated-cell-6"><pre class="sourceCode numberSource stan code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode stan"><a class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-6-1" class="code-annotation-target"><a href="#annotated-cell-6-1"></a><span class="kw">model</span> {</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-6-2" class="code-annotation-target"><a href="#annotated-cell-6-2"></a>    p ~ dirichlet(alpha);</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-6-3" class="code-annotation-target"><a href="#annotated-cell-6-3"></a>    y ~ multinomial(q);</span>
<span id="annotated-cell-6-4"><a href="#annotated-cell-6-4"></a>}</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-6" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="1,4" data-code-annotation="1">The <code>model</code> block contains all sampling statements contributing to log density</span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="2" data-code-annotation="2">Dirichlet prior on allele frequencies <code>p</code></span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="3" data-code-annotation="3">Multinomial likelihood over phenotype counts with probabilities <code>q</code>.</span>
</dd>
</dl>
</section>
<section id="stan-model-step-5-generated-quantities" class="slide level2">
<h2>Stan Model (Step 5: Generated Quantities)</h2>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="annotated-cell-7"><pre class="sourceCode numberSource stan code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode stan"><a class="code-annotation-anchor" data-target-cell="annotated-cell-7" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-7-1" class="code-annotation-target"><a href="#annotated-cell-7-1"></a><span class="kw">generated quantities</span> {</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-7" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-7-2" class="code-annotation-target"><a href="#annotated-cell-7-2"></a>    <span class="dt">real</span> log_lik = multinomial_lpmf(y | q);</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-7" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-7-3" class="code-annotation-target"><a href="#annotated-cell-7-3"></a>    <span class="dt">vector</span>[<span class="dv">6</span>] geno_freq;</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-7" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-7-4" class="code-annotation-target"><a href="#annotated-cell-7-4"></a>    <span class="dt">real</span> pA = p[<span class="dv">1</span>]; <span class="dt">real</span> pB = p[<span class="dv">2</span>]; <span class="dt">real</span> pO = p[<span class="dv">3</span>];</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-7" data-target-annotation="5" onclick="event.preventDefault();">5</a><span id="annotated-cell-7-5" class="code-annotation-target"><a href="#annotated-cell-7-5"></a>    geno_freq[<span class="dv">1</span>] = pA * pA;        <span class="co"># AA</span></span>
<span id="annotated-cell-7-6"><a href="#annotated-cell-7-6"></a>    geno_freq[<span class="dv">2</span>] = <span class="dv">2</span> * pA * pO;    <span class="co"># AO</span></span>
<span id="annotated-cell-7-7"><a href="#annotated-cell-7-7"></a>    geno_freq[<span class="dv">3</span>] = pB * pB;        <span class="co"># BB</span></span>
<span id="annotated-cell-7-8"><a href="#annotated-cell-7-8"></a>    geno_freq[<span class="dv">4</span>] = <span class="dv">2</span> * pB * pO;    <span class="co"># BO</span></span>
<span id="annotated-cell-7-9"><a href="#annotated-cell-7-9"></a>    geno_freq[<span class="dv">5</span>] = <span class="dv">2</span> * pA * pB;    <span class="co"># AB</span></span>
<span id="annotated-cell-7-10"><a href="#annotated-cell-7-10"></a>    geno_freq[<span class="dv">6</span>] = pO * pO;        <span class="co"># OO</span></span>
<span id="annotated-cell-7-11"><a href="#annotated-cell-7-11"></a>}</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-7" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-7" data-code-lines="1,11" data-code-annotation="1">Post-processing: quantities saved per posterior draw</span>
</dd>
<dt data-target-cell="annotated-cell-7" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-7" data-code-lines="2" data-code-annotation="2">Store log-likelihood for model comparison / LOO / WAIC</span>
</dd>
<dt data-target-cell="annotated-cell-7" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-7" data-code-lines="3" data-code-annotation="3">Allocate genotype frequency vector</span>
</dd>
<dt data-target-cell="annotated-cell-7" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-7" data-code-lines="4" data-code-annotation="4">Local aliases improve clarity when computing genotype frequencies</span>
</dd>
<dt data-target-cell="annotated-cell-7" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-7" data-code-lines="5,6,7,8,9,10" data-code-annotation="5">Hardy–Weinberg genotype probabilities</span>
</dd>
</dl>
</section>
<section id="compile-stan-model" class="slide level2">
<h2>Compile Stan Model</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a></a>mod_path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"stan"</span>, <span class="st">"abo_multinomial.stan"</span>)</span>
<span id="cb1-2"><a></a>abo_mod <span class="ot">&lt;-</span> <span class="fu">cmdstan_model</span>(mod_path)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<ul>
<li>Stan uses C++ on the backend, so we need to compile the model once before fitting</li>
<li><code>cmdstan_model()</code> handles compilation and returns a model object for sampling</li>
</ul>
</section>
<section id="fit-weak-prior-dirichlet111" class="slide level2">
<h2>Fit: Weak Prior (Dirichlet(1,1,1))</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-9"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><span id="annotated-cell-9-1"><a href="#annotated-cell-9-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">nA =</span> <span class="dv">725</span>, <span class="at">nAB =</span> <span class="dv">72</span>, <span class="at">nB =</span> <span class="dv">258</span>, <span class="at">nO =</span> <span class="dv">1073</span>)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-9" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-9-2" class="code-annotation-target"><a href="#annotated-cell-9-2"></a>fit_weak <span class="ot">&lt;-</span> abo_mod<span class="sc">$</span><span class="fu">sample</span>(</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-9" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-9-3" class="code-annotation-target"><a href="#annotated-cell-9-3"></a>    <span class="at">data =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-9-4"><a href="#annotated-cell-9-4"></a>        <span class="at">nA =</span> counts<span class="sc">$</span>nA, <span class="at">nAB =</span> counts<span class="sc">$</span>nAB, <span class="at">nB =</span> counts<span class="sc">$</span>nB, <span class="at">nO =</span> counts<span class="sc">$</span>nO,</span>
<span id="annotated-cell-9-5"><a href="#annotated-cell-9-5"></a>        <span class="at">alpha =</span> <span class="fu">rep</span>(<span class="dv">1</span>, <span class="dv">3</span>)</span>
<span id="annotated-cell-9-6"><a href="#annotated-cell-9-6"></a>    ),</span>
<span id="annotated-cell-9-7"><a href="#annotated-cell-9-7"></a>    <span class="at">seed =</span> <span class="dv">8878</span>,</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-9" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-9-8" class="code-annotation-target"><a href="#annotated-cell-9-8"></a>    <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">parallel_chains =</span> <span class="dv">4</span>,</span>
<span id="annotated-cell-9-9"><a href="#annotated-cell-9-9"></a>    <span class="at">iter_warmup =</span> <span class="dv">1000</span>, <span class="at">iter_sampling =</span> <span class="dv">1000</span>,</span>
<span id="annotated-cell-9-10"><a href="#annotated-cell-9-10"></a>    <span class="at">refresh =</span> <span class="dv">0</span></span>
<span id="annotated-cell-9-11"><a href="#annotated-cell-9-11"></a>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-9" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-9" data-code-lines="2" data-code-annotation="1">Execute the MCMC sampler using the <code>sample()</code> method on the compiled model object</span>
</dd>
<dt data-target-cell="annotated-cell-9" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-9" data-code-lines="3,4,5,6" data-code-annotation="2">Pass data from R to Stan as a named list</span>
</dd>
<dt data-target-cell="annotated-cell-9" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-9" data-code-lines="8,9,10" data-code-annotation="3">Specify MCMC sampler settings</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Running MCMC with 4 parallel chains...

Chain 1 finished in 0.0 seconds.
Chain 2 finished in 0.0 seconds.
Chain 3 finished in 0.0 seconds.
Chain 4 finished in 0.0 seconds.

All 4 chains finished successfully.
Mean chain execution time: 0.0 seconds.
Total execution time: 0.2 seconds.</code></pre>
</div>
</div>
</section>
<section id="model-diagnostics" class="slide level2">
<h2>Model Diagnostics</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a></a>fit_weak<span class="sc">$</span><span class="fu">summary</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 18 × 10
   variable       mean   median      sd     mad       q5      q95  rhat ess_bulk
   &lt;chr&gt;         &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;
 1 lp__       -2.31e+3 -2.31e+3 9.83e-1 7.26e-1 -2.31e+3 -2.31e+3  1.00    1880.
 2 p[1]        2.09e-1  2.09e-1 6.60e-3 6.52e-3  1.99e-1  2.20e-1  1.00    4495.
 3 p[2]        8.10e-2  8.09e-2 4.23e-3 4.26e-3  7.41e-2  8.81e-2  1.00    2125.
 4 p[3]        7.10e-1  7.10e-1 7.27e-3 7.18e-3  6.98e-1  7.22e-1  1.00    3368.
 5 q[1]        3.41e-1  3.41e-1 9.80e-3 9.84e-3  3.25e-1  3.57e-1  1.00    4376.
 6 q[2]        3.39e-2  3.39e-2 1.92e-3 1.92e-3  3.08e-2  3.70e-2  1.00    2338.
 7 q[3]        1.22e-1  1.21e-1 6.27e-3 6.21e-3  1.11e-1  1.32e-1  1.00    2210.
 8 q[4]        5.04e-1  5.04e-1 1.03e-2 1.02e-2  4.87e-1  5.21e-1  1.00    3368.
 9 log_lik    -1.16e+1 -1.13e+1 9.81e-1 7.14e-1 -1.36e+1 -1.07e+1  1.00    1867.
10 geno_freq…  4.38e-2  4.37e-2 2.76e-3 2.73e-3  3.94e-2  4.85e-2  1.00    4495.
11 geno_freq…  2.97e-1  2.97e-1 7.10e-3 7.08e-3  2.85e-1  3.08e-1  1.00    4242.
12 geno_freq…  6.58e-3  6.55e-3 6.87e-4 6.90e-4  5.50e-3  7.76e-3  1.00    2125.
13 geno_freq…  1.15e-1  1.15e-1 5.59e-3 5.53e-3  1.06e-1  1.24e-1  1.00    2225.
14 geno_freq…  3.39e-2  3.39e-2 1.92e-3 1.92e-3  3.08e-2  3.70e-2  1.00    2338.
15 geno_freq…  5.04e-1  5.04e-1 1.03e-2 1.02e-2  4.87e-1  5.21e-1  1.00    3368.
16 pA          2.09e-1  2.09e-1 6.60e-3 6.52e-3  1.99e-1  2.20e-1  1.00    4495.
17 pB          8.10e-2  8.09e-2 4.23e-3 4.26e-3  7.41e-2  8.81e-2  1.00    2125.
18 pO          7.10e-1  7.10e-1 7.27e-3 7.18e-3  6.98e-1  7.22e-1  1.00    3368.
# ℹ 1 more variable: ess_tail &lt;dbl&gt;</code></pre>
</div>
</div>
</section>
<section id="posterior-summaries-intervals-weak-prior" class="slide level2">
<h2>Posterior Summaries &amp; Intervals (Weak Prior)</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a></a>post_weak <span class="ot">&lt;-</span> fit_weak<span class="sc">$</span><span class="fu">draws</span>(<span class="at">variables =</span> <span class="fu">c</span>(<span class="st">"p"</span>)) <span class="sc">|&gt;</span> <span class="fu">as_draws_df</span>()</span>
<span id="cb5-2"><a></a>weak_summ <span class="ot">&lt;-</span> post_weak <span class="sc">%&gt;%</span> <span class="fu">summarise</span>(</span>
<span id="cb5-3"><a></a>    <span class="at">mean_pA =</span> <span class="fu">mean</span>(<span class="st">`</span><span class="at">p[1]</span><span class="st">`</span>), <span class="at">mean_pB =</span> <span class="fu">mean</span>(<span class="st">`</span><span class="at">p[2]</span><span class="st">`</span>), <span class="at">mean_pO =</span> <span class="fu">mean</span>(<span class="st">`</span><span class="at">p[3]</span><span class="st">`</span>),</span>
<span id="cb5-4"><a></a>    <span class="at">sd_pA =</span> <span class="fu">sd</span>(<span class="st">`</span><span class="at">p[1]</span><span class="st">`</span>), <span class="at">sd_pB =</span> <span class="fu">sd</span>(<span class="st">`</span><span class="at">p[2]</span><span class="st">`</span>), <span class="at">sd_pO =</span> <span class="fu">sd</span>(<span class="st">`</span><span class="at">p[3]</span><span class="st">`</span>)</span>
<span id="cb5-5"><a></a>)</span>
<span id="cb5-6"><a></a>weak_ci <span class="ot">&lt;-</span> post_weak <span class="sc">%&gt;%</span> <span class="fu">summarise</span>(</span>
<span id="cb5-7"><a></a>    <span class="at">pA_low =</span> <span class="fu">quantile</span>(<span class="st">`</span><span class="at">p[1]</span><span class="st">`</span>, <span class="fl">0.025</span>), <span class="at">pA_high =</span> <span class="fu">quantile</span>(<span class="st">`</span><span class="at">p[1]</span><span class="st">`</span>, <span class="fl">0.975</span>),</span>
<span id="cb5-8"><a></a>    <span class="at">pB_low =</span> <span class="fu">quantile</span>(<span class="st">`</span><span class="at">p[2]</span><span class="st">`</span>, <span class="fl">0.025</span>), <span class="at">pB_high =</span> <span class="fu">quantile</span>(<span class="st">`</span><span class="at">p[2]</span><span class="st">`</span>, <span class="fl">0.975</span>),</span>
<span id="cb5-9"><a></a>    <span class="at">pO_low =</span> <span class="fu">quantile</span>(<span class="st">`</span><span class="at">p[3]</span><span class="st">`</span>, <span class="fl">0.025</span>), <span class="at">pO_high =</span> <span class="fu">quantile</span>(<span class="st">`</span><span class="at">p[3]</span><span class="st">`</span>, <span class="fl">0.975</span>)</span>
<span id="cb5-10"><a></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="posterior-histograms-weak-prior" class="slide level2">
<h2>Posterior Histograms (Weak Prior)</h2>

<img data-src="lecture-04_files/figure-revealjs/unnamed-chunk-8-1.png" width="960" class="r-stretch"></section>
<section id="em-vs-bayesian-point-estimates" class="slide level2">
<h2>EM vs Bayesian Point Estimates</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 5
  method            pA     pB    pO loglik
  &lt;chr&gt;          &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;
1 EM             0.209 0.0808 0.710 -2304.
2 Posterior Mean 0.209 0.0810 0.710    NA </code></pre>
</div>
</div>
</section>
<section id="consolidated-posterior-comparison" class="slide level2">
<h2>Consolidated Posterior Comparison</h2>

<img data-src="lecture-04_files/figure-revealjs/unnamed-chunk-11-1.png" width="960" class="r-stretch"></section></section>
<section>
<section id="population-structure" class="title-slide slide level1 center">
<h1>Population Structure</h1>

</section>
<section id="population-substructure" class="slide level2">
<h2>Population Substructure</h2>
<ul>
<li>Features of a population which result from variation of expected allele frequencies across individuals</li>
<li>Standard allele counting (<span class="math inline">\hat{p} = (2n_{AA} + n_{Aa}/2n)</span>) will still be unbiased</li>
<li>But, not all subjects may have the same probability of being represented in the sample</li>
<li>Variance of estimate will be effected</li>
</ul>
</section>
<section id="population-stratification" class="slide level2">
<h2>Population Stratification</h2>
<ul>
<li>Individuals in a population can be subdivided into mutually exclusive strata</li>
<li>Within each strata the allele frequency is the same for all individuals</li>
<li>Intuitively, we are partitioning a large dataset into multiple smaller datasets</li>
</ul>
</section>
<section id="population-admixture" class="slide level2">
<h2>Population Admixture</h2>
<ul>
<li>When individuals in a population have a mixture of different genetic ancestries due to prior mixing of two or more populations</li>
<li>Often result of migration</li>
</ul>
</section>
<section id="population-admixture-1" class="slide level2">
<h2>Population Admixture</h2>

<img data-src="images/admixture.png" class="r-stretch quarto-figure-center"><p class="caption">From <span class="citation" data-cites="korunesHumanGeneticAdmixture2021">Korunes and Goldberg (<a href="#/reference" role="doc-biblioref" onclick="">2021</a>)</span></p></section>
<section id="population-inbreeding" class="slide level2">
<h2>Population Inbreeding</h2>
<ul>
<li>Occurs when there is a preference for mating among relatives in a population or because geographic isolation of subgroups restricts mating choices</li>
<li>Possibility that an offspring will inherit two copies of the same ancestral allele</li>
<li>Define <span class="math inline">F</span>, the inbreeding coefficient, as the probability that a random individual in the population inherits two copies of the same allele from a common ancestor</li>
</ul>
</section>
<section id="admixture-as-a-confounder" class="slide level2">
<h2>Admixture as a Confounder</h2>
<ul>
<li>Consider the problem of estimating the effect of a SNP on a disease phenotype: <span class="math inline">\beta</span> in <span class="math inline">P(Y=1) = \text{logit}^{-1}(\alpha + \beta G)</span></li>
<li>Recent admixture mixes ancestries within individuals: genotype is a convex combination of source populations</li>
<li>If phenotype prevalence differs by ancestry, local or global ancestry proportions act like hidden covariates</li>
<li>Association tests must separate causal signal from ancestry-driven allele frequency differences</li>
</ul>
</section>
<section id="principal-components-for-structure" class="slide level2">
<h2>Principal Components for Structure</h2>
<ul>
<li>Construct the standardized genotype matrix <span class="math inline">Z</span> and compute <span class="math inline">Z^T Z / M</span> (with <span class="math inline">M</span> markers)</li>
<li>Top eigenvectors capture major ancestry gradients</li>
<li>Use the leading PCs as covariates in association tests or to stratify downstream analyses</li>
</ul>
</section>
<section id="principal-components-for-structure-1" class="slide level2">
<h2>Principal Components for Structure</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a></a><span class="fu">data</span>(<span class="st">"eHGDP"</span>)</span>
<span id="cb7-2"><a></a>eHGDP</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>/// GENIND OBJECT /////////

 // 1,350 individuals; 678 loci; 8,170 alleles; size: 44.1 Mb

 // Basic content
   @tab:  1350 x 8170 matrix of allele counts
   @loc.n.all: number of alleles per locus (range: 5-35)
   @loc.fac: locus factor for the 8170 columns of @tab
   @all.names: list of allele names for each locus
   @ploidy: ploidy of each individual  (range: 2-2)
   @type:  codom
   @call: read.fstat(file = file, missing = missing, quiet = quiet)

 // Optional content
   @pop: population of each individual (group size range: 3-50)
   @other: a list containing: popInfo </code></pre>
</div>
</div>
</section>
<section id="principal-components-for-structure-2" class="slide level2">
<h2>Principal Components for Structure</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a></a>hgdp_df <span class="ot">&lt;-</span> <span class="fu">genind2df</span>(eHGDP, <span class="at">sep =</span> <span class="st">"/"</span>)</span>
<span id="cb9-2"><a></a></span>
<span id="cb9-3"><a></a><span class="co"># Convert to allele count matrix, center columns (replace missing with locus means)</span></span>
<span id="cb9-4"><a></a>geno_mat <span class="ot">&lt;-</span> <span class="fu">scaleGen</span>(eHGDP, <span class="at">center =</span> <span class="cn">TRUE</span>, <span class="at">scale =</span> <span class="cn">FALSE</span>, <span class="at">NA.method =</span> <span class="st">"mean"</span>)</span>
<span id="cb9-5"><a></a>pc_fit <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(geno_mat, <span class="at">center =</span> <span class="cn">FALSE</span>, <span class="at">scale. =</span> <span class="cn">FALSE</span>)</span>
<span id="cb9-6"><a></a></span>
<span id="cb9-7"><a></a><span class="co"># Map individuals to geographic regions for coloring</span></span>
<span id="cb9-8"><a></a>pop_info <span class="ot">&lt;-</span> eHGDP<span class="sc">@</span>other<span class="sc">$</span>popInfo</span>
<span id="cb9-9"><a></a>pop_index <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(<span class="fu">pop</span>(eHGDP))</span>
<span id="cb9-10"><a></a>region <span class="ot">&lt;-</span> pop_info<span class="sc">$</span>Region[pop_index]</span>
<span id="cb9-11"><a></a></span>
<span id="cb9-12"><a></a>plot_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">PC1 =</span> pc_fit<span class="sc">$</span>x[, <span class="dv">1</span>], <span class="at">PC2 =</span> pc_fit<span class="sc">$</span>x[, <span class="dv">2</span>], <span class="at">Region =</span> region)</span>
<span id="cb9-13"><a></a><span class="fu">ggplot</span>(plot_df, <span class="fu">aes</span>(PC1, PC2, <span class="at">color =</span> Region)) <span class="sc">+</span></span>
<span id="cb9-14"><a></a>    <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.7</span>, <span class="at">size =</span> <span class="fl">1.5</span>) <span class="sc">+</span></span>
<span id="cb9-15"><a></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Population structure diagnostic"</span>, <span class="at">x =</span> <span class="st">"PC1"</span>, <span class="at">y =</span> <span class="st">"PC2"</span>) <span class="sc">+</span></span>
<span id="cb9-16"><a></a>    <span class="fu">theme_minimal</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="principal-components-for-structure-3" class="slide level2">
<h2>Principal Components for Structure</h2>

<img data-src="lecture-04_files/figure-revealjs/unnamed-chunk-14-1.png" width="960" class="r-stretch"></section>
<section id="structure-admixture" class="slide level2">
<h2>STRUCTURE &amp; ADMIXTURE</h2>
<ul>
<li>Model-based clustering methods for population structure inference</li>
<li>Assume <span class="math inline">K</span> latent populations with distinct allele frequencies</li>
<li>Each individual has ancestry proportions <span class="math inline">\boldsymbol{\pi}_i</span> across <span class="math inline">K</span> populations</li>
<li>Each genotype is drawn from a mixture of population-specific allele frequencies</li>
<li>Use maximum likelihood (ADMIXTURE) or Bayesian inference (STRUCTURE) to estimate parameters</li>
</ul>
</section>
<section id="assumptions" class="slide level2">
<h2>Assumptions</h2>
<ul>
<li>Hardy-Weinberg equilibrium within each ancestral population</li>
<li>Linkage equilibrium between loci within each ancestral population</li>
<li>Loci are unlinked (or weakly linked)</li>
</ul>
</section>
<section id="example-output" class="slide level2">
<h2>Example Output</h2>

<img data-src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41437-022-00535-z/MediaObjects/41437_2022_535_Fig4_HTML.png" class="r-stretch"></section>
<section id="bayesian-admixture" class="slide level2">
<h2>Bayesian Admixture</h2>
<ul>
<li>Latent populations <span class="math inline">k = 1,\ldots,K</span> possess allele frequencies <span class="math inline">\theta_{k\ell}</span> at locus <span class="math inline">\ell</span></li>
<li>Individual ancestry proportions <span class="math inline">\boldsymbol{\pi}_i \sim \text{Dirichlet}(\boldsymbol{\alpha})</span></li>
<li>Genotype <span class="math inline">y_{i\ell} \sim \text{Binomial}\left(2, \sum_{k} \pi_{ik} \theta_{k\ell}\right)</span> assuming HWE within each ancestral population</li>
<li>Posterior draws propagate ancestry/allele-frequency uncertainty into association testing, local ancestry, and polygenic prediction</li>
</ul>
</section>
<section id="identifiability-label-switching" class="slide level2">
<h2>Identifiability &amp; Label Switching</h2>
<ul>
<li>Problem: Mixture components are exchangeable → posterior multimodality</li>
<li>Solution: Anchor loci with informative priors break symmetry
<ul>
<li>Populations defined by genetic signatures, not arbitrary labels</li>
<li>Multiple anchors provide robustness against weak signals</li>
</ul></li>
</ul>
</section>
<section id="stan-model-data-parameters" class="slide level2">
<h2>Stan Model: Data &amp; Parameters</h2>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb10"><pre class="sourceCode numberSource stan number-lines code-with-copy"><code class="sourceCode stan"><span id="cb10-1"><a></a><span class="kw">data</span> {</span>
<span id="cb10-2"><a></a>  <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">1</span>&gt; N;                                <span class="co">// individuals</span></span>
<span id="cb10-3"><a></a>  <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">1</span>&gt; L;                                <span class="co">// loci</span></span>
<span id="cb10-4"><a></a>  <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">1</span>&gt; K;                                <span class="co">// ancestral pops</span></span>
<span id="cb10-5"><a></a>  <span class="dt">array</span>[N, L] <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>, <span class="kw">upper</span>=<span class="dv">2</span>&gt; y;           <span class="co">// genotypes (0,1,2)</span></span>
<span id="cb10-6"><a></a>  <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; L_soft;</span>
<span id="cb10-7"><a></a>  <span class="dt">array</span>[L_soft] <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">1</span>, <span class="kw">upper</span>=L&gt; soft_idx;  <span class="co">// e.g., {2}</span></span>
<span id="cb10-8"><a></a>  <span class="dt">array</span>[K] <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; a_theta;                <span class="co">// Beta 'a' for non-anchor loci</span></span>
<span id="cb10-9"><a></a>  <span class="dt">array</span>[K] <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; b_theta;                <span class="co">// Beta 'b' for non-anchor loci</span></span>
<span id="cb10-10"><a></a>  <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">1</span>, <span class="kw">upper</span>=L&gt; l_star;                  <span class="co">// anchor locus index</span></span>
<span id="cb10-11"><a></a>  <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; conc_pi;                         <span class="co">// shared Dirichlet conc. for pi</span></span>
<span id="cb10-12"><a></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="stan-model-parameters-transformed-parameters" class="slide level2">
<h2>Stan Model: Parameters &amp; Transformed Parameters</h2>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb11"><pre class="sourceCode numberSource stan number-lines code-with-copy"><code class="sourceCode stan"><span id="cb11-1"><a></a><span class="kw">parameters</span> {</span>
<span id="cb11-2"><a></a>  <span class="dt">array</span>[N] <span class="dt">simplex</span>[K] pi;                        <span class="co">// ancestry proportions</span></span>
<span id="cb11-3"><a></a>  <span class="dt">ordered</span>[K] eta;                                <span class="co">// ordered logits at anchor locus</span></span>
<span id="cb11-4"><a></a>  <span class="dt">matrix</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>, <span class="kw">upper</span>=<span class="dv">1</span>&gt;[K, L<span class="dv">-1</span>] theta_rest;   <span class="co">// allele freqs for non-anchor loci</span></span>
<span id="cb11-5"><a></a>}</span>
<span id="cb11-6"><a></a></span>
<span id="cb11-7"><a></a></span>
<span id="cb11-8"><a></a><span class="kw">transformed parameters</span> {</span>
<span id="cb11-9"><a></a>  <span class="dt">matrix</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>, <span class="kw">upper</span>=<span class="dv">1</span>&gt;[K, L] theta;          <span class="co">// full allele-frequency matrix</span></span>
<span id="cb11-10"><a></a>  <span class="dt">matrix</span>[N, L] p_mix;                            <span class="co">// mixed allele frequency</span></span>
<span id="cb11-11"><a></a></span>
<span id="cb11-12"><a></a>  <span class="co">// anchor column (ordered)</span></span>
<span id="cb11-13"><a></a>  <span class="cf">for</span> (k <span class="cf">in</span> <span class="dv">1</span>:K) theta[k, l_star] = inv_logit(eta[k]);</span>
<span id="cb11-14"><a></a></span>
<span id="cb11-15"><a></a>  <span class="co">// fill remaining columns</span></span>
<span id="cb11-16"><a></a>  {</span>
<span id="cb11-17"><a></a>    <span class="dt">int</span> c = <span class="dv">1</span>;</span>
<span id="cb11-18"><a></a>    <span class="cf">for</span> (l <span class="cf">in</span> <span class="dv">1</span>:L) {</span>
<span id="cb11-19"><a></a>      <span class="cf">if</span> (l == l_star) <span class="cf">continue</span>;</span>
<span id="cb11-20"><a></a>      <span class="cf">for</span> (k <span class="cf">in</span> <span class="dv">1</span>:K) theta[k, l] = theta_rest[k, c];</span>
<span id="cb11-21"><a></a>      c += <span class="dv">1</span>;</span>
<span id="cb11-22"><a></a>    }</span>
<span id="cb11-23"><a></a>  }</span>
<span id="cb11-24"><a></a></span>
<span id="cb11-25"><a></a>  <span class="co">// mixture expectations</span></span>
<span id="cb11-26"><a></a>  <span class="cf">for</span> (n <span class="cf">in</span> <span class="dv">1</span>:N)</span>
<span id="cb11-27"><a></a>    <span class="cf">for</span> (l <span class="cf">in</span> <span class="dv">1</span>:L)</span>
<span id="cb11-28"><a></a>      p_mix[n, l] = dot_product(pi[n], col(theta, l));</span>
<span id="cb11-29"><a></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="stan-model-priors-likelihood" class="slide level2">
<h2>Stan Model: Priors &amp; Likelihood</h2>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb12"><pre class="sourceCode numberSource stan number-lines code-with-copy"><code class="sourceCode stan"><span id="cb12-1"><a></a><span class="kw">model</span> {</span>
<span id="cb12-2"><a></a>  <span class="co">// priors</span></span>
<span id="cb12-3"><a></a>  eta ~ normal(<span class="dv">0</span>, <span class="fl">2.5</span>);                          <span class="co">// weak prior; ordering gives ID</span></span>
<span id="cb12-4"><a></a>  <span class="cf">for</span> (k <span class="cf">in</span> <span class="dv">1</span>:K)</span>
<span id="cb12-5"><a></a>    <span class="cf">for</span> (c <span class="cf">in</span> <span class="dv">1</span>:(L - <span class="dv">1</span>))</span>
<span id="cb12-6"><a></a>      theta_rest[k, c] ~ beta(a_theta[k], b_theta[k]);</span>
<span id="cb12-7"><a></a></span>
<span id="cb12-8"><a></a>  <span class="cf">for</span> (n <span class="cf">in</span> <span class="dv">1</span>:N)</span>
<span id="cb12-9"><a></a>    pi[n] ~ dirichlet(rep_vector(conc_pi, K));</span>
<span id="cb12-10"><a></a></span>
<span id="cb12-11"><a></a>  <span class="cf">for</span> (c <span class="cf">in</span> <span class="dv">1</span>:L_soft) {</span>
<span id="cb12-12"><a></a>  <span class="dt">int</span> l = soft_idx[c];</span>
<span id="cb12-13"><a></a>  <span class="cf">if</span> (l != l_star) {</span>
<span id="cb12-14"><a></a>    <span class="co">// comp 1 LOW, comp 2 HIGH at these loci (gentle)</span></span>
<span id="cb12-15"><a></a>    <span class="kw">target +=</span> beta_lpdf(theta[<span class="dv">1</span>, l] | <span class="dv">2</span>, <span class="dv">8</span>);</span>
<span id="cb12-16"><a></a>    <span class="kw">target +=</span> beta_lpdf(theta[<span class="dv">2</span>, l] | <span class="dv">8</span>, <span class="dv">2</span>);</span>
<span id="cb12-17"><a></a>  }</span>
<span id="cb12-18"><a></a>}</span>
<span id="cb12-19"><a></a></span>
<span id="cb12-20"><a></a>  <span class="co">// likelihood</span></span>
<span id="cb12-21"><a></a>  <span class="cf">for</span> (n <span class="cf">in</span> <span class="dv">1</span>:N)</span>
<span id="cb12-22"><a></a>    <span class="cf">for</span> (l <span class="cf">in</span> <span class="dv">1</span>:L)</span>
<span id="cb12-23"><a></a>      y[n, l] ~ binomial(<span class="dv">2</span>, p_mix[n, l]);</span>
<span id="cb12-24"><a></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="stan-model-generated-quantities" class="slide level2">
<h2>Stan Model: Generated Quantities</h2>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb13"><pre class="sourceCode numberSource stan number-lines code-with-copy"><code class="sourceCode stan"><span id="cb13-1"><a></a><span class="kw">generated quantities</span> {</span>
<span id="cb13-2"><a></a>  <span class="dt">matrix</span>[N, K] logit_pi;</span>
<span id="cb13-3"><a></a>  <span class="dt">array</span>[N, L] <span class="dt">int</span> y_rep;</span>
<span id="cb13-4"><a></a></span>
<span id="cb13-5"><a></a>  <span class="cf">for</span> (n <span class="cf">in</span> <span class="dv">1</span>:N)</span>
<span id="cb13-6"><a></a>    <span class="cf">for</span> (k <span class="cf">in</span> <span class="dv">1</span>:K)</span>
<span id="cb13-7"><a></a>      logit_pi[n, k] = logit(pi[n, k]);</span>
<span id="cb13-8"><a></a></span>
<span id="cb13-9"><a></a>  <span class="cf">for</span> (n <span class="cf">in</span> <span class="dv">1</span>:N)</span>
<span id="cb13-10"><a></a>    <span class="cf">for</span> (l <span class="cf">in</span> <span class="dv">1</span>:L)</span>
<span id="cb13-11"><a></a>      y_rep[n, l] = binomial_rng(<span class="dv">2</span>, p_mix[n, l]);</span>
<span id="cb13-12"><a></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="mathematical-framework" class="slide level2">
<h2>Mathematical Framework</h2>
<h3 id="individual-locus-allele-frequency">Individual-Locus Allele Frequency</h3>
<p>For individual <span class="math inline">i</span> at locus <span class="math inline">\ell</span>, expected allele frequency: <span class="math display">p_{i\ell} = \sum_{k=1}^K \pi_{ik} \theta_{k\ell}</span></p>
<h3 id="information-borrowing-across-loci">Information Borrowing Across Loci</h3>
<p><strong>Key insight:</strong> Same <span class="math inline">\boldsymbol{\pi}_i</span> parameters appear in likelihood for all loci</p>
<p><span class="math display">L(\boldsymbol{\pi}_i, \boldsymbol{\Theta}) = \prod_{\ell=1}^L \text{Binomial}\left(y_{i\ell} \mid 2, \sum_{k=1}^K \pi_{ik} \theta_{k\ell}\right)</span></p>
</section>
<section id="mathematical-framework-1" class="slide level2">
<h2>Mathematical Framework</h2>
<h3 id="hierarchical-learning">Hierarchical Learning</h3>
<ul>
<li>Anchor loci provide strong identification signal</li>
<li>Remaining loci contribute cumulative evidence</li>
<li>Posterior uncertainty propagates through all parameters</li>
</ul>
</section>
<section id="simulating-admixed-genotypes" class="slide level2">
<h2>Simulating Admixed Genotypes</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-18"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-18" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-18-1" class="code-annotation-target"><a href="#annotated-cell-18-1"></a><span class="fu">set.seed</span>(<span class="dv">887804</span>)</span>
<span id="annotated-cell-18-2"><a href="#annotated-cell-18-2"></a></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-18" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-18-3" class="code-annotation-target"><a href="#annotated-cell-18-3"></a>rdirichlet <span class="ot">&lt;-</span> <span class="cf">function</span>(n, alpha) {</span>
<span id="annotated-cell-18-4"><a href="#annotated-cell-18-4"></a>    k <span class="ot">&lt;-</span> <span class="fu">length</span>(alpha)</span>
<span id="annotated-cell-18-5"><a href="#annotated-cell-18-5"></a>    gamma_draws <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rgamma</span>(n <span class="sc">*</span> k, <span class="at">shape =</span> alpha, <span class="at">rate =</span> <span class="dv">1</span>),</span>
<span id="annotated-cell-18-6"><a href="#annotated-cell-18-6"></a>                          <span class="at">ncol =</span> k, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span>
<span id="annotated-cell-18-7"><a href="#annotated-cell-18-7"></a>    <span class="fu">sweep</span>(gamma_draws, <span class="dv">1</span>, <span class="fu">rowSums</span>(gamma_draws), <span class="st">"/"</span>)</span>
<span id="annotated-cell-18-8"><a href="#annotated-cell-18-8"></a>}</span>
<span id="annotated-cell-18-9"><a href="#annotated-cell-18-9"></a></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-18" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-18-10" class="code-annotation-target"><a href="#annotated-cell-18-10"></a>N <span class="ot">&lt;-</span> <span class="dv">40</span>   <span class="co"># individuals</span></span>
<span id="annotated-cell-18-11"><a href="#annotated-cell-18-11"></a>L <span class="ot">&lt;-</span> <span class="dv">20</span>   <span class="co"># loci</span></span>
<span id="annotated-cell-18-12"><a href="#annotated-cell-18-12"></a>K <span class="ot">&lt;-</span> <span class="dv">2</span>    <span class="co"># ancestral populations</span></span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-18" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-18" data-code-lines="1" data-code-annotation="1">Fixed random seed for consistent results across sessions</span>
</dd>
<dt data-target-cell="annotated-cell-18" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-18" data-code-lines="3,7" data-code-annotation="2">Normalized gamma variates generate simplex-constrained ancestry proportions</span>
</dd>
<dt data-target-cell="annotated-cell-18" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-18" data-code-lines="10,11,12" data-code-annotation="3">Set simulation parameters</span>
</dd>
</dl>
</div>
</div>
</section>
<section id="simulating-admixed-genotypes-1" class="slide level2">
<h2>Simulating Admixed Genotypes</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-19"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-19" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-19-1" class="code-annotation-target"><a href="#annotated-cell-19-1"></a><span class="co"># Create very clear population differentiation</span></span>
<span id="annotated-cell-19-2"><a href="#annotated-cell-19-2"></a>theta_true <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>, <span class="at">nrow =</span> K, <span class="at">ncol =</span> L)</span>
<span id="annotated-cell-19-3"><a href="#annotated-cell-19-3"></a></span>
<span id="annotated-cell-19-4"><a href="#annotated-cell-19-4"></a><span class="co"># Multiple anchor loci: very strong differentiation</span></span>
<span id="annotated-cell-19-5"><a href="#annotated-cell-19-5"></a>theta_true[<span class="dv">1</span>, <span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="fl">0.1</span>  <span class="co"># Pop 1: very low frequency at anchor 1</span></span>
<span id="annotated-cell-19-6"><a href="#annotated-cell-19-6"></a>theta_true[<span class="dv">2</span>, <span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="fl">0.9</span>  <span class="co"># Pop 2: very high frequency at anchor 1</span></span>
<span id="annotated-cell-19-7"><a href="#annotated-cell-19-7"></a>theta_true[<span class="dv">1</span>, <span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="fl">0.1</span>  <span class="co"># Pop 1: very low frequency at anchor 2</span></span>
<span id="annotated-cell-19-8"><a href="#annotated-cell-19-8"></a>theta_true[<span class="dv">2</span>, <span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="fl">0.9</span>  <span class="co"># Pop 2: very high frequency at anchor 2</span></span>
<span id="annotated-cell-19-9"><a href="#annotated-cell-19-9"></a></span>
<span id="annotated-cell-19-10"><a href="#annotated-cell-19-10"></a><span class="co"># Other loci: strong differentiation </span></span>
<span id="annotated-cell-19-11"><a href="#annotated-cell-19-11"></a>theta_true[<span class="dv">1</span>, <span class="dv">3</span><span class="sc">:</span>L] <span class="ot">&lt;-</span> <span class="fu">rbeta</span>(L<span class="dv">-2</span>, <span class="dv">2</span>, <span class="dv">8</span>)   <span class="co"># Pop 1: much lower overall</span></span>
<span id="annotated-cell-19-12"><a href="#annotated-cell-19-12"></a>theta_true[<span class="dv">2</span>, <span class="dv">3</span><span class="sc">:</span>L] <span class="ot">&lt;-</span> <span class="fu">rbeta</span>(L<span class="dv">-2</span>, <span class="dv">8</span>, <span class="dv">2</span>)   <span class="co"># Pop 2: much higher overall</span></span>
<span id="annotated-cell-19-13"><a href="#annotated-cell-19-13"></a></span>
<span id="annotated-cell-19-14"><a href="#annotated-cell-19-14"></a></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-19" data-target-annotation="5" onclick="event.preventDefault();">5</a><span id="annotated-cell-19-15" class="code-annotation-target"><a href="#annotated-cell-19-15"></a>pi_true <span class="ot">&lt;-</span> <span class="fu">rbind</span>(</span>
<span id="annotated-cell-19-16"><a href="#annotated-cell-19-16"></a>  <span class="fu">rdirichlet</span>(<span class="dv">12</span>, <span class="fu">c</span>(<span class="dv">90</span>, <span class="dv">10</span>)),  <span class="co"># pop 1</span></span>
<span id="annotated-cell-19-17"><a href="#annotated-cell-19-17"></a>  <span class="fu">rdirichlet</span>(<span class="dv">12</span>, <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">90</span>)),  <span class="co"># Very pure pop 2</span></span>
<span id="annotated-cell-19-18"><a href="#annotated-cell-19-18"></a>  <span class="fu">rdirichlet</span>(<span class="dv">16</span>, <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">10</span>))                                    <span class="co"># Admixed individuals</span></span>
<span id="annotated-cell-19-19"><a href="#annotated-cell-19-19"></a>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-19" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-19" data-code-lines="1" data-code-annotation="4">Strong genetic signatures with multiple anchor loci</span>
</dd>
<dt data-target-cell="annotated-cell-19" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-19" data-code-lines="15" data-code-annotation="5">Mix of unadmixed founders and admixed descendants</span>
</dd>
</dl>
</div>
</div>
</section>
<section id="simulating-admixed-genotypes-2" class="slide level2">
<h2>Simulating Admixed Genotypes</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-20"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-20" data-target-annotation="6" onclick="event.preventDefault();">6</a><span id="annotated-cell-20-1" class="code-annotation-target"><a href="#annotated-cell-20-1"></a>y <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>L, <span class="at">nrow =</span> N, <span class="at">ncol =</span> L)</span>
<span id="annotated-cell-20-2"><a href="#annotated-cell-20-2"></a><span class="cf">for</span> (n <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>N) {</span>
<span id="annotated-cell-20-3"><a href="#annotated-cell-20-3"></a>    <span class="cf">for</span> (l <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>L) {</span>
<span id="annotated-cell-20-4"><a href="#annotated-cell-20-4"></a>        p_mix <span class="ot">&lt;-</span> <span class="fu">sum</span>(pi_true[n, ] <span class="sc">*</span> theta_true[, l])</span>
<span id="annotated-cell-20-5"><a href="#annotated-cell-20-5"></a>        y[n, l] <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">prob =</span> p_mix)</span>
<span id="annotated-cell-20-6"><a href="#annotated-cell-20-6"></a>    }</span>
<span id="annotated-cell-20-7"><a href="#annotated-cell-20-7"></a>}</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-20" data-target-annotation="6">6</dt>
<dd>
<span data-code-cell="annotated-cell-20" data-code-lines="1,4,5" data-code-annotation="6">Genotypes drawn from Binomial(2, p_mix) per individual and locus</span>
</dd>
</dl>
</div>
</div>
</section>
<section id="pca-on-simulated-genotypes" class="slide level2">
<h2>PCA On Simulated Genotypes</h2>
<ul>
<li>Center genotypes by <span class="math inline">2 \times \hat{p}</span> and scale by <span class="math inline">\sqrt{2 \cdot \hat{p} \cdot (1 - \hat{p})}</span></li>
</ul>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a></a>p_hat <span class="ot">&lt;-</span> <span class="fu">colMeans</span>(y) <span class="sc">/</span> <span class="dv">2</span></span>
<span id="cb14-2"><a></a>sd_hat <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(<span class="fu">pmax</span>(<span class="fl">1e-6</span>, <span class="dv">2</span> <span class="sc">*</span> p_hat <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> p_hat)))</span>
<span id="cb14-3"><a></a>Z <span class="ot">&lt;-</span> <span class="fu">scale</span>(y, <span class="at">center =</span> <span class="dv">2</span> <span class="sc">*</span> p_hat, <span class="at">scale =</span> sd_hat)</span>
<span id="cb14-4"><a></a>pc <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(Z, <span class="at">center =</span> <span class="cn">FALSE</span>, <span class="at">scale. =</span> <span class="cn">FALSE</span>)</span>
<span id="cb14-5"><a></a></span>
<span id="cb14-6"><a></a>pc_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">PC1 =</span> pc<span class="sc">$</span>x[, <span class="dv">1</span>], <span class="at">PC2 =</span> pc<span class="sc">$</span>x[, <span class="dv">2</span>], <span class="at">pi1 =</span> pi_true[, <span class="dv">1</span>])</span>
<span id="cb14-7"><a></a><span class="fu">ggplot</span>(pc_df, <span class="fu">aes</span>(PC1, PC2, <span class="at">color =</span> pi1)) <span class="sc">+</span></span>
<span id="cb14-8"><a></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.7</span>, <span class="at">size =</span> <span class="fl">1.6</span>) <span class="sc">+</span></span>
<span id="cb14-9"><a></a>  <span class="fu">scale_color_viridis_c</span>(<span class="at">end =</span> .<span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb14-10"><a></a>  <span class="fu">labs</span>(<span class="at">color =</span> <span class="fu">expression</span>(pi[<span class="dv">1</span>]), <span class="at">x =</span> <span class="st">"PC1"</span>, <span class="at">y =</span> <span class="st">"PC2"</span>) <span class="sc">+</span></span>
<span id="cb14-11"><a></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fl">2.75</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"gray40"</span>, <span class="at">linewidth =</span> <span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb14-12"><a></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="sc">-</span><span class="fl">2.75</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"gray40"</span>, <span class="at">linewidth =</span> <span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb14-13"><a></a>  <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">x =</span> <span class="fl">4.5</span>, <span class="at">y =</span> <span class="dv">6</span>, <span class="at">label =</span> <span class="st">"Pop 1"</span>, <span class="at">color =</span> <span class="st">"gray20"</span>, <span class="at">size =</span> <span class="dv">6</span>) <span class="sc">+</span></span>
<span id="cb14-14"><a></a>  <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">x =</span> <span class="sc">-</span><span class="fl">4.5</span>, <span class="at">y =</span> <span class="dv">6</span>, <span class="at">label =</span> <span class="st">"Pop 2"</span>, <span class="at">color =</span> <span class="st">"gray20"</span>, <span class="at">size =</span> <span class="dv">6</span>) <span class="sc">+</span></span>
<span id="cb14-15"><a></a>  <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">x =</span> <span class="dv">0</span>, <span class="at">y =</span> <span class="dv">6</span>, <span class="at">label =</span> <span class="st">"Admixed"</span>, <span class="at">color =</span> <span class="st">"gray20"</span>, <span class="at">size =</span> <span class="dv">6</span>) <span class="sc">+</span></span>
<span id="cb14-16"><a></a>  <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">14</span>) <span class="sc">+</span></span>
<span id="cb14-17"><a></a>  <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>()) </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="pca-on-simulated-genotypes-1" class="slide level2">
<h2>PCA On Simulated Genotypes</h2>

<img data-src="lecture-04_files/figure-revealjs/unnamed-chunk-19-1.png" width="960" class="r-stretch"></section>
<section id="fitting-the-model" class="slide level2">
<h2>Fitting the Model</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a></a>admix_mod <span class="ot">&lt;-</span> <span class="fu">cmdstan_model</span>(<span class="fu">file.path</span>(<span class="st">"stan"</span>, <span class="st">"structure_admixture.stan"</span>))</span>
<span id="cb15-2"><a></a></span>
<span id="cb15-3"><a></a>l_star <span class="ot">&lt;-</span> <span class="dv">1</span>  <span class="co"># choose your anchor locus (e.g., 1)</span></span>
<span id="cb15-4"><a></a></span>
<span id="cb15-5"><a></a>admix_data <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb15-6"><a></a>    <span class="at">N =</span> N, <span class="at">L =</span> L, <span class="at">K =</span> K, <span class="at">y =</span> y,</span>
<span id="cb15-7"><a></a>    <span class="at">a_theta =</span> <span class="fu">rep</span>(<span class="fl">1.5</span>, K), <span class="at">b_theta =</span> <span class="fu">rep</span>(<span class="fl">1.5</span>, K),</span>
<span id="cb15-8"><a></a>    <span class="at">l_star =</span> l_star, <span class="at">L_soft =</span> <span class="dv">1</span>, <span class="at">soft_idx =</span> <span class="fu">c</span>(<span class="dv">2</span>),</span>
<span id="cb15-9"><a></a>    <span class="at">conc_pi =</span> <span class="fl">2.0</span>      </span>
<span id="cb15-10"><a></a>)</span>
<span id="cb15-11"><a></a></span>
<span id="cb15-12"><a></a>admix_fit <span class="ot">&lt;-</span> admix_mod<span class="sc">$</span><span class="fu">sample</span>(</span>
<span id="cb15-13"><a></a>  <span class="at">data =</span> admix_data,</span>
<span id="cb15-14"><a></a>  <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">parallel_chains =</span> <span class="dv">4</span>,</span>
<span id="cb15-15"><a></a>  <span class="at">iter_warmup =</span> <span class="dv">2000</span>, <span class="at">iter_sampling =</span> <span class="dv">2000</span>,</span>
<span id="cb15-16"><a></a>  <span class="at">seed =</span> <span class="dv">887804</span>, <span class="at">refresh =</span> <span class="dv">0</span></span>
<span id="cb15-17"><a></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Running MCMC with 4 parallel chains...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Chain 2 Informational Message: The current Metropolis proposal is about to be rejected because of the following issue:</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Chain 2 Exception: binomial_lpmf: Probability parameter is 1, but must be in the interval [0, 1] (in '/var/folders/3f/7lk7ddbn19j4f1rzxtlx3z9w0000gn/T/RtmpX9Wapd/model-d8207aefdca0.stan', line 65, column 6 to column 41)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Chain 2 If this warning occurs sporadically, such as for highly constrained variable types like covariance matrices, then the sampler is fine,</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Chain 2 but if this warning occurs often then your model may be either severely ill-conditioned or misspecified.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Chain 2 </code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Chain 4 Informational Message: The current Metropolis proposal is about to be rejected because of the following issue:</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Chain 4 Exception: binomial_lpmf: Probability parameter is 1, but must be in the interval [0, 1] (in '/var/folders/3f/7lk7ddbn19j4f1rzxtlx3z9w0000gn/T/RtmpX9Wapd/model-d8207aefdca0.stan', line 65, column 6 to column 41)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Chain 4 If this warning occurs sporadically, such as for highly constrained variable types like covariance matrices, then the sampler is fine,</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Chain 4 but if this warning occurs often then your model may be either severely ill-conditioned or misspecified.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Chain 4 </code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Chain 2 finished in 8.6 seconds.
Chain 1 finished in 8.7 seconds.
Chain 4 finished in 8.6 seconds.
Chain 3 finished in 9.0 seconds.

All 4 chains finished successfully.
Mean chain execution time: 8.7 seconds.
Total execution time: 9.1 seconds.</code></pre>
</div>
</div>
</section>
<section id="stan-output" class="slide level2">
<h2>Stan Output</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb28"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a></a>dm <span class="ot">&lt;-</span> admix_fit<span class="sc">$</span><span class="fu">draws</span>(<span class="at">variables =</span> <span class="fu">c</span>(<span class="st">"theta"</span>,<span class="st">"pi"</span>), <span class="at">format =</span> <span class="st">"draws_df"</span>)</span>
<span id="cb28-2"><a></a></span>
<span id="cb28-3"><a></a>lstar <span class="ot">&lt;-</span> admix_data<span class="sc">$</span>l_star</span>
<span id="cb28-4"><a></a>L <span class="ot">&lt;-</span> <span class="fu">ncol</span>(y)</span>
<span id="cb28-5"><a></a>non_anchor <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(<span class="dv">1</span><span class="sc">:</span>L, lstar)</span>
<span id="cb28-6"><a></a></span>
<span id="cb28-7"><a></a><span class="co"># Orientation score per draw: average sign over non-anchor loci</span></span>
<span id="cb28-8"><a></a>sign_per_draw <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(<span class="fu">sapply</span>(non_anchor, <span class="cf">function</span>(l)</span>
<span id="cb28-9"><a></a>  <span class="fu">sign</span>(dm[[<span class="fu">sprintf</span>(<span class="st">"theta[2,%d]"</span>, l)]] <span class="sc">-</span> dm[[<span class="fu">sprintf</span>(<span class="st">"theta[1,%d]"</span>, l)]])))</span>
<span id="cb28-10"><a></a></span>
<span id="cb28-11"><a></a><span class="co"># Reference orientation: the majority sign across all draws</span></span>
<span id="cb28-12"><a></a>ref_sign <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(<span class="fu">mean</span>(sign_per_draw) <span class="sc">&gt;=</span> <span class="dv">0</span>, <span class="sc">-</span><span class="dv">1</span>, <span class="dv">1</span>)</span>
<span id="cb28-13"><a></a></span>
<span id="cb28-14"><a></a><span class="co"># Decide which DRAWS to flip (not just which chains)</span></span>
<span id="cb28-15"><a></a>flip_draw <span class="ot">&lt;-</span> sign_per_draw <span class="sc">*</span> ref_sign <span class="sc">&lt;</span> <span class="dv">0</span></span>
<span id="cb28-16"><a></a></span>
<span id="cb28-17"><a></a><span class="co"># Helper: swap theta rows and pi columns for those draws</span></span>
<span id="cb28-18"><a></a>swap_block <span class="ot">&lt;-</span> <span class="cf">function</span>(df, pat1, pat2, idx){</span>
<span id="cb28-19"><a></a>  i1 <span class="ot">&lt;-</span> <span class="fu">grep</span>(pat1, <span class="fu">names</span>(df)); i2 <span class="ot">&lt;-</span> <span class="fu">grep</span>(pat2, <span class="fu">names</span>(df))</span>
<span id="cb28-20"><a></a>  tmp <span class="ot">&lt;-</span> df[idx, i1, drop<span class="ot">=</span><span class="cn">FALSE</span>]</span>
<span id="cb28-21"><a></a>  df[idx, i1] <span class="ot">&lt;-</span> df[idx, i2, drop<span class="ot">=</span><span class="cn">FALSE</span>]</span>
<span id="cb28-22"><a></a>  df[idx, i2] <span class="ot">&lt;-</span> tmp</span>
<span id="cb28-23"><a></a>  df</span>
<span id="cb28-24"><a></a>}</span>
<span id="cb28-25"><a></a></span>
<span id="cb28-26"><a></a><span class="co"># Swap theta rows</span></span>
<span id="cb28-27"><a></a>dm <span class="ot">&lt;-</span> <span class="fu">swap_block</span>(dm, <span class="st">"^theta</span><span class="sc">\\</span><span class="st">[1,"</span>, <span class="st">"^theta</span><span class="sc">\\</span><span class="st">[2,"</span>, flip_draw)</span>
<span id="cb28-28"><a></a><span class="co"># Swap pi columns</span></span>
<span id="cb28-29"><a></a>dm <span class="ot">&lt;-</span> <span class="fu">swap_block</span>(dm, <span class="st">"^pi</span><span class="sc">\\</span><span class="st">[[0-9]+,1</span><span class="sc">\\</span><span class="st">]$"</span>, <span class="st">"^pi</span><span class="sc">\\</span><span class="st">[[0-9]+,2</span><span class="sc">\\</span><span class="st">]$"</span>, flip_draw)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="aligned-posteriors" class="slide level2">
<h2>Aligned Posteriors</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb29"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a></a><span class="co"># Plot aligned posteriors with true values</span></span>
<span id="cb29-2"><a></a>true_line <span class="ot">&lt;-</span> <span class="cf">function</span>(pars, truths) <span class="fu">data.frame</span>(<span class="at">variable =</span> pars, <span class="at">truth =</span> truths)</span>
<span id="cb29-3"><a></a></span>
<span id="cb29-4"><a></a><span class="fu">mcmc_hist</span>(dm, <span class="at">pars =</span> <span class="fu">c</span>(<span class="st">"pi[1,1]"</span>)) <span class="sc">+</span> </span>
<span id="cb29-5"><a></a>  <span class="fu">geom_vline</span>(<span class="at">data =</span> <span class="fu">true_line</span>(<span class="fu">c</span>(<span class="st">"pi[1,1]"</span>), <span class="fu">c</span>(pi_true[<span class="dv">1</span>,<span class="dv">1</span>])),</span>
<span id="cb29-6"><a></a>             <span class="fu">aes</span>(<span class="at">xintercept =</span> truth), <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">colour =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb29-7"><a></a><span class="fu">xlim</span>(<span class="dv">0</span>, <span class="dv">1</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>

</div>
<img data-src="lecture-04_files/figure-revealjs/unnamed-chunk-22-1.png" width="960" class="r-stretch"></section>
<section id="credible-intervals" class="slide level2">
<h2>Credible Intervals</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb30"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a></a><span class="fu">mcmc_intervals</span>(dm, <span class="at">pars =</span> <span class="fu">c</span>(<span class="st">"pi[1,1]"</span>, <span class="st">"pi[3,1]"</span>, <span class="st">"pi[14,1]"</span>, <span class="st">"pi[38,1]"</span>)) </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>

</div>
<img data-src="lecture-04_files/figure-revealjs/unnamed-chunk-23-1.png" width="960" class="r-stretch"></section>
<section id="summary-next-steps" class="slide level2">
<h2>Summary &amp; Next Steps</h2>
<ul>
<li>Population structure influences genetic analyses</li>
<li>Bayesian framing exposes prior choices, enables posterior uncertainty on structure (ABO example, admixture Stan model)</li>
<li>Bayesian modeling as attempting to capture the <strong>data-generating process</strong></li>
</ul>
</section>
<section id="reference" class="slide level2 smaller scrollable">
<h2>Reference</h2>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-carpenterStanProbabilisticProgramming2017" class="csl-entry" role="listitem">
Carpenter,B. <em>et al.</em> (2017) <a href="https://doi.org/10.18637/jss.v076.i01">Stan: A probabilistic programming language</a>. <em>Journal of Statistical Software</em>, <strong>76</strong>, 1–32.
</div>
<div id="ref-gelmanBayesianDataAnalysis2013" class="csl-entry" role="listitem">
Gelman,A. <em>et al.</em> (2013) <a href="https://doi.org/10.1201/b16018">Bayesian data analysis</a> 3rd ed. Chapman; Hall/CRC, Boca Raton, FL.
</div>
<div id="ref-korunesHumanGeneticAdmixture2021" class="csl-entry" role="listitem">
Korunes,K.L. and Goldberg,A. (2021) <a href="https://doi.org/10.1371/journal.pgen.1009374">Human genetic admixture</a>. <em>PLOS Genetics</em>, <strong>17</strong>, e1009374.
</div>
<div id="ref-mcelreathStatisticalRethinkingBayesian2020" class="csl-entry" role="listitem">
McElreath,R. (2020) <a href="https://doi.org/10.1201/9780429029608">Statistical rethinking: A bayesian course with examples in r and stan</a> 2nd ed. Chapman; Hall/CRC, Boca Raton, FL.
</div>
<div id="ref-mourant1976" class="csl-entry" role="listitem">
Mourant,A.E. <em>et al.</em> (1976) The distribution of the human blood groups and other polymorphisms 2nd ed. Oxford University Press, London.
</div>
<div id="ref-yamamoto2012abo" class="csl-entry" role="listitem">
Yamamoto,F. <em>et al.</em> (2012) <a href="https://doi.org/10.1016/j.tmrv.2011.08.002">ABO research in the modern era of genomics</a>. <em>Transfusion Medicine Reviews</em>, <strong>26</strong>, 103–118.
</div>
</div>
</section></section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">

</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
          const outerScaffold = trigger.parentElement.cloneNode(true);
          const codeEl = outerScaffold.querySelector('code');
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp('/' + window.location.host + '/');
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    

</body></html>