<!DOCTYPE html>
<html lang="en"><head>
<link href="../course_logo.png" rel="icon" type="image/png">
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-html/tabby.min.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.8.25">

  <title>Statistical Genetics – Lecture 06: Prediction models in genetics</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    html { -webkit-text-size-adjust: 100%; }
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; font-weight: bold; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="../site_libs/revealjs/dist/theme/quarto-a8f461879e61fe6cdedd4a5ed206ce40.css">
  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.css">
  <link href="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Lecture 06: Prediction models in genetics</h1>
  <p class="subtitle">PUBH 8878, Statistical Genetics</p>

<div class="quarto-title-authors">
</div>

</section>
<section id="inference-vs-prediction" class="slide level2">
<h2>Inference vs Prediction</h2>
<ul>
<li>So far, we have talked about estimation of certain quantities of interest
<ul>
<li><span class="math inline">p</span>: Allele frequencies</li>
<li><span class="math inline">\beta</span>: Association effect sizes</li>
<li><span class="math inline">\pi</span>: Ancestry proportions</li>
</ul></li>
</ul>
</section>
<section id="inference-vs-prediction-1" class="slide level2">
<h2>Inference vs Prediction</h2>
<ul>
<li>Some tasks, however, focus on predicting outcomes rather than estimating parameters.
<ul>
<li>Disease risk classification (polygenic risk scores, case vs control)</li>
<li>Absolute/lifetime risk and time-to-event predictions</li>
<li>Drug response/toxicity prediction</li>
</ul></li>
</ul>
</section>
<section id="inference-vs-prediction-2" class="slide level2">
<h2>Inference vs Prediction</h2>
<p>While not mutually exclusive, these tasks require different approaches and metrics</p>
<ul>
<li>Association
<ul>
<li>Goal: identify variants related to trait.</li>
<li>Target: low FDR/valid inference, effect estimates with SEs.</li>
<li>Tools: single‑variant tests, LMMs, fine‑mapping</li>
</ul></li>
<li>Prediction
<ul>
<li>Goal: accurate out‑of‑sample trait/risk estimates.</li>
<li>Target: minimize expected loss (MSE, log loss)</li>
<li>Tools: penalized GLMs, BLUP/GBLUP, ensembles, NNs</li>
</ul></li>
</ul>
</section>
<section id="bait-and-switch" class="slide level2">
<h2>Bait and Switch</h2>
<p>Today we will be talking about baseball! (Briefly)</p>

<img data-src="https://upload.wikimedia.org/wikipedia/commons/thumb/0/0e/Pittsburgh_Pirates_park_%28Unsplash%29.jpg/1599px-Pittsburgh_Pirates_park_%28Unsplash%29.jpg?20170823100406" class="r-stretch quarto-figure-center"><p class="caption"><a href="https://commons.wikimedia.org/wiki/File:Pittsburgh_Pirates_park_(Unsplash).jpg">Joshua Peacock jcpeacock, CC0, via Wikimedia Commons</a></p></section>
<section id="prediction-in-baseball" class="slide level2">
<h2>Prediction in baseball</h2>
<ul>
<li>Task: predict a later‑season batting rate using the first 45 at‑bats for each player (following Efron–Morris)</li>
</ul>
</section>
<section id="our-data" class="slide level2">
<h2>Our data</h2>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top">
<thead>
<tr class="header">
<th style="text-align: left;">name</th>
<th style="text-align: left;">team</th>
<th style="text-align: left;">league</th>
<th style="text-align: right;">hits</th>
<th style="text-align: right;">at_bats</th>
<th style="text-align: right;">early_avg</th>
<th style="text-align: right;">final_avg</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Roberto Clemente</td>
<td style="text-align: left;">Pitts</td>
<td style="text-align: left;">NL</td>
<td style="text-align: right;">18</td>
<td style="text-align: right;">45</td>
<td style="text-align: right;">0.400</td>
<td style="text-align: right;">0.346</td>
</tr>
<tr class="even">
<td style="text-align: left;">Frank Robinson</td>
<td style="text-align: left;">Balt</td>
<td style="text-align: left;">AL</td>
<td style="text-align: right;">17</td>
<td style="text-align: right;">45</td>
<td style="text-align: right;">0.378</td>
<td style="text-align: right;">0.298</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Frank Howard</td>
<td style="text-align: left;">Wash</td>
<td style="text-align: left;">AL</td>
<td style="text-align: right;">16</td>
<td style="text-align: right;">45</td>
<td style="text-align: right;">0.356</td>
<td style="text-align: right;">0.276</td>
</tr>
<tr class="even">
<td style="text-align: left;">Jay Johnstone</td>
<td style="text-align: left;">Cal</td>
<td style="text-align: left;">AL</td>
<td style="text-align: right;">15</td>
<td style="text-align: right;">45</td>
<td style="text-align: right;">0.333</td>
<td style="text-align: right;">0.222</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Ken Berry</td>
<td style="text-align: left;">Chi</td>
<td style="text-align: left;">AL</td>
<td style="text-align: right;">14</td>
<td style="text-align: right;">45</td>
<td style="text-align: right;">0.311</td>
<td style="text-align: right;">0.273</td>
</tr>
<tr class="even">
<td style="text-align: left;">Jim Spencer</td>
<td style="text-align: left;">Cal</td>
<td style="text-align: left;">AL</td>
<td style="text-align: right;">14</td>
<td style="text-align: right;">45</td>
<td style="text-align: right;">0.311</td>
<td style="text-align: right;">0.270</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="estimating-batting-averages" class="slide level2">
<h2>Estimating batting averages</h2>
<ul>
<li>If we assume that each player’s batting average comes from a binomial distribution, then the MLE is the sample proportion <span class="math inline">\hat{p_i} = h_i/n_i</span></li>
<li>A different approach is to use a beta prior centered at the league average
<ul>
<li>The prior is <span class="math inline">p_i \sim \mathrm{Beta}(\alpha, \beta)</span> with <span class="math inline">\alpha = \bar p \kappa</span> and <span class="math inline">\beta = (1 - \bar p)\kappa</span>, where <span class="math inline">\bar p</span> is the league average</li>
</ul></li>
</ul>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a></a>league_avg <span class="ot">&lt;-</span> <span class="fu">mean</span>(batting_averages<span class="sc">$</span>hits <span class="sc">/</span> batting_averages<span class="sc">$</span>at_bats)</span>
<span id="cb1-2"><a></a></span>
<span id="cb1-3"><a></a>k_grid <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">k =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">50</span>, <span class="at">by =</span> <span class="dv">1</span>))</span>
<span id="cb1-4"><a></a></span>
<span id="cb1-5"><a></a>batting_avg_long <span class="ot">&lt;-</span> tidyr<span class="sc">::</span><span class="fu">crossing</span>(batting_averages, k_grid) <span class="sc">|&gt;</span></span>
<span id="cb1-6"><a></a>    <span class="fu">mutate</span>(<span class="at">alpha =</span> league_avg <span class="sc">*</span> k,</span>
<span id="cb1-7"><a></a>           <span class="at">beta  =</span> (<span class="dv">1</span> <span class="sc">-</span> league_avg) <span class="sc">*</span> k,</span>
<span id="cb1-8"><a></a>           <span class="at">p_k   =</span> (alpha <span class="sc">+</span> hits) <span class="sc">/</span> (alpha <span class="sc">+</span> beta <span class="sc">+</span> at_bats),</span>
<span id="cb1-9"><a></a>           <span class="at">estimator =</span> <span class="fu">paste0</span>(<span class="st">"k="</span>, k)</span>
<span id="cb1-10"><a></a>           ) </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="prediction-performance" class="slide level2">
<h2>Prediction performance</h2>

<img data-src="lecture-06_files/figure-revealjs/unnamed-chunk-4-1.png" width="960" class="r-stretch"></section>
<section id="shrinkage-visualization" class="slide level2">
<h2>Shrinkage visualization</h2>

<img data-src="lecture-06_files/figure-revealjs/unnamed-chunk-5-1.png" width="960" class="r-stretch"></section>
<section id="what-are-we-doing-here" class="slide level2">
<h2>What are we doing here?</h2>
<ul>
<li>Our beta-binomial model makes an assumption that players regress toward the league average</li>
<li>More statistically, we are assuming that the true batting averages are drawn from a common prior distribution centered at the league average</li>
</ul>
</section>
<section id="back-to-genetics" class="slide level2">
<h2>Back to genetics</h2>
<ul>
<li>The same principles apply when performing prediction in genetics</li>
<li>Many problems involve a large number of genetic variants (p) and relatively small sample sizes (n)</li>
<li>When <span class="math inline">p &gt; n</span>, linear regression is singular
<ul>
<li>This means the design matrix does not have full rank, leading to non-unique solutions</li>
<li>We solve this by imposing strong structure (assumptions) on effects</li>
</ul></li>
</ul>
</section>
<section id="p-n-structural-assumptions-we-use" class="slide level2">
<h2><span class="math inline">p &gt;&gt; n</span>: structural assumptions we use</h2>
<ul>
<li>Sparsity: most variant effects are 0 (lasso, debiased lasso for inference)</li>
<li>Gaussian shrinkage: many small effects (ridge, BLUP/GBLUP, PRS-CS)</li>
<li>Low-rank/latent structure: factors, PCs, random effects (LMMs)</li>
<li>Smoothness/blocks: genomic locality/LD (fused/TV penalties, windowed methods)</li>
<li>Hybrid/empirical Bayes: estimate prior from data + partial pooling</li>
</ul>
</section>
<section id="examples-of-shrinkagesparse-estimators" class="slide level2">
<h2>Examples of Shrinkage/Sparse Estimators</h2>
<ul>
<li>Ridge regression</li>
<li>Lasso regression</li>
<li>Elastic net</li>
<li>Bayesian linear regression with Gaussian priors</li>
<li>BLUP/GBLUP in genetics</li>
<li>DESeq2 (transcriptomics)</li>
</ul>
</section>
<section id="what-is-blup" class="slide level2">
<h2>What is BLUP?</h2>
<ul>
<li>BLUP = Best Linear Unbiased Prediction of random effects in a mixed model.</li>
<li>Model: <span class="math inline">y = X\beta + Zu + \varepsilon</span>, with <span class="math inline">u \sim \mathcal N(0, \sigma_g^2 K)</span> and <span class="math inline">\varepsilon \sim \mathcal N(0, \sigma_e^2 I)</span>.
<ul>
<li><span class="math inline">u</span> collect latent genetic effects, <span class="math inline">K</span> encodes genetic similarity among individuals.</li>
</ul></li>
<li>BLUP shrinks predictions toward a common mean
<ul>
<li>The ratio <span class="math inline">\sigma_g^2/\sigma_e^2</span> determines shrinkage strength</li>
<li>As <span class="math inline">\sigma_g^2/\sigma_e^2 \to 0</span>, predictions shrink to the mean</li>
</ul></li>
<li>Pedigree-based BLUP: use <span class="math inline">K = A</span>, the expected relatedness matrix from pedigrees.</li>
<li>Genomic BLUP: use <span class="math inline">K = G</span>, the realized similarity matrix from marker data.</li>
</ul>
</section>
<section id="relationship-matrices-a-and-g-how-we-build-them" class="slide level2">
<h2>Relationship matrices A and G (how we build them)</h2>
<ul>
<li>Pedigree <span class="math inline">A</span>:
<ul>
<li><span class="math inline">A_{ij}</span> = expected proportion of alleles identical‑by‑descent</li>
<li>Requires accurate pedigrees.</li>
</ul></li>
<li>Genomic <span class="math inline">G</span> (unified with ridge via standardized markers):
<ul>
<li>Build a standardized marker matrix <span class="math inline">Z</span> by centering each SNP by its allele frequency and scaling by its standard deviation.
<ul>
<li>0/1 SNPs (DArT): <span class="math inline">Z_{ij} = (X_{ij} - p_j)/\sqrt{p_j(1-p_j)}</span>.</li>
<li>0/1/2 SNPs: <span class="math inline">Z_{ij} = (X_{ij} - 2p_j)/\sqrt{2p_j(1-p_j)}</span>.</li>
</ul></li>
<li>Set <span class="math inline">G = ZZ^\top / m</span> with <span class="math inline">m</span> markers; then <span class="math inline">\operatorname{E}[\operatorname{diag}(G)] \approx 1</span>.</li>
<li>Fit ridge on the same <span class="math inline">Z</span> with <code>standardize=FALSE</code> so ridge and GBLUP are directly comparable.</li>
</ul></li>
</ul>
</section>
<section id="ridge-kernel-ridge-and-gblup-equivalence" class="slide level2">
<h2>Ridge, Kernel Ridge, and GBLUP: Equivalence</h2>
<ul>
<li>Setup: <span class="math inline">Z\in\mathbb R^{n\times p}</span> (standardized), <span class="math inline">K=ZZ^\top/p</span> (with <span class="math inline">p</span> markers), <span class="math inline">\tilde y = y - \bar y\,\mathbf 1</span>.</li>
<li>Primal ridge: <span class="math inline">\hat\beta = (Z^\top Z + \lambda I)^{-1} Z^\top\tilde y</span>, predict <span class="math inline">\hat y = \bar y\,\mathbf 1 + Z\hat\beta</span>.</li>
<li>Dual/kernel ridge: <span class="math inline">\hat\alpha = (K + \lambda_K I)^{-1}\tilde y</span>, predict <span class="math inline">\hat y = \bar y\,\mathbf 1 + K\hat\alpha</span>.</li>
<li>Identity: <span class="math inline">(Z^\top Z + \lambda I)^{-1} Z^\top = Z^\top (ZZ^\top + \lambda I)^{-1}</span> ⇒ <span class="math inline">\hat\beta = Z^\top\hat\alpha</span> and <span class="math inline">Z\hat\beta = K\hat\alpha</span> ⇒ same <span class="math inline">\hat y</span>.</li>
<li>Scaling: if <span class="math inline">K=ZZ^\top/p</span>, use <span class="math inline">\lambda_K=\lambda/p</span> to match primal.</li>
<li>GBLUP: <span class="math inline">y=\mathbf 1\mu + g + \varepsilon</span>, <span class="math inline">g\sim\mathcal N(0,\sigma_g^2 K)</span>, <span class="math inline">\varepsilon\sim\mathcal N(0,\sigma_e^2 I)</span> ⇒ <span class="math display">\hat y = \hat\mu\,\mathbf 1 + K\big(K + (\sigma_e^2/\sigma_g^2)I\big)^{-1}(y-\hat\mu\,\mathbf 1),</span> i.e., kernel ridge on the same <span class="math inline">K</span> with <span class="math inline">\lambda_K=\sigma_e^2/\sigma_g^2</span>.</li>
</ul>
</section>
<section id="prediction-for-crop-yield" class="slide level2">
<h2>Prediction for crop yield</h2>
<ul>
<li>Consider the problem of predicting grain yield for different wheat lines across multiple environments</li>
<li>We are interested in modeling the genetic and environmental effects on yield</li>
</ul>
<p>Dataset at a glance</p>
<ul>
<li>Lines: 599 historical wheat lines</li>
<li>Environments: 4 mega-environments (columns in Y)</li>
<li>Phenotype (Y): average grain yield (one column per environment)</li>
<li>Markers (X): (0/1 after QC + imputation)</li>
<li>Pedigree (A): 599×599 additive relationship from multi-generation pedigree</li>
</ul>
</section>
<section id="breeding-context-what-are-we-predicting" class="slide level2">
<h2>Breeding context: what are we predicting?</h2>
<ul>
<li>A “line” = inbred or DH line to be advanced or discarded; objective is to predict its genetic value (GEBV) for yield.</li>
<li>“Environment (Env)” = location–year–management combination; yield varies by Env and by G×E (line×environment interaction)</li>
<li>Target Population of Environments (TPE): we want models that generalize to the environments we care about</li>
</ul>
</section>
<section id="crossvalidation-that-matches-breeding-decisions" class="slide level2">
<h2>Cross‑validation that matches breeding decisions</h2>
<ul>
<li>Use line‑wise folds so the same line never appears in both train and test.</li>
<li>For multi‑environment data:
<ul>
<li>Goal “new line, known environments” <span class="math inline">\to</span> mask whole lines across all environments in test.</li>
<li>Goal “known lines, new environment” <span class="math inline">\to</span> environment‑wise holdouts and G×E models.</li>
</ul></li>
</ul>
</section>
<section id="loading-data" class="slide level2">
<h2>Loading data</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a></a><span class="fu">set.seed</span>(<span class="dv">2025</span>)</span>
<span id="cb2-2"><a></a></span>
<span id="cb2-3"><a></a><span class="co"># Load data (BGLR wheat)</span></span>
<span id="cb2-4"><a></a><span class="fu">data</span>(wheat)</span>
<span id="cb2-5"><a></a>X <span class="ot">&lt;-</span> wheat.X</span>
<span id="cb2-6"><a></a>Y <span class="ot">&lt;-</span> wheat.Y</span>
<span id="cb2-7"><a></a>A <span class="ot">&lt;-</span> wheat.A</span>
<span id="cb2-8"><a></a>n <span class="ot">&lt;-</span> <span class="fu">nrow</span>(X)</span>
<span id="cb2-9"><a></a>m <span class="ot">&lt;-</span> <span class="fu">ncol</span>(X)</span>
<span id="cb2-10"><a></a>e <span class="ot">&lt;-</span> <span class="fu">ncol</span>(Y)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="looking-at-our-data" class="slide level2">
<h2>Looking at our data</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a></a>X[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>     wPt.0538 wPt.8463 wPt.6348
[1,]        0        1        1
[2,]        1        1        1
[3,]        1        1        1</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a></a>Y[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>              1          2          4          5
775   1.6716295 -1.7274699 -1.8902848  0.0509159
2166 -0.2527028  0.4095224  0.3093855 -1.7387588
2167  0.3418151 -0.6486263 -0.7995592 -1.0535691</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a></a>A[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>        775   2166   2167
775  1.9698 0.5742 0.5742
2166 0.5742 1.9930 1.9930
2167 0.5742 1.9930 1.9966</code></pre>
</div>
</div>
</section>
<section id="standardizing-markers" class="slide level2">
<h2>Standardizing markers</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-6"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-6-1" class="code-annotation-target"><a href="#annotated-cell-6-1"></a>p <span class="ot">&lt;-</span> <span class="fu">colMeans</span>(X)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-6-2" class="code-annotation-target"><a href="#annotated-cell-6-2"></a>s <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(p <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> p))</span>
<span id="annotated-cell-6-3"><a href="#annotated-cell-6-3"></a>s[s <span class="sc">==</span> <span class="dv">0</span>] <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-6-4" class="code-annotation-target"><a href="#annotated-cell-6-4"></a>Z <span class="ot">&lt;-</span> <span class="fu">sweep</span>(<span class="fu">sweep</span>(X, <span class="dv">2</span>, p, <span class="st">"-"</span>), <span class="dv">2</span>, s, <span class="st">"/"</span>)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-6-5" class="code-annotation-target"><a href="#annotated-cell-6-5"></a>G <span class="ot">&lt;-</span> <span class="fu">tcrossprod</span>(Z) <span class="sc">/</span> <span class="fu">ncol</span>(Z)</span>
<span id="annotated-cell-6-6"><a href="#annotated-cell-6-6"></a></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="5" onclick="event.preventDefault();">5</a><span id="annotated-cell-6-7" class="code-annotation-target"><a href="#annotated-cell-6-7"></a>G <span class="ot">&lt;-</span> G <span class="sc">/</span> <span class="fu">mean</span>(<span class="fu">diag</span>(G))</span>
<span id="annotated-cell-6-8"><a href="#annotated-cell-6-8"></a>A <span class="ot">&lt;-</span> A <span class="sc">/</span> <span class="fu">mean</span>(<span class="fu">diag</span>(A))</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-6" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="1" data-code-annotation="1">Calculate allele frequencies per marker</span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="2" data-code-annotation="2">Calculate standard deviations for each marker</span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="4" data-code-annotation="3">Standardize markers into <span class="math inline">Z</span></span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="5" data-code-annotation="4">Build genomic relationship matrix <span class="math inline">G = ZZ^\top / m</span></span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="7,8" data-code-annotation="5">Normalize <span class="math inline">A</span> and <span class="math inline">G</span></span>
</dd>
</dl>
</div>
</div>
</section>
<section id="cross-validation-folds" class="slide level2">
<h2>Cross-validation folds</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a></a>make_folds <span class="ot">&lt;-</span> <span class="cf">function</span>(n, <span class="at">K =</span> <span class="dv">5</span>, <span class="at">seed =</span> <span class="dv">2025</span>) { </span>
<span id="cb9-2"><a></a>  <span class="fu">set.seed</span>(seed)</span>
<span id="cb9-3"><a></a>  <span class="fu">sample</span>(<span class="fu">rep</span>(<span class="fu">seq_len</span>(K), <span class="at">length.out =</span> n)) </span>
<span id="cb9-4"><a></a>}</span>
<span id="cb9-5"><a></a></span>
<span id="cb9-6"><a></a>K <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb9-7"><a></a>fold <span class="ot">&lt;-</span> <span class="fu">make_folds</span>(n, K)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<ul>
<li>Create a function to generate cross-validation folds</li>
<li>If we have multiple observations for the same line, ensure they are in the same fold</li>
</ul>
</section>
<section id="metrics-functions" class="slide level2">
<h2>Metrics functions</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a></a>metrics <span class="ot">&lt;-</span> <span class="cf">function</span>(obs, pred) { </span>
<span id="cb10-2"><a></a>  accuracy <span class="ot">&lt;-</span> <span class="fu">suppressWarnings</span>(<span class="fu">cor</span>(obs, pred))</span>
<span id="cb10-3"><a></a>  rmse <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(<span class="fu">mean</span>((obs <span class="sc">-</span> pred)<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb10-4"><a></a>  <span class="fu">c</span>(<span class="at">Accuracy =</span> accuracy, <span class="at">RMSE =</span> rmse)</span>
<span id="cb10-5"><a></a>}</span>
<span id="cb10-6"><a></a></span>
<span id="cb10-7"><a></a>by_fold_metrics <span class="ot">&lt;-</span> <span class="cf">function</span>(y, pred, fold, model_label){</span>
<span id="cb10-8"><a></a>  idx <span class="ot">&lt;-</span> <span class="fu">split</span>(<span class="fu">seq_along</span>(y), fold)</span>
<span id="cb10-9"><a></a>  acc <span class="ot">&lt;-</span> <span class="fu">sapply</span>(idx, <span class="cf">function</span>(ii) <span class="fu">suppressWarnings</span>(<span class="fu">cor</span>(y[ii], pred[ii])))</span>
<span id="cb10-10"><a></a>  rmse <span class="ot">&lt;-</span> <span class="fu">sapply</span>(idx, <span class="cf">function</span>(ii) <span class="fu">sqrt</span>(<span class="fu">mean</span>((y[ii] <span class="sc">-</span> pred[ii])<span class="sc">^</span><span class="dv">2</span>)))</span>
<span id="cb10-11"><a></a>  <span class="fu">tibble</span>(<span class="at">Model =</span> model_label, <span class="at">Fold =</span> <span class="fu">seq_along</span>(idx), <span class="at">Accuracy =</span> acc, <span class="at">RMSE =</span> rmse)</span>
<span id="cb10-12"><a></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<ul>
<li>Create functions to compute metrics</li>
</ul>
</section>
<section id="ridge-regression" class="slide level2">
<h2>Ridge regression</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a></a>cv_ridge_Z <span class="ot">&lt;-</span> <span class="cf">function</span>(y, Z, fold, <span class="at">inner_nfolds =</span> <span class="dv">5</span>){</span>
<span id="cb11-2"><a></a>  K <span class="ot">&lt;-</span> <span class="fu">max</span>(fold)</span>
<span id="cb11-3"><a></a>  pred <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA_real_</span>, <span class="fu">length</span>(y))</span>
<span id="cb11-4"><a></a>  lambda <span class="ot">&lt;-</span> <span class="fu">numeric</span>(K)</span>
<span id="cb11-5"><a></a>  <span class="cf">for</span>(k <span class="cf">in</span> <span class="fu">seq_len</span>(K)){</span>
<span id="cb11-6"><a></a>    tr <span class="ot">&lt;-</span> fold <span class="sc">!=</span> k; te <span class="ot">&lt;-</span> <span class="sc">!</span>tr</span>
<span id="cb11-7"><a></a>    fit <span class="ot">&lt;-</span> <span class="fu">cv.glmnet</span>(Z[tr, , <span class="at">drop =</span> <span class="cn">FALSE</span>], y[tr],</span>
<span id="cb11-8"><a></a>                     <span class="at">alpha =</span> <span class="dv">0</span>, <span class="at">standardize =</span> <span class="cn">FALSE</span>, <span class="at">intercept =</span> <span class="cn">TRUE</span>,</span>
<span id="cb11-9"><a></a>                     <span class="at">nfolds =</span> inner_nfolds)</span>
<span id="cb11-10"><a></a>    pred[te] <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">predict</span>(fit, <span class="at">newx =</span> Z[te, , <span class="at">drop =</span> <span class="cn">FALSE</span>], <span class="at">s =</span> <span class="st">"lambda.min"</span>))</span>
<span id="cb11-11"><a></a>    lambda[k] <span class="ot">&lt;-</span> fit<span class="sc">$</span>lambda.min</span>
<span id="cb11-12"><a></a>  }</span>
<span id="cb11-13"><a></a>  <span class="fu">list</span>(<span class="at">pred =</span> pred, <span class="at">lambda =</span> lambda)</span>
<span id="cb11-14"><a></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<ul>
<li>Create a function for ridge regression using <code>cv.glmnet</code></li>
<li>Uses <code>Z</code>, the standardized marker matrix</li>
</ul>
</section>
<section id="ag-blup-models" class="slide level2">
<h2>A/G-BLUP Models</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a></a>cv_bglr_kernel <span class="ot">&lt;-</span> <span class="cf">function</span>(y, K, fold, <span class="at">nIter =</span> <span class="dv">6000</span>, <span class="at">burnIn =</span> <span class="dv">1000</span>){</span>
<span id="cb12-2"><a></a>  pred <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA_real_</span>, <span class="fu">length</span>(y))</span>
<span id="cb12-3"><a></a>  <span class="cf">for</span>(k <span class="cf">in</span> <span class="fu">seq_len</span>(<span class="fu">max</span>(fold))){</span>
<span id="cb12-4"><a></a>    ytr <span class="ot">&lt;-</span> y; ytr[fold <span class="sc">==</span> k] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb12-5"><a></a>    fit <span class="ot">&lt;-</span> <span class="fu">BGLR</span>(<span class="at">y =</span> ytr, <span class="at">ETA =</span> <span class="fu">list</span>(<span class="fu">list</span>(<span class="at">K =</span> K, <span class="at">model =</span> <span class="st">"RKHS"</span>)),</span>
<span id="cb12-6"><a></a>                <span class="at">nIter =</span> nIter, <span class="at">burnIn =</span> burnIn, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb12-7"><a></a>    pred[fold <span class="sc">==</span> k] <span class="ot">&lt;-</span> fit<span class="sc">$</span>yHat[fold <span class="sc">==</span> k]</span>
<span id="cb12-8"><a></a>  }</span>
<span id="cb12-9"><a></a>  <span class="fu">list</span>(<span class="at">pred =</span> pred)</span>
<span id="cb12-10"><a></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<ul>
<li>Use <code>BGLR</code> for kernel ridge regression with a given kernel matrix</li>
<li>The <code>"RKHS"</code> option directs BGLR to fit a kernel ridge model with covariance <span class="math inline">\sigma_g^2 K</span>, so shrinkage follows the supplied similarity matrix</li>
</ul>
</section>
<section id="ag-blup-models-1" class="slide level2">
<h2>A+G-BLUP Models</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a></a>cv_bglr_AplusG <span class="ot">&lt;-</span> <span class="cf">function</span>(y, A, G, fold, <span class="at">nIter =</span> <span class="dv">6000</span>, <span class="at">burnIn =</span> <span class="dv">1000</span>){</span>
<span id="cb13-2"><a></a>  pred <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA_real_</span>, <span class="fu">length</span>(y))</span>
<span id="cb13-3"><a></a>  <span class="cf">for</span>(k <span class="cf">in</span> <span class="fu">seq_len</span>(<span class="fu">max</span>(fold))){</span>
<span id="cb13-4"><a></a>    ytr <span class="ot">&lt;-</span> y; ytr[fold <span class="sc">==</span> k] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb13-5"><a></a>    fit <span class="ot">&lt;-</span> <span class="fu">BGLR</span>(<span class="at">y =</span> ytr,</span>
<span id="cb13-6"><a></a>                <span class="at">ETA =</span> <span class="fu">list</span>(<span class="fu">list</span>(<span class="at">K =</span> A, <span class="at">model =</span> <span class="st">"RKHS"</span>),</span>
<span id="cb13-7"><a></a>                           <span class="fu">list</span>(<span class="at">K =</span> G, <span class="at">model =</span> <span class="st">"RKHS"</span>)),</span>
<span id="cb13-8"><a></a>                <span class="at">nIter =</span> nIter, <span class="at">burnIn =</span> burnIn, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb13-9"><a></a>    pred[fold <span class="sc">==</span> k] <span class="ot">&lt;-</span> fit<span class="sc">$</span>yHat[fold <span class="sc">==</span> k]</span>
<span id="cb13-10"><a></a>  }</span>
<span id="cb13-11"><a></a>  <span class="fu">list</span>(<span class="at">pred =</span> pred)</span>
<span id="cb13-12"><a></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<ul>
<li>Combine pedigree and genomic relationship matrices in a two-kernel model</li>
<li>This allows modeling both additive genetic effects from pedigree and genomic information</li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  Model            avg_acc sd_acc avg_rmse sd_rmse
  &lt;chr&gt;              &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;
1 A+G (two-kernel)   0.546 0.0336    0.845  0.0358
2 A-BLUP (A)         0.439 0.0391    0.903  0.0503
3 GBLUP (G)          0.535 0.0543    0.854  0.0402
4 Ridge(Z)           0.533 0.0539    0.854  0.0412</code></pre>
</div>
</div>
</section>
<section id="results" class="slide level2">
<h2>Results</h2>



<img data-src="lecture-06_files/figure-revealjs/unnamed-chunk-15-1.png" width="960" class="r-stretch"></section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">

</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
          const outerScaffold = trigger.parentElement.cloneNode(true);
          const codeEl = outerScaffold.querySelector('code');
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp('/' + window.location.host + '/');
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    

</body></html>