<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Lecture 05: GWAS in R – Statistical Genetics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../course_logo.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-8991b0faf62225856095ff2fe360da1f.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.css">

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN" && texText && texText.data) {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-sidebar docked quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Lecture 05: GWAS in R</li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
      <a href="../index.html" class="sidebar-logo-link">
      <img src="../course_logo.png" alt="" class="sidebar-logo light-content py-0 d-lg-inline d-none">
      <img src="../course_logo.png" alt="" class="sidebar-logo dark-content py-0 d-lg-inline d-none">
      </a>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link"><i class="bi bi-calendar" role="img">
</i> 
 <span class="menu-text">Schedule</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../syllabus.html" class="sidebar-item-text sidebar-link"><i class="bi bi-book" role="img">
</i> 
 <span class="menu-text">Syllabus</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../faq.html" class="sidebar-item-text sidebar-link"><i class="bi bi-question" role="img">
</i> 
 <span class="menu-text">FAQ</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../links.html" class="sidebar-item-text sidebar-link"><i class="bi bi-link" role="img">
</i> 
 <span class="menu-text">Useful Links</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://github.com/gwcbi/StatGen" class="sidebar-item-text sidebar-link" target="_blank"><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text">GitHub</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://www.zotero.org/groups/6113424/pubh-8878/library" class="sidebar-item-text sidebar-link" target="_blank"><i class="bi bi-book" role="img">
</i> 
 <span class="menu-text">Zotero Library</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#overview" id="toc-overview" class="nav-link active" data-scroll-target="#overview">Overview</a></li>
  <li><a href="#setup" id="toc-setup" class="nav-link" data-scroll-target="#setup">Setup</a></li>
  <li><a href="#step-1-accessing-and-loading-data" id="toc-step-1-accessing-and-loading-data" class="nav-link" data-scroll-target="#step-1-accessing-and-loading-data">Step 1: Accessing and Loading Data</a>
  <ul class="collapse">
  <li><a href="#hapmap-3" id="toc-hapmap-3" class="nav-link" data-scroll-target="#hapmap-3">HapMap 3</a></li>
  </ul></li>
  <li><a href="#step-2-pre-modeling-qc" id="toc-step-2-pre-modeling-qc" class="nav-link" data-scroll-target="#step-2-pre-modeling-qc">Step 2: Pre-Modeling QC</a>
  <ul class="collapse">
  <li><a href="#genotype-quality-control" id="toc-genotype-quality-control" class="nav-link" data-scroll-target="#genotype-quality-control">Genotype quality control</a></li>
  <li><a href="#sample-level-missingness-heterozygosity" id="toc-sample-level-missingness-heterozygosity" class="nav-link" data-scroll-target="#sample-level-missingness-heterozygosity">Sample-level missingness &amp; heterozygosity</a></li>
  <li><a href="#thresholding" id="toc-thresholding" class="nav-link" data-scroll-target="#thresholding">Thresholding</a></li>
  <li><a href="#snp-correlation-structure" id="toc-snp-correlation-structure" class="nav-link" data-scroll-target="#snp-correlation-structure">SNP correlation structure</a></li>
  <li><a href="#ld-pruning" id="toc-ld-pruning" class="nav-link" data-scroll-target="#ld-pruning">LD pruning</a></li>
  <li><a href="#relatedness" id="toc-relatedness" class="nav-link" data-scroll-target="#relatedness">Relatedness</a></li>
  <li><a href="#hwe-within-ancestry-groups-metadata-unrelateds" id="toc-hwe-within-ancestry-groups-metadata-unrelateds" class="nav-link" data-scroll-target="#hwe-within-ancestry-groups-metadata-unrelateds">HWE within ancestry groups (metadata + unrelateds)</a></li>
  <li><a href="#ancestry-pcs" id="toc-ancestry-pcs" class="nav-link" data-scroll-target="#ancestry-pcs">Ancestry PCs</a></li>
  <li><a href="#ancestryadjusted-relatedness" id="toc-ancestryadjusted-relatedness" class="nav-link" data-scroll-target="#ancestryadjusted-relatedness">Ancestry‑adjusted relatedness</a></li>
  <li><a href="#construct-phenotype-and-merge-covariates" id="toc-construct-phenotype-and-merge-covariates" class="nav-link" data-scroll-target="#construct-phenotype-and-merge-covariates">Construct phenotype and merge covariates</a></li>
  </ul></li>
  <li><a href="#step-3-modeling" id="toc-step-3-modeling" class="nav-link" data-scroll-target="#step-3-modeling">Step 3: Modeling</a>
  <ul class="collapse">
  <li><a href="#null-model" id="toc-null-model" class="nav-link" data-scroll-target="#null-model">Null model</a></li>
  <li><a href="#association-testing" id="toc-association-testing" class="nav-link" data-scroll-target="#association-testing">Association testing</a></li>
  </ul></li>
  <li><a href="#step-4-postgwas" id="toc-step-4-postgwas" class="nav-link" data-scroll-target="#step-4-postgwas">Step 4: Post‑GWAS</a>
  <ul class="collapse">
  <li><a href="#extracting-results" id="toc-extracting-results" class="nav-link" data-scroll-target="#extracting-results">Extracting results</a></li>
  <li><a href="#manhattan-plot" id="toc-manhattan-plot" class="nav-link" data-scroll-target="#manhattan-plot">Manhattan plot</a></li>
  <li><a href="#qq-plot" id="toc-qq-plot" class="nav-link" data-scroll-target="#qq-plot">QQ Plot</a></li>
  <li><a href="#genomic-control-inflation-factor" id="toc-genomic-control-inflation-factor" class="nav-link" data-scroll-target="#genomic-control-inflation-factor">Genomic-control inflation factor</a></li>
  <li><a href="#cleanup" id="toc-cleanup" class="nav-link" data-scroll-target="#cleanup">Cleanup</a></li>
  </ul></li>
  <li><a href="#when-to-use-a-different-tool" id="toc-when-to-use-a-different-tool" class="nav-link" data-scroll-target="#when-to-use-a-different-tool">When to use a different tool?</a></li>
  <li><a href="#further-reading" id="toc-further-reading" class="nav-link" data-scroll-target="#further-reading">Further reading</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">Session Info</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Lecture 05: GWAS in R</h1>
<p class="subtitle lead">PUBH 8878, Statistical Genetics</p>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="overview" class="level2">
<h2 class="anchored" data-anchor-id="overview">Overview</h2>
<p>This session walks through a minimal GWAS workflow in R using <code>SNPRelate</code> and <code>GENESIS</code>. There exist many tools for GWAS, including <code>PLINK</code>, <code>BOLT-LMM</code>, <code>REGENIE</code>, <code>SAIGE</code>, and others. Here, we focus on <code>GENESIS</code> to illustrate the key steps.</p>
<p>We work with a small example <code>GDS</code> bundled with GENESIS to keep runtime short while illustrating the full pipeline:</p>
<ol type="1">
<li>Quality control (variant- and sample‑level)</li>
<li>LD pruning for structure/relatedness tasks</li>
<li>Relatedness (KING‑robust) and population structure (PC‑AiR)</li>
<li>Null model fitting (linear mixed model)</li>
<li>Single‑variant association and diagnostics (QQ, λGC)</li>
</ol>
</section>
<section id="setup" class="level2">
<h2 class="anchored" data-anchor-id="setup">Setup</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-1"><pre class="sourceCode r code-annotation-code code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-1" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-1-1" class="code-annotation-target"><a href="#annotated-cell-1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SNPRelate)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-1" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-1-2" class="code-annotation-target"><a href="#annotated-cell-1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GENESIS)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-1" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-1-3" class="code-annotation-target"><a href="#annotated-cell-1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GWASTools)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-1" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-1-4" class="code-annotation-target"><a href="#annotated-cell-1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(qqman)</span>
<span id="annotated-cell-1-5"><a href="#annotated-cell-1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="annotated-cell-1-6"><a href="#annotated-cell-1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="annotated-cell-1-7"><a href="#annotated-cell-1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="annotated-cell-1-8"><a href="#annotated-cell-1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">8878</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-1" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-1" data-code-lines="1" data-code-annotation="1"><code>SNPRelate</code> (Zheng et al., 2012) for GDS file handling and genotype data management <span class="citation" data-cites="zhengHighperformanceComputingToolset2012">(<a href="#ref-zhengHighperformanceComputingToolset2012" role="doc-biblioref">Zheng <em>et al.</em>, 2012</a>)</span></span>
</dd>
<dt data-target-cell="annotated-cell-1" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-1" data-code-lines="2" data-code-annotation="2"><code>GENESIS</code> for genome-wide association analysis and relatedness estimation <span class="citation" data-cites="manichaikulRobustRelationshipInference2010 conomosPCAiR2015">(<a href="#ref-manichaikulRobustRelationshipInference2010" role="doc-biblioref">Manichaikul <em>et al.</em>, 2010</a>; <a href="#ref-conomosPCAiR2015" role="doc-biblioref">Conomos <em>et al.</em>, 2015</a>)</span></span>
</dd>
<dt data-target-cell="annotated-cell-1" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-1" data-code-lines="3" data-code-annotation="3"><code>GWASTools</code> for genotype data management and quality control</span>
</dd>
<dt data-target-cell="annotated-cell-1" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-1" data-code-lines="4" data-code-annotation="4"><code>qqman</code> for creating Q-Q and Manhattan plots</span>
</dd>
</dl>
</div>
</div>
</section>
<section id="step-1-accessing-and-loading-data" class="level2">
<h2 class="anchored" data-anchor-id="step-1-accessing-and-loading-data">Step 1: Accessing and Loading Data</h2>
<section id="hapmap-3" class="level3">
<h3 class="anchored" data-anchor-id="hapmap-3">HapMap 3</h3>
<ul>
<li>HapMap3 is an integrated reference of common and low-frequency alleles. The project genotyped 1.6 million common SNPs in 1,184 reference individuals from 11 global populations. <span class="citation" data-cites="hapmap3IntegratingCommonRare2010">(<a href="#ref-hapmap3IntegratingCommonRare2010" role="doc-biblioref">The International HapMap 3 Consortium, 2010</a>)</span></li>
<li>The project mapped linkage disequilibrium (LD) patterns used for tag‑SNP selection and served as an early imputation reference. <span class="citation" data-cites="hapmap3IntegratingCommonRare2010 internationalHapMapConsortium2005">(<a href="#ref-hapmap3IntegratingCommonRare2010" role="doc-biblioref">The International HapMap 3 Consortium, 2010</a>; <a href="#ref-internationalHapMapConsortium2005" role="doc-biblioref">The International HapMap Consortium, 2005</a>)</span></li>
<li>Here, we will use a small subset from ASW (African ancestry in the Southwest USA) and MXL (Mexican ancestry in Los Angeles) samples</li>
</ul>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-2"><pre class="sourceCode r code-annotation-code code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-2-1" class="code-annotation-target"><a href="#annotated-cell-2-1" aria-hidden="true" tabindex="-1"></a>gdsfile <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"HapMap_ASW_MXL_geno.gds"</span>, <span class="at">package =</span> <span class="st">"GENESIS"</span>)</span>
<span id="annotated-cell-2-2"><a href="#annotated-cell-2-2" aria-hidden="true" tabindex="-1"></a></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-2-3" class="code-annotation-target"><a href="#annotated-cell-2-3" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> SNPRelate<span class="sc">::</span><span class="fu">snpgdsOpen</span>(gdsfile)</span>
<span id="annotated-cell-2-4"><a href="#annotated-cell-2-4" aria-hidden="true" tabindex="-1"></a></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-2-5" class="code-annotation-target"><a href="#annotated-cell-2-5" aria-hidden="true" tabindex="-1"></a>samp.id <span class="ot">&lt;-</span> <span class="fu">read.gdsn</span>(<span class="fu">index.gdsn</span>(g, <span class="st">"sample.id"</span>))</span>
<span id="annotated-cell-2-6"><a href="#annotated-cell-2-6" aria-hidden="true" tabindex="-1"></a>snp.id <span class="ot">&lt;-</span> <span class="fu">read.gdsn</span>(<span class="fu">index.gdsn</span>(g, <span class="st">"snp.id"</span>))</span>
<span id="annotated-cell-2-7"><a href="#annotated-cell-2-7" aria-hidden="true" tabindex="-1"></a>chr <span class="ot">&lt;-</span> <span class="fu">read.gdsn</span>(<span class="fu">index.gdsn</span>(g, <span class="st">"snp.chromosome"</span>))</span>
<span id="annotated-cell-2-8"><a href="#annotated-cell-2-8" aria-hidden="true" tabindex="-1"></a>pos <span class="ot">&lt;-</span> <span class="fu">read.gdsn</span>(<span class="fu">index.gdsn</span>(g, <span class="st">"snp.position"</span>))</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-2" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="1" data-code-annotation="1">Locate the example GDS file included with the GENESIS package</span>
</dd>
<dt data-target-cell="annotated-cell-2" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="3" data-code-annotation="2">Open the GDS file for random access to genotype data</span>
</dd>
<dt data-target-cell="annotated-cell-2" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="5,6,7,8" data-code-annotation="3">Read sample IDs, SNP IDs, chromosome numbers, and positions from the GDS file</span>
</dd>
</dl>
</div>
</div>
<p>We can also download the associated metadata</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_tsv</span>(<span class="st">"https://ftp.ncbi.nlm.nih.gov/hapmap/genotypes/hapmap3_r3/relationships_w_pops_041510.txt"</span>, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(metadata)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 7
  FID   IID     dad     mom       sex pheno population
  &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;     
1 2427  NA19919 NA19908 NA19909     1     0 ASW       
2 2431  NA19916 0       0           1     0 ASW       
3 2424  NA19835 0       0           2     0 ASW       
4 2469  NA20282 0       0           2     0 ASW       
5 2368  NA19703 0       0           1     0 ASW       
6 2425  NA19902 NA19900 NA19901     2     0 ASW       </code></pre>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="sc">|&gt;</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">population =</span> <span class="fu">case_when</span>(population <span class="sc">==</span> <span class="st">"MEX"</span> <span class="sc">~</span> <span class="st">"MXL"</span>, </span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> population))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,397 × 7
   FID   IID     dad     mom       sex pheno population
   &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;     
 1 2427  NA19919 NA19908 NA19909     1     0 ASW       
 2 2431  NA19916 0       0           1     0 ASW       
 3 2424  NA19835 0       0           2     0 ASW       
 4 2469  NA20282 0       0           2     0 ASW       
 5 2368  NA19703 0       0           1     0 ASW       
 6 2425  NA19902 NA19900 NA19901     2     0 ASW       
 7 2425  NA19901 0       0           2     0 ASW       
 8 2427  NA19908 0       0           1     0 ASW       
 9 2430  NA19914 0       0           2     0 ASW       
10 2470  NA20287 0       0           2     0 ASW       
# ℹ 1,387 more rows</code></pre>
</div>
</div>
</section>
</section>
<section id="step-2-pre-modeling-qc" class="level2">
<h2 class="anchored" data-anchor-id="step-2-pre-modeling-qc">Step 2: Pre-Modeling QC</h2>
<section id="genotype-quality-control" class="level3">
<h3 class="anchored" data-anchor-id="genotype-quality-control">Genotype quality control</h3>
<section id="filtering-for-minor-allele-frequency-maf" class="level4">
<h4 class="anchored" data-anchor-id="filtering-for-minor-allele-frequency-maf">Filtering for minor allele frequency (MAF)</h4>
<ul>
<li>SNP-chips often determine genotypes based on intensities for each allele</li>
<li>When MAF is low, genotype clusters can overlap, leading to genotyping errors</li>
</ul>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lecture-05_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can compute various statistics regarding allele frequency using <code>snpgdsSNPRateFreq</code></p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-5"><pre class="sourceCode r code-annotation-code code-with-copy"><code class="sourceCode r"><span id="annotated-cell-5-1"><a href="#annotated-cell-5-1" aria-hidden="true" tabindex="-1"></a>af <span class="ot">&lt;-</span> <span class="fu">snpgdsSNPRateFreq</span>(g)</span>
<span id="annotated-cell-5-2"><a href="#annotated-cell-5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(af)</span>
<span id="annotated-cell-5-3"><a href="#annotated-cell-5-3" aria-hidden="true" tabindex="-1"></a>maf <span class="ot">&lt;-</span> af<span class="sc">$</span>MinorFreq</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 3
 $ AlleleFreq : num [1:20000] 0.39 0.494 0.101 0.486 0.445 ...
 $ MinorFreq  : num [1:20000] 0.39 0.494 0.101 0.486 0.445 ...
 $ MissingRate: num [1:20000] 0 0 0 0 0.00578 ...</code></pre>
</div>
</div>
<p>We can see that <code>af</code> contains both minor allele frequency and missing rate per SNP.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(<span class="at">maf =</span> maf) <span class="sc">|&gt;</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">maf_le05 =</span> maf <span class="sc">&lt;</span> <span class="fl">0.05</span>) <span class="sc">|&gt;</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> maf, <span class="at">fill =</span> maf_le05)) <span class="sc">+</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="fl">0.01</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="fl">0.05</span>), <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>    viridis<span class="sc">::</span><span class="fu">scale_fill_viridis</span>(<span class="at">discrete =</span> <span class="cn">TRUE</span>, <span class="at">end =</span> .<span class="dv">9</span>, <span class="at">begin =</span> .<span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">"MAF &lt; .05"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lecture-05_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="variant-missingness" class="level4">
<h4 class="anchored" data-anchor-id="variant-missingness">Variant missingness</h4>
<ul>
<li>Missingness can indicate genotyping errors or batch effects</li>
<li>Missingness can be non-random with respect to phenotype or ancestry</li>
<li>Typically filter variants with <span class="math inline">&gt; 2%</span> missingness</li>
<li>We can use the missingness results from our <code>af</code> object</li>
<li>Additionally, we can retrieve the genotype matrix itself, and manually compute the rate per SNP</li>
</ul>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>G <span class="ot">&lt;-</span> <span class="fu">snpgdsGetGeno</span>(g, <span class="at">with.id =</span> <span class="cn">TRUE</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Genotype matrix: 20000 SNPs X 173 samples</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>geno_mat <span class="ot">&lt;-</span> G<span class="sc">$</span>genotype</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(geno_mat)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 20000   173</code></pre>
</div>
</div>
<p>Indexing in <code>R</code> goes rows-by-columns. So, For <code>geno_mat</code>, the SNPs are rows and the samples are columns. We can inspect subset of the matrix:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>geno_mat[<span class="dv">10</span><span class="sc">:</span><span class="dv">20</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
 [1,]    2    0    0    1    0    1    0    1    2     0
 [2,]    0    1    1    1    1    0    0    0    2     0
 [3,]    1    1    1    1    0    0    1    0    2     1
 [4,]    0    0    0    0    0    0    0    0    0     1
 [5,]    0    0    0    0    0    0    0    0    0     0
 [6,]    1    0    1    1    0    0    0    2    1     0
 [7,]    2    1    1    1    0    0    0    2    0     1
 [8,]    0    0    0    0    1    1    0    0    2     0
 [9,]    1    0    0   NA    0    1    0    2    0     0
[10,]    1    1    0    1    1    1    1    0    0     0
[11,]    1    0    1    0    1    1    0    2    1     0</code></pre>
</div>
</div>
<p><code>SNPRelate</code> allows for four possible values stored in the genotype matrix: 0, 1, 2, 3. Consider a bi-allelic SNP site with possible alleles <span class="math inline">A</span> and <span class="math inline">B</span>.</p>
<ul>
<li>0 indicates <span class="math inline">BB</span></li>
<li>1 indicates <span class="math inline">AB</span></li>
<li>2 indicates <span class="math inline">AA</span></li>
<li>3 indicates a missing genotype</li>
</ul>
<p>For multi-allelic sites, it is a count of the reference allele.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>miss_var_manual <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(<span class="fu">is.na</span>(geno_mat))</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>miss_var <span class="ot">&lt;-</span> af<span class="sc">$</span>MissingRate </span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="fu">identical</span>(miss_var, miss_var_manual)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(<span class="at">miss_var =</span> miss_var) <span class="sc">|&gt;</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">miss_var_gt02 =</span> miss_var <span class="sc">&gt;</span> <span class="fl">0.02</span>) <span class="sc">|&gt;</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> miss_var, <span class="at">fill =</span> miss_var_gt02)) <span class="sc">+</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="fl">0.001</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="fl">0.02</span>), <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>    viridis<span class="sc">::</span><span class="fu">scale_fill_viridis</span>(<span class="at">discrete =</span> <span class="cn">TRUE</span>, <span class="at">end =</span> .<span class="dv">9</span>, <span class="at">begin =</span> .<span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">"Missing &gt; .02"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lecture-05_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="sample-level-missingness-heterozygosity" class="level3">
<h3 class="anchored" data-anchor-id="sample-level-missingness-heterozygosity">Sample-level missingness &amp; heterozygosity</h3>
<ul>
<li>Samples with high missingness may have poor DNA quality or technical issues</li>
<li>Samples with extreme heterozygosity rates may be contaminated or mislabelled</li>
</ul>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>samp_miss <span class="ot">&lt;-</span> <span class="fu">snpgdsSampMissRate</span>(g)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(<span class="at">samp_miss =</span> samp_miss) <span class="sc">|&gt;</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">samp_outlier =</span> samp_miss <span class="sc">&gt;</span> <span class="fl">0.02</span>) <span class="sc">|&gt;</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> samp_miss, <span class="at">fill =</span> samp_outlier)) <span class="sc">+</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="fl">0.01</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="fl">0.02</span>), <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Sample missingness"</span>, <span class="at">x =</span> <span class="st">"Missing rate"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lecture-05_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>het_rate <span class="ot">&lt;-</span> <span class="fu">colMeans</span>(geno_mat <span class="sc">==</span> <span class="dv">1</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(<span class="at">het_rate =</span> het_rate) <span class="sc">|&gt;</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">het_outlier =</span> <span class="fu">abs</span>(<span class="fu">scale</span>(het_rate)) <span class="sc">&gt;</span> <span class="dv">3</span>) <span class="sc">|&gt;</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> het_rate, <span class="at">fill =</span> het_outlier)) <span class="sc">+</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="fl">0.01</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="fu">mean</span>(het_rate) <span class="sc">+</span> <span class="dv">3</span> <span class="sc">*</span> <span class="fu">sd</span>(het_rate),</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>                              <span class="fu">mean</span>(het_rate) <span class="sc">-</span> <span class="dv">3</span> <span class="sc">*</span> <span class="fu">sd</span>(het_rate)),</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>               <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Heterozygosity rate"</span>, <span class="at">x =</span> <span class="st">"Heterozygosity rate"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lecture-05_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="thresholding" class="level3">
<h3 class="anchored" data-anchor-id="thresholding">Thresholding</h3>
<p>Putting it together: below we (i) filter samples on missingness/heterozygosity, (ii) apply preliminary variant filters (missingness + MAF), (iii) LD‑prune and estimate relatedness (KING) on that prelim set, and later (iv) compute HWE within ancestry‑homogeneous unrelated subsets using the metadata to finalize variant inclusion. This order avoids false HWE failures from population mixture/relatedness. <span class="citation" data-cites="andersonQC2010">(<a href="#ref-andersonQC2010" role="doc-biblioref">Anderson <em>et al.</em>, 2010</a>)</span></p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>thr_samp_miss <span class="ot">&lt;-</span> <span class="fl">0.02</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>thr_var_miss <span class="ot">&lt;-</span> <span class="fl">0.02</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>thr_maf <span class="ot">&lt;-</span> <span class="fl">0.05</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>thr_het_z <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>keep_sample <span class="ot">&lt;-</span> <span class="fu">which</span>(samp_miss <span class="sc">&lt;=</span> thr_samp_miss <span class="sc">&amp;</span> <span class="fu">abs</span>(<span class="fu">scale</span>(het_rate)) <span class="sc">&lt;</span> thr_het_z)</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>samp.keep.id <span class="ot">&lt;-</span> samp.id[keep_sample]</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>keep_snp_prelim <span class="ot">&lt;-</span> <span class="fu">which</span>(miss_var <span class="sc">&lt;=</span> thr_var_miss <span class="sc">&amp;</span> maf <span class="sc">&gt;=</span> thr_maf)</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>snp.prelim.id <span class="ot">&lt;-</span> snp.id[keep_snp_prelim]</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(snp.prelim.id)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 18960</code></pre>
</div>
</div>
</section>
<section id="snp-correlation-structure" class="level3">
<h3 class="anchored" data-anchor-id="snp-correlation-structure">SNP correlation structure</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>max_plot_snps <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>ld_snp_ids <span class="ot">&lt;-</span> snp.prelim.id[<span class="fu">seq_len</span>(max_plot_snps)]</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>ld_mat <span class="ot">&lt;-</span> <span class="fu">snpgdsLDMat</span>(</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>    g,</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">sample.id =</span> samp.keep.id,</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">snp.id =</span> ld_snp_ids,</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">slide =</span> <span class="sc">-</span><span class="dv">1</span>,</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">method =</span> <span class="st">"corr"</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Linkage Disequilibrium (LD) estimation on genotypes:
    # of samples: 167
    # of SNPs: 100
    using 1 thread
    method: correlation
LD matrix:    the sum of all selected genotypes (0,1,2) = 9659</code></pre>
</div>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lecture-05_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="ld-pruning" class="level3">
<h3 class="anchored" data-anchor-id="ld-pruning">LD pruning</h3>
<p>LD pruning removes variants in high LD to yield a subset of approximately independent variants. This is useful for population structure and relatedness estimation, which can be distorted by large blocks of correlated SNPs.</p>
<p>The algorithm works as follows <span class="citation" data-cites="zhengHighperformanceComputingToolset2012">(<a href="#ref-zhengHighperformanceComputingToolset2012" role="doc-biblioref">Zheng <em>et al.</em>, 2012</a>)</span>:</p>
<ol type="1">
<li>Randomly select a starting position i (<code>start.pos="random"</code>), i=1 if <code>start.pos="first"</code>, or i=last if <code>start.pos="last"</code>; and let the current SNP set S={ i };</li>
<li>For each right position j from i+1 to n: if any LD between j and k is greater than <code>ld.threshold</code>, where k belongs to S, and both of j and k are in the sliding window, then skip j; otherwise, let S be S + { j };</li>
<li>For each left position j from i-1 to 1: if any LD between j and k is greater than <code>ld.threshold</code>, where k belongs to S, and both of j and k are in the sliding window, then skip j; otherwise, let S be S + { j };</li>
<li>Output S, the final selection of SNPs.</li>
</ol>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>ld_prune <span class="ot">&lt;-</span> <span class="fu">snpgdsLDpruning</span>(g,</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">sample.id =</span> samp.keep.id, </span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">snp.id =</span> snp.prelim.id,</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">autosome.only =</span> <span class="cn">TRUE</span>,</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">method =</span> <span class="st">"r"</span>,</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">slide.max.bp =</span> <span class="dv">500000</span>,</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">ld.threshold =</span> <span class="fu">sqrt</span>(.<span class="dv">1</span>)</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>SNP pruning based on LD:
Excluding 1,040 SNPs (non-autosomes or non-selection)
Excluding 0 SNP (monomorphic: TRUE, MAF: 0.005, missing rate: 0.05)
    # of samples: 167
    # of SNPs: 18,960
    using 1 thread
    sliding window: 500,000 basepairs, Inf SNPs
    |LD| threshold: 0.316228
    method: R
Chrom 1: |====================|====================|
    21.27%, 4,255 / 20,000 (Fri Oct  3 16:11:37 2025)
4,255 markers are selected in total.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>snps_pruned <span class="ot">&lt;-</span> <span class="fu">unlist</span>(ld_prune, <span class="at">use.names =</span> <span class="cn">FALSE</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="relatedness" class="level3">
<h3 class="anchored" data-anchor-id="relatedness">Relatedness</h3>
<p>We need to account for relatedness to avoid confounding in PCA and association testing. KING (Kinship‑based INference for GWAS) estimates pairwise kinship robust to population structure <span class="citation" data-cites="manichaikulRobustRelationshipInference2010">(<a href="#ref-manichaikulRobustRelationshipInference2010" role="doc-biblioref">Manichaikul <em>et al.</em>, 2010</a>)</span>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-18"><pre class="sourceCode r code-annotation-code code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-18" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-18-1" class="code-annotation-target"><a href="#annotated-cell-18-1" aria-hidden="true" tabindex="-1"></a>king <span class="ot">&lt;-</span> <span class="fu">snpgdsIBDKING</span>(g,</span>
<span id="annotated-cell-18-2"><a href="#annotated-cell-18-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">sample.id =</span> samp.keep.id,</span>
<span id="annotated-cell-18-3"><a href="#annotated-cell-18-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">snp.id =</span> snps_pruned,</span>
<span id="annotated-cell-18-4"><a href="#annotated-cell-18-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">type =</span> <span class="st">"KING-robust"</span>,</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-18" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-18-5" class="code-annotation-target"><a href="#annotated-cell-18-5" aria-hidden="true" tabindex="-1"></a>                      <span class="at">family.id =</span> metadata<span class="sc">$</span>FID[<span class="fu">match</span>(samp.keep.id, metadata<span class="sc">$</span>IID)])</span>
<span id="annotated-cell-18-6"><a href="#annotated-cell-18-6" aria-hidden="true" tabindex="-1"></a></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-18" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-18-7" class="code-annotation-target"><a href="#annotated-cell-18-7" aria-hidden="true" tabindex="-1"></a>KINGmat <span class="ot">&lt;-</span> GENESIS<span class="sc">::</span><span class="fu">kingToMatrix</span>(king)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-18" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-18" data-code-lines="1,2,3" data-code-annotation="1">Run KING on our dataset with the filtered samples and LD‑pruned SNPs.</span>
</dd>
<dt data-target-cell="annotated-cell-18" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-18" data-code-lines="5" data-code-annotation="2">Model is robust to misspecifications in population structure.</span>
</dd>
<dt data-target-cell="annotated-cell-18" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-18" data-code-lines="7" data-code-annotation="3">Convert into a kinship matrix we can use for downstream tasks.</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Using 167 samples provided</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Identifying clusters of relatives...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>    154 relatives in 29 clusters; largest cluster = 49</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Creating block matrices for clusters...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>13 samples with no relatives included</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Putting all samples together into one block diagonal matrix</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>IBD analysis (KING method of moment) on genotypes:
Excluding 15,745 SNPs (non-autosomes or non-selection)
Excluding 0 SNP (monomorphic: TRUE, MAF: NaN, missing rate: NaN)
    # of samples: 167
    # of SNPs: 4,255
    using 1 thread
# of families: 73, and within- and between-family relationship are estimated differently.
Relationship inference in the presence of population stratification.
KING IBD:    the sum of all selected genotypes (0,1,2) = 365234
CPU capabilities:
Fri Oct  3 16:11:37 2025    (internal increment: 65536)

[..................................................]  0%, ETC: ---        
[==================================================] 100%, completed, 0s
Fri Oct  3 16:11:37 2025    Done.</code></pre>
</div>
</div>
</section>
<section id="hwe-within-ancestry-groups-metadata-unrelateds" class="level3">
<h3 class="anchored" data-anchor-id="hwe-within-ancestry-groups-metadata-unrelateds">HWE within ancestry groups (metadata + unrelateds)</h3>
<p>Under random mating and no technical artifacts, genotype frequencies should follow the familiar proportions under HWE for allele frequency p.&nbsp;A simple check is whether observed counts depart more than sampling noise would predict. In GWAS array data, notable departures often flag lab/array problems rather than biology. Common culprits are poor clustering for one allele, batch effects, or plate swaps (all of which distort heterozygote vs.&nbsp;homozygote balance).</p>
<p>For case-control studies, we compute HWE in controls. True disease associations can produce real departures in cases (e.g., risk allele homozygotes enriched), so filtering on cases risks discarding biological signal. <span class="citation" data-cites="andersonQC2010">(<a href="#ref-andersonQC2010" role="doc-biblioref">Anderson <em>et al.</em>, 2010</a>)</span></p>
<p>We will use an exact test for bi‑allelic SNPs. Asymptotic chi‑square tests can mis‑calibrate at low genotype counts or low MAF, while exact tests keep Type I error in check across MAF ranges. <span class="citation" data-cites="wiggintonExactHWE2005 guoThompson1992">(<a href="#ref-wiggintonExactHWE2005" role="doc-biblioref">Wigginton <em>et al.</em>, 2005</a>; <a href="#ref-guoThompson1992" role="doc-biblioref">Guo and Thompson, 1992</a>)</span></p>
<p>Regarding a specific threshold, large studies typically use $p &lt; 10^{−6} as a conservative filter. Smaller studies sometimes use <span class="math inline">p &lt; 10^{-4}</span>. Because power to detect departures depends on MAF we can consider MAF‑stratified thresholds to avoid over‑penalizing rare variants. <span class="citation" data-cites="andersonQC2010">(<a href="#ref-andersonQC2010" role="doc-biblioref">Anderson <em>et al.</em>, 2010</a>)</span></p>
<p>X/sex chromosomes require specific sex‑aware tests (not used here; single autosome). <span class="citation" data-cites="andersonQC2010">(<a href="#ref-andersonQC2010" role="doc-biblioref">Anderson <em>et al.</em>, 2010</a>)</span></p>
<p>To avoid the Wahlund effect (heterozygote deficit from pooling distinct ancestries) and non‑independence from relatives, we compute HWE within ancestry‑homogeneous, unrelated subsets. We use KING to define unrelateds and the metadata population labels (ASW, MXL) for grouping.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-19"><pre class="sourceCode r code-annotation-code code-with-copy code-annotated"><code class="sourceCode r"><span id="annotated-cell-19-1"><a href="#annotated-cell-19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify unrelateds (3rd‑degree or closer considered related)</span></span>
<span id="annotated-cell-19-2"><a href="#annotated-cell-19-2" aria-hidden="true" tabindex="-1"></a>km <span class="ot">&lt;-</span> KINGmat <span class="sc">|&gt;</span> <span class="fu">as.matrix</span>()</span>
<span id="annotated-cell-19-3"><a href="#annotated-cell-19-3" aria-hidden="true" tabindex="-1"></a><span class="fu">diag</span>(km) <span class="ot">&lt;-</span> <span class="dv">0</span> </span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-19" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-19-4" class="code-annotation-target"><a href="#annotated-cell-19-4" aria-hidden="true" tabindex="-1"></a>unrel_ids <span class="ot">&lt;-</span> <span class="fu">rownames</span>(km)[<span class="fu">rowSums</span>(km <span class="sc">&gt;=</span> <span class="fl">0.0442</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>) <span class="sc">==</span> <span class="dv">0</span>]</span>
<span id="annotated-cell-19-5"><a href="#annotated-cell-19-5" aria-hidden="true" tabindex="-1"></a>unrel_ids <span class="ot">&lt;-</span> <span class="fu">intersect</span>(unrel_ids, samp.keep.id)</span>
<span id="annotated-cell-19-6"><a href="#annotated-cell-19-6" aria-hidden="true" tabindex="-1"></a></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-19" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-19-7" class="code-annotation-target"><a href="#annotated-cell-19-7" aria-hidden="true" tabindex="-1"></a>grp <span class="ot">&lt;-</span> <span class="fu">split</span>(unrel_ids, metadata<span class="sc">$</span>population[<span class="fu">match</span>(unrel_ids, metadata<span class="sc">$</span>IID)])</span>
<span id="annotated-cell-19-8"><a href="#annotated-cell-19-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-19-9"><a href="#annotated-cell-19-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Exact HWE p-values per group on prelim variants</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-19" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-19-10" class="code-annotation-target"><a href="#annotated-cell-19-10" aria-hidden="true" tabindex="-1"></a>p_by_grp <span class="ot">&lt;-</span> <span class="fu">lapply</span>(grp, <span class="cf">function</span>(ids) {</span>
<span id="annotated-cell-19-11"><a href="#annotated-cell-19-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">snpgdsHWE</span>(g, <span class="at">snp.id =</span> snp.prelim.id, <span class="at">sample.id =</span> ids)</span>
<span id="annotated-cell-19-12"><a href="#annotated-cell-19-12" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="annotated-cell-19-13"><a href="#annotated-cell-19-13" aria-hidden="true" tabindex="-1"></a>p_mat <span class="ot">&lt;-</span> <span class="fu">do.call</span>(cbind, p_by_grp)</span>
<span id="annotated-cell-19-14"><a href="#annotated-cell-19-14" aria-hidden="true" tabindex="-1"></a></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-19" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-19-15" class="code-annotation-target"><a href="#annotated-cell-19-15" aria-hidden="true" tabindex="-1"></a>thr_hwe <span class="ot">&lt;-</span> <span class="fl">1e-6</span></span>
<span id="annotated-cell-19-16"><a href="#annotated-cell-19-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-19-17"><a href="#annotated-cell-19-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Conservative rule: pass only if all groups pass</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-19" data-target-annotation="5" onclick="event.preventDefault();">5</a><span id="annotated-cell-19-18" class="code-annotation-target"><a href="#annotated-cell-19-18" aria-hidden="true" tabindex="-1"></a>pass_all <span class="ot">&lt;-</span> <span class="fu">apply</span>(p_mat, <span class="dv">1</span>, <span class="cf">function</span>(p) <span class="fu">all</span>(<span class="fu">is.na</span>(p) <span class="sc">|</span> p <span class="sc">&gt;</span> thr_hwe))</span>
<span id="annotated-cell-19-19"><a href="#annotated-cell-19-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-19-20"><a href="#annotated-cell-19-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Final variant set after HWE</span></span>
<span id="annotated-cell-19-21"><a href="#annotated-cell-19-21" aria-hidden="true" tabindex="-1"></a>keep_snp <span class="ot">&lt;-</span> <span class="fu">which</span>(pass_all)</span>
<span id="annotated-cell-19-22"><a href="#annotated-cell-19-22" aria-hidden="true" tabindex="-1"></a>snp.keep.id <span class="ot">&lt;-</span> snp.prelim.id[keep_snp]</span>
<span id="annotated-cell-19-23"><a href="#annotated-cell-19-23" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(snp.prelim.id)</span>
<span id="annotated-cell-19-24"><a href="#annotated-cell-19-24" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(snp.keep.id)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-19" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-19" data-code-lines="4" data-code-annotation="1">Define unrelateds as those with kinship &lt; 0.0442 (3rd-degree or closer) to any other sample.</span>
</dd>
<dt data-target-cell="annotated-cell-19" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-19" data-code-lines="7" data-code-annotation="2">Split unrelateds by population using metadata</span>
</dd>
<dt data-target-cell="annotated-cell-19" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-19" data-code-lines="10" data-code-annotation="3">Compute exact HWE p-values for each group on the preliminary variant set</span>
</dd>
<dt data-target-cell="annotated-cell-19" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-19" data-code-lines="15" data-code-annotation="4">Set a stringent HWE p-value threshold</span>
</dd>
<dt data-target-cell="annotated-cell-19" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-19" data-code-lines="18" data-code-annotation="5">Retain variants that pass HWE in all groups</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 18960
[1] 18960</code></pre>
</div>
</div>
<p>Note that we end up not filtering any variants with this dataset.</p>
</section>
<section id="ancestry-pcs" class="level3">
<h3 class="anchored" data-anchor-id="ancestry-pcs">Ancestry PCs</h3>
<p>With a kinship matrix in hand, we estimate ancestry PCs to use as covariates. PC‑AiR computes PCs on a subset of unrelated individuals and projects relateds onto that space, providing robust structure estimates in the presence of relatedness <span class="citation" data-cites="conomosPCAiR2015">(<a href="#ref-conomosPCAiR2015" role="doc-biblioref">Conomos <em>et al.</em>, 2015</a>)</span>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-20"><pre class="sourceCode r code-annotation-code code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-20" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-20-1" class="code-annotation-target"><a href="#annotated-cell-20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">snpgdsClose</span>(g)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-20" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-20-2" class="code-annotation-target"><a href="#annotated-cell-20-2" aria-hidden="true" tabindex="-1"></a>geno_reader <span class="ot">&lt;-</span> GWASTools<span class="sc">::</span><span class="fu">GdsGenotypeReader</span>(gdsfile)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-20" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-20-3" class="code-annotation-target"><a href="#annotated-cell-20-3" aria-hidden="true" tabindex="-1"></a>geno_data <span class="ot">&lt;-</span> <span class="fu">GenotypeData</span>(geno_reader)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-20" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-20-4" class="code-annotation-target"><a href="#annotated-cell-20-4" aria-hidden="true" tabindex="-1"></a>pcair_out <span class="ot">&lt;-</span> <span class="fu">pcair</span>(geno_data,</span>
<span id="annotated-cell-20-5"><a href="#annotated-cell-20-5" aria-hidden="true" tabindex="-1"></a>                   <span class="at">kinobj =</span> KINGmat,</span>
<span id="annotated-cell-20-6"><a href="#annotated-cell-20-6" aria-hidden="true" tabindex="-1"></a>                   <span class="at">divobj =</span> KINGmat,</span>
<span id="annotated-cell-20-7"><a href="#annotated-cell-20-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">snp.include =</span> snps_pruned,</span>
<span id="annotated-cell-20-8"><a href="#annotated-cell-20-8" aria-hidden="true" tabindex="-1"></a>                   <span class="at">sample.include =</span> samp.keep.id)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-20" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-20" data-code-lines="1" data-code-annotation="1">We need to close the SNPRelate connection before opening with GWASTools</span>
</dd>
<dt data-target-cell="annotated-cell-20" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-20" data-code-lines="2" data-code-annotation="2">Open the GDS file with GWASTools</span>
</dd>
<dt data-target-cell="annotated-cell-20" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-20" data-code-lines="3" data-code-annotation="3">Create a GenotypeData object for analysis</span>
</dd>
<dt data-target-cell="annotated-cell-20" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-20" data-code-lines="4,5,6,7,8" data-code-annotation="4">Run PC-AiR using the KING kinship matrix to identify unrelateds and compute PCs</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Using kinobj and divobj to partition samples into unrelated and related sets</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Working with 167 samples</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Identifying relatives for each sample using kinship threshold 0.0220970869120796</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Identifying pairs of divergent samples using divergence threshold -0.0220970869120796</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Partitioning samples into unrelated and related sets...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Unrelated Set: 92 Samples 
Related Set: 75 Samples</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Performing PCA on the Unrelated Set...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Predicting PC Values for the Related Set...</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Principal Component Analysis (PCA) on genotypes:
Excluding 15,745 SNPs (non-autosomes or non-selection)
Excluding 0 SNP (monomorphic: TRUE, MAF: NaN, missing rate: NaN)
    # of samples: 92
    # of SNPs: 4,255
    using 1 thread
    # of principal components: 32
PCA:    the sum of all selected genotypes (0,1,2) = 201672
CPU capabilities:
Fri Oct  3 16:11:37 2025    (internal increment: 5340)

[..................................................]  0%, ETC: ---        
[==================================================] 100%, completed, 0s
Fri Oct  3 16:11:37 2025    Begin (eigenvalues and eigenvectors)
Fri Oct  3 16:11:37 2025    Done.
SNP Loading:
    # of samples: 92
    # of SNPs: 4,255
    using 1 thread
    using the top 32 eigenvectors
SNP Loading:    the sum of all selected genotypes (0,1,2) = 201672
Fri Oct  3 16:11:37 2025    (internal increment: 42740)

[..................................................]  0%, ETC: ---        
[==================================================] 100%, completed, 0s
Fri Oct  3 16:11:37 2025    Done.
Sample Loading:
    # of samples: 75
    # of SNPs: 4,255
    using 1 thread
    using the top 32 eigenvectors
Sample Loading:    the sum of all selected genotypes (0,1,2) = 163562
Fri Oct  3 16:11:37 2025    (internal increment: 52428)

[..................................................]  0%, ETC: ---        
[==================================================] 100%, completed, 0s
Fri Oct  3 16:11:37 2025    Done.</code></pre>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>pcair_tb <span class="ot">&lt;-</span> pcair_out<span class="sc">$</span>vectors <span class="sc">|&gt;</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as_tibble</span>() <span class="sc">|&gt;</span></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">bind_cols</span>(<span class="st">"IID"</span> <span class="ot">=</span> pcair_out<span class="sc">$</span>sample.id) <span class="sc">|&gt;</span></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">left_join</span>(metadata, <span class="at">by =</span> <span class="st">"IID"</span>)</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>pcair_tb <span class="sc">|&gt;</span></span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(V1, V2, <span class="at">col =</span> population)) <span class="sc">+</span></span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lecture-05_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can use a screeplot to visualize the variance explained by each PC. Depending on where we see an “elbow” in the screeplot, we might choose to include the corresponding PCs in our model.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>pcair_out<span class="sc">$</span>values <span class="sc">|&gt;</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as_tibble</span>() <span class="sc">|&gt;</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">PC =</span> <span class="fu">row_number</span>()) <span class="sc">|&gt;</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(PC, value)) <span class="sc">+</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">1</span>, <span class="dv">10</span>, <span class="at">by =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Screeplot of PC-AiR eigenvalues"</span>, <span class="at">x =</span> <span class="st">"Principal Component"</span>, <span class="at">y =</span> <span class="st">"Eigenvalue"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lecture-05_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="ancestryadjusted-relatedness" class="level3">
<h3 class="anchored" data-anchor-id="ancestryadjusted-relatedness">Ancestry‑adjusted relatedness</h3>
<p>Now that we have ancestry PCs, we estimate relatedness adjusted for ancestry using PC‑Relate and convert it to a GRM suitable for the null model. Training on the unrelated set stabilizes estimates. <span class="citation" data-cites="conomosPCRelate2016">(<a href="#ref-conomosPCRelate2016" role="doc-biblioref">Conomos <em>et al.</em>, 2016</a>)</span></p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>genodata_iter <span class="ot">&lt;-</span> <span class="fu">GenotypeBlockIterator</span>(geno_data, <span class="at">snpInclude=</span>snps_pruned)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>pcrel <span class="ot">&lt;-</span> GENESIS<span class="sc">::</span><span class="fu">pcrelate</span>(</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">gdsobj =</span> genodata_iter,</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">pcs =</span> pcair_out<span class="sc">$</span>vectors[,<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>],</span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">training.set =</span> <span class="fu">intersect</span>(unrel_ids, pcair_out<span class="sc">$</span>sample.id),</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample.include =</span> samp.keep.id,</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Using 6 CPU cores</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>167 samples to be included in the analysis...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Betas for 3 PC(s) will be calculated using 20 samples in training.set...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Running PC-Relate analysis for 167 samples using 4255 SNPs in 1 blocks...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Performing Small Sample Correction...</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>pcrelMat <span class="ot">&lt;-</span> GENESIS<span class="sc">::</span><span class="fu">pcrelateToMatrix</span>(pcrel, <span class="at">scaleKin =</span> <span class="dv">2</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Using 167 samples provided</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Identifying clusters of relatives...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>    167 relatives in 1 clusters; largest cluster = 167</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Creating block matrices for clusters...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>0 samples with no relatives included</code></pre>
</div>
</div>
</section>
<section id="construct-phenotype-and-merge-covariates" class="level3">
<h3 class="anchored" data-anchor-id="construct-phenotype-and-merge-covariates">Construct phenotype and merge covariates</h3>
<p>For the sake of this demo, we will simulate a quantitative trait with a known genetic effect</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">99</span>)</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="fu">length</span>(samp.keep.id)</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Use recorded sex from metadata for covariates (single chromosome: no sex-chr analysis needed)</span></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>sex <span class="ot">&lt;-</span> metadata <span class="sc">|&gt;</span></span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(IID <span class="sc">%in%</span> samp.keep.id) <span class="sc">|&gt;</span></span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">sex =</span> <span class="fu">case_when</span>(sex <span class="sc">==</span> <span class="dv">1</span> <span class="sc">~</span> <span class="st">"M"</span>,</span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a>                           <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"F"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb56-8"><a href="#cb56-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(sex)</span>
<span id="cb56-9"><a href="#cb56-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-10"><a href="#cb56-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Choose one moderately frequent SNP and simulate a quantitative trait with a planted effect</span></span>
<span id="cb56-11"><a href="#cb56-11" aria-hidden="true" tabindex="-1"></a>maf_keep <span class="ot">&lt;-</span> maf[snp.keep.id]</span>
<span id="cb56-12"><a href="#cb56-12" aria-hidden="true" tabindex="-1"></a>ix <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="sc">!</span><span class="fu">is.na</span>(maf_keep) <span class="sc">&amp;</span> maf_keep <span class="sc">&gt;</span> <span class="fl">0.2</span> <span class="sc">&amp;</span> maf_keep <span class="sc">&lt;</span> <span class="fl">0.3</span>)[<span class="dv">1</span>]</span>
<span id="cb56-13"><a href="#cb56-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-14"><a href="#cb56-14" aria-hidden="true" tabindex="-1"></a>G1 <span class="ot">&lt;-</span> GWASTools<span class="sc">::</span><span class="fu">getGenotypeSelection</span>(geno_data, <span class="at">snpID =</span> snp.keep.id[ix], <span class="at">scanID =</span> samp.keep.id)</span>
<span id="cb56-15"><a href="#cb56-15" aria-hidden="true" tabindex="-1"></a>G1 <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(G1) </span>
<span id="cb56-16"><a href="#cb56-16" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fl">0.4</span><span class="sc">*</span> <span class="fu">scale</span>(G1) <span class="sc">+</span> <span class="fl">0.05</span> <span class="sc">*</span> (sex <span class="sc">==</span> <span class="st">"M"</span>) <span class="sc">+</span> <span class="fu">rnorm</span>(N))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>We now have a phenotype vector <code>y</code> and covariates. We need to construct a <code>ScanAnnotationDataFrame</code> object for the null model fitting.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>pcs_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">scanID =</span> pcair_out<span class="sc">$</span>sample.id,</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">PC1 =</span> pcair_out<span class="sc">$</span>vectors[, <span class="dv">1</span>],</span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">PC2 =</span> pcair_out<span class="sc">$</span>vectors[, <span class="dv">2</span>],</span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">PC3 =</span> pcair_out<span class="sc">$</span>vectors[, <span class="dv">3</span>]</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a>scanDF <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">left_join</span>(</span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>(<span class="at">scanID =</span> samp.keep.id, <span class="at">y =</span> y, <span class="at">sex =</span> sex),</span>
<span id="cb57-10"><a href="#cb57-10" aria-hidden="true" tabindex="-1"></a>  pcs_df,</span>
<span id="cb57-11"><a href="#cb57-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">by =</span> <span class="st">"scanID"</span></span>
<span id="cb57-12"><a href="#cb57-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb57-13"><a href="#cb57-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-14"><a href="#cb57-14" aria-hidden="true" tabindex="-1"></a>scanAnnot2 <span class="ot">&lt;-</span> GWASTools<span class="sc">::</span><span class="fu">ScanAnnotationDataFrame</span>(scanDF)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
</section>
<section id="step-3-modeling" class="level2">
<h2 class="anchored" data-anchor-id="step-3-modeling">Step 3: Modeling</h2>
<section id="null-model" class="level3">
<h3 class="anchored" data-anchor-id="null-model">Null model</h3>
<p>We first fit the mixed model under the null hypothesis that each SNP has no effect. The null includes covariates (age, sex, ancestry PCs) and a random genetic effect capturing relatedness; it does not include per‑SNP fixed effects.</p>
<p>The model:</p>
<p><span class="math display">
y = X\beta + g + \epsilon
</span></p>
<p>Where <span class="math inline">y</span> is the phenotype, <span class="math inline">X\beta</span> are fixed‑effect covariates, <span class="math inline">g \sim \mathcal{N}(0, \sigma^2_g K)</span> is the random genetic effect with relatedness matrix <span class="math inline">K</span>, and <span class="math inline">\epsilon \sim \mathcal{N}(0, \sigma^2_e I)</span> is residual error.</p>
<section id="notes-on-estimation" class="level4">
<h4 class="anchored" data-anchor-id="notes-on-estimation">Notes on estimation</h4>
<ul>
<li>Covariance under the null: <span class="math inline">\operatorname{Var}(y) = V = \sigma^2_e I + \sigma^2_g K</span>.</li>
<li>We pass an ancestry‑adjusted GRM from PC‑Relate
<ul>
<li>This yields diagonals near 1 (via <code>pcrelateToMatrix(scaleKin=2)</code>) and controls for population structure in kinship estimation</li>
</ul></li>
<li>Fixed effects (GLS) estimator given variance components: <span class="math inline">\hat{\beta} = (X^\top V^{-1} X)^{-1} X^\top V^{-1} y</span>.</li>
<li>Variance components for Gaussian outcomes are estimated by REML.</li>
<li>Heritability estimate: <span class="math inline">h^2 = \sigma^2_g/(\sigma^2_g + \sigma^2_e)</span>.</li>
</ul>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>nullmod <span class="ot">&lt;-</span> GENESIS<span class="sc">::</span><span class="fu">fitNullModel</span>(</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>    scanAnnot2,</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">outcome =</span> <span class="st">"y"</span>,</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">covars =</span> <span class="fu">c</span>(<span class="st">"sex"</span>, <span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span>),</span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">cov.mat =</span> pcrelMat, <span class="at">family =</span> <span class="st">"gaussian"</span></span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing Variance Component Estimates...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Sigma^2_A     log-lik     RSS</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]    0.5717426    0.5717426 -243.7704375    1.0182557
[1]    0.1884426    0.8753057 -242.1753414    1.0461782
[1]    0.2420265    0.8685482 -242.1442635    1.0026855
[1]    0.2408589    0.8726307 -242.1442236    1.0000083
[1]    0.2411044    0.8724022 -242.1442228    1.0000000</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>nullmod<span class="sc">$</span>varComp</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>        V_A V_resid.var 
  0.2411044   0.8724022 </code></pre>
</div>
</div>
<p><code>varComp</code> reports the variance components from the mixed model. The entry <code>V_A</code> is the additive‑genetic variance captured by our GRM (PC‑Relate), and <code>V_resid.var</code> is the residual variance after adjusting for age, sex, and ancestry PCs. A convenient summary is the mixed‑model heritability on this working scale, <span class="math display">h^2 = \frac{V_A}{V_A + V_{resid}}</span></p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>varc <span class="ot">&lt;-</span> nullmod<span class="sc">$</span>varComp</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>h2 <span class="ot">&lt;-</span> <span class="fu">unname</span>(varc[<span class="st">"V_A"</span>] <span class="sc">/</span> <span class="fu">sum</span>(varc))</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>tibble<span class="sc">::</span><span class="fu">tibble</span>(</span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">component =</span> <span class="fu">c</span>(<span class="st">"V_A"</span>, <span class="st">"V_resid"</span>, <span class="st">"h2_mixed_model"</span>),</span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">value =</span> <span class="fu">c</span>(varc[<span class="st">"V_A"</span>], varc[<span class="st">"V_resid.var"</span>], h2)</span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  component      value
  &lt;chr&gt;          &lt;dbl&gt;
1 V_A            0.241
2 V_resid        0.872
3 h2_mixed_model 0.217</code></pre>
</div>
</div>
</section>
</section>
<section id="association-testing" class="level3">
<h3 class="anchored" data-anchor-id="association-testing">Association testing</h3>
<p>We now test each SNP for association with the phenotype using a score test. The score test is computationally efficient because it only requires fitting the null model once, rather than refitting for each SNP as in a Wald or likelihood ratio test.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Genotype iterator (block-wise scan of all SNPs kept by QC)</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>it <span class="ot">&lt;-</span> GWASTools<span class="sc">::</span><span class="fu">GenotypeBlockIterator</span>(</span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>    geno_data, </span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">snpInclude =</span> snp.keep.id, </span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">snpBlock =</span> <span class="dv">5000</span></span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a>assoc <span class="ot">&lt;-</span> GENESIS<span class="sc">::</span><span class="fu">assocTestSingle</span>(it, <span class="at">null.model =</span> nullmod, <span class="at">test =</span> <span class="st">"Score"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Using 6 CPU cores</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">SNP =</span> assoc<span class="sc">$</span>variant.id,</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">CHR =</span> <span class="fu">as.numeric</span>(assoc<span class="sc">$</span>chr),</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">BP =</span> assoc<span class="sc">$</span>pos,</span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">P =</span> assoc<span class="sc">$</span>Score.pval,</span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">BETA =</span> assoc<span class="sc">$</span>Est,</span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">SE =</span> assoc<span class="sc">$</span>Est.SE</span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
</section>
<section id="step-4-postgwas" class="level2">
<h2 class="anchored" data-anchor-id="step-4-postgwas">Step 4: Post‑GWAS</h2>
<section id="extracting-results" class="level3">
<h3 class="anchored" data-anchor-id="extracting-results">Extracting results</h3>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>        SNP    BP       BETA        SE            P
6         6     6  0.8396265 0.1568205 8.600221e-08
4628   4892  4892 -0.5765968 0.1501810 1.233630e-04
9363   9895  9895 -0.5772324 0.1532537 1.655421e-04
1796   1878  1878 -0.5223076 0.1415787 2.249887e-04
17050 17968 17968  0.4745057 0.1312115 2.987953e-04</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6</code></pre>
</div>
</div>
<p>Note that our top SNP is the one we simulated to have an effect.</p>
</section>
<section id="manhattan-plot" class="level3">
<h3 class="anchored" data-anchor-id="manhattan-plot">Manhattan plot</h3>
<p>We visualize the genome‑wide p-values using a Manhattan plot. The blue line indicates a <span class="math inline">p</span>-value threshold of <span class="math inline">1 \times 10^{-5}</span>. Note that this is equivalent to a Bonferroni correction for 5,000 independent tests. The red line at <span class="math inline">5 \times 10^{-8}</span> is a common genome‑wide significance threshold for GWAS, though it is conservative for smaller studies or those with fewer variants.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="co"># qqman expects columns named: CHR, BP, SNP, P</span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>qqman<span class="sc">::</span><span class="fu">manhattan</span>(</span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>  res,</span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">chr =</span> <span class="st">"CHR"</span>, <span class="at">bp =</span> <span class="st">"BP"</span>, <span class="at">snp =</span> <span class="st">"SNP"</span>, <span class="at">p =</span> <span class="st">"P"</span>,</span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">suggestiveline =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">1e-5</span>)</span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lecture-05_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="qq-plot" class="level3">
<h3 class="anchored" data-anchor-id="qq-plot">QQ Plot</h3>
<p>We compare observed p‑values to the null expectation to assess calibration. Few true signals should deviate strongly.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>qqman<span class="sc">::</span><span class="fu">qq</span>(res<span class="sc">$</span>P)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lecture-05_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="genomic-control-inflation-factor" class="level3">
<h3 class="anchored" data-anchor-id="genomic-control-inflation-factor">Genomic-control inflation factor</h3>
<p>The genomic-control inflation factor is a single-number summary (<span class="math inline">\lambda_{GC}</span>) of how inflated or conservative the genome‑wide test statistics are relative to the null. Values near 1 indicate well‑calibrated tests, &gt;1 suggests inflation (e.g., population structure, relatedness, batch), &lt;1 suggests conservative/underpowered tests</p>
<p>For 1‑df tests, the null statistic follows <span class="math inline">\chi^2_1</span>, whose median is <span class="math inline">qchisq(0.5, 1) \approx 0.456</span>. We will convert our <span class="math inline">p</span>‑values to <span class="math inline">\chi^2</span> statistics <span class="math inline">T_i = \texttt{qchisq}(1 − p_i, df = 1)</span>. Then <span class="math inline">\lambda_{GC} = \text{median}(T_i) / m_0</span>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>lambda_gc <span class="ot">&lt;-</span> <span class="fu">median</span>(stats<span class="sc">::</span><span class="fu">qchisq</span>(<span class="dv">1</span> <span class="sc">-</span> res<span class="sc">$</span>P, <span class="at">df =</span> <span class="dv">1</span>), <span class="at">na.rm =</span> <span class="cn">TRUE</span>) <span class="sc">/</span> <span class="fl">0.456</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>lambda_gc</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.9851403</code></pre>
</div>
</div>
</section>
<section id="cleanup" class="level3">
<h3 class="anchored" data-anchor-id="cleanup">Cleanup</h3>
<p>When done, close the GDS connection</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>GWASTools<span class="sc">::</span><span class="fu">close</span>(geno_reader)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
</section>
<section id="when-to-use-a-different-tool" class="level2">
<h2 class="anchored" data-anchor-id="when-to-use-a-different-tool">When to use a different tool?</h2>
<ul>
<li>For very large datasets (e.g., UK Biobank), specialized tools like <code>BOLT-LMM</code>, <code>REGENIE</code>, or <code>SAIGE</code> may be more efficient</li>
<li>For binary traits, consider tools that handle case-control imbalance and relatedness effectively, such as <code>SAIGE</code></li>
<li>For rare variant analysis, consider burden tests or SKAT implemented in packages like <code>SKAT</code> or <code>seqMeta</code></li>
</ul>
</section>
<section id="further-reading" class="level2">
<h2 class="anchored" data-anchor-id="further-reading">Further reading</h2>
<ul>
<li><a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/GENESIS/inst/doc/pcair.html">Population Structure and Relatedness Inference using the GENESIS Package</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/vignettes/GENESIS/inst/doc/assoc_test.html#output">Genetic Association Testing using the GENESIS Package</a></li>
<li><a href="https://plink.readthedocs.io/en/latest/GWAS/">Using PLINK for GWAS</a></li>
</ul>
</section>
<section id="session-info" class="level2">
<h2 class="anchored" data-anchor-id="session-info">Session Info</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.5.1 (2025-06-13)
Platform: aarch64-apple-darwin20
Running under: macOS Tahoe 26.0.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1

locale:
[1] C.UTF-8/C.UTF-8/C.UTF-8/C/C.UTF-8/C.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
 [1] readr_2.1.5         dplyr_1.1.4         ggplot2_3.5.2      
 [4] qqman_0.1.9         GWASTools_1.54.0    Biobase_2.68.0     
 [7] BiocGenerics_0.54.0 generics_0.1.4      GENESIS_2.38.0     
[10] SNPRelate_1.42.0    gdsfmt_1.44.1      

loaded via a namespace (and not attached):
  [1] Rdpack_2.6.4            DBI_1.2.3               gridExtra_2.3          
  [4] sandwich_3.1-1          rlang_1.1.6             magrittr_2.0.3         
  [7] SeqVarTools_1.46.0      compiler_4.5.1          RSQLite_2.4.3          
 [10] mgcv_1.9-3              reshape2_1.4.4          vctrs_0.6.5            
 [13] stringr_1.5.1           quantreg_6.1            pkgconfig_2.0.3        
 [16] shape_1.4.6.1           crayon_1.5.3            fastmap_1.2.0          
 [19] backports_1.5.0         XVector_0.48.0          labeling_0.4.3         
 [22] utf8_1.2.6              rmarkdown_2.29          tzdb_0.5.0             
 [25] UCSC.utils_1.4.0        nloptr_2.2.1            MatrixModels_0.5-4     
 [28] purrr_1.0.4             bit_4.6.0               xfun_0.52              
 [31] glmnet_4.1-10           jomo_2.7-6              logistf_1.26.1         
 [34] cachem_1.1.0            GenomeInfoDb_1.44.3     jsonlite_2.0.0         
 [37] blob_1.2.4              pan_1.9                 BiocParallel_1.42.2    
 [40] broom_1.0.8             parallel_4.5.1          R6_2.6.1               
 [43] stringi_1.8.7           RColorBrewer_1.1-3      boot_1.3-31            
 [46] DNAcopy_1.82.0          rpart_4.1.24            lmtest_0.9-40          
 [49] GenomicRanges_1.60.0    Rcpp_1.1.0              iterators_1.0.14       
 [52] knitr_1.50              zoo_1.8-14              IRanges_2.42.0         
 [55] igraph_2.1.4            Matrix_1.7-3            splines_4.5.1          
 [58] nnet_7.3-20             tidyselect_1.2.1        viridis_0.6.5          
 [61] yaml_2.3.10             codetools_0.2-20        curl_6.2.3             
 [64] plyr_1.8.9              lattice_0.22-7          tibble_3.3.0           
 [67] quantsmooth_1.74.0      withr_3.0.2             evaluate_1.0.3         
 [70] survival_3.8-3          Biostrings_2.76.0       pillar_1.11.0          
 [73] BiocManager_1.30.26     mice_3.18.0             renv_1.1.5             
 [76] foreach_1.5.2           stats4_4.5.1            reformulas_0.4.1       
 [79] vroom_1.6.5             S4Vectors_0.46.0        hms_1.1.3              
 [82] scales_1.4.0            minqa_1.2.8             calibrate_1.7.7        
 [85] GWASExactHW_1.2         glue_1.8.0              tools_4.5.1            
 [88] data.table_1.17.4       lme4_1.1-37             SparseM_1.84-2         
 [91] cowplot_1.2.0           grid_4.5.1              tidyr_1.3.1            
 [94] rbibutils_2.3           nlme_3.1-168            formula.tools_1.7.1    
 [97] GenomeInfoDbData_1.2.14 cli_3.6.5               SeqArray_1.48.3        
[100] viridisLite_0.4.2       gtable_0.3.6            digest_0.6.37          
[103] operator.tools_1.6.3    farver_2.1.2            memoise_2.0.1          
[106] htmltools_0.5.8.1       lifecycle_1.0.4         httr_1.4.7             
[109] mitml_0.4-5             bit64_4.6.0-1           MASS_7.3-65            </code></pre>
</div>
</div>
</section>
<section id="references" class="level2">




</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-andersonQC2010" class="csl-entry" role="listitem">
Anderson,C.A. <em>et al.</em> (2010) <a href="https://doi.org/10.1038/nprot.2010.116">Data quality control in genetic case–control association studies</a>. <em>Nature Protocols</em>, <strong>5</strong>, 1564–1573.
</div>
<div id="ref-conomosPCRelate2016" class="csl-entry" role="listitem">
Conomos,M.P. <em>et al.</em> (2016) <a href="https://doi.org/10.1016/j.ajhg.2015.11.022">Model-free estimation of recent genetic relatedness</a>. <em>The American Journal of Human Genetics</em>, <strong>98</strong>, 127–148.
</div>
<div id="ref-conomosPCAiR2015" class="csl-entry" role="listitem">
Conomos,M.P. <em>et al.</em> (2015) <a href="https://doi.org/10.1002/gepi.21896">Robust inference of population structure for ancestry prediction and correction of stratification in the presence of relatedness</a>. <em>Genetic Epidemiology</em>, <strong>39</strong>, 276–293.
</div>
<div id="ref-guoThompson1992" class="csl-entry" role="listitem">
Guo,S.W. and Thompson,E.A. (1992) <a href="https://doi.org/10.2307/2532296">Performing the exact test of <span>Hardy</span>–<span>Weinberg</span> proportion for multiple alleles</a>. <em>Biometrics</em>, <strong>48</strong>, 361–372.
</div>
<div id="ref-manichaikulRobustRelationshipInference2010" class="csl-entry" role="listitem">
Manichaikul,A. <em>et al.</em> (2010) <a href="https://doi.org/10.1093/bioinformatics/btq559">Robust relationship inference in genome-wide association studies</a>. <em>Bioinformatics</em>, <strong>26</strong>, 2867–2873.
</div>
<div id="ref-hapmap3IntegratingCommonRare2010" class="csl-entry" role="listitem">
The International HapMap 3 Consortium (2010) <a href="https://doi.org/10.1038/nature09298">Integrating common and rare genetic variation in diverse human populations</a>. <em>Nature</em>, <strong>467</strong>, 52–58.
</div>
<div id="ref-internationalHapMapConsortium2005" class="csl-entry" role="listitem">
The International HapMap Consortium (2005) <a href="https://doi.org/10.1038/nature04226">A haplotype map of the human genome</a>. <em>Nature</em>, <strong>437</strong>, 1299–1320.
</div>
<div id="ref-wiggintonExactHWE2005" class="csl-entry" role="listitem">
Wigginton,J.E. <em>et al.</em> (2005) <a href="https://doi.org/10.1086/429864">A note on exact tests of <span>Hardy</span>–<span>Weinberg</span> equilibrium</a>. <em>The American Journal of Human Genetics</em>, <strong>76</strong>, 887–893.
</div>
<div id="ref-zhengHighperformanceComputingToolset2012" class="csl-entry" role="listitem">
Zheng,X. <em>et al.</em> (2012) <a href="https://doi.org/10.1093/bioinformatics/bts606">A high-performance computing toolset for relatedness and principal component analysis of SNP data</a>. <em>Bioinformatics</em>, <strong>28</strong>, 3326–3328.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>