<!DOCTYPE html>
<html lang="en"><head>
<link href="../course_logo.png" rel="icon" type="image/png">
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-html/tabby.min.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-ea385d0e468b0dd5ea5bf0780b1290d9.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.7.33">

  <title>Statistical Genetics – Lecture 01: Introduction and Foundations</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    html { -webkit-text-size-adjust: 100%; }
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; font-weight: bold; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="../site_libs/revealjs/dist/theme/quarto-fe3b3be043834d1b7bb052cb1587cec0.css">
  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.css">
  <link href="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Lecture 01: Introduction and Foundations</h1>
  <p class="subtitle">PUBH 8878, Statistical Genetics</p>

<div class="quarto-title-authors">
</div>

</section>
<section id="about-the-instructors" class="slide level2">
<h2>About the Instructors</h2>
<div class="quarto-layout-panel" data-layout="[100, 70, 66.5]">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 42.3%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://publichealth.gwu.edu/sites/g/files/zaxdzs4586/files/styles/d07/public/2023-02/keith-crandall-up-2014-jmc_3352lowres.jpeg"></p>
<figcaption>Dr.&nbsp;Keith Crandall</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 29.6%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://publichealth.gwu.edu/sites/g/files/zaxdzs4586/files/styles/d07/public/2024-11/ali_rahnavard_102024.jpg"></p>
<figcaption>Dr.&nbsp;Ali Rahnavard</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 28.1%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://chiraaggohel.com/favicon.jpeg"></p>
<figcaption>Chiraag Gohel</figcaption>
</figure>
</div>
</div>
</div>
</div>
</section>
<section id="introductions" class="slide level2">
<h2>Introductions</h2>
</section>
<section id="course-site" class="slide level2">
<h2>Course Site</h2>
<p><a href="https://gwcbi.github.io/StatGen/">https://gwcbi.github.io/StatGen/</a></p>
<iframe src="https://gwcbi.github.io/StatGen/" width="100%" height="600px">
</iframe>
</section>
<section id="zotero" class="slide level2">
<h2>Zotero</h2>

<img data-src="images/paste-1.png" class="r-stretch"></section>
<section id="what-is-statistical-genetics" class="slide level2">
<h2>What is statistical genetics?</h2>
<div class="quarto-layout-panel" data-layout-ncol="3">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 33.3%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://upload.wikimedia.org/wikipedia/commons/thumb/a/ae/Darwin%27s_finches_by_Gould.jpg/1590px-Darwin%27s_finches_by_Gould.jpg"></p>
<figcaption>Population Genetics</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 33.3%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41576-022-00478-5/MediaObjects/41576_2022_478_Fig2_HTML.png"></p>
<figcaption>Genetic Epidemiology</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 33.3%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://brightspotcdn.byu.edu/dims4/default/6658c6f/2147483647/strip/true/crop/1602x2400+0+0/resize/534x800!/quality/90/?url=https%3A%2F%2Fbrigham-young-brightspot-us-east-2.s3.us-east-2.amazonaws.com%2F9b%2Faf%2F9f5a135aaaf234b6420456b7312c%2Fbradley-measure.jpg"></p>
<figcaption>Quantitative Genetics</figcaption>
</figure>
</div>
</div>
</div>
</div>
</section>
<section id="a-statistical-geneticist-may-want-to-know" class="slide level2">
<h2>A statistical geneticist may want to know</h2>
<ol type="1">
<li class="fragment">Is there a genetic component contributing to the total variance of these traits?</li>
<li class="fragment">Is the genetic component of the traits driven by a few genes located on a particular chromosome, or are there many genes scattered across many chromosomes? How many genes are involved and is this a scientifically sensible question?</li>
<li class="fragment">Are the genes detected protein-coding genes, or are there also noncoding genes involved in gene regulation?</li>
<li class="fragment">How is the strength of the signals captured in a statistical analysis related to the two types of genes? What fraction of the total genetic variation is allocated to both types of genes?</li>
<li class="fragment">What are the frequencies of the genes in the sample? Are the frequencies associated with the magnitude of their effects on the traits?</li>
<li class="fragment">What is the mode of action of the genes?</li>
</ol>
</section>
<section id="a-statistical-geneticist-may-want-to-know-1" class="slide level2">
<h2>A statistical geneticist may want to know</h2>
<ol start="7" type="1">
<li class="fragment">What proportion of the genetic variance estimated in 1 can be explained by the discovered genes?</li>
<li class="fragment">Given the information on the set of genes carried by an individual, will a genetic score constructed before observing the trait help with early diagnosis and prevention?</li>
<li class="fragment">How should the predictive ability of the score be measured?</li>
<li class="fragment">Are there other non-genetic factors that affect the traits, such as smoking behavior, alcohol consumption, blood pressure measurements, body mass index and level of physical exercise?</li>
<li class="fragment">Could the predictive ability of the genetic score be improved by incorporation of these non-genetic sources of information, either additively or considering interactions? What is the relative contribution from the different sources of information?</li>
</ol>
</section>
<section id="what-does-the-data-look-like" class="slide level2">
<h2>What does the data look like?</h2>
<p><strong>Family/Pedigree Studies</strong></p>

<img data-src="https://www.genome.gov/sites/default/files/media/images/tg/Pedigree.jpg" class="r-stretch quarto-figure-center"><p class="caption">Family/Pedigree Studies</p></section>
<section id="what-does-the-data-look-like-1" class="slide level2">
<h2>What does the data look like?</h2>
<p><strong>Genome-Wide Association Studies (GWAS)</strong></p>

<img data-src="https://upload.wikimedia.org/wikipedia/commons/4/4f/Manhattan_plot_from_a_GWAS_of_kidney_stone_disease.png" style="width:60.0%" class="r-stretch quarto-figure-center"><p class="caption">Genome-Wide Association Studies</p></section>
<section id="core-abstractions-for-quantitative-biology" class="slide level2">
<h2>Core abstractions for quantitative biology</h2>
<ol type="1">
<li><strong>Model-Building</strong>
<ul>
<li>Formulate generative models linking genotype, environment, and phenotype (e.g., linear mixed, Bayesian hierarchical, non-parametric kernels).</li>
<li>Encode biological structure: linkage &amp; LD, population stratification, dominance/epistasis, multi-omics priors.</li>
<li>Balance realism and tractability to enable scalable computation on genome-scale data.</li>
</ul></li>
</ol>

<img data-src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41467-021-25756-4/MediaObjects/41467_2021_25756_Fig1_HTML.png" class="r-stretch"></section>
<section id="core-abstractions-for-quantitative-biology-1" class="slide level2">
<h2>Core abstractions for quantitative biology</h2>
<ol start="2" type="1">
<li><strong>Inference</strong>
<ul>
<li>Estimate unknown parameters and latent effects via likelihood maximisation, EM, MCMC, or SGD</li>
<li>Quantify uncertainty with standard errors, posterior intervals, and credible sets</li>
<li>Control false discoveries across millions of tests with FDR/Q-value, permutation, and empirical-Bayes shrinkage.</li>
</ul></li>
</ol>

<img data-src="https://www.cs.auckland.ac.nz/~georgy/research/texture/thesis-html/img133.png" class="r-stretch"></section>
<section id="core-abstractions-for-quantitative-biology-2" class="slide level2">
<h2>Core abstractions for quantitative biology</h2>
<ol start="3" type="1">
<li><strong>Prediction</strong>
<ul>
<li>Use fitted models for out-of-sample trait prediction: BLUP/GBLUP, ridge/lasso/elastic-net, Bayesian whole-genome regressions, random forests, neural nets.</li>
<li>Evaluate accuracy (MSE, AUC), bias–variance trade-off, calibration, and portability across ancestries or cell types.</li>
<li>Translate genomic predictions into actionable scores for breeding, risk stratification, and drug-target prioritisation.</li>
</ul></li>
</ol>

<img data-src="https://www.cell.com/cms/10.1016/j.xgen.2024.100523/asset/29efca68-0928-4eed-94b9-9799e7cedda2/main.assets/gr5_lrg.jpg" class="r-stretch"></section>
<section id="core-abstractions-for-quantitative-biology-3" class="slide level2">
<h2>Core abstractions for quantitative biology</h2>
<ol start="4" type="1">
<li><strong>Interpretation &amp; Validation</strong>
<ul>
<li>Integrate functional annotations, eQTL, and single-cell data to refine biological mechanisms.</li>
<li>Perform replication, cross-cohort meta-analysis, and sensitivity analyses to population assumptions.</li>
<li>Communicate findings with clear visualisations and reproducible workflows (R/Bioconductor, Git, notebooks).</li>
</ul></li>
</ol>
<div class="quarto-layout-panel" data-layout="[1.5, -.2 ,1, -.2 ,2]">
<div class="quarto-layout-row quarto-layout-valign-center">
<div class="quarto-layout-cell" style="flex-basis: 30.6%;justify-content: center;">
<p><img data-src="https://upload.wikimedia.org/wikipedia/commons/thumb/d/db/Nature_journal_logo.svg/1200px-Nature_journal_logo.svg.png"></p>
</div>
<div class="quarto-figure-spacer quarto-layout-cell" style="flex-basis: 4.1%;justify-content: flex-start;">
<p>&nbsp;</p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 20.4%;justify-content: center;">
<p><img data-src="https://upload.wikimedia.org/wikipedia/commons/thumb/3/38/Jupyter_logo.svg/1200px-Jupyter_logo.svg.png"></p>
</div>
<div class="quarto-figure-spacer quarto-layout-cell" style="flex-basis: 4.1%;justify-content: flex-start;">
<p>&nbsp;</p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 40.8%;justify-content: center;">
<p><img data-src="https://www.bioconductor.org/images/logo/png/bioconductor_logo_rgb.png"></p>
</div>
</div>
</div>
</section>
<section id="some-vocabulary" class="slide level2">
<h2>Some vocabulary</h2>
<ul>
<li><strong>Trait/Phenotype</strong>: A measurable characteristic of an organism, such as height, weight, or disease status.</li>
<li><strong>Genotype</strong>: The genetic constitution of an individual, often represented by specific alleles at particular loci.</li>
<li><strong>Allele</strong>: A variant form of a gene that can exist at a specific locus on a chromosome.</li>
<li><strong>Locus</strong>: A specific, fixed position on a chromosome where a particular gene or genetic marker is located.</li>
<li><strong>Polymorphism</strong>: The occurrence of two or more genetically determined forms in a population, such as single nucleotide polymorphisms (SNPs) or copy number variations (CNVs).</li>
<li><strong>Genetic Marker</strong>: A specific DNA sequence with a known location on a chromosome that can be used to identify individuals or species, often used in genetic mapping or association studies.</li>
<li><strong>Linkage Disequilibrium (LD)</strong>: The non-random association of alleles at different loci, indicating that certain allele combinations occur together more frequently than expected by chance.</li>
<li><strong>Heritability</strong>: The proportion of phenotypic variance in a trait that can be attributed to genetic variance, often estimated through twin or family studies.</li>
</ul>
</section>
<section id="some-vocabulary-1" class="slide level2">
<h2>Some vocabulary</h2>
<ul>
<li><strong>Genome-Wide Association Study (GWAS)</strong>: A study that looks for associations between genetic variants across the genome and specific traits or diseases in a large population.</li>
<li><strong>Polygenic Score (PGS)</strong>: A score that aggregates the effects of multiple genetic variants to predict an individual’s genetic predisposition to a trait or disease.</li>
<li><strong>Quantitative Trait Locus (QTL)</strong>: A region of the genome that is associated with a quantitative trait, often identified through linkage or association mapping.</li>
<li><strong>Epistasis</strong>: The interaction between genes where the effect of one gene is modified by one or more other genes, influencing the expression of a trait.<br>
</li>
<li><strong>Epigenetics</strong>: The study of heritable changes in gene expression that do not involve changes to the underlying DNA sequence, often influenced by environmental factors.</li>
<li><strong>Functional Annotation</strong>: The process of identifying the functional elements in the genome, such as genes, regulatory regions, and non-coding RNAs, and understanding their roles in biological processes.</li>
</ul>
</section>
<section id="probability-refresher" class="slide level2">
<h2>Probability Refresher</h2>
<h3 id="random-variables-and-distributions">Random Variables and Distributions</h3>
<ul>
<li class="fragment"><strong>Random Variable</strong>: <span class="math inline">X: \Omega \to \mathbb{R}</span> – A function mapping sample space <span class="math inline">\Omega</span> to real numbers</li>
<li class="fragment"><strong>Discrete RV</strong>: <span class="math inline">X \in \{x_1, x_2, \ldots\}</span> (e.g., genotype counts, number of mutations)
<ul>
<li class="fragment"><strong>Probability Mass Function (PMF)</strong>: <span class="math inline">p_X(x) = P(X = x)</span></li>
<li class="fragment"><strong>Cumulative Distribution Function (CDF)</strong>: <span class="math inline">F_X(x) = P(X \leq x) = \sum_{x_i \leq x} p_X(x_i)</span></li>
</ul></li>
<li class="fragment"><strong>Continuous RV</strong>: <span class="math inline">X \in \mathbb{R}</span> or interval (e.g., gene expression, quantitative traits)
<ul>
<li class="fragment"><strong>Probability Density Function (PDF)</strong>: <span class="math inline">f_X(x)</span> where <span class="math inline">P(a &lt; X &lt; b) = \int_a^b f_X(x) dx</span></li>
<li class="fragment"><strong>Cumulative Distribution Function (CDF)</strong>: <span class="math inline">F_X(x) = P(X \leq x) = \int_{-\infty}^x f_X(t) dt</span></li>
</ul></li>
<li class="fragment"><strong>Key Properties</strong>: <span class="math inline">F_X(-\infty) = 0</span>, <span class="math inline">F_X(\infty) = 1</span>, <span class="math inline">f_X(x) = \frac{d}{dx}F_X(x)</span></li>
</ul>
</section>
<section id="probability-refresher-1" class="slide level2">
<h2>Probability Refresher</h2>
<h3 id="key-distributions-in-statistical-genetics">Key Distributions in Statistical Genetics</h3>
<div class="quarto-layout-panel" data-layout-ncol="2">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p><strong>Binomial Distribution</strong> <span class="math display">X \sim \text{Binomial}(n, p)</span> <span class="math display">P(X = k) = \binom{n}{k} p^k (1-p)^{n-k}</span></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p>Used for: Allele counts, genotype frequencies</p>
</div>
</div>
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p><strong>Normal Distribution</strong> <span class="math display">X \sim N(\mu, \sigma^2)</span> <span class="math display">f(x) = \frac{1}{\sqrt{2\pi\sigma^2}} e^{-\frac{(x-\mu)^2}{2\sigma^2}}</span></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p>Used for: Quantitative traits, effect sizes</p>
</div>
</div>
</div>
</section>
<section id="probability-refresher-2" class="slide level2">
<h2>Probability Refresher</h2>
<h3 id="expected-value-and-variance">Expected Value and Variance</h3>
<ul>
<li class="fragment"><strong>Expected Value</strong>: <span class="math inline">E[X] = \sum x \cdot P(X = x)</span> (discrete) or <span class="math inline">E[X] = \int x f(x) dx</span> (continuous)</li>
<li class="fragment"><strong>Variance</strong>: <span class="math inline">\text{Var}(X) = E[(X - E[X])^2] = E[X^2] - (E[X])^2</span></li>
<li class="fragment"><strong>Standard Error</strong>: <span class="math inline">\text{SE} = \frac{\sigma}{\sqrt{n}}</span> (uncertainty in sample estimates)</li>
<li class="fragment"><strong>Central Limit Theorem</strong>: Sample means approach normality as <span class="math inline">n \to \infty</span></li>
</ul>
</section>
<section id="probability-refresher-3" class="slide level2">
<h2>Probability Refresher</h2>
<h3 id="conditional-probability-and-independence">Conditional Probability and Independence</h3>
<ul>
<li class="fragment"><strong>Conditional Probability</strong>: <span class="math inline">P(A|B) = \frac{P(A \cap B)}{P(B)}</span></li>
<li class="fragment"><strong>Bayes’ Theorem</strong>: <span class="math inline">P(A|B) = \frac{P(B|A)P(A)}{P(B)}</span></li>
<li class="fragment"><strong>Independence</strong>: <span class="math inline">P(A \cap B) = P(A) \cdot P(B)</span> iff <span class="math inline">A</span> and <span class="math inline">B</span> are independent</li>
</ul>
</section>
<section id="early-work-in-the-field" class="slide level2">
<h2>Early work in the field</h2>
<div class="quarto-layout-panel" data-layout="[70, 30]">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 70.0%;justify-content: center;">
<p><img data-src="https://cdn.kastatic.org/ka-perseus-images/b966bcd0e55933e24836c4afe4a2a9889247579c.png"></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 30.0%;justify-content: center;">
<p><img data-src="https://cdn.kastatic.org/ka-perseus-images/97ef4b490d977c359786b040ead67b9cdaeac1aa.png"></p>
</div>
</div>
</div>
</section>
<section id="genetics-statistics-and-eugenics-have-always-been-closely-linked" class="slide level2">
<h2>Genetics, statistics, and eugenics, have always been closely linked</h2>
<div class="quarto-layout-panel" data-layout="[35, 65]">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 35.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://upload.wikimedia.org/wikipedia/commons/thumb/a/a7/PSM_V60_D232_Standard_eugenics_scheme_of_descent.png/960px-PSM_V60_D232_Standard_eugenics_scheme_of_descent.png"></p>
<figcaption>Standard eugenics scheme of descent</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 65.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/galton.png"></p>
<figcaption>Francis Galton and his work, “Inquires into Human Faculty and its Development”</figcaption>
</figure>
</div>
</div>
</div>
</div>
</section>
<section id="the-ethics-of-genetics-and-genomics-iswas-of-great-importance" class="slide level2">
<h2>The ethics of genetics and genomics is/(was) of great importance!</h2>

<img data-src="images/nhgri_ethics.png" class="r-stretch"></section>
<section id="parameter-estimation" class="slide level2">
<h2>Parameter Estimation</h2>
<ul>
<li>Imagine a geneticist is studying <strong>allele ages</strong> (i.e., how many generations ago an allele arose via mutation) under a simplified model where new mutations arise randomly and uniformly over a fixed window — say, the last <span class="math inline">\theta</span> generations.</li>
<li>We could model <span class="math inline">\theta \sim \textsf{Uniform}(0, \theta)</span>.</li>
<li>We have our observed sample <span class="math inline">\{x_1,x_2, \ldots, x_n\}</span>, ages in generations of <span class="math inline">n</span> observed alleles.
<ul>
<li>For concreteness, let our sample be <span class="math inline">\{0, 1, 8, 4, 3, 1, 1, 5\}</span>.</li>
</ul></li>
<li>How should we estimate <span class="math inline">\theta</span> ?</li>
</ul>
<div class="fragment">
<div class="cell quarto-layout-panel" data-layout-ncol="2">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img data-src="lecture-01_files/figure-revealjs/unnamed-chunk-2-1.png" width="960"></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img data-src="lecture-01_files/figure-revealjs/unnamed-chunk-2-2.png" width="960"></p>
</div>
</div>
</div>
</div>
</section>
<section id="sampling-distributions-in-genetics" class="slide level2">
<h2>Sampling Distributions in Genetics</h2>
<h3 id="from-population-to-sample-allele-frequency-estimation">From Population to Sample: Allele Frequency Estimation</h3>
<p><strong>True population parameter</strong>: <span class="math inline">p = 0.3</span> (allele frequency) <strong>Sample estimates</strong>: <span class="math inline">\hat{p} = \frac{\text{# A alleles observed}}{\text{total alleles sampled}}</span></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a></a><span class="co"># Simulate sampling from population with p = 0.3</span></span>
<span id="cb1-2"><a></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb1-3"><a></a>true_p <span class="ot">&lt;-</span> <span class="fl">0.3</span></span>
<span id="cb1-4"><a></a>sample_sizes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">20</span>, <span class="dv">100</span>, <span class="dv">500</span>)</span>
<span id="cb1-5"><a></a>n_sims <span class="ot">&lt;-</span> <span class="dv">1000</span></span>
<span id="cb1-6"><a></a></span>
<span id="cb1-7"><a></a>sampling_results <span class="ot">&lt;-</span> <span class="fu">map_dfr</span>(sample_sizes, <span class="cf">function</span>(n) {</span>
<span id="cb1-8"><a></a>  p_hats <span class="ot">&lt;-</span> <span class="fu">replicate</span>(n_sims, {</span>
<span id="cb1-9"><a></a>    <span class="co"># Sample n individuals (2n alleles total)</span></span>
<span id="cb1-10"><a></a>    alleles <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(n, <span class="dv">2</span>, true_p)  <span class="co"># Each individual contributes 2 alleles</span></span>
<span id="cb1-11"><a></a>    <span class="fu">sum</span>(alleles) <span class="sc">/</span> (<span class="dv">2</span> <span class="sc">*</span> n)  <span class="co"># Sample allele frequency</span></span>
<span id="cb1-12"><a></a>  })</span>
<span id="cb1-13"><a></a>  </span>
<span id="cb1-14"><a></a>  <span class="fu">data.frame</span>(</span>
<span id="cb1-15"><a></a>    <span class="at">sample_size =</span> n,</span>
<span id="cb1-16"><a></a>    <span class="at">p_hat =</span> p_hats,</span>
<span id="cb1-17"><a></a>    <span class="at">theoretical_se =</span> <span class="fu">sqrt</span>(true_p <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> true_p) <span class="sc">/</span> (<span class="dv">2</span> <span class="sc">*</span> n))</span>
<span id="cb1-18"><a></a>  )</span>
<span id="cb1-19"><a></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<h3 id="sampling-distribution-properties">Sampling Distribution Properties</h3>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a></a><span class="co"># Plot sampling distributions</span></span>
<span id="cb2-2"><a></a><span class="fu">ggplot</span>(sampling_results, <span class="fu">aes</span>(<span class="at">x =</span> p_hat)) <span class="sc">+</span></span>
<span id="cb2-3"><a></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">y =</span> ..density..), <span class="at">bins =</span> <span class="dv">30</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb2-4"><a></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> true_p, <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb2-5"><a></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span><span class="fu">paste</span>(<span class="st">"n ="</span>, sample_size), <span class="at">scales =</span> <span class="st">"free_y"</span>) <span class="sc">+</span></span>
<span id="cb2-6"><a></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Sampling Distribution of p̂"</span>, </span>
<span id="cb2-7"><a></a>       <span class="at">x =</span> <span class="st">"Sample Allele Frequency (p̂)"</span>, <span class="at">y =</span> <span class="st">"Density"</span>) <span class="sc">+</span></span>
<span id="cb2-8"><a></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
ℹ Please use `linewidth` instead.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The dot-dot notation (`..density..`) was deprecated in ggplot2 3.4.0.
ℹ Please use `after_stat(density)` instead.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a></a><span class="co"># Compare theoretical vs empirical standard errors</span></span>
<span id="cb5-2"><a></a>se_comparison <span class="ot">&lt;-</span> sampling_results <span class="sc">%&gt;%</span></span>
<span id="cb5-3"><a></a>  <span class="fu">group_by</span>(sample_size) <span class="sc">%&gt;%</span></span>
<span id="cb5-4"><a></a>  <span class="fu">summarise</span>(</span>
<span id="cb5-5"><a></a>    <span class="at">empirical_se =</span> <span class="fu">sd</span>(p_hat),</span>
<span id="cb5-6"><a></a>    <span class="at">theoretical_se =</span> <span class="fu">first</span>(theoretical_se),</span>
<span id="cb5-7"><a></a>    <span class="at">empirical_mean =</span> <span class="fu">mean</span>(p_hat)</span>
<span id="cb5-8"><a></a>  )</span>
<span id="cb5-9"><a></a></span>
<span id="cb5-10"><a></a><span class="fu">ggplot</span>(se_comparison, <span class="fu">aes</span>(<span class="at">x =</span> sample_size)) <span class="sc">+</span></span>
<span id="cb5-11"><a></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">y =</span> empirical_se, <span class="at">color =</span> <span class="st">"Empirical"</span>), <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb5-12"><a></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">y =</span> theoretical_se, <span class="at">color =</span> <span class="st">"Theoretical"</span>), <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb5-13"><a></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> theoretical_se, <span class="at">color =</span> <span class="st">"Theoretical"</span>)) <span class="sc">+</span></span>
<span id="cb5-14"><a></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Standard Error: Theory vs Practice"</span>,</span>
<span id="cb5-15"><a></a>       <span class="at">x =</span> <span class="st">"Sample Size"</span>, <span class="at">y =</span> <span class="st">"Standard Error of p̂"</span>) <span class="sc">+</span></span>
<span id="cb5-16"><a></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"Empirical"</span> <span class="ot">=</span> <span class="st">"blue"</span>, <span class="st">"Theoretical"</span> <span class="ot">=</span> <span class="st">"red"</span>)) <span class="sc">+</span></span>
<span id="cb5-17"><a></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell quarto-layout-panel" data-layout-ncol="2">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img data-src="lecture-01_files/figure-revealjs/unnamed-chunk-4-1.png" width="960"></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img data-src="lecture-01_files/figure-revealjs/unnamed-chunk-4-2.png" width="960"></p>
</div>
</div>
</div>
</section>
<section id="population-vs-sample-key-insights" class="slide level2">
<h2>Population vs Sample: Key Insights</h2>
<ul>
<li class="fragment"><strong>Unbiased Estimator</strong>: <span class="math inline">E[\hat{p}] = p</span> (sample mean equals population parameter)</li>
<li class="fragment"><strong>Standard Error</strong>: <span class="math inline">SE(\hat{p}) = \sqrt{\frac{p(1-p)}{2n}}</span> decreases with sample size</li>
<li class="fragment"><strong>Normal Approximation</strong>: For large <span class="math inline">n</span>, <span class="math inline">\hat{p} \sim N\left(p, \frac{p(1-p)}{2n}\right)</span></li>
<li class="fragment"><strong>Confidence Intervals</strong>: <span class="math inline">\hat{p} \pm 1.96 \times SE(\hat{p})</span> covers true <span class="math inline">p</span> ~95% of time</li>
<li class="fragment"><strong>Power</strong>: Larger samples = better ability to detect true differences between populations</li>
</ul>
</section>
<section id="sampling-distribution-hardy-weinberg-testing" class="slide level2">
<h2>Sampling Distribution: Hardy-Weinberg Testing</h2>
<p><strong>Population</strong>: HWE with <span class="math inline">p = 0.6</span>, genotype frequencies: <span class="math inline">p^2 = 0.36</span>, <span class="math inline">2pq = 0.48</span>, <span class="math inline">q^2 = 0.16</span></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a></a><span class="co"># Simulate HWE sampling and testing</span></span>
<span id="cb6-2"><a></a>simulate_hwe_test <span class="ot">&lt;-</span> <span class="cf">function</span>(n, p_true) {</span>
<span id="cb6-3"><a></a>  <span class="co"># Expected HWE frequencies</span></span>
<span id="cb6-4"><a></a>  exp_AA <span class="ot">&lt;-</span> n <span class="sc">*</span> p_true<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb6-5"><a></a>  exp_Aa <span class="ot">&lt;-</span> n <span class="sc">*</span> <span class="dv">2</span> <span class="sc">*</span> p_true <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> p_true)  </span>
<span id="cb6-6"><a></a>  exp_aa <span class="ot">&lt;-</span> n <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> p_true)<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb6-7"><a></a>  </span>
<span id="cb6-8"><a></a>  <span class="co"># Sample genotypes under HWE</span></span>
<span id="cb6-9"><a></a>  genotypes <span class="ot">&lt;-</span> <span class="fu">rmultinom</span>(<span class="dv">1</span>, n, <span class="fu">c</span>(p_true<span class="sc">^</span><span class="dv">2</span>, <span class="dv">2</span><span class="sc">*</span>p_true<span class="sc">*</span>(<span class="dv">1</span><span class="sc">-</span>p_true), (<span class="dv">1</span><span class="sc">-</span>p_true)<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb6-10"><a></a>  </span>
<span id="cb6-11"><a></a>  <span class="co"># Estimate allele frequency from sample</span></span>
<span id="cb6-12"><a></a>  p_hat <span class="ot">&lt;-</span> (<span class="dv">2</span><span class="sc">*</span>genotypes[<span class="dv">1</span>] <span class="sc">+</span> genotypes[<span class="dv">2</span>]) <span class="sc">/</span> (<span class="dv">2</span><span class="sc">*</span>n)</span>
<span id="cb6-13"><a></a>  </span>
<span id="cb6-14"><a></a>  <span class="co"># Recalculate expected frequencies with estimated p</span></span>
<span id="cb6-15"><a></a>  exp_AA_hat <span class="ot">&lt;-</span> n <span class="sc">*</span> p_hat<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb6-16"><a></a>  exp_Aa_hat <span class="ot">&lt;-</span> n <span class="sc">*</span> <span class="dv">2</span> <span class="sc">*</span> p_hat <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> p_hat)</span>
<span id="cb6-17"><a></a>  exp_aa_hat <span class="ot">&lt;-</span> n <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> p_hat)<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb6-18"><a></a>  </span>
<span id="cb6-19"><a></a>  <span class="co"># Chi-square test statistic</span></span>
<span id="cb6-20"><a></a>  chi_sq <span class="ot">&lt;-</span> <span class="fu">sum</span>((genotypes <span class="sc">-</span> <span class="fu">c</span>(exp_AA_hat, exp_Aa_hat, exp_aa_hat))<span class="sc">^</span><span class="dv">2</span> <span class="sc">/</span> </span>
<span id="cb6-21"><a></a>                <span class="fu">c</span>(exp_AA_hat, exp_Aa_hat, exp_aa_hat))</span>
<span id="cb6-22"><a></a>  </span>
<span id="cb6-23"><a></a>  p_value <span class="ot">&lt;-</span> <span class="fu">pchisq</span>(chi_sq, <span class="at">df =</span> <span class="dv">1</span>, <span class="at">lower.tail =</span> <span class="cn">FALSE</span>)</span>
<span id="cb6-24"><a></a>  </span>
<span id="cb6-25"><a></a>  <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">p_hat =</span> p_hat, <span class="at">chi_sq =</span> chi_sq, <span class="at">p_value =</span> p_value))</span>
<span id="cb6-26"><a></a>}</span>
<span id="cb6-27"><a></a></span>
<span id="cb6-28"><a></a><span class="co"># Run simulation</span></span>
<span id="cb6-29"><a></a><span class="fu">set.seed</span>(<span class="dv">456</span>)</span>
<span id="cb6-30"><a></a>n_sample <span class="ot">&lt;-</span> <span class="dv">200</span></span>
<span id="cb6-31"><a></a>hwe_results <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">1000</span>, <span class="fu">simulate_hwe_test</span>(n_sample, <span class="fl">0.6</span>), <span class="at">simplify =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<h3 id="sampling-distribution-of-hwe-tests">Sampling Distribution of HWE Tests</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a></a>p_values <span class="ot">&lt;-</span> <span class="fu">map_dbl</span>(hwe_results, <span class="sc">~</span>.<span class="sc">$</span>p_value)</span>
<span id="cb7-2"><a></a>p_hats <span class="ot">&lt;-</span> <span class="fu">map_dbl</span>(hwe_results, <span class="sc">~</span>.<span class="sc">$</span>p_hat)</span>
<span id="cb7-3"><a></a></span>
<span id="cb7-4"><a></a><span class="fu">cat</span>(<span class="st">"Proportion of tests rejecting HWE (α = 0.05):"</span>, <span class="fu">mean</span>(p_values <span class="sc">&lt;</span> <span class="fl">0.05</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Proportion of tests rejecting HWE (α = 0.05): 0.053</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Mean estimated allele frequency:"</span>, <span class="fu">round</span>(<span class="fu">mean</span>(p_hats), <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Mean estimated allele frequency: 0.599</code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">SD of estimated allele frequency:"</span>, <span class="fu">round</span>(<span class="fu">sd</span>(p_hats), <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
SD of estimated allele frequency: 0.025</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Theoretical SE:"</span>, <span class="fu">round</span>(<span class="fu">sqrt</span>(<span class="fl">0.6</span> <span class="sc">*</span> <span class="fl">0.4</span> <span class="sc">/</span> (<span class="dv">2</span> <span class="sc">*</span> n_sample)), <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Theoretical SE: 0.024</code></pre>
</div>
</div>
</section>
<section id="importance-of-statistical-theory" class="slide level2">
<h2>Importance of statistical theory</h2>
<ul>
<li>Understanding random variables, sampling distributions, and bias/variance of estimators helps motivate biological questions, and understand how they can be answered</li>
<li>Let y be the expression of a trait, G be the additive contribution of genetic variables, and an environmental value E;</li>
<li>What assumptions do we make via the following models?</li>
</ul>
<div class="quarto-layout-panel" data-layout="[44, -12, 44]">
<div class="quarto-layout-row">
<div class="column quarto-layout-cell" style="flex-basis: 44.0%;justify-content: flex-start;">
<p><span class="math display">\begin{align}

y &amp;= G + E \\
y &amp;= \beta_0 + \beta_1 G + \beta_2 E + \epsilon \\
y &amp;= \beta_0 + \beta_1 G + \beta_2 E + \epsilon, \quad \epsilon \sim N(0, \sigma^2) \\
y &amp;= \beta_0 + \beta_1 G + \beta_2 E + \beta_3 \left(G \times E\right) + \epsilon, \quad \epsilon \sim N(0, \sigma^2)

\end{align}</span></p>
</div><div class="quarto-figure-spacer quarto-layout-cell" style="flex-basis: 12.0%;justify-content: flex-start;">
<p>&nbsp;</p>
</div>
<div class="column quarto-layout-cell" style="flex-basis: 44.0%;justify-content: center;">
<p><img data-src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fsj.tpj.6500085/MediaObjects/41397_2002_Article_BF6500085_Fig1_HTML.jpg"></p>
</div></div>
</div>
</section>
<section id="molecular-genetics" class="slide level2">
<h2>Molecular Genetics</h2>

<img data-src="https://s3.amazonaws.com/libapps/accounts/14039/images/karyotype-human-female.jpg" class="r-stretch"></section>
<section id="genetic-variants" class="slide level2">
<h2>Genetic Variants</h2>

<img data-src="https://upload.wikimedia.org/wikipedia/commons/3/37/Single_nucleotide_polymorphism_substitution_mutation_diagram_-_cytosine_to_thymine.png" class="r-stretch"></section>
<section id="the-sampling-distribution-of-a-random-variable" class="slide level2">
<h2>The sampling distribution of a random variable</h2>
<ul>
<li>A random variable is a function that assigns a numerical value to each outcome in a sample space.</li>
<li>A probability distribution describes how the probabilities are distributed over the values of the random variable.</li>
<li>The likelihood function is a function of the parameters of a statistical model given specific observed data.</li>
</ul>
</section>
<section id="modeling" class="slide level2">
<h2>Modeling</h2>
<ul>
<li>Imagine that we have a sample size of <span class="math inline">n</span> unrelated haploid individuals from some population</li>
<li>We want to estimate allele frequencies for a biallelic SNP, sy A/a</li>
<li>In our sample, we observe <span class="math inline">x</span> individuals with allele A and <span class="math inline">n - x</span> individuals with allele a</li>
<li>Let <span class="math inline">p</span> be the frequency of allele A and <span class="math inline">q = 1 - p</span> be the frequency of allele a</li>
</ul>
<p><span class="math display">Pr(X = x| n, p) = \binom{n}{x} p^x q^{n - x}</span></p>
<ul>
<li>Lets say we observe <span class="math inline">n = 27</span> and <span class="math inline">x = 11</span>. How do we estimate <span class="math inline">p</span>?</li>
</ul>
</section>
<section id="maximum-likelihood-estimator" class="slide level2">
<h2>Maximum Likelihood Estimator</h2>
<p>Our likelihood function for <span class="math inline">p</span> is <span class="math display">
L(P|x = 11, n = 27) = {27 \choose 11} p^{11} q^{27-11}
</span></p>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a></a>n <span class="ot">&lt;-</span> <span class="dv">27</span></span>
<span id="cb15-2"><a></a>x <span class="ot">&lt;-</span> <span class="dv">11</span></span>
<span id="cb15-3"><a></a>p_seq <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">length.out =</span> <span class="dv">200</span>)</span>
<span id="cb15-4"><a></a>likelihood <span class="ot">&lt;-</span> <span class="fu">dbinom</span>(x, n, p_seq)</span>
<span id="cb15-5"><a></a>plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb15-6"><a></a>    <span class="fu">data.frame</span>(<span class="at">p =</span> p_seq, <span class="at">likelihood =</span> likelihood),</span>
<span id="cb15-7"><a></a>    <span class="fu">aes</span>(<span class="at">x =</span> p, <span class="at">y =</span> likelihood)</span>
<span id="cb15-8"><a></a>) <span class="sc">+</span></span>
<span id="cb15-9"><a></a>    <span class="fu">geom_line</span>(<span class="at">linewidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb15-10"><a></a>    <span class="fu">labs</span>(</span>
<span id="cb15-11"><a></a>        <span class="at">x =</span> <span class="st">"Probability of Success (p)"</span>,</span>
<span id="cb15-12"><a></a>        <span class="at">y =</span> <span class="st">"Likelihood"</span>,</span>
<span id="cb15-13"><a></a>        <span class="at">title =</span> <span class="st">"Binomial Likelihood Function"</span></span>
<span id="cb15-14"><a></a>    ) <span class="sc">+</span></span>
<span id="cb15-15"><a></a>    <span class="fu">geom_vline</span>(</span>
<span id="cb15-16"><a></a>        <span class="at">xintercept =</span> x <span class="sc">/</span> n,</span>
<span id="cb15-17"><a></a>        <span class="at">color =</span> <span class="st">"red"</span>,</span>
<span id="cb15-18"><a></a>        <span class="at">linetype =</span> <span class="st">"dashed"</span></span>
<span id="cb15-19"><a></a>    ) <span class="sc">+</span></span>
<span id="cb15-20"><a></a>    <span class="fu">theme_light</span>(<span class="at">base_size =</span> <span class="dv">18</span>) <span class="sc">+</span></span>
<span id="cb15-21"><a></a>    <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure>
<p><img data-src="lecture-01_files/figure-revealjs/unnamed-chunk-7-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</div></div>
</section>
<section id="building-one-locus-likelihood-functions" class="slide level2">
<h2>Building One-Locus Likelihood Functions</h2>
<h3 id="the-general-framework">The General Framework</h3>
<ul>
<li class="fragment"><strong>Genetic Data</strong>: Observed phenotypes/genotypes from individuals or families</li>
<li class="fragment"><strong>Genetic Model</strong>: Hypotheses about mode of inheritance, penetrance, allele frequencies</li>
<li class="fragment"><strong>Likelihood Function</strong>: <span class="math inline">L(\boldsymbol{\theta} | \text{data}) = P(\text{data} | \boldsymbol{\theta})</span></li>
<li class="fragment"><strong>Parameters</strong> <span class="math inline">\boldsymbol{\theta}</span>: May include allele frequencies, penetrance values, recombination rates</li>
</ul>
</section>
<section id="building-one-locus-likelihood-functions-1" class="slide level2">
<h2>Building One-Locus Likelihood Functions</h2>
<h3 id="example-1-dominant-disease-model">Example 1: Dominant Disease Model</h3>
<p>Consider a rare dominant disease in a family pedigree:</p>
<div class="quarto-layout-panel" data-layout-ncol="2">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p><strong>Family Data</strong> - Parents: Affected father, unaffected mother - 4 offspring: 2 affected, 2 unaffected - Disease allele frequency: <span class="math inline">p</span> (rare, so <span class="math inline">p \approx 0</span>)</p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p><strong>Likelihood Construction</strong> 1. <strong>Parental genotypes</strong>: <span class="math inline">Dd \times dd</span> 2. <strong>Expected offspring ratio</strong>: 1:1 (affected:unaffected) 3. <strong>Observed data</strong>: 2 affected, 2 unaffected 4. <strong>Likelihood</strong>: <span class="math inline">L(p) = \binom{4}{2} \left(\frac{1}{2}\right)^2 \left(\frac{1}{2}\right)^2 = \frac{6}{16}</span></p>
</div>
</div>
</div>
</section>
<section id="building-one-locus-likelihood-functions-2" class="slide level2">
<h2>Building One-Locus Likelihood Functions</h2>
<h3 id="example-2-penetrance-model">Example 2: Penetrance Model</h3>
<p>Disease with reduced penetrance:</p>
<ul>
<li class="fragment"><strong>Genotype-specific risks</strong>: <span class="math inline">P(\text{disease} | DD) = f_2</span>, <span class="math inline">P(\text{disease} | Dd) = f_1</span>, <span class="math inline">P(\text{disease} | dd) = f_0</span></li>
<li class="fragment"><strong>Observed data</strong>: <span class="math inline">n</span> individuals, <span class="math inline">k</span> affected</li>
<li class="fragment"><strong>Genotype frequencies</strong>: Hardy-Weinberg proportions <span class="math inline">p^2, 2pq, q^2</span></li>
<li class="fragment"><strong>Population disease risk</strong>: <span class="math inline">K = p^2 f_2 + 2pq f_1 + q^2 f_0</span></li>
</ul>
<p><span class="math display">L(p, f_0, f_1, f_2) = \binom{n}{k} K^k (1-K)^{n-k}</span></p>
</section>
<section id="building-one-locus-likelihood-functions-3" class="slide level2">
<h2>Building One-Locus Likelihood Functions</h2>
<h3 id="example-3-family-based-likelihood">Example 3: Family-Based Likelihood</h3>
<p>Nuclear family with 2 parents, 3 offspring:</p>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a></a><span class="co"># Genotype probabilities for parents</span></span>
<span id="cb16-2"><a></a>parent_probs <span class="ot">&lt;-</span> <span class="cf">function</span>(p, genotype) {</span>
<span id="cb16-3"><a></a>  <span class="fu">case_when</span>(</span>
<span id="cb16-4"><a></a>    genotype <span class="sc">==</span> <span class="st">"AA"</span> <span class="sc">~</span> p<span class="sc">^</span><span class="dv">2</span>,</span>
<span id="cb16-5"><a></a>    genotype <span class="sc">==</span> <span class="st">"Aa"</span> <span class="sc">~</span> <span class="dv">2</span><span class="sc">*</span>p<span class="sc">*</span>(<span class="dv">1</span><span class="sc">-</span>p),</span>
<span id="cb16-6"><a></a>    genotype <span class="sc">==</span> <span class="st">"aa"</span> <span class="sc">~</span> (<span class="dv">1</span><span class="sc">-</span>p)<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb16-7"><a></a>  )</span>
<span id="cb16-8"><a></a>}</span>
<span id="cb16-9"><a></a></span>
<span id="cb16-10"><a></a><span class="co"># Offspring probability given parents</span></span>
<span id="cb16-11"><a></a>offspring_prob <span class="ot">&lt;-</span> <span class="cf">function</span>(p1_geno, p2_geno, </span>
<span id="cb16-12"><a></a>                          offspring_geno) {</span>
<span id="cb16-13"><a></a>  <span class="co"># Mendelian transmission probabilities</span></span>
<span id="cb16-14"><a></a>  <span class="co"># Implementation depends on parental genotypes</span></span>
<span id="cb16-15"><a></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div><div class="column" style="width:50%;">
<p><strong>Likelihood Components</strong> - <span class="math inline">L_{\text{parents}} = P(G_1, G_2 | p)</span> - <span class="math inline">L_{\text{offspring}} = \prod_{i=1}^3 P(G_i | G_1, G_2)</span> - <strong>Total</strong>: <span class="math inline">L(p) = L_{\text{parents}} \times L_{\text{offspring}}</span></p>
<p><strong>Interpretation</strong>: Peak likelihood value gives MLE for allele frequency</p>
</div></div>
</section>
<section id="building-one-locus-likelihood-functions-4" class="slide level2">
<h2>Building One-Locus Likelihood Functions</h2>
<h3 id="likelihood-surfaces-and-interpretation">Likelihood Surfaces and Interpretation</h3>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The following aesthetics were dropped during statistical transformation: fill.
ℹ This can happen when ggplot fails to infer the correct grouping structure in
  the data.
ℹ Did you forget to specify a `group` aesthetic or to convert a numerical
  variable into a factor?</code></pre>
</div>

</div>
<img data-src="lecture-01_files/figure-revealjs/unnamed-chunk-9-1.png" width="960" class="r-stretch"></section>
<section id="building-one-locus-likelihood-functions-5" class="slide level2">
<h2>Building One-Locus Likelihood Functions</h2>
<h3 id="key-principles-for-likelihood-construction">Key Principles for Likelihood Construction</h3>
<ol type="1">
<li class="fragment"><strong>Start with the data</strong>: What did you observe?</li>
<li class="fragment"><strong>Specify the model</strong>: Mode of inheritance, penetrance assumptions</li>
<li class="fragment"><strong>Write conditional probabilities</strong>: <span class="math inline">P(\text{data} | \text{parameters})</span></li>
<li class="fragment"><strong>Account for ascertainment</strong>: How were families/individuals selected?</li>
<li class="fragment"><strong>Maximize or explore</strong>: Find MLE or examine likelihood surface</li>
<li class="fragment"><strong>Interpret biologically</strong>: What do parameter estimates tell us about the genetics?</li>
</ol>
</section>
<section id="modeling-but-make-it-bayesian" class="slide level2">
<h2>Modeling, but make it bayesian</h2>
<ul>
<li>Let’s say we have run a previous experiment and observed <span class="math inline">n = 20</span> individuals with allele A and <span class="math inline">x = 3</span> individuals with allele A</li>
<li>We can use this information to inform our prior distribution for <span class="math inline">p</span></li>
</ul>
<div class="r-stack">
<div class="fragment fade-in-then-out">
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a></a>p <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">length.out =</span> <span class="dv">1000</span>)</span>
<span id="cb18-2"><a></a><span class="co"># Prior (Beta distribution from previous experiment)</span></span>
<span id="cb18-3"><a></a>prior <span class="ot">&lt;-</span> <span class="fu">dbeta</span>(p, <span class="dv">3</span> <span class="sc">+</span> <span class="dv">1</span>, <span class="dv">17</span> <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb18-4"><a></a><span class="co"># Likelihood (Beta distribution from current data)</span></span>
<span id="cb18-5"><a></a>likelihood <span class="ot">&lt;-</span> <span class="fu">dbeta</span>(p, <span class="dv">11</span> <span class="sc">+</span> <span class="dv">1</span>, <span class="dv">16</span> <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb18-6"><a></a><span class="co"># Posterior (Beta distribution combining both)</span></span>
<span id="cb18-7"><a></a>posterior <span class="ot">&lt;-</span> <span class="fu">dbeta</span>(p, <span class="dv">14</span> <span class="sc">+</span> <span class="dv">1</span>, <span class="dv">33</span> <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb18-8"><a></a><span class="co"># Combine into a data frame</span></span>
<span id="cb18-9"><a></a>plot_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb18-10"><a></a>    <span class="at">p =</span> <span class="fu">rep</span>(p, <span class="dv">3</span>),</span>
<span id="cb18-11"><a></a>    <span class="at">density =</span> <span class="fu">c</span>(prior, likelihood, posterior),</span>
<span id="cb18-12"><a></a>    <span class="at">type =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"Prior"</span>, <span class="st">"Likelihood"</span>, <span class="st">"Posterior"</span>), <span class="at">each =</span> <span class="fu">length</span>(p))</span>
<span id="cb18-13"><a></a>)</span>
<span id="cb18-14"><a></a><span class="co"># Add flat prior distribution</span></span>
<span id="cb18-15"><a></a>flat_prior <span class="ot">&lt;-</span> <span class="fu">dbeta</span>(p, <span class="dv">1</span>, <span class="dv">1</span>)</span>
<span id="cb18-16"><a></a>flat_posterior <span class="ot">&lt;-</span> <span class="fu">dbeta</span>(p, <span class="dv">12</span>, <span class="dv">17</span>)</span>
<span id="cb18-17"><a></a>plot_data <span class="ot">&lt;-</span> <span class="fu">rbind</span>(</span>
<span id="cb18-18"><a></a>    plot_data,</span>
<span id="cb18-19"><a></a>    <span class="fu">data.frame</span>(</span>
<span id="cb18-20"><a></a>        <span class="at">p =</span> <span class="fu">rep</span>(p, <span class="dv">3</span>),</span>
<span id="cb18-21"><a></a>        <span class="at">density =</span> <span class="fu">c</span>(flat_prior, likelihood, flat_posterior),</span>
<span id="cb18-22"><a></a>        <span class="at">type =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"Prior"</span>, <span class="st">"Likelihood"</span>, <span class="st">"Posterior"</span>), <span class="at">each =</span> <span class="fu">length</span>(p))</span>
<span id="cb18-23"><a></a>    )</span>
<span id="cb18-24"><a></a>)</span>
<span id="cb18-25"><a></a><span class="co"># Add a column to distinguish between informative and flat scenarios</span></span>
<span id="cb18-26"><a></a>plot_data<span class="sc">$</span>scenario <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"Informative"</span>, <span class="dv">3000</span>), <span class="fu">rep</span>(<span class="st">"Flat"</span>, <span class="dv">3000</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="fragment fade-in-then-out">
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb19-2"><a></a>    plot_data <span class="sc">%&gt;%</span> <span class="fu">filter</span>(scenario <span class="sc">==</span> <span class="st">"Informative"</span>),</span>
<span id="cb19-3"><a></a>    <span class="fu">aes</span>(<span class="at">x =</span> p, <span class="at">y =</span> density, <span class="at">color =</span> type)</span>
<span id="cb19-4"><a></a>) <span class="sc">+</span></span>
<span id="cb19-5"><a></a>    <span class="fu">geom_line</span>(<span class="at">linewidth =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb19-6"><a></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">11</span> <span class="sc">/</span> <span class="dv">27</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb19-7"><a></a>    <span class="fu">labs</span>(</span>
<span id="cb19-8"><a></a>        <span class="at">x =</span> <span class="st">"Allele Frequency (p)"</span>,</span>
<span id="cb19-9"><a></a>        <span class="at">y =</span> <span class="st">"Density"</span>,</span>
<span id="cb19-10"><a></a>        <span class="at">title =</span> <span class="st">"Bayesian Update with Informative Prior"</span></span>
<span id="cb19-11"><a></a>    ) <span class="sc">+</span></span>
<span id="cb19-12"><a></a>    <span class="fu">scale_color_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb19-13"><a></a>    <span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb20-2"><a></a>  plot_data <span class="sc">%&gt;%</span> <span class="fu">filter</span>(scenario <span class="sc">==</span> <span class="st">"Flat"</span>),</span>
<span id="cb20-3"><a></a>  <span class="fu">aes</span>(<span class="at">x =</span> p, <span class="at">y =</span> density, <span class="at">color =</span> type, <span class="at">linetype =</span> type)</span>
<span id="cb20-4"><a></a>) <span class="sc">+</span></span>
<span id="cb20-5"><a></a>  <span class="fu">geom_line</span>(<span class="at">linewidth =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb20-6"><a></a>  <span class="fu">scale_linetype_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"Prior"</span> <span class="ot">=</span> <span class="st">"dotted"</span>, <span class="st">"Likelihood"</span> <span class="ot">=</span> <span class="st">"solid"</span>, <span class="st">"Posterior"</span> <span class="ot">=</span> <span class="st">"dashed"</span>)) <span class="sc">+</span></span>
<span id="cb20-7"><a></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">11</span> <span class="sc">/</span> <span class="dv">27</span>, <span class="at">linetype =</span> <span class="st">"dotdash"</span>, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb20-8"><a></a>  <span class="fu">labs</span>(</span>
<span id="cb20-9"><a></a>    <span class="at">x =</span> <span class="st">"Allele Frequency (p)"</span>,</span>
<span id="cb20-10"><a></a>    <span class="at">y =</span> <span class="st">"Density"</span>,</span>
<span id="cb20-11"><a></a>    <span class="at">title =</span> <span class="st">"Bayesian Update with Flat Prior"</span></span>
<span id="cb20-12"><a></a>  ) <span class="sc">+</span></span>
<span id="cb20-13"><a></a>  <span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">15</span>) <span class="sc">+</span></span>
<span id="cb20-14"><a></a>  <span class="fu">scale_color_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb20-15"><a></a>  <span class="fu">annotate</span>(<span class="st">"text"</span>,</span>
<span id="cb20-16"><a></a>    <span class="at">x =</span> <span class="fl">0.7</span>, <span class="at">y =</span> <span class="fu">max</span>(flat_posterior) <span class="sc">*</span> <span class="fl">0.8</span>,</span>
<span id="cb20-17"><a></a>    <span class="at">label =</span> <span class="st">"Likelihood ≈ Posterior"</span>, <span class="at">color =</span> <span class="st">"gray30"</span></span>
<span id="cb20-18"><a></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div></div>
</div>
<div class="fragment">
<div class="cell quarto-layout-panel" data-layout-ncol="2">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img data-src="lecture-01_files/figure-revealjs/unnamed-chunk-13-1.png" width="960"></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img data-src="lecture-01_files/figure-revealjs/unnamed-chunk-13-2.png" width="960"></p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="modeling-but-make-it-bayesian-1" class="slide level2">
<h2>Modeling, but make it Bayesian</h2>

<img src="scripts/_output/sequential_bayesian_update.gif" alt="Bayesian updating" class="r-stretch"></section>
<section id="mendels-laws" class="slide level2">
<h2>Mendel’s Laws</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><strong>Mendel’s First Law (Segregation)</strong>: One allele of each parent is randomly and independently selected, with probability 1/2 for transmission to the offspring; the alleles unite randomly to form the offspring’s genotype.</p>
<p><strong>Mendel’s Second Law (Independent Assortment)</strong>: The alleles underlying two or more different traits are transmitted to offspring independently of each other; the transmission of each trait separately follows the first law of segregation.</p>
</div><div class="column" style="width:50%;">
<p><img data-src="https://upload.wikimedia.org/wikipedia/commons/8/8e/Intermediate_inheritance_P_-_F1_-_F2.png" style="width:100.0%"></p>
</div></div>
</section>
<section id="worked-example-segregation-31-ratio" class="slide level2">
<h2>Worked Example: Segregation (3:1 Ratio)</h2>
<h3 id="cross-aa-aa">Cross: Aa × Aa</h3>
<div class="quarto-layout-panel" data-layout-ncol="2">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p><strong>Punnett Square</strong></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<table class="caption-top">
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;"><strong>A</strong></th>
<th style="text-align: left;"><strong>a</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><strong>A</strong></td>
<td style="text-align: left;">AA</td>
<td style="text-align: left;">Aa</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>a</strong></td>
<td style="text-align: left;">Aa</td>
<td style="text-align: left;">aa</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p><strong>Expected Ratios</strong> - <strong>Genotypic</strong>: 1 AA : 2 Aa : 1 aa - <strong>Phenotypic</strong>: 3 dominant : 1 recessive - <strong>Probabilities</strong>: <span class="math inline">P(AA) = \frac{1}{4}</span>, <span class="math inline">P(Aa) = \frac{1}{2}</span>, <span class="math inline">P(aa) = \frac{1}{4}</span></p>
</div>
</div>
</div>
</section>
<section id="testing-segregation-ratios" class="slide level2">
<h2>Testing Segregation Ratios</h2>
<p>Observed offspring from Aa × Aa cross: 240 total - 180 dominant phenotype, 60 recessive phenotype - <strong>Expected under 3:1 ratio</strong>: 180 dominant, 60 recessive</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a></a><span class="co"># Test for 3:1 segregation ratio</span></span>
<span id="cb21-2"><a></a>observed <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">180</span>, <span class="dv">60</span>)</span>
<span id="cb21-3"><a></a>expected <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">180</span>, <span class="dv">60</span>)  <span class="co"># 3:1 ratio from 240 total</span></span>
<span id="cb21-4"><a></a>chi_sq <span class="ot">&lt;-</span> <span class="fu">sum</span>((observed <span class="sc">-</span> expected)<span class="sc">^</span><span class="dv">2</span> <span class="sc">/</span> expected)</span>
<span id="cb21-5"><a></a>p_value <span class="ot">&lt;-</span> <span class="fu">pchisq</span>(chi_sq, <span class="at">df =</span> <span class="dv">1</span>, <span class="at">lower.tail =</span> <span class="cn">FALSE</span>)</span>
<span id="cb21-6"><a></a><span class="fu">cat</span>(<span class="st">"χ² ="</span>, chi_sq, <span class="st">", p-value ="</span>, p_value)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>χ² = 0 , p-value = 1</code></pre>
</div>
</div>
<p><strong>Conclusion</strong>: Perfect fit to Mendelian expectations (p = 1.0)</p>
</section>
<section id="test-cross-example" class="slide level2">
<h2>Test Cross Example</h2>
<h3 id="cross-aa-aa-11-expected">Cross: Aa × aa (1:1 Expected)</h3>
<div class="quarto-layout-panel" data-layout-ncol="2">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p><strong>Punnett Square</strong></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<table class="caption-top">
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;"><strong>A</strong></th>
<th style="text-align: left;"><strong>a</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><strong>a</strong></td>
<td style="text-align: left;">Aa</td>
<td style="text-align: left;">aa</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>a</strong></td>
<td style="text-align: left;">Aa</td>
<td style="text-align: left;">aa</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p><strong>Purpose &amp; Results</strong> - <strong>Use</strong>: Test if unknown dominant is AA or Aa - <strong>Expected</strong>: 1:1 ratio (50% each phenotype) - <strong>Observed</strong>: 23 dominant, 27 recessive (n=50)</p>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a></a>observed_testcross <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">23</span>, <span class="dv">27</span>)</span>
<span id="cb23-2"><a></a>expected_testcross <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">25</span>, <span class="dv">25</span>)  <span class="co"># 1:1 ratio</span></span>
<span id="cb23-3"><a></a>chi_sq_test <span class="ot">&lt;-</span> <span class="fu">sum</span>((observed_testcross <span class="sc">-</span> expected_testcross)<span class="sc">^</span><span class="dv">2</span> <span class="sc">/</span> expected_testcross)</span>
<span id="cb23-4"><a></a>p_val_test <span class="ot">&lt;-</span> <span class="fu">pchisq</span>(chi_sq_test, <span class="at">df =</span> <span class="dv">1</span>, <span class="at">lower.tail =</span> <span class="cn">FALSE</span>)</span>
<span id="cb23-5"><a></a><span class="fu">cat</span>(<span class="st">"χ² ="</span>, chi_sq_test, <span class="st">", p-value ="</span>, p_val_test)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>χ² = 0.32 , p-value = 0.5716076</code></pre>
</div>
</div>
</section>
<section id="dihybrid-cross-independent-assortment" class="slide level2">
<h2>Dihybrid Cross: Independent Assortment</h2>
<h3 id="aabb-aabb-9331-expected">AaBb × AaBb (9:3:3:1 Expected)</h3>
<p><strong>Two traits</strong>: Seed shape (A/a) and color (B/b)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a></a><span class="co"># Calculate expected probabilities</span></span>
<span id="cb25-2"><a></a>prob_AB <span class="ot">&lt;-</span> (<span class="dv">3</span><span class="sc">/</span><span class="dv">4</span>) <span class="sc">*</span> (<span class="dv">3</span><span class="sc">/</span><span class="dv">4</span>)  <span class="co"># 9/16 Round Yellow</span></span>
<span id="cb25-3"><a></a>prob_Ab <span class="ot">&lt;-</span> (<span class="dv">3</span><span class="sc">/</span><span class="dv">4</span>) <span class="sc">*</span> (<span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>)  <span class="co"># 3/16 Round Green</span></span>
<span id="cb25-4"><a></a>prob_aB <span class="ot">&lt;-</span> (<span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>) <span class="sc">*</span> (<span class="dv">3</span><span class="sc">/</span><span class="dv">4</span>)  <span class="co"># 3/16 Wrinkled Yellow  </span></span>
<span id="cb25-5"><a></a>prob_ab <span class="ot">&lt;-</span> (<span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>) <span class="sc">*</span> (<span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>)  <span class="co"># 1/16 Wrinkled Green</span></span>
<span id="cb25-6"><a></a></span>
<span id="cb25-7"><a></a><span class="fu">cat</span>(<span class="st">"Expected ratios:"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Expected ratios:</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Round Yellow:"</span>, <span class="dv">9</span>, <span class="st">"("</span>, prob_AB, <span class="st">")"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Round Yellow: 9 ( 0.5625 )</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Round Green:"</span>, <span class="dv">3</span>, <span class="st">"("</span>, prob_Ab, <span class="st">")"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Round Green: 3 ( 0.1875 )</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Wrinkled Yellow:"</span>, <span class="dv">3</span>, <span class="st">"("</span>, prob_aB, <span class="st">")"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Wrinkled Yellow: 3 ( 0.1875 )</code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Wrinkled Green:"</span>, <span class="dv">1</span>, <span class="st">"("</span>, prob_ab, <span class="st">")"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Wrinkled Green: 1 ( 0.0625 )</code></pre>
</div>
</div>
</section>
<section id="testing-independent-assortment" class="slide level2">
<h2>Testing Independent Assortment</h2>
<p><strong>Mendel’s observed data</strong>: 556 F₂ offspring - Round Yellow: 315, Round Green: 108<br>
- Wrinkled Yellow: 101, Wrinkled Green: 32</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a></a>total_f2 <span class="ot">&lt;-</span> <span class="dv">556</span></span>
<span id="cb35-2"><a></a>observed_dihybrid <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">315</span>, <span class="dv">108</span>, <span class="dv">101</span>, <span class="dv">32</span>)</span>
<span id="cb35-3"><a></a>expected_dihybrid <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb35-4"><a></a>  total_f2 <span class="sc">*</span> <span class="dv">9</span><span class="sc">/</span><span class="dv">16</span>,  <span class="co"># Round Yellow: 312.75</span></span>
<span id="cb35-5"><a></a>  total_f2 <span class="sc">*</span> <span class="dv">3</span><span class="sc">/</span><span class="dv">16</span>,  <span class="co"># Round Green: 104.25</span></span>
<span id="cb35-6"><a></a>  total_f2 <span class="sc">*</span> <span class="dv">3</span><span class="sc">/</span><span class="dv">16</span>,  <span class="co"># Wrinkled Yellow: 104.25</span></span>
<span id="cb35-7"><a></a>  total_f2 <span class="sc">*</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">16</span>   <span class="co"># Wrinkled Green: 34.75</span></span>
<span id="cb35-8"><a></a>)</span>
<span id="cb35-9"><a></a></span>
<span id="cb35-10"><a></a>chi_sq_dihybrid <span class="ot">&lt;-</span> <span class="fu">sum</span>((observed_dihybrid <span class="sc">-</span> expected_dihybrid)<span class="sc">^</span><span class="dv">2</span> <span class="sc">/</span> expected_dihybrid)</span>
<span id="cb35-11"><a></a>p_val_dihybrid <span class="ot">&lt;-</span> <span class="fu">pchisq</span>(chi_sq_dihybrid, <span class="at">df =</span> <span class="dv">3</span>, <span class="at">lower.tail =</span> <span class="cn">FALSE</span>)</span>
<span id="cb35-12"><a></a></span>
<span id="cb35-13"><a></a><span class="fu">cat</span>(<span class="st">"Expected:"</span>, <span class="fu">round</span>(expected_dihybrid, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Expected: 312.8 104.2 104.2 34.8</code></pre>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Observed:"</span>, observed_dihybrid)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Observed: 315 108 101 32</code></pre>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">χ² ="</span>, <span class="fu">round</span>(chi_sq_dihybrid, <span class="dv">3</span>), <span class="st">", p-value ="</span>, <span class="fu">round</span>(p_val_dihybrid, <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
χ² = 0.47 , p-value = 0.925</code></pre>
</div>
</div>
<p><strong>Conclusion</strong>: Excellent fit supports independent assortment</p>
</section>
<section id="genetic-models" class="slide level2">
<h2>Genetic Models</h2>
<h3 id="mendelian">Mendelian</h3>
<div title="Dominant">
<div class="callout callout-none no-icon callout-titled callout-style-simple">
<div class="callout-body">
<div class="callout-title">
<p><strong>Dominant</strong></p>
</div>
<div class="callout-content">
<p><span class="math display">P(Y = 1 | DD) = P(Y = 1 | Dd) = 1</span> <span class="math display">P(Y = 1 | dd) = 0</span></p>
</div>
</div>
</div>
</div>
<div title="Recessive">
<div class="callout callout-none no-icon callout-titled callout-style-simple">
<div class="callout-body">
<div class="callout-title">
<p><strong>Recessive</strong></p>
</div>
<div class="callout-content">
<p><span class="math display">P(Y = 1 | DD) = 1</span> <span class="math display">P(Y = 1 | Dd) = P(Y = 1 | dd) = 0</span></p>
</div>
</div>
</div>
</div>
<p>These models are not reasonable for non rare/mendelian diseases</p>
</section>
<section id="penetrance" class="slide level2">
<h2>Penetrance</h2>
<p>A genetic model is of <strong>reduced penetrance</strong> if the probability of disease is less than 1 for individuals with a disease genotype</p>
<ul>
<li>Recessive: <span class="math inline">P(Y = 1 | DD) &lt; 1</span></li>
<li>Dominant: <span class="math inline">P(Y = 1 | DD) &lt; 1, P(Y = 1 | Dd) &lt; 1</span></li>
</ul>
</section>
<section id="continuous-penetrance-functions" class="slide level2">
<h2>Continuous penetrance functions</h2>
<p>What is the difference in what these two plots are showing?</p>
<div class="quarto-layout-panel" data-layout-ncol="2">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img data-src="images/penetrance-01.png"></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img data-src="images/penetrance-02.jpg"></p>
</div>
</div>
</div>
<p><span class="column-margin margin-aside" style="font-size:.6em;">From Goodrich, J. K et al.&nbsp;(2021), “Determinants of penetrance and variable expressivity in monogenic metabolic conditions across 77,184 exomes,” Nature Communications, Nature Publishing Group, 12, 3505.</span></p>
</section>
<section id="mode-of-inheritance" class="slide level2">
<h2>Mode of Inheritance</h2>
<p>For quantitative traits, there exist four modes of inheritance:</p>
<ol type="1">
<li>Dominant, when only one copy of the allele is required to induce an effect on the disease phenotype</li>
<li>Recessive, when two copies of the allele are required to induce an effect on the disease phenotype</li>
<li>Additive, when the effect of the allele is proportional to the number of copies present</li>
<li>Codominant, when the effect of the allele is equal to the sum of the effects of each copy</li>
</ol>
</section>
<section>
<section id="detour" class="title-slide slide level1 center">
<h1>DETOUR</h1>

</section>
<section id="generalized-linear-models" class="slide level2">
<h2>Generalized Linear Models</h2>
<div title="Simple Linear Model">
<div class="callout callout-none no-icon callout-titled callout-style-simple">
<div class="callout-body">
<div class="callout-title">
<p><strong>Simple Linear Model</strong></p>
</div>
<div class="callout-content">
<p><span class="math display">
y_i = \beta_0 + \beta_1 x_i + \epsilon_i
</span></p>
</div>
</div>
</div>
</div>
<div title="Generalized Linear Model">
<div class="callout callout-none no-icon callout-titled callout-style-simple">
<div class="callout-body">
<div class="callout-title">
<p><strong>Generalized Linear Model</strong></p>
</div>
<div class="callout-content">
<p><span class="math display">
g(y_i) = \beta_0 + \beta_1 x_i + \epsilon_i
</span></p>
<p>Where</p>
<ul>
<li><span class="math inline">g(\cdot)</span> is a link function that relates the linear predictor to the mean of the response variable</li>
<li><span class="math inline">V(\cdot)</span> is a variance function that relates the mean of the response variable to its variance</li>
</ul>
</div>
</div>
</div>
</div>
</section></section>
<section>
<section id="population-genetics" class="title-slide slide level1 center">
<h1>Population Genetics</h1>

</section>
<section id="estimating-allele-frequencies" class="slide level2">
<h2>Estimating Allele Frequencies</h2>
<p>Consider a sample of <span class="math inline">n</span> individuals, and let</p>
<ul>
<li><span class="math inline">n_{AA}</span> be the number of individuals with genotype <span class="math inline">AA</span></li>
<li><span class="math inline">n_{Aa}</span> be the number of individuals with genotype <span class="math inline">Aa</span></li>
<li><span class="math inline">n_{aa}</span> be the number of individuals with genotype <span class="math inline">aa</span></li>
</ul>
<ol type="1">
<li>What is <span class="math inline">n_{AA} + n_{Aa} + n_{aa}</span>?</li>
<li>What is the frequency of allele <span class="math inline">A</span> in the population?</li>
<li>How many alleles are there in total?</li>
<li>What is the frequency of allele <span class="math inline">a</span> in the population?</li>
<li>What is the standard error of the allele frequency estimate?</li>
</ol>
</section>
<section id="population-substructure" class="slide level2">
<h2>Population Substructure</h2>

<img data-src="https://media.springernature.com/full/springer-static/image/art%3A10.1186%2Fgb-2011-12-5-223/MediaObjects/13059_2011_Article_2523_Fig1_HTML.jpg" class="r-stretch quarto-figure-center"><p class="caption">Shriner, D., Adeyemo, A., Ramos, E., Chen, G., and Rotimi, C. N. (2011), <a href="https://doi.org/10.1186/gb-2011-12-5-223">“Mapping of disease-associated variants in admixed populations,”</a> Genome Biology, BioMed Central, 12, 1–8.</p></section>
<section id="hardy-weinberg-equilibrium" class="slide level2">
<h2>Hardy-Weinberg Equilibrium</h2>

<img data-src="https://cdn.kastatic.org/ka-perseus-images/010efe45f63d8da65c92d7af25c503c0bc91f105.png" class="r-stretch quarto-figure-center"><p class="caption">Khan Academy</p></section>
<section id="testing-for-hwe" class="slide level2" data-auto-animate="true">
<h2 data-id="quarto-animate-title">Testing for HWE</h2>
<ol type="1">
<li>Estimate allele frequencies from the sample</li>
<li>Estimate <span class="math inline">\hat{p}</span>, the frequency of allele A</li>
<li>Calculate expected genotype frequencies under HWE</li>
<li>Compare observed and expected genotype frequencies using a statistical test (e.g., Pearson goodness-of-fit test)</li>
</ol>
</section>
<section id="testing-for-hwe-1" class="slide level2" data-auto-animate="true">
<h2 data-id="quarto-animate-title">Testing for HWE</h2>
<ol type="1">
<li><strong>Estimate allele frequencies from the sample</strong></li>
<li>Estimate <span class="math inline">\hat{p}</span>, the frequency of allele A</li>
<li>Calculate expected genotype frequencies under HWE</li>
<li>Compare observed and expected genotype frequencies using a statistical test (e.g., Pearson goodness-of-fit test)</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode numberSource r number-lines code-with-copy" data-id="quarto-animate-code"><code class="sourceCode r hljs"><span id="cb41-1" class="hljs-ln-code"><a></a>n <span class="ot">&lt;-</span> <span class="dv">212</span></span>
<span id="cb41-2" class="hljs-ln-code"><a></a>n_AA <span class="ot">&lt;-</span> <span class="dv">175</span></span>
<span id="cb41-3" class="hljs-ln-code"><a></a>n_Aa <span class="ot">&lt;-</span> <span class="dv">33</span></span>
<span id="cb41-4" class="hljs-ln-code"><a></a>n_aa <span class="ot">&lt;-</span> <span class="dv">4</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="testing-for-hwe-2" class="slide level2" data-auto-animate="true">
<h2 data-id="quarto-animate-title">Testing for HWE</h2>
<ol type="1">
<li>Estimate allele frequencies from the sample</li>
<li><strong>Estimate</strong> <span class="math inline">\hat{p}</span>, the frequency of allele A</li>
<li>Calculate expected genotype frequencies under HWE</li>
<li>Compare observed and expected genotype frequencies using a statistical test (e.g., Pearson goodness-of-fit test)</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode numberSource r number-lines code-with-copy" data-id="quarto-animate-code"><code class="sourceCode r hljs"><span id="cb42-1" class="hljs-ln-code"><a></a>n <span class="ot">&lt;-</span> <span class="dv">212</span></span>
<span id="cb42-2" class="hljs-ln-code"><a></a>n_AA <span class="ot">&lt;-</span> <span class="dv">175</span></span>
<span id="cb42-3" class="hljs-ln-code"><a></a>n_Aa <span class="ot">&lt;-</span> <span class="dv">33</span></span>
<span id="cb42-4" class="hljs-ln-code"><a></a>n_aa <span class="ot">&lt;-</span> <span class="dv">4</span></span>
<span id="cb42-5" class="hljs-ln-code"><a></a></span>
<span id="cb42-6" class="hljs-ln-code"><a></a>p_hat <span class="ot">&lt;-</span> (<span class="dv">2</span><span class="sc">*</span> n_AA <span class="sc">+</span> n_Aa) <span class="sc">/</span> (<span class="dv">2</span><span class="sc">*</span>n)</span>
<span id="cb42-7" class="hljs-ln-code"><a></a>p_hat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.9033019</code></pre>
</div>
</div>
</section>
<section id="testing-for-hwe-3" class="slide level2" data-auto-animate="true">
<h2 data-id="quarto-animate-title">Testing for HWE</h2>
<ol type="1">
<li>Estimate allele frequencies from the sample</li>
<li>Estimate <span class="math inline">\hat{p}</span>, the frequency of allele A</li>
<li><strong>Calculate expected genotype frequencies under HWE</strong></li>
<li>Compare observed and expected genotype frequencies using a statistical test (e.g., Pearson goodness-of-fit test)</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a></a>n <span class="ot">&lt;-</span> <span class="dv">212</span></span>
<span id="cb44-2"><a></a>n_AA <span class="ot">&lt;-</span> <span class="dv">175</span></span>
<span id="cb44-3"><a></a>n_Aa <span class="ot">&lt;-</span> <span class="dv">33</span></span>
<span id="cb44-4"><a></a>n_aa <span class="ot">&lt;-</span> <span class="dv">4</span></span>
<span id="cb44-5"><a></a></span>
<span id="cb44-6"><a></a>p_hat <span class="ot">&lt;-</span> (<span class="dv">2</span><span class="sc">*</span> n_AA <span class="sc">+</span> n_Aa) <span class="sc">/</span> (<span class="dv">2</span><span class="sc">*</span>n)</span>
<span id="cb44-7"><a></a>p_hat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.9033019</code></pre>
</div>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a></a>e_AA <span class="ot">&lt;-</span> <span class="fu">round</span>(n <span class="sc">*</span> p_hat<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb46-2"><a></a>e_Aa <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="dv">2</span> <span class="sc">*</span> n <span class="sc">*</span> p_hat <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> p_hat))</span>
<span id="cb46-3"><a></a>e_aa <span class="ot">&lt;-</span> <span class="fu">round</span>(n <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> p_hat)<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb46-4"><a></a><span class="fu">c</span>(e_AA, e_Aa, e_aa)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 173  37   2</code></pre>
</div>
</div>
</section>
<section id="testing-for-hwe-4" class="slide level2" data-auto-animate="true">
<h2 data-id="quarto-animate-title">Testing for HWE</h2>
<ol type="1">
<li>Estimate allele frequencies from the sample</li>
<li>Estimate <span class="math inline">\hat{p}</span>, the frequency of allele A</li>
<li>Calculate expected genotype frequencies under HWE</li>
<li><strong>Compare observed and expected genotype frequencies using a statistical test (e.g., Pearson goodness-of-fit test)</strong></li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a></a>n <span class="ot">&lt;-</span> <span class="dv">212</span></span>
<span id="cb48-2"><a></a>n_AA <span class="ot">&lt;-</span> <span class="dv">175</span></span>
<span id="cb48-3"><a></a>n_Aa <span class="ot">&lt;-</span> <span class="dv">33</span></span>
<span id="cb48-4"><a></a>n_aa <span class="ot">&lt;-</span> <span class="dv">4</span></span>
<span id="cb48-5"><a></a></span>
<span id="cb48-6"><a></a>p_hat <span class="ot">&lt;-</span> (<span class="dv">2</span><span class="sc">*</span> n_AA <span class="sc">+</span> n_Aa) <span class="sc">/</span> (<span class="dv">2</span><span class="sc">*</span>n)</span>
<span id="cb48-7"><a></a>p_hat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.9033019</code></pre>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a></a>e_AA <span class="ot">&lt;-</span> <span class="fu">round</span>(n <span class="sc">*</span> p_hat<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb50-2"><a></a>e_Aa <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="dv">2</span> <span class="sc">*</span> n <span class="sc">*</span> p_hat <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> p_hat))</span>
<span id="cb50-3"><a></a>e_aa <span class="ot">&lt;-</span> <span class="fu">round</span>(n <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> p_hat)<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb50-4"><a></a><span class="fu">c</span>(e_AA, e_Aa, e_aa)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 173  37   2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a></a>observed <span class="ot">&lt;-</span> <span class="fu">c</span>(n_AA, n_Aa, n_aa)</span>
<span id="cb52-2"><a></a>expected <span class="ot">&lt;-</span> <span class="fu">c</span>(e_AA, e_Aa, e_aa)</span>
<span id="cb52-3"><a></a>test_stat <span class="ot">&lt;-</span> <span class="fu">sum</span>((observed <span class="sc">-</span> expected)<span class="sc">^</span><span class="dv">2</span> <span class="sc">/</span> expected)</span>
<span id="cb52-4"><a></a><span class="fu">pchisq</span>(test_stat, <span class="at">df =</span> <span class="dv">1</span>, <span class="at">lower.tail =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1171099</code></pre>
</div>
</div>


</section></section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">

</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"8\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
            const codeEl = trigger.previousElementSibling.cloneNode(true);
            for (const childEl of codeEl.children) {
              if (isCodeAnnotation(childEl)) {
                childEl.remove();
              }
            }
            return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp('/' + window.location.host + '/');
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    

</body></html>