<!DOCTYPE html>
<html lang="en"><head>
<link href="../course_logo.png" rel="icon" type="image/png">
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-html/tabby.min.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.8.25">

  <title>Statistical Genetics – Lecture 01: Introduction and Foundations</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    html { -webkit-text-size-adjust: 100%; }
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; font-weight: bold; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="../site_libs/revealjs/dist/theme/quarto-a8f461879e61fe6cdedd4a5ed206ce40.css">
  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.css">
  <link href="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Lecture 01: Introduction and Foundations</h1>
  <p class="subtitle">PUBH 8878, Statistical Genetics</p>

<div class="quarto-title-authors">
</div>

</section>
<section id="about-the-instructors" class="slide level2">
<h2>About the Instructors</h2>
<div class="quarto-layout-panel" data-layout="[100, 70, 66.5]">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 42.3%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://publichealth.gwu.edu/sites/g/files/zaxdzs4586/files/styles/d07/public/2023-02/keith-crandall-up-2014-jmc_3352lowres.jpeg"></p>
<figcaption>Dr.&nbsp;Keith Crandall</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 29.6%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://publichealth.gwu.edu/sites/g/files/zaxdzs4586/files/styles/d07/public/2024-11/ali_rahnavard_102024.jpg"></p>
<figcaption>Dr.&nbsp;Ali Rahnavard</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 28.1%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://chiraaggohel.com/favicon.jpeg"></p>
<figcaption>Chiraag Gohel</figcaption>
</figure>
</div>
</div>
</div>
</div>
</section>
<section id="introductions" class="slide level2">
<h2>Introductions</h2>
<ol type="1">
<li>Name, Degree, any other ways you like to define yourself (favorite hot dog brand, etc.)</li>
<li>Current research focus or research interests.</li>
<li>What you hope to get out of this course.</li>
</ol>
</section>
<section id="course-site" class="slide level2">
<h2>Course Site</h2>
<p><a href="https://gwcbi.github.io/StatGen/" class="uri">https://gwcbi.github.io/StatGen/</a></p>
<iframe src="https://gwcbi.github.io/StatGen/" width="100%" height="600px">
</iframe>
</section>
<section id="zotero" class="slide level2">
<h2>Zotero</h2>

<img data-src="images/paste-1.png" class="r-stretch"></section>
<section id="course-logistics-and-expectations" class="slide level2">
<h2>Course Logistics and Expectations</h2>
<h3 id="grading-breakdown">Grading Breakdown</h3>
<table class="caption-top">
<thead>
<tr class="header">
<th>Assignment Type</th>
<th>% of Grade</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Problem Sets</td>
<td>60%</td>
</tr>
<tr class="even">
<td>Class Participation</td>
<td>20%</td>
</tr>
<tr class="odd">
<td>Research Project</td>
<td>20%</td>
</tr>
</tbody>
</table>
<p><strong>Total Workload</strong>: 112.5 hours (5+ hrs/week independent + 6 hrs/week class/async)</p>
</section>
<section id="problem-sets-60-of-grade" class="slide level2">
<h2>Problem Sets (60% of grade)</h2>
<ul>
<li><strong>Weekly assignments</strong> blending mathematical derivations with coding</li>
<li><strong>Collaborative time</strong> provided in class</li>
<li><strong>Individual work</strong> required – substantial time outside class</li>
<li><strong>Late penalty</strong>: 1% per hour (first 5 hours), then 5% per day</li>
<li><strong>Format</strong>: Mix of theory problems and R/Bioconductor analysis</li>
</ul>
</section>
<section id="class-participation-20-of-grade" class="slide level2">
<h2>Class Participation (20% of grade)</h2>
<p><strong>More than just showing up!</strong></p>
<ul>
<li>Come <strong>prepared</strong> with questions from async materials</li>
<li><strong>Engage thoughtfully</strong> in discussions and problem-solving</li>
<li><strong>Demonstrate mastery</strong> through insightful contributions</li>
<li><strong>Collaborate respectfully</strong> - help others understand concepts</li>
</ul>
<p><strong>Quantitative + Qualitative Assessment</strong>: Async responses + live session engagement</p>
</section>
<section id="research-project-20-of-grade" class="slide level2">
<h2>Research Project (20% of grade)</h2>
<p><strong>Final third of semester</strong> - Choose your own adventure!</p>
<div class="quarto-layout-panel" data-layout-ncol="2">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p><strong>Project Types</strong></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<ul>
<li>Computational methods</li>
<li>Applied analysis</li>
<li>Methodological development</li>
<li>All based on primary literature</li>
</ul>
</div>
</div>
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p><strong>Deliverables</strong></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<ul>
<li>Scientific paper (Genetics journal format)</li>
<li>Conference-style presentation</li>
<li>Groups of up to 3 encouraged</li>
<li>Individual contributions must be clear</li>
</ul>
</div>
</div>
</div>
</section>
<section id="course-structure-overview" class="slide level2">
<h2>Course Structure Overview</h2>
<p><strong>Foundation → Scale → Specialization</strong> (10 weeks)</p>
<ul>
<li><strong>Weeks 1-4</strong>: Foundations (Mendelian genetics, heritability, likelihood algorithms, population structure &amp; Bayesian methods)</li>
<li><strong>Week 5</strong>: GWAS at scale (design, linear mixed models, meta-analysis)</li>
<li><strong>Weeks 6-7</strong>: Prediction models &amp; multiple testing control</li>
<li><strong>Weeks 8-9</strong>: Binary traits &amp; causal inference (Mendelian randomization)</li>
<li><strong>Week 10</strong>: Advanced AI topics in statistical genetics</li>
</ul>
<p><strong>Lab Component</strong>: Hands-on R/Bioconductor workflows parallel to lectures</p>
</section>
<section id="technology-setup-requirements" class="slide level2">
<h2>Technology Setup Requirements</h2>
<ul>
<li class="fragment">R (4.5+) and IDE installation</li>
<li class="fragment">Git setup and basic commands</li>
<li class="fragment">Zotero for reference management</li>
<li class="fragment">Zoom</li>
<li class="fragment">Optional: HPC access</li>
</ul>
</section>
<section id="course-textbook" class="slide level2">
<h2>Course Textbook</h2>

<img data-src="images/paste-6.png" class="r-stretch quarto-figure-center"><p class="caption"><span class="citation" data-cites="sorensenStatisticalLearningGenetics2025">Sorensen (<a href="#/references" role="doc-biblioref" onclick="">2025</a>)</span></p></section>
<section id="what-is-statistical-genetics" class="slide level2">
<h2>What is statistical genetics?</h2>
<div class="quarto-layout-panel" data-layout-ncol="3">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 33.3%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://upload.wikimedia.org/wikipedia/commons/thumb/a/ae/Darwin%27s_finches_by_Gould.jpg/1590px-Darwin%27s_finches_by_Gould.jpg"></p>
<figcaption>Population Genetics</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 33.3%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41576-022-00478-5/MediaObjects/41576_2022_478_Fig2_HTML.png"></p>
<figcaption>Genetic Epidemiology</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 33.3%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://brightspotcdn.byu.edu/dims4/default/6658c6f/2147483647/strip/true/crop/1602x2400+0+0/resize/534x800!/quality/90/?url=https%3A%2F%2Fbrigham-young-brightspot-us-east-2.s3.us-east-2.amazonaws.com%2F9b%2Faf%2F9f5a135aaaf234b6420456b7312c%2Fbradley-measure.jpg"></p>
<figcaption>Quantitative Genetics</figcaption>
</figure>
</div>
</div>
</div>
</div>
<aside class="notes">
<ul>
<li>Population genetics is the study of evolutionary processes affecting genetic variation between organisms. A classic example is Darwin’s finches, where beak shape and size vary between islands in the Galapagos, driven by natural selection.</li>
<li>Genetic Epidemiology is the study of how genetic factors contribute to health and disease in populations. A classic example is the study of genetic variants associated with diseases like diabetes or heart disease.</li>
<li>Quantitative genetics is the study of continuous traits influenced by multiple genes and environmental factors. A classic example is human height, which is influenced by many genetic variants and environmental factors like nutrition. Shawn Bradley was a basketball player who as 7’6 tall, The study found Bradley had 198 more height-associated genetic variants than the average person in the sample</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="a-statistical-geneticist-may-want-to-know" class="slide level2">
<h2>A statistical geneticist may want to know</h2>
<ol type="1">
<li>Is there a genetic component contributing to the total variance of these traits?</li>
<li>Is the genetic component of the traits driven by a few genes located on a particular chromosome, or are there many genes scattered across many chromosomes? How many genes are involved and is this a scientifically sensible question?</li>
<li>Are the genes detected protein-coding genes, or are there also noncoding genes involved in gene regulation?</li>
<li>How is the strength of the signals captured in a statistical analysis related to the two types of genes? What fraction of the total genetic variation is allocated to both types of genes?</li>
<li>What are the frequencies of the genes in the sample? Are the frequencies associated with the magnitude of their effects on the traits?</li>
<li>What is the mode of action of the genes?</li>
</ol>
</section>
<section id="a-statistical-geneticist-may-want-to-know-1" class="slide level2">
<h2>A statistical geneticist may want to know</h2>
<ol start="7" type="1">
<li>What proportion of the genetic variance estimated in 1 can be explained by the discovered genes?</li>
<li>Given the information on the set of genes carried by an individual, will a genetic score constructed before observing the trait help with early diagnosis and prevention?</li>
<li>How should the predictive ability of the score be measured?</li>
<li>Are there other non-genetic factors that affect the traits, such as smoking behavior, alcohol consumption, blood pressure measurements, body mass index and level of physical exercise?</li>
<li>Could the predictive ability of the genetic score be improved by incorporation of these non-genetic sources of information, either additively or considering interactions? What is the relative contribution from the different sources of information?</li>
</ol>
</section>
<section id="what-does-the-data-look-like" class="slide level2">
<h2>What does the data look like?</h2>
<p><strong>Family/Pedigree Studies</strong></p>

<img data-src="https://www.genome.gov/sites/default/files/media/images/tg/Pedigree.jpg" class="r-stretch quarto-figure-center"><p class="caption">Family/Pedigree Studies</p><aside class="notes">
<ul>
<li>Family studies are used to understand the inheritance patterns of traits and diseases within families. They can help identify genetic factors contributing to diseases by analyzing the occurrence of traits in related individuals.</li>
<li>Pedigree charts visually represent family relationships and the transmission of genetic traits across generations. They are useful for tracking the inheritance of specific traits or diseases within a family.</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="what-does-the-data-look-like-1" class="slide level2">
<h2>What does the data look like?</h2>
<p><strong>Genome-Wide Association Studies (GWAS)</strong></p>

<img data-src="https://upload.wikimedia.org/wikipedia/commons/4/4f/Manhattan_plot_from_a_GWAS_of_kidney_stone_disease.png" style="width:60.0%" class="r-stretch quarto-figure-center"><p class="caption">Genome-Wide Association Studies</p><aside class="notes">
<ul>
<li>GWAS are studies that look for associations between genetic variants across the genome and specific traits or diseases in a large population. They can help identify genetic factors contributing to complex diseases by analyzing the frequency of genetic variants in affected and unaffected individuals.</li>
<li>Manhattan plots are visual representations of GWAS results, where each point represents a genetic variant and its association with the trait of interest. The x-axis represents the genomic position, while the y-axis represents the significance of the association (usually -log10(p-value)). Peaks in the plot indicate regions of the genome that may be associated with the trait.</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="core-abstractions-for-quantitative-biology" class="slide level2">
<h2>Core abstractions for quantitative biology</h2>
<ol type="1">
<li><strong>Model-Building</strong>
<ul>
<li>Formulate generative models linking genotype, environment, and phenotype (e.g., linear mixed, Bayesian hierarchical, non-parametric kernels).</li>
<li>Encode biological structure: linkage &amp; LD, population stratification, dominance/epistasis, multi-omics priors.</li>
<li>Balance realism and tractability to enable scalable computation on genome-scale data.</li>
</ul></li>
</ol>

<img data-src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41467-021-25756-4/MediaObjects/41467_2021_25756_Fig1_HTML.png" class="r-stretch"></section>
<section id="core-abstractions-for-quantitative-biology-1" class="slide level2">
<h2>Core abstractions for quantitative biology</h2>
<ol start="2" type="1">
<li><strong>Inference</strong>
<ul>
<li>Estimate unknown parameters and latent effects via likelihood maximisation, EM, MCMC, or SGD</li>
<li>Quantify uncertainty with standard errors, posterior intervals, and credible sets</li>
<li>Control false discoveries across millions of tests with FDR/Q-value, permutation, and empirical-Bayes shrinkage.</li>
</ul></li>
</ol>

<img data-src="https://www.cs.auckland.ac.nz/~georgy/research/texture/thesis-html/img133.png" class="r-stretch"></section>
<section id="core-abstractions-for-quantitative-biology-2" class="slide level2">
<h2>Core abstractions for quantitative biology</h2>
<ol start="3" type="1">
<li><strong>Prediction</strong>
<ul>
<li>Use fitted models for out-of-sample trait prediction: BLUP/GBLUP, ridge/lasso/elastic-net, Bayesian whole-genome regressions, random forests, neural nets.</li>
<li>Evaluate accuracy (MSE, AUC), bias–variance trade-off, calibration, and portability across ancestries or cell types.</li>
<li>Translate genomic predictions into actionable scores for breeding, risk stratification, and drug-target prioritisation.</li>
</ul></li>
</ol>

<img data-src="https://www.cell.com/cms/10.1016/j.xgen.2024.100523/asset/29efca68-0928-4eed-94b9-9799e7cedda2/main.assets/gr5_lrg.jpg" class="r-stretch"></section>
<section id="core-abstractions-for-quantitative-biology-3" class="slide level2">
<h2>Core abstractions for quantitative biology</h2>
<ol start="4" type="1">
<li><strong>Interpretation &amp; Validation</strong>
<ul>
<li>Integrate functional annotations, eQTL, and single-cell data to refine biological mechanisms.</li>
<li>Perform replication, cross-cohort meta-analysis, and sensitivity analyses to population assumptions.</li>
<li>Communicate findings with clear visualisations and reproducible workflows (R/Bioconductor, Git, notebooks).</li>
</ul></li>
</ol>
<div class="quarto-layout-panel" data-layout="[1.5, -.2 ,1, -.2 ,2]">
<div class="quarto-layout-row quarto-layout-valign-center">
<div class="quarto-layout-cell" style="flex-basis: 30.6%;justify-content: center;">
<p><img data-src="https://upload.wikimedia.org/wikipedia/commons/thumb/d/db/Nature_journal_logo.svg/1200px-Nature_journal_logo.svg.png"></p>
</div>
<div class="quarto-figure-spacer quarto-layout-cell" style="flex-basis: 4.1%;justify-content: flex-start;">
<p>&nbsp;</p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 20.4%;justify-content: center;">
<p><img data-src="https://upload.wikimedia.org/wikipedia/commons/thumb/3/38/Jupyter_logo.svg/1200px-Jupyter_logo.svg.png"></p>
</div>
<div class="quarto-figure-spacer quarto-layout-cell" style="flex-basis: 4.1%;justify-content: flex-start;">
<p>&nbsp;</p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 40.8%;justify-content: center;">
<p><img data-src="https://www.bioconductor.org/images/logo/png/bioconductor_logo_rgb.png"></p>
</div>
</div>
</div>
</section>
<section id="early-work-in-the-field" class="slide level2">
<h2>Early work in the field</h2>
<div class="quarto-layout-panel" data-layout="[70, 30]">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 70.0%;justify-content: center;">
<p><img data-src="https://cdn.kastatic.org/ka-perseus-images/b966bcd0e55933e24836c4afe4a2a9889247579c.png"></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 30.0%;justify-content: center;">
<p><img data-src="https://cdn.kastatic.org/ka-perseus-images/97ef4b490d977c359786b040ead67b9cdaeac1aa.png"></p>
</div>
</div>
</div>
<aside class="notes">
<ul>
<li>Gregor Mendel (1822-1884) was an Austrian monk and scientist who is considered the father of modern genetics. He conducted experiments on pea plants and discovered the fundamental laws of inheritance, including the concepts of dominant and recessive traits, segregation, and independent assortment.</li>
<li>Mendel’s work laid the foundation for the field of genetics, but it was largely ignored during his lifetime. It wasn’t until the early 20th century that his findings were rediscovered and integrated into the broader understanding of biology.</li>
<li>It is hypothesized that Charles Darwin would have begun his work on evolution much earlier if he had access to Mendel’s work on inheritance.</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="genetics-statistics-and-eugenics-have-always-been-closely-linked" class="slide level2">
<h2>Genetics, statistics, and eugenics, have always been closely linked</h2>
<div class="quarto-layout-panel" data-layout="[35, 65]">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 35.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://upload.wikimedia.org/wikipedia/commons/thumb/a/a7/PSM_V60_D232_Standard_eugenics_scheme_of_descent.png/960px-PSM_V60_D232_Standard_eugenics_scheme_of_descent.png"></p>
<figcaption>Standard eugenics scheme of descent</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 65.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/galton.png"></p>
<figcaption>Francis Galton and his work, “Inquires into Human Faculty and its Development”</figcaption>
</figure>
</div>
</div>
</div>
</div>
<aside class="notes">
<ul>
<li>Francis Galton (1822-1911) was an English polymath who made significant contributions to various fields, including statistics, psychology, and anthropology. He is best known for his work on eugenics, a now-discredited movement that aimed to improve the genetic quality of the human population through selective breeding.</li>
<li>Galton was a cousin of Charles Darwin and was influenced by Darwin’s theory of evolution. He believed that intelligence and other traits were hereditary and could be measured and quantified. He developed statistical methods, such as correlation and regression, to study the inheritance of traits.</li>
<li>Galton’s work on eugenics has been widely criticized for its ethical implications and its association with discriminatory practices, such as forced sterilizations and racial segregation</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="the-ethics-of-genetics-and-genomics-iswas-of-great-importance" class="slide level2">
<h2>The ethics of genetics and genomics is/(was) of great importance!</h2>

<img data-src="images/nhgri_ethics.png" class="r-stretch"></section>
<section id="some-vocabulary" class="slide level2">
<h2>Some vocabulary</h2>
<ul>
<li><strong>Trait/Phenotype</strong>: A measurable characteristic of an organism, such as height, weight, or disease status.</li>
<li><strong>Gene</strong>: Unit of inheritance</li>
<li><strong>Genotype</strong>: The genetic constitution of an individual, often represented by specific alleles at particular loci.</li>
<li><strong>Allele</strong>: A variant form of a gene that can exist at a specific locus on a chromosome.</li>
<li><strong>Locus</strong>: A specific, fixed position on a chromosome where a particular gene or genetic marker is located.</li>
<li><strong>Diploid</strong>: Having two sets of chromosomes (one from each parent)</li>
<li><strong>Penetrance</strong>: Likelihood of expressing a phenotype given a genotype</li>
<li><strong>Polymorphism</strong>: The occurrence of two or more genetically determined forms in a population, such as single nucleotide polymorphisms (SNPs) or copy number variations (CNVs).</li>
<li><strong>Genetic Marker</strong>: A specific DNA sequence with a known location on a chromosome that can be used to identify individuals or species, often used in genetic mapping or association studies.</li>
</ul>
</section>
<section id="some-vocabulary-1" class="slide level2">
<h2>Some vocabulary</h2>
<ul>
<li><strong>Mendelian Disease</strong>: A disease caused by a mutation in a single gene, following Mendelian inheritance patterns (dominant, recessive, X-linked).</li>
<li><strong>Linkage Disequilibrium (LD)</strong>: The non-random association of alleles at different loci, indicating that certain allele combinations occur together more frequently than expected by chance.</li>
<li><strong>Heritability</strong>: The proportion of phenotypic variance in a trait that can be attributed to genetic variance, often estimated through twin or family studies.</li>
<li><strong>Genome-Wide Association Study (GWAS)</strong>: A study that looks for associations between genetic variants across the genome and specific traits or diseases in a large population.</li>
<li><strong>Polygenic Score (PGS)</strong>: A score that aggregates the effects of multiple genetic variants to predict an individual’s genetic predisposition to a trait or disease.</li>
<li><strong>Quantitative Trait Locus (QTL)</strong>: A region of the genome that is associated with a quantitative trait, often identified through linkage or association mapping.</li>
<li><strong>Epistasis</strong>: The interaction between genes where the effect of one gene is modified by one or more other genes, influencing the expression of a trait.</li>
</ul>
</section>
<section id="genetic-variant-example" class="slide level2">
<h2>Genetic Variant Example</h2>

<img data-src="images/paste-3.png" class="r-stretch quarto-figure-center"><p class="caption"><span class="citation" data-cites="laird2011">Laird and Lange (<a href="#/references" role="doc-biblioref" onclick="">2011</a>)</span></p></section>
<section id="probability-refresher" class="slide level2">
<h2>Probability Refresher</h2>
<h3 id="random-variables-and-distributions">Random Variables and Distributions</h3>
<ul>
<li class="fragment">A <strong>random variable</strong> is a numerical summary of randomness (e.g., the count of (A) alleles in a sample). We will denote random variables with uppercase letters (e.g., <span class="math inline">X</span>, <span class="math inline">Y</span>).</li>
<li class="fragment">We use a <strong>model</strong> (e.g., binomial/multinomial) to describe how data vary from sample to sample.
<ul>
<li class="fragment">This is otherwise known as a probability mass function (pmf) for discrete variables, or a probability density function (pdf) for continuous variables</li>
</ul></li>
<li class="fragment">Key ideas:
<ul>
<li class="fragment"><strong>Expectation (mean)</strong> and <strong>variance</strong>: long‑run average and spread.
<ul>
<li class="fragment"><span class="math inline">E[X] = \sum x \cdot P(X = x)</span> (discrete) or <span class="math inline">E[X] = \int x f(x) dx</span> (continuous)</li>
<li class="fragment"><span class="math inline">\textsf{Var}(X) = E[(X - E[X])^2] = E[X^2] - (E[X])^2</span></li>
</ul></li>
<li class="fragment"><strong>Standard error (SE)</strong>: typical sampling variability of an estimator.
<ul>
<li class="fragment"><span class="math inline">\textsf{SE} = \frac{\sigma}{\sqrt{n}}</span> (uncertainty in sample estimates)</li>
</ul></li>
<li class="fragment"><strong>Central Limit Theorem (CLT)</strong>: averages and proportions are roughly normal for large (n).
<ul>
<li class="fragment"><span class="math inline">\bar{X} \sim N\left(\mu, \frac{\sigma^2}{n}\right)</span> as <span class="math inline">n \to \infty</span></li>
</ul></li>
</ul></li>
</ul>
</section>
<section id="probability-distributions-vs-sampling-distributions" class="slide level2">
<h2>Probability Distributions vs Sampling Distributions</h2>
<ul>
<li>A <strong>probability distribution</strong> describes how a random variable behaves in the population.
<ul>
<li>Example: <span class="math inline">X \sim \textsf{Binomial}(n, p)</span></li>
</ul></li>
<li>A <strong>sampling distribution</strong> describes how a statistic (e.g., sample mean, sample proportion) behaves across repeated samples from the population.
<ul>
<li>Example: <span class="math inline">\hat{p} \sim N\left(p, \frac{(p(1-p))}{n}\right)</span> as <span class="math inline">n \to \infty</span></li>
</ul></li>
</ul>
</section>
<section id="probability-refresher-1" class="slide level2">
<h2>Probability Refresher</h2>
<h3 id="key-distributions-in-statistical-genetics">Key Distributions in Statistical Genetics</h3>
<div class="quarto-layout-panel" data-layout-ncol="2">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p><strong>Binomial Distribution</strong> <span class="math display">X \sim \textsf{Binomial}(n, p)</span> <span class="math display">P(X = k) = \binom{n}{k} p^k (1-p)^{n-k}</span></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p>Used for: Allele counts</p>
</div>
</div>
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p><strong>Normal Distribution</strong> <span class="math display">X \sim N(\mu, \sigma^2)</span> <span class="math display">f(x) = \frac{1}{\sqrt{2\pi\sigma^2}} e^{-\frac{(x-\mu)^2}{2\sigma^2}}</span></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p>Used for: Quantitative traits, effect sizes</p>
</div>
</div>
</div>
</section>
<section id="conditional-probability-and-independence" class="slide level2">
<h2>Conditional Probability and Independence</h2>
<ul>
<li><strong>Conditional Probability</strong>: <span class="math inline">P(A|B) = \frac{P(A \cap B)}{P(B)}</span></li>
<li><strong>Bayes’ Theorem</strong>: <span class="math inline">P(A|B) = \frac{P(B|A)P(A)}{P(B)}</span></li>
<li><strong>Independence</strong>: <span class="math inline">P(A \cap B) = P(A) \cdot P(B)</span> <span class="math inline">\iff</span> <span class="math inline">A</span> and <span class="math inline">B</span> are independent</li>
</ul>
<p>Conditional probability and Bayes’ Theorem are crucial for understanding rare disease genetics! For example, consider a rare disease with a prevalence of 1 in 10,000. If a genetic test has a sensitivity of 99% and a specificity of 99%, we can use Bayes’ Theorem to calculate the probability that an individual has the disease given a positive test result.</p>
<div class="fragment">
<p><span class="math display">\begin{align*}
P(Disease|Positive) &amp;= \frac{P(Positive|Disease)P(Disease)}{P(Positive)} \\
&amp;= \frac{(0.99)(0.0001)}{(0.99)(0.0001) + (0.01)(0.9999)} \\
&amp;\approx 0.0098
\end{align*}</span></p>
</div>
</section>
<section id="parameter-estimation" class="slide level2">
<h2>Parameter Estimation</h2>
<ul>
<li>Imagine a geneticist is studying <strong>allele ages</strong> (i.e., how many generations ago an allele arose via mutation) under a simplified model where new mutations arise randomly and uniformly over a fixed window — say, the last <span class="math inline">\theta</span> generations.</li>
<li>We could model <span class="math inline">X \sim \textsf{Uniform}(0, \theta)</span>.</li>
<li>We have our observed sample <span class="math inline">\{x_1,x_2, \ldots, x_n\}</span>, ages in generations of <span class="math inline">n</span> observed alleles.
<ul>
<li>For concreteness, let our sample be <span class="math inline">\{0, 1, 8, 4, 3, 7, 7, 6\}</span>.</li>
</ul></li>
<li>How should we estimate <span class="math inline">\theta</span> ?</li>
</ul>
<div class="fragment">
<div class="cell quarto-layout-panel" data-layout-ncol="2">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img data-src="lecture-01_files/figure-revealjs/unnamed-chunk-2-1.png" width="960"></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img data-src="lecture-01_files/figure-revealjs/unnamed-chunk-2-2.png" width="960"></p>
</div>
</div>
</div>
</div>
</section>
<section id="genetic-models" class="slide level2">
<h2>Genetic Models</h2>
<ul>
<li>Understanding random variables, sampling distributions, and bias/variance of estimators helps motivate biological questions, and understand how they can be answered</li>
<li>Let y be the expression of a trait, G be the additive contribution of genetic variables, and an environmental value E;</li>
<li>What assumptions do we make via the following models?</li>
</ul>
<div class="quarto-layout-panel" data-layout="[44, -12, 44]">
<div class="quarto-layout-row">
<div class="column quarto-layout-cell" style="flex-basis: 44.0%;justify-content: flex-start;">
<p><span class="math display">\begin{align}

y &amp;= G + E \\
y &amp;= \beta_0 + \beta_1 G + \beta_2 E + \epsilon \\
y &amp;= \beta_0 + \beta_1 G + \beta_2 E + \epsilon, \quad \epsilon \sim N(0, \sigma^2) \\
y &amp;= \beta_0 + \beta_1 G + \beta_2 E + \beta_3 \left(G \times E\right) + \epsilon, \quad \epsilon \sim N(0, \sigma^2)

\end{align}</span></p>
</div><div class="quarto-figure-spacer quarto-layout-cell" style="flex-basis: 12.0%;justify-content: flex-start;">
<p>&nbsp;</p>
</div>
<div class="column quarto-layout-cell" style="flex-basis: 44.0%;justify-content: center;">
<p><img data-src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fsj.tpj.6500085/MediaObjects/41397_2002_Article_BF6500085_Fig1_HTML.jpg"></p>
</div></div>
</div>
</section>
<section id="modeling" class="slide level2">
<h2>Modeling</h2>
<ul>
<li>Imagine that we have a sample size of <span class="math inline">n</span> unrelated haploid individuals from some population</li>
<li>We want to estimate allele frequencies for a biallelic SNP, say A/a</li>
<li>In our sample, we observe <span class="math inline">x</span> individuals with allele A and <span class="math inline">n - x</span> individuals with allele a</li>
<li>Let <span class="math inline">p</span> be the frequency of allele A and <span class="math inline">q = 1 - p</span> be the frequency of allele a</li>
</ul>
<p><span class="math display">Pr(X = x| n, p) = \binom{n}{x} p^x q^{n - x}</span></p>
<ul>
<li>Lets say we observe <span class="math inline">n = 27</span> and <span class="math inline">x = 11</span>. How do we estimate <span class="math inline">p</span>?</li>
</ul>
</section>
<section id="likelihood-functions" class="slide level2">
<h2>Likelihood Functions</h2>
<ul>
<li><p>The <strong>likelihood function</strong> is a function of the parameters of a statistical model, given specific observed data.</p></li>
<li><p>It represents the plausibility of different parameter values based on the observed data.</p></li>
<li><p>For a given set of data, the likelihood function is defined as: <span class="math display">L(\theta | data) = P(data | \theta)</span></p></li>
<li><p>In our case, the likelihood function for <span class="math inline">p</span> is: <span class="math display">L(p | x = 11, n = 27) = P(X = 11 | n = 27, p) = \binom{27}{11} p^{11} (1 - p)^{16}</span></p></li>
<li><p>We are asking, “how likely is it that we would observe this data for different values of <span class="math inline">p</span>?”</p></li>
</ul>
</section>
<section id="maximum-likelihood-estimator" class="slide level2">
<h2>Maximum Likelihood Estimator</h2>
<p>Our likelihood function for <span class="math inline">p</span> is</p>
<p><span class="math display">L(P|x = 11, n = 27) = {27 \choose 11} p^{11} q^{27-11}</span></p>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a></a>n <span class="ot">&lt;-</span> <span class="dv">27</span></span>
<span id="cb1-2"><a></a>x <span class="ot">&lt;-</span> <span class="dv">11</span></span>
<span id="cb1-3"><a></a>p_seq <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">length.out =</span> <span class="dv">200</span>)</span>
<span id="cb1-4"><a></a>likelihood <span class="ot">&lt;-</span> <span class="fu">dbinom</span>(x, n, p_seq)</span>
<span id="cb1-5"><a></a>plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb1-6"><a></a>    <span class="fu">data.frame</span>(<span class="at">p =</span> p_seq, <span class="at">likelihood =</span> likelihood),</span>
<span id="cb1-7"><a></a>    <span class="fu">aes</span>(<span class="at">x =</span> p, <span class="at">y =</span> likelihood)</span>
<span id="cb1-8"><a></a>) <span class="sc">+</span></span>
<span id="cb1-9"><a></a>    <span class="fu">geom_line</span>(<span class="at">linewidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb1-10"><a></a>    <span class="fu">labs</span>(</span>
<span id="cb1-11"><a></a>        <span class="at">x =</span> <span class="st">"Probability of Success (p)"</span>,</span>
<span id="cb1-12"><a></a>        <span class="at">y =</span> <span class="st">"Likelihood"</span>,</span>
<span id="cb1-13"><a></a>        <span class="at">title =</span> <span class="st">"Binomial Likelihood Function"</span></span>
<span id="cb1-14"><a></a>    ) <span class="sc">+</span></span>
<span id="cb1-15"><a></a>    <span class="fu">geom_vline</span>(</span>
<span id="cb1-16"><a></a>        <span class="at">xintercept =</span> x <span class="sc">/</span> n,</span>
<span id="cb1-17"><a></a>        <span class="at">color =</span> <span class="st">"red"</span>,</span>
<span id="cb1-18"><a></a>        <span class="at">linetype =</span> <span class="st">"dashed"</span></span>
<span id="cb1-19"><a></a>    ) <span class="sc">+</span></span>
<span id="cb1-20"><a></a>    <span class="fu">theme_light</span>(<span class="at">base_size =</span> <span class="dv">18</span>) <span class="sc">+</span></span>
<span id="cb1-21"><a></a>    <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>())</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure>
<p><img data-src="lecture-01_files/figure-revealjs/unnamed-chunk-3-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</div></div>
</section>
<section id="modeling-but-make-it-bayesian" class="slide level2">
<h2>Modeling, but make it bayesian</h2>
<ul>
<li>In Bayesian statistics, we treat parameters as random variables and use probability distributions to represent our uncertainty about them.</li>
<li>We start with a <strong>prior distribution</strong> that reflects our beliefs about the parameter before seeing the data.</li>
<li>We then use the observed data to update our beliefs and obtain a <strong>posterior distribution</strong>.</li>
<li>The posterior distribution combines the information from the prior and the likelihood of the observed data using Bayes’ theorem: <span class="math display">P(p | data) = \frac{P(data | p) P(p)}{P(data)}</span></li>
<li>Let’s say we have run a previous experiment and observed <span class="math inline">n = 20</span> individuals and <span class="math inline">x = 3</span> individuals with allele A (again, bilelic SNP).</li>
<li>We can use this information to inform our prior distribution for <span class="math inline">p</span></li>
</ul>
</section>
<section id="modeling-but-make-it-bayesian-1" class="slide level2">
<h2>Modeling, but make it bayesian</h2>
<div class="r-stack">
<div class="fragment fade-in-then-out">
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a></a>p <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">length.out =</span> <span class="dv">1000</span>)</span>
<span id="cb2-2"><a></a><span class="co"># Prior (Beta distribution from previous experiment)</span></span>
<span id="cb2-3"><a></a>prior <span class="ot">&lt;-</span> <span class="fu">dbeta</span>(p, <span class="dv">3</span> <span class="sc">+</span> <span class="dv">1</span>, <span class="dv">17</span> <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb2-4"><a></a><span class="co"># Likelihood (Beta distribution from current data)</span></span>
<span id="cb2-5"><a></a>likelihood <span class="ot">&lt;-</span> <span class="fu">dbeta</span>(p, <span class="dv">11</span> <span class="sc">+</span> <span class="dv">1</span>, <span class="dv">16</span> <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb2-6"><a></a><span class="co"># Posterior (Beta distribution combining both)</span></span>
<span id="cb2-7"><a></a>posterior <span class="ot">&lt;-</span> <span class="fu">dbeta</span>(p, <span class="dv">14</span> <span class="sc">+</span> <span class="dv">1</span>, <span class="dv">33</span> <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb2-8"><a></a><span class="co"># Combine into a data frame</span></span>
<span id="cb2-9"><a></a>plot_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb2-10"><a></a>    <span class="at">p =</span> <span class="fu">rep</span>(p, <span class="dv">3</span>),</span>
<span id="cb2-11"><a></a>    <span class="at">density =</span> <span class="fu">c</span>(prior, likelihood, posterior),</span>
<span id="cb2-12"><a></a>    <span class="at">type =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"Prior"</span>, <span class="st">"Likelihood"</span>, <span class="st">"Posterior"</span>), <span class="at">each =</span> <span class="fu">length</span>(p))</span>
<span id="cb2-13"><a></a>)</span>
<span id="cb2-14"><a></a><span class="co"># Add flat prior distribution</span></span>
<span id="cb2-15"><a></a>flat_prior <span class="ot">&lt;-</span> <span class="fu">dbeta</span>(p, <span class="dv">1</span>, <span class="dv">1</span>)</span>
<span id="cb2-16"><a></a>flat_posterior <span class="ot">&lt;-</span> <span class="fu">dbeta</span>(p, <span class="dv">12</span>, <span class="dv">17</span>)</span>
<span id="cb2-17"><a></a>plot_data <span class="ot">&lt;-</span> <span class="fu">rbind</span>(</span>
<span id="cb2-18"><a></a>    plot_data,</span>
<span id="cb2-19"><a></a>    <span class="fu">data.frame</span>(</span>
<span id="cb2-20"><a></a>        <span class="at">p =</span> <span class="fu">rep</span>(p, <span class="dv">3</span>),</span>
<span id="cb2-21"><a></a>        <span class="at">density =</span> <span class="fu">c</span>(flat_prior, likelihood, flat_posterior),</span>
<span id="cb2-22"><a></a>        <span class="at">type =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"Prior"</span>, <span class="st">"Likelihood"</span>, <span class="st">"Posterior"</span>), <span class="at">each =</span> <span class="fu">length</span>(p))</span>
<span id="cb2-23"><a></a>    )</span>
<span id="cb2-24"><a></a>)</span>
<span id="cb2-25"><a></a><span class="co"># Add a column to distinguish between informative and flat scenarios</span></span>
<span id="cb2-26"><a></a>plot_data<span class="sc">$</span>scenario <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"Informative"</span>, <span class="dv">3000</span>), <span class="fu">rep</span>(<span class="st">"Flat"</span>, <span class="dv">3000</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</div>
<div class="fragment fade-in-then-out">
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb3-2"><a></a>    plot_data <span class="sc">%&gt;%</span> <span class="fu">filter</span>(scenario <span class="sc">==</span> <span class="st">"Informative"</span>),</span>
<span id="cb3-3"><a></a>    <span class="fu">aes</span>(<span class="at">x =</span> p, <span class="at">y =</span> density, <span class="at">color =</span> type)</span>
<span id="cb3-4"><a></a>) <span class="sc">+</span></span>
<span id="cb3-5"><a></a>    <span class="fu">geom_line</span>(<span class="at">linewidth =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb3-6"><a></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">11</span> <span class="sc">/</span> <span class="dv">27</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb3-7"><a></a>    <span class="fu">labs</span>(</span>
<span id="cb3-8"><a></a>        <span class="at">x =</span> <span class="st">"Allele Frequency (p)"</span>,</span>
<span id="cb3-9"><a></a>        <span class="at">y =</span> <span class="st">"Density"</span>,</span>
<span id="cb3-10"><a></a>        <span class="at">title =</span> <span class="st">"Bayesian Update with Informative Prior"</span></span>
<span id="cb3-11"><a></a>    ) <span class="sc">+</span></span>
<span id="cb3-12"><a></a>    <span class="fu">scale_color_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb3-13"><a></a>    <span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">15</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb4-2"><a></a>  plot_data <span class="sc">%&gt;%</span> <span class="fu">filter</span>(scenario <span class="sc">==</span> <span class="st">"Flat"</span>),</span>
<span id="cb4-3"><a></a>  <span class="fu">aes</span>(<span class="at">x =</span> p, <span class="at">y =</span> density, <span class="at">color =</span> type, <span class="at">linetype =</span> type)</span>
<span id="cb4-4"><a></a>) <span class="sc">+</span></span>
<span id="cb4-5"><a></a>  <span class="fu">geom_line</span>(<span class="at">linewidth =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb4-6"><a></a>  <span class="fu">scale_linetype_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"Prior"</span> <span class="ot">=</span> <span class="st">"dotted"</span>, <span class="st">"Likelihood"</span> <span class="ot">=</span> <span class="st">"solid"</span>, <span class="st">"Posterior"</span> <span class="ot">=</span> <span class="st">"dashed"</span>)) <span class="sc">+</span></span>
<span id="cb4-7"><a></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">11</span> <span class="sc">/</span> <span class="dv">27</span>, <span class="at">linetype =</span> <span class="st">"dotdash"</span>, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb4-8"><a></a>  <span class="fu">labs</span>(</span>
<span id="cb4-9"><a></a>    <span class="at">x =</span> <span class="st">"Allele Frequency (p)"</span>,</span>
<span id="cb4-10"><a></a>    <span class="at">y =</span> <span class="st">"Density"</span>,</span>
<span id="cb4-11"><a></a>    <span class="at">title =</span> <span class="st">"Bayesian Update with Flat Prior"</span></span>
<span id="cb4-12"><a></a>  ) <span class="sc">+</span></span>
<span id="cb4-13"><a></a>  <span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">15</span>) <span class="sc">+</span></span>
<span id="cb4-14"><a></a>  <span class="fu">scale_color_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb4-15"><a></a>  <span class="fu">annotate</span>(<span class="st">"text"</span>,</span>
<span id="cb4-16"><a></a>    <span class="at">x =</span> <span class="fl">0.7</span>, <span class="at">y =</span> <span class="fu">max</span>(flat_posterior) <span class="sc">*</span> <span class="fl">0.8</span>,</span>
<span id="cb4-17"><a></a>    <span class="at">label =</span> <span class="st">"Likelihood ≈ Posterior"</span>, <span class="at">color =</span> <span class="st">"gray30"</span></span>
<span id="cb4-18"><a></a>  )</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</div></div>
</div>
<div class="fragment">
<div class="cell quarto-layout-panel" data-layout-ncol="2">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img data-src="lecture-01_files/figure-revealjs/unnamed-chunk-7-1.png" width="960"></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img data-src="lecture-01_files/figure-revealjs/unnamed-chunk-7-2.png" width="960"></p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="mendels-laws" class="slide level2">
<h2>Mendel’s Laws</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><strong>Mendel’s First Law (Segregation)</strong><br>
One allele of each parent is randomly transmitted, with probability 1/2, to the offspring; the alleles unite randomly to form the offspring’s genotype.</p>
<p><strong>Mendel’s Second Law (Independent Assortment)</strong><br>
At different loci, alleles are transmitted independently (when loci are unlinked).</p>
</div><div class="column" style="width:50%;">
<p><img data-src="https://upload.wikimedia.org/wikipedia/commons/8/8e/Intermediate_inheritance_P_-_F1_-_F2.png" style="width:100.0%" alt="Punnett-style diagram illustrating Mendelian segregation and independent assortment"></p>
</div></div>
<aside class="notes">
<ul>
<li>Mendel’s First Law, the Law of Segregation, states that during the formation of gametes (sperm and egg cells), the two alleles for a given gene separate, so that each gamete receives only one allele. This means that each parent contributes one allele to their offspring, and the combination of these alleles determines the offspring’s genotype.</li>
<li>Mendel’s Second Law, the Law of Independent Assortment, states that alleles for different genes are transmitted independently of one another, provided the genes are located on different chromosomes or are far apart on the same chromosome. This means that the inheritance of one trait does not influence the inheritance of another trait, allowing for a variety of genetic combinations in the offspring.</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="worked-example-segregation-31-ratio" class="slide level2">
<h2>Worked Example: Segregation (3:1 Ratio)</h2>
<h3 id="cross-aa-aa">Cross: Aa × Aa</h3>
<div class="quarto-layout-panel" data-layout-ncol="2">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p><strong>Punnett Square</strong></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<table class="caption-top">
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;"><strong>A</strong></th>
<th style="text-align: left;"><strong>a</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><strong>A</strong></td>
<td style="text-align: left;">AA</td>
<td style="text-align: left;">Aa</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>a</strong></td>
<td style="text-align: left;">Aa</td>
<td style="text-align: left;">aa</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<p><strong>Expected Ratios</strong></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<ul>
<li><strong>Genotypic</strong>: 1 AA : 2 Aa : 1 aa</li>
<li><strong>Phenotypic</strong>: 3 dominant : 1 recessive</li>
<li><strong>Probabilities</strong>: <span class="math inline">P(AA)=\frac14, P(Aa)=\frac12, P(aa)=\frac14</span></li>
</ul>
</div>
</div>
</div>
</section>
<section id="test-cross-example" class="slide level2">
<h2>Test Cross Example</h2>
<p>Proving law of segregation with a test cross</p>
<p>Example data:</p>
<ul>
<li>Cross: Aa × aa</li>
<li>Observed: 21 dominant, 28 recessive</li>
<li>Null hypothesis: Probability of transmission is 1:1.</li>
<li>Alternative hypothesis: Probability of transmission is not 1:1.</li>
<li>How to test this?</li>
</ul>
<p>Our question is related to observation of <strong>proportions</strong> –&gt; use a chi-squared goodness-of-fit test or a binomial test.</p>
</section>
<section id="test-cross-example-1" class="slide level2">
<h2>Test Cross Example</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a></a>observed_testcross <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">21</span>, <span class="dv">28</span>)     <span class="co"># Dominant, Recessive</span></span>
<span id="cb5-2"><a></a>chisq_test <span class="ot">&lt;-</span> <span class="fu">chisq.test</span>(observed_testcross, <span class="at">p =</span> <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">0.5</span>))</span>
<span id="cb5-3"><a></a>chisq_test</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Chi-squared test for given probabilities

data:  observed_testcross
X-squared = 1, df = 1, p-value = 0.3173</code></pre>
</div>
</div>
</section>
<section id="testing-via-simulation" class="slide level2">
<h2>Testing via Simulation</h2>
<p>We could also simulate our idea of the null hypothesis to get a sense of how extreme our observed data is.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb7-2"><a></a>n_sim <span class="ot">&lt;-</span> <span class="dv">10000</span></span>
<span id="cb7-3"><a></a>N <span class="ot">&lt;-</span> <span class="dv">49</span></span>
<span id="cb7-4"><a></a>simulated <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(n_sim, N, <span class="fl">0.5</span>)</span>
<span id="cb7-5"><a></a>p_value_sim <span class="ot">&lt;-</span> <span class="fu">mean</span>(simulated <span class="sc">&lt;=</span> <span class="dv">21</span> <span class="sc">|</span> simulated <span class="sc">&gt;=</span> <span class="dv">28</span>)</span>
<span id="cb7-6"><a></a>p_value_sim</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.3868</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a></a><span class="fu">hist</span>(simulated, <span class="at">breaks =</span> <span class="dv">30</span>, <span class="at">main =</span> <span class="st">"Simulated Test Cross Outcomes"</span>, <span class="at">xlab =</span> <span class="st">"Number of Dominant Offspring"</span>)</span>
<span id="cb9-2"><a></a><span class="fu">abline</span>(<span class="at">v =</span> <span class="fu">c</span>(<span class="dv">21</span>, <span class="dv">28</span>), <span class="at">col =</span> <span class="st">"red"</span>, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>

</div>
<img data-src="lecture-01_files/figure-revealjs/unnamed-chunk-9-1.png" width="960" class="r-stretch"></section>
<section id="on-simulations" class="slide level2">
<h2>On Simulations</h2>

<img data-src="images/paste-5.png" class="r-stretch quarto-figure-center"><p class="caption"><span class="citation" data-cites="downey2016">Downey (<a href="#/references" role="doc-biblioref" onclick="">2015</a>)</span></p></section>
<section id="estimation-of-allele-frequencies" class="slide level2">
<h2>Estimation of Allele Frequencies</h2>
<p><span class="math inline">n_{AA}</span> = Number of individuals with genotype AA <span class="math inline">n_{Aa}</span> = Number of individuals with genotype Aa <span class="math inline">n_{aa}</span> = Number of individuals with genotype aa</p>
<p>where, <span class="math inline">n = n_{AA} + n_{Aa} + n_{aa} = N</span>.</p>
<p>The sample proportions of allele <span class="math inline">A</span> is:</p>
<p><span class="math display">
\hat{p} = \frac{2n_{AA} + n_{Aa}}{2n}
</span></p>
<p>The usual standard error for a proportion is <span class="math inline">\sqrt{\hat{p}(1 - \hat{p})/2n}</span>, but this assumes independence of the <span class="math inline">2n</span> sampled chromosomes.</p>
</section>
<section id="hardyweinberg-equilibrium-hwe-assumptions" class="slide level2">
<h2>Hardy–Weinberg Equilibrium (HWE): Assumptions</h2>
<p>Let’s reframe the problem in terms of <strong>genotype frequencies</strong>. First, some assumptions:</p>
<ul>
<li><strong>Random mating</strong></li>
<li><strong>Large population</strong> (no strong drift in one generation)</li>
<li><strong>No selection</strong>, <strong>no migration</strong>, <strong>no mutation</strong> (in the generation under study)</li>
<li><strong>Accurate genotypes</strong> (no systematic genotyping error)</li>
</ul>
<p><strong>If these hold</strong>, and the allele frequency of <span class="math inline">A</span> is <span class="math inline">p</span> (so <span class="math inline">q = 1-p</span>), we say that the population is in <strong>Hardy-Weinberg Equilibrium (HWE)</strong>. The <strong>population genotype probabilities</strong> are:</p>
<p><span class="math display">
P(AA)=p^2 \quad P(Aa)=2pq \quad P(aa)=q^2
</span></p>
<p>As a consequence, the <strong>genotype frequencies</strong> in a random sample of <span class="math inline">n</span> individuals are approximately multinomially distributed with cell probabilities <span class="math inline">(p^2, 2pq, q^2)</span>. Thus, the sampling error of <span class="math inline">\hat{p}</span> is approximately <span class="math inline">\sqrt{p(1-p)/2n}</span>.</p>
<p>This confidence interval is known as a <strong>Wald confidence interval</strong>.</p>
</section>
<section id="testing-for-hwe" class="slide level2">
<h2>Testing for HWE</h2>
<ul>
<li>With large samples, we can test for HWE using a <strong>chi-squared goodness-of-fit test</strong>.</li>
</ul>

<img data-src="images/paste-4.png" class="r-stretch quarto-figure-center"><p class="caption"><span class="citation" data-cites="laird2011">Laird and Lange (<a href="#/references" role="doc-biblioref" onclick="">2011</a>)</span></p></section>
<section id="hwe-worked-example" class="slide level2">
<h2>HWE: Worked Example</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a></a>n <span class="ot">&lt;-</span> <span class="dv">212</span></span>
<span id="cb10-2"><a></a>n_AA <span class="ot">&lt;-</span> <span class="dv">175</span></span>
<span id="cb10-3"><a></a>n_Aa <span class="ot">&lt;-</span> <span class="dv">33</span></span>
<span id="cb10-4"><a></a>n_aa <span class="ot">&lt;-</span> <span class="dv">4</span></span>
<span id="cb10-5"><a></a></span>
<span id="cb10-6"><a></a><span class="co"># Estimate allele frequency</span></span>
<span id="cb10-7"><a></a>p_hat <span class="ot">&lt;-</span> (<span class="dv">2</span><span class="sc">*</span>n_AA <span class="sc">+</span> n_Aa) <span class="sc">/</span> (<span class="dv">2</span><span class="sc">*</span>n)</span>
<span id="cb10-8"><a></a>q_hat <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> p_hat</span>
<span id="cb10-9"><a></a>p_hat</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.9033019</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a></a><span class="co"># Expected counts</span></span>
<span id="cb12-2"><a></a>expected <span class="ot">&lt;-</span> <span class="fu">c</span>(n <span class="sc">*</span> p_hat<span class="sc">^</span><span class="dv">2</span>, <span class="dv">2</span><span class="sc">*</span>n <span class="sc">*</span> p_hat<span class="sc">*</span>q_hat, n <span class="sc">*</span> q_hat<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb12-3"><a></a>observed <span class="ot">&lt;-</span> <span class="fu">c</span>(n_AA, n_Aa, n_aa)</span>
<span id="cb12-4"><a></a>expected</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 172.982311  37.035377   1.982311</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a></a><span class="co"># Pearson chi-squared</span></span>
<span id="cb14-2"><a></a>chisq_val <span class="ot">&lt;-</span> <span class="fu">sum</span>((observed <span class="sc">-</span> expected)<span class="sc">^</span><span class="dv">2</span> <span class="sc">/</span> expected)</span>
<span id="cb14-3"><a></a>df <span class="ot">&lt;-</span> <span class="dv">3</span> <span class="sc">-</span> <span class="dv">1</span> <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb14-4"><a></a>p_val <span class="ot">&lt;-</span> <span class="fu">pchisq</span>(chisq_val, <span class="at">df =</span> df, <span class="at">lower.tail =</span> <span class="cn">FALSE</span>)</span>
<span id="cb14-5"><a></a>p_val</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1126299</code></pre>
</div>
</div>
</section>
<section id="hwe-in-practice" class="slide level2">
<h2>HWE in Practice</h2>
<ul>
<li>HWE is a useful null model for quality control of genotype data.</li>
<li>Deviations from HWE can indicate genotyping errors, population stratification, or selection</li>
</ul>
</section>
<section id="sampling-distributions-intuition-via-simulation" class="slide level2">
<h2>Sampling distributions: intuition via simulation</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb16-2"><a></a>true_p <span class="ot">&lt;-</span> <span class="fl">0.3</span></span>
<span id="cb16-3"><a></a>sample_sizes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">10</span>, <span class="dv">100</span>, <span class="dv">500</span>)</span>
<span id="cb16-4"><a></a>n_sims <span class="ot">&lt;-</span> <span class="dv">1000</span></span>
<span id="cb16-5"><a></a></span>
<span id="cb16-6"><a></a>sampling_results <span class="ot">&lt;-</span> <span class="fu">map_dfr</span>(sample_sizes, <span class="cf">function</span>(n) {</span>
<span id="cb16-7"><a></a>  p_hats <span class="ot">&lt;-</span> <span class="fu">replicate</span>(n_sims, {</span>
<span id="cb16-8"><a></a>    <span class="co"># Sample n individuals (2n alleles total)</span></span>
<span id="cb16-9"><a></a>    alleles <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(n, <span class="dv">2</span>, true_p)  <span class="co"># each individual contributes 0,1,2 A alleles</span></span>
<span id="cb16-10"><a></a>    <span class="fu">sum</span>(alleles) <span class="sc">/</span> (<span class="dv">2</span> <span class="sc">*</span> n)           <span class="co"># sample allele frequency</span></span>
<span id="cb16-11"><a></a>  })</span>
<span id="cb16-12"><a></a>  <span class="fu">tibble</span>(<span class="at">sample_size =</span> n, <span class="at">p_hat =</span> p_hats,</span>
<span id="cb16-13"><a></a>         <span class="at">theoretical_se =</span> <span class="fu">sqrt</span>(true_p <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> true_p) <span class="sc">/</span> (<span class="dv">2</span> <span class="sc">*</span> n)))</span>
<span id="cb16-14"><a></a>})</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="sampling-distributions-intuition-via-simulation-1" class="slide level2">
<h2>Sampling distributions: intuition via simulation</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a></a><span class="co"># Plot sampling distributions</span></span>
<span id="cb17-2"><a></a><span class="fu">ggplot</span>(sampling_results, <span class="fu">aes</span>(<span class="at">x =</span> p_hat)) <span class="sc">+</span></span>
<span id="cb17-3"><a></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">y =</span> <span class="fu">after_stat</span>(density)), <span class="at">bins =</span> <span class="dv">30</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb17-4"><a></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> true_p, <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">linewidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb17-5"><a></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>sample_size_name, <span class="at">scales =</span> <span class="st">"free_y"</span>) <span class="sc">+</span></span>
<span id="cb17-6"><a></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Sampling Distribution of p̂"</span>, <span class="at">x =</span> <span class="st">"Sample Allele Frequency (p̂)"</span>, <span class="at">y =</span> <span class="st">"Density"</span>) <span class="sc">+</span></span>
<span id="cb17-7"><a></a>  <span class="fu">theme_minimal</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>

</div>
<img data-src="lecture-01_files/figure-revealjs/unnamed-chunk-13-1.png" width="960" class="r-stretch"></section>
<section id="references" class="slide level2 smaller scrollable">
<h2>References</h2>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-downey2016" class="csl-entry" role="listitem">
Downey,A.B. (2015) <a href="https://greenteapress.com/thinkstats2/">Think stats: Exploratory data analysis in python</a> 2nd ed. O’Reilly Media, Sebastopol, CA.
</div>
<div id="ref-laird2011" class="csl-entry" role="listitem">
Laird,N.M. and Lange,C. (2011) <a href="https://doi.org/10.1007/978-1-4419-7338-2">The fundamentals of modern statistical genetics</a> Springer, New York, NY.
</div>
<div id="ref-sorensenStatisticalLearningGenetics2025" class="csl-entry" role="listitem">
Sorensen,D. (2025) <a href="https://doi.org/10.1007/978-3-031-86274-8">Statistical learning in genetics: An introduction using r</a> 2nd ed. Springer, Cham.
</div>
</div>
</section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">

</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"8\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
          const outerScaffold = trigger.parentElement.cloneNode(true);
          const codeEl = outerScaffold.querySelector('code');
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp('/' + window.location.host + '/');
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    

</body></html>