<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Lecture 03: Fitting Likelihoods, Linkage, and Association – Statistical Genetics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../course_logo.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-8991b0faf62225856095ff2fe360da1f.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN" && texText && texText.data) {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-sidebar docked quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Lecture 03: Fitting Likelihoods, Linkage, and Association</li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
      <a href="../index.html" class="sidebar-logo-link">
      <img src="../course_logo.png" alt="" class="sidebar-logo light-content py-0 d-lg-inline d-none">
      <img src="../course_logo.png" alt="" class="sidebar-logo dark-content py-0 d-lg-inline d-none">
      </a>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link"><i class="bi bi-calendar" role="img">
</i> 
 <span class="menu-text">Schedule</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../syllabus.html" class="sidebar-item-text sidebar-link"><i class="bi bi-book" role="img">
</i> 
 <span class="menu-text">Syllabus</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../faq.html" class="sidebar-item-text sidebar-link"><i class="bi bi-question" role="img">
</i> 
 <span class="menu-text">FAQ</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../links.html" class="sidebar-item-text sidebar-link"><i class="bi bi-link" role="img">
</i> 
 <span class="menu-text">Useful Links</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://github.com/gwcbi/StatGen" class="sidebar-item-text sidebar-link" target="_blank"><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text">GitHub</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://www.zotero.org/groups/6113424/pubh-8878/library" class="sidebar-item-text sidebar-link" target="_blank"><i class="bi bi-book" role="img">
</i> 
 <span class="menu-text">Zotero Library</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#review-likelihood-function-basics" id="toc-review-likelihood-function-basics" class="nav-link active" data-scroll-target="#review-likelihood-function-basics">Review: Likelihood Function Basics</a></li>
  <li><a href="#agenda" id="toc-agenda" class="nav-link" data-scroll-target="#agenda">Agenda</a></li>
  <li><a href="#newton-raphson-method-for-mle" id="toc-newton-raphson-method-for-mle" class="nav-link" data-scroll-target="#newton-raphson-method-for-mle">Newton-Raphson Method for MLE</a></li>
  <li><a href="#gradient-descent" id="toc-gradient-descent" class="nav-link" data-scroll-target="#gradient-descent">Gradient Descent</a></li>
  <li><a href="#the-em-algorithm" id="toc-the-em-algorithm" class="nav-link" data-scroll-target="#the-em-algorithm">The EM Algorithm</a></li>
  <li><a href="#em-for-gene-frequencies" id="toc-em-for-gene-frequencies" class="nav-link" data-scroll-target="#em-for-gene-frequencies">EM for Gene Frequencies</a></li>
  <li><a href="#em-for-gene-frequencies-1" id="toc-em-for-gene-frequencies-1" class="nav-link" data-scroll-target="#em-for-gene-frequencies-1">EM for Gene Frequencies</a></li>
  <li><a href="#em-for-gene-frequencies-2" id="toc-em-for-gene-frequencies-2" class="nav-link" data-scroll-target="#em-for-gene-frequencies-2">EM for Gene Frequencies</a></li>
  <li><a href="#em-for-gene-frequencies-3" id="toc-em-for-gene-frequencies-3" class="nav-link" data-scroll-target="#em-for-gene-frequencies-3">EM for Gene Frequencies</a></li>
  <li><a href="#em-for-gene-frequencies-4" id="toc-em-for-gene-frequencies-4" class="nav-link" data-scroll-target="#em-for-gene-frequencies-4">EM for Gene Frequencies</a></li>
  <li><a href="#em-starting-values-matter" id="toc-em-starting-values-matter" class="nav-link" data-scroll-target="#em-starting-values-matter">EM: Starting Values Matter</a></li>
  <li><a href="#em-starting-values-matter-1" id="toc-em-starting-values-matter-1" class="nav-link" data-scroll-target="#em-starting-values-matter-1">EM: Starting Values Matter</a></li>
  <li><a href="#em-parameter-trajectories" id="toc-em-parameter-trajectories" class="nav-link" data-scroll-target="#em-parameter-trajectories">EM Parameter Trajectories</a></li>
  <li><a href="#when-does-em-fail" id="toc-when-does-em-fail" class="nav-link" data-scroll-target="#when-does-em-fail">When does EM fail?</a></li>
  <li><a href="#linkage-vs-association-testing" id="toc-linkage-vs-association-testing" class="nav-link" data-scroll-target="#linkage-vs-association-testing">Linkage vs Association Testing</a></li>
  <li><a href="#two-point-linkage-terms" id="toc-two-point-linkage-terms" class="nav-link" data-scroll-target="#two-point-linkage-terms">Two-Point Linkage: Terms</a></li>
  <li><a href="#two-point-linkage-terms-1" id="toc-two-point-linkage-terms-1" class="nav-link" data-scroll-target="#two-point-linkage-terms-1">Two-Point Linkage: Terms</a></li>
  <li><a href="#two-point-linkage-repulsion-vs-coupling" id="toc-two-point-linkage-repulsion-vs-coupling" class="nav-link" data-scroll-target="#two-point-linkage-repulsion-vs-coupling">Two-Point Linkage: Repulsion vs Coupling</a></li>
  <li><a href="#two-point-linkage-direct-counting" id="toc-two-point-linkage-direct-counting" class="nav-link" data-scroll-target="#two-point-linkage-direct-counting">Two-Point Linkage: Direct Counting</a></li>
  <li><a href="#worked-example-direct-counting-lod" id="toc-worked-example-direct-counting-lod" class="nav-link" data-scroll-target="#worked-example-direct-counting-lod">Worked Example: Direct Counting LOD</a></li>
  <li><a href="#worked-example-direct-counting-lod-1" id="toc-worked-example-direct-counting-lod-1" class="nav-link" data-scroll-target="#worked-example-direct-counting-lod-1">Worked Example: Direct Counting LOD</a></li>
  <li><a href="#worked-example-direct-counting-lod-2" id="toc-worked-example-direct-counting-lod-2" class="nav-link" data-scroll-target="#worked-example-direct-counting-lod-2">Worked Example: Direct Counting LOD</a></li>
  <li><a href="#two-point-linkage-em-with-unknown-phase" id="toc-two-point-linkage-em-with-unknown-phase" class="nav-link" data-scroll-target="#two-point-linkage-em-with-unknown-phase">Two-Point Linkage: EM with Unknown Phase</a></li>
  <li><a href="#two-point-linkage-em-with-unknown-phase-1" id="toc-two-point-linkage-em-with-unknown-phase-1" class="nav-link" data-scroll-target="#two-point-linkage-em-with-unknown-phase-1">Two-Point Linkage: EM with Unknown Phase</a></li>
  <li><a href="#two-point-linkage-em-with-unknown-phase-2" id="toc-two-point-linkage-em-with-unknown-phase-2" class="nav-link" data-scroll-target="#two-point-linkage-em-with-unknown-phase-2">Two-Point Linkage: EM with Unknown Phase</a></li>
  <li><a href="#from-two-point-em-to-multipoint" id="toc-from-two-point-em-to-multipoint" class="nav-link" data-scroll-target="#from-two-point-em-to-multipoint">From Two-Point EM to Multipoint</a></li>
  <li><a href="#multipoint-linkage-hmm" id="toc-multipoint-linkage-hmm" class="nav-link" data-scroll-target="#multipoint-linkage-hmm">Multipoint Linkage (HMM)</a></li>
  <li><a href="#missing-data-em-in-linkage" id="toc-missing-data-em-in-linkage" class="nav-link" data-scroll-target="#missing-data-em-in-linkage">Missing-Data EM in Linkage</a></li>
  <li><a href="#practical-issues-in-linkage" id="toc-practical-issues-in-linkage" class="nav-link" data-scroll-target="#practical-issues-in-linkage">Practical Issues in Linkage</a></li>
  <li><a href="#linkage-disequilibrium" id="toc-linkage-disequilibrium" class="nav-link" data-scroll-target="#linkage-disequilibrium">Linkage Disequilibrium</a></li>
  <li><a href="#linkage-disequilibrium-1" id="toc-linkage-disequilibrium-1" class="nav-link" data-scroll-target="#linkage-disequilibrium-1">Linkage Disequilibrium</a></li>
  <li><a href="#linkage-disequilibrium-2" id="toc-linkage-disequilibrium-2" class="nav-link" data-scroll-target="#linkage-disequilibrium-2">Linkage Disequilibrium</a></li>
  <li><a href="#linkage-disequilibrium-3" id="toc-linkage-disequilibrium-3" class="nav-link" data-scroll-target="#linkage-disequilibrium-3">Linkage Disequilibrium</a></li>
  <li><a href="#goal-of-association-ld-link" id="toc-goal-of-association-ld-link" class="nav-link" data-scroll-target="#goal-of-association-ld-link">Goal of Association &amp; LD Link</a></li>
  <li><a href="#single-marker-association-core-test-linear" id="toc-single-marker-association-core-test-linear" class="nav-link" data-scroll-target="#single-marker-association-core-test-linear">Single-Marker Association: Core Test (Linear)</a></li>
  <li><a href="#basic-qc-pipeline-variant-level" id="toc-basic-qc-pipeline-variant-level" class="nav-link" data-scroll-target="#basic-qc-pipeline-variant-level">Basic QC Pipeline (Variant-level)</a></li>
  <li><a href="#basic-qc-pipeline-sample-level" id="toc-basic-qc-pipeline-sample-level" class="nav-link" data-scroll-target="#basic-qc-pipeline-sample-level">Basic QC Pipeline (Sample-level)</a></li>
  <li><a href="#summary-key-takeaways" id="toc-summary-key-takeaways" class="nav-link" data-scroll-target="#summary-key-takeaways">Summary &amp; Key Takeaways</a></li>
  <li><a href="#hands-on-lab-03-applied" id="toc-hands-on-lab-03-applied" class="nav-link" data-scroll-target="#hands-on-lab-03-applied">Hands-On: Lab 03 (Applied)</a></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="lecture-03.html"><i class="bi bi-file-slides"></i>RevealJS</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Lecture 03: Fitting Likelihoods, Linkage, and Association</h1>
<p class="subtitle lead">PUBH 8878, Statistical Genetics</p>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="review-likelihood-function-basics" class="level2">
<h2 class="anchored" data-anchor-id="review-likelihood-function-basics">Review: Likelihood Function Basics</h2>
<ul>
<li>Likelihood function: <span class="math inline">\(L(\boldsymbol{\theta} \mid \text{data}) = P(\text{data} \mid \boldsymbol{\theta})\)</span></li>
<li>Maximum Likelihood Estimation (MLE)</li>
<li>Simple one-parameter examples</li>
</ul>
<p>However often direct MLE is not possible because</p>
<ul>
<li>Closed-form solution does not exist</li>
<li>Missing or latent data</li>
<li>High-dimensional or constrained parameter space leads to non-convex surface with multiple local maxima</li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“Let’s start by aligning on what <em>likelihood</em> means. We take the sampling model <span class="math inline">\(p(\text{data}\mid\theta)\)</span> and, after we’ve observed the data, we view it as a function of the parameter <span class="math inline">\(\theta\)</span>. That function is the likelihood <span class="math inline">\(L(\theta)\)</span>. The MLE is simply the value of <span class="math inline">\(\theta\)</span> that maximizes <span class="math inline">\(L\)</span>.</p>
<p>Two reminders that will be useful all semester: first, the <strong>invariance</strong> property—if <span class="math inline">\(\hat\theta\)</span> maximizes <span class="math inline">\(L(\theta)\)</span>, then <span class="math inline">\(g(\hat\theta)\)</span> maximizes the likelihood of <span class="math inline">\(g(\theta)\)</span>. Second, the <strong>curvature</strong> of the log-likelihood, via Fisher information, anticipates precision.</p>
<p>In practice, closed forms fail us for three common reasons: there’s no algebraic solution; there’s <strong>missing or latent data</strong>; or the parameter space is constrained or high-dimensional so the surface is bumpy. Today we’ll build the numerical toolkit that handles these realities.”</p>
</div>
</section>
<section id="agenda" class="level2">
<h2 class="anchored" data-anchor-id="agenda">Agenda</h2>
<ul>
<li>Optimization for MLE: Newton–Raphson, Gradient Descent, SGD</li>
<li>EM algorithm: gene frequencies and incomplete data</li>
<li>Linkage analysis: two-point, unknown phase EM, multipoint HMM</li>
<li>Linkage disequilibrium (LD) and its measures (D, D’, r)</li>
<li>Association: single-marker/haplotype tests and basic QC</li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“Here’s the plan. We’ll warm up with Newton–Raphson and first‑order methods—how to actually <em>move</em> on a likelihood surface. We’ll then introduce EM as a principled way to deal with missing or latent pieces, anchored by an ABO example.</p>
<p>Next, we pivot to <strong>linkage</strong>: first two‑point LOD, then an EM formulation for <strong>unknown phase</strong>, and finally the multipoint <strong>HMM</strong> perspective. We’ll connect that to <strong>LD</strong>, clarify <span class="math inline">\(D\)</span>, <span class="math inline">\(D'\)</span> and <span class="math inline">\(r\)</span>, and finish with a single‑marker association demo and the <strong>QC</strong> that keeps results honest.”</p>
</div>
</section>
<section id="newton-raphson-method-for-mle" class="level2">
<h2 class="anchored" data-anchor-id="newton-raphson-method-for-mle">Newton-Raphson Method for MLE</h2>
<ul>
<li>General procedure to find <span class="math inline">\(x\)</span> such that <span class="math inline">\(g(x) = 0\)</span></li>
<li>Uses Taylor Expansion, for <span class="math inline">\(g(x)\)</span> about a point <span class="math inline">\(x = a\)</span>:</li>
</ul>
<p><span class="math display">\[\begin{align*}
g(x) = g(a) &amp;+
\frac{g'(a)(x-a)}{1!} + \frac{g''(a)(x-a)^2}{2!} \\
&amp;+ ... + \frac{g^{(n)}(a)(x-a)^n}{n!}
\end{align*}\]</span></p>
<div class="notes">
<p><strong>Script:</strong><br>
“In MLE we set <span class="math inline">\(g(\theta)=\ell'(\theta)\)</span>, the <strong>score</strong>, and look for a root. Newton’s update is<br>
\[ \theta<sup>{(t+1)}=</sup>{(t)}-.\]<br>
In multiple dimensions it’s <span class="math inline">\(\boldsymbol{\theta}^{(t+1)}=\boldsymbol{\theta}^{(t)}-\mathbf{H}^{-1}\mathbf{g}\)</span>, where <span class="math inline">\(\mathbf{g}\)</span> is the gradient and <span class="math inline">\(\mathbf{H}\)</span> the Hessian.</p>
<p>Three practical notes. First, Newton is <strong>quadratically convergent</strong> when you’re close to the solution—that’s why it’s fast. Second, guard against wild steps with <strong>damping or line search</strong>, and keep feasibility in mind if your parameter must lie on a simplex or in <span class="math inline">\([0,0.5]\)</span>. Third, if the Hessian misbehaves, use <strong>Fisher scoring</strong> (replace <span class="math inline">\(-\mathbf{H}\)</span> with expected information) or a <strong>trust region</strong>.”</p>
</div>
</section>
<section id="gradient-descent" class="level2">
<h2 class="anchored" data-anchor-id="gradient-descent">Gradient Descent</h2>
<ul>
<li>First-order iterative optimization algorithm</li>
<li>Does not rely on matrix inversions <span class="math inline">\(\longrightarrow\)</span> better for high dimensional problems</li>
<li><a href="https://uclaacm.github.io/gradient-descent-visualiser/#playground">Visualizer from UCLA</a></li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“First‑order methods move along the <strong>steepest descent</strong> direction without ever inverting a matrix. That’s a superpower in high‑dimensional genetics models where Hessians are huge. The tradeoff: convergence is typically <strong>linear</strong>, not quadratic.</p>
<p>What matters in practice is the <strong>step size</strong>: constant steps are brittle; backtracking or scheduled decay is safer. Stochastic and mini‑batch variants are essential when your objective decomposes over people, markers, or families. The mental model is simple: noisy but cheap steps that converge ‘on average’.”</p>
</div>
</section>
<section id="the-em-algorithm" class="level2">
<h2 class="anchored" data-anchor-id="the-em-algorithm">The EM Algorithm</h2>
<ul>
<li>Expectation-Maximization Algorithm</li>
<li>Key idea: transform an <strong>incomplete</strong> data problem into a <strong>complete</strong> data problem</li>
</ul>
<p><span class="math display">\[
p\left(y \mid \theta \right) = \frac{p\left(y, z \mid \theta \right)}{p\left(z \mid y, \theta\right)}
\]</span></p>
<div class="notes">
<p><strong>Script:</strong><br>
“EM turns missingness into bookkeeping. We invent latent variables <span class="math inline">\(z\)</span> so that the <strong>complete‑data</strong> log-likelihood <span class="math inline">\(\log p(y,z\mid\theta)\)</span> is easy to maximize. At iterate <span class="math inline">\(t\)</span>:</p>
<ul>
<li><strong>E‑step:</strong> compute <span class="math inline">\(Q(\theta\mid\theta^{(t)})=\mathbb{E}_{z\mid y,\theta^{(t)}}[\log p(y,z\mid\theta)]\)</span>.<br>
</li>
<li><strong>M‑step:</strong> update <span class="math inline">\(\theta^{(t+1)}=\arg\max_\theta Q(\theta\mid\theta^{(t)})\)</span>.</li>
</ul>
<p>Why does it work? By Jensen’s inequality, EM <strong>monotonically increases</strong> the observed log-likelihood. The feel is ‘stable but sometimes slow’; we’ll flag accelerations like ECME and PX‑EM later.”</p>
</div>
</section>
<section id="em-for-gene-frequencies" class="level2">
<h2 class="anchored" data-anchor-id="em-for-gene-frequencies">EM for Gene Frequencies</h2>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;">Genotype</th>
<th style="text-align: left;">Phenotype</th>
<th style="text-align: left;">Observed Counts</th>
<th style="text-align: left;">Genotype Frequency</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><span class="math inline">\(AA\)</span></td>
<td style="text-align: left;"><span class="math inline">\(A\)</span></td>
<td style="text-align: left;"><span class="math inline">\(n_A\)</span></td>
<td style="text-align: left;"><span class="math inline">\(p_A^2\)</span></td>
</tr>
<tr class="even">
<td style="text-align: left;"><span class="math inline">\(AO\)</span></td>
<td style="text-align: left;"><span class="math inline">\(A\)</span></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"><span class="math inline">\(2 p_A p_O\)</span></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><span class="math inline">\(AB\)</span></td>
<td style="text-align: left;"><span class="math inline">\(AB\)</span></td>
<td style="text-align: left;"><span class="math inline">\(n_{AB}\)</span></td>
<td style="text-align: left;"><span class="math inline">\(2 p_A p_B\)</span></td>
</tr>
<tr class="even">
<td style="text-align: left;"><span class="math inline">\(BB\)</span></td>
<td style="text-align: left;"><span class="math inline">\(B\)</span></td>
<td style="text-align: left;"><span class="math inline">\(n_B\)</span></td>
<td style="text-align: left;"><span class="math inline">\(p_B^2\)</span></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><span class="math inline">\(BO\)</span></td>
<td style="text-align: left;"><span class="math inline">\(B\)</span></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"><span class="math inline">\(2 p_B p_O\)</span></td>
</tr>
<tr class="even">
<td style="text-align: left;"><span class="math inline">\(OO\)</span></td>
<td style="text-align: left;"><span class="math inline">\(O\)</span></td>
<td style="text-align: left;"><span class="math inline">\(n_O\)</span></td>
<td style="text-align: left;"><span class="math inline">\(p_O^2\)</span></td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="notes">
<p><strong>Script:</strong><br>
“Here’s the ABO mapping. The assumption is Hardy–Weinberg and random mating, so genotype frequencies are the usual <span class="math inline">\(p^2\)</span>, <span class="math inline">\(2pq\)</span>, <span class="math inline">\(q^2\)</span>. The ambiguity sits exactly in the <strong>A</strong> and <strong>B</strong> phenotypes: each collapses over two genotypes. EM will split those cells into expected counts using the current allele frequencies and then re‑estimate the frequencies as if those expected counts were observed.”</p>
</div>
</section>
<section id="em-for-gene-frequencies-1" class="level2">
<h2 class="anchored" data-anchor-id="em-for-gene-frequencies-1">EM for Gene Frequencies</h2>
<ul>
<li>We don’t know if <span class="math inline">\(n_A\)</span> is from <span class="math inline">\(AA\)</span> or <span class="math inline">\(AO\)</span>! (And similarly for <span class="math inline">\(n_B\)</span>)</li>
<li>However, given our data model, we can find values of <span class="math inline">\((\hat{p}_A, \hat{p}_B)\)</span> via EM</li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“Think of each A phenotype as a <strong>soft label</strong> between AA and AO, with probabilities determined by the current <span class="math inline">\((p_A,p_O)\)</span>. Same story for B. That’s the essence of the E‑step.”</p>
</div>
</section>
<section id="em-for-gene-frequencies-2" class="level2">
<h2 class="anchored" data-anchor-id="em-for-gene-frequencies-2">EM for Gene Frequencies</h2>
<ul>
<li><p>Start with initial <span class="math inline">\((p_A^{(0)}, p_B^{(0)})\)</span> and set <span class="math inline">\(p_O^{(0)} = 1 - p_A^{(0)} - p_B^{(0)}\)</span>.</p></li>
<li><p>E-step: allocate ambiguous phenotype counts:</p>
<ul>
<li><span class="math inline">\(A\)</span> phenotype: <span class="math inline">\(\tilde n_{AA} = n_A \frac{(p_A^{(t)})^2}{(p_A^{(t)})^2 + 2 p_A^{(t)} p_O^{(t)}}\)</span>, <span class="math inline">\(\tilde n_{AO} = n_A \frac{2 p_A^{(t)} p_O^{(t)}}{(p_A^{(t)})^2 + 2 p_A^{(t)} p_O^{(t)}}\)</span>.</li>
<li><span class="math inline">\(B\)</span> phenotype: <span class="math inline">\(\tilde n_{BB} = n_B \frac{(p_B^{(t)})^2}{(p_B^{(t)})^2 + 2 p_B^{(t)} p_O^{(t)}}\)</span>, <span class="math inline">\(\tilde n_{BO} = n_B \frac{2 p_B^{(t)} p_O^{(t)}}{(p_B^{(t)})^2 + 2 p_B^{(t)} p_O^{(t)}}\)</span>.</li>
</ul></li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“These are just conditional probabilities. Given the person has phenotype A, what’s the chance they’re AA versus AO under the current <span class="math inline">\((p_A,p_O)\)</span>? Multiply by <span class="math inline">\(n_A\)</span> to get <strong>expected</strong> genotype counts <span class="math inline">\(\tilde n_{AA}\)</span> and <span class="math inline">\(\tilde n_{AO}\)</span>. Do the same for B. That’s the E‑step.”</p>
</div>
</section>
<section id="em-for-gene-frequencies-3" class="level2">
<h2 class="anchored" data-anchor-id="em-for-gene-frequencies-3">EM for Gene Frequencies</h2>
<ul>
<li>M-step (with <span class="math inline">\(N\)</span> total individuals):</li>
</ul>
<span class="math display">\[\begin{aligned}
p_A^{(t+1)} &amp;= \frac{2\tilde n_{AA} + \tilde n_{AO} + n_{AB}}{2N},\\
p_B^{(t+1)} &amp;= \frac{2\tilde n_{BB} + \tilde n_{BO} + n_{AB}}{2N},\\
p_O^{(t+1)} &amp;= 1 - p_A^{(t+1)} - p_B^{(t+1)}
\end{aligned}\]</span>
<ul>
<li>Iterate until parameter changes are below a tolerance.</li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“Now treat the expected genotype table as complete and <strong>recount alleles</strong>: two from each homozygote and one from each heterozygote, divided by <span class="math inline">\(2N\)</span>. That gives updated <span class="math inline">\((p_A,p_B,p_O)\)</span>. Iterate E and M until the changes are tiny and the observed log‑likelihood is no longer increasing.”</p>
</div>
</section>
<section id="em-for-gene-frequencies-4" class="level2">
<h2 class="anchored" data-anchor-id="em-for-gene-frequencies-4">EM for Gene Frequencies</h2>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;">Genotype</th>
<th style="text-align: left;">Phenotype</th>
<th style="text-align: left;">Observed Counts</th>
<th style="text-align: left;">Genotype Frequency</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><span class="math inline">\(AA\)</span></td>
<td style="text-align: left;"><span class="math inline">\(A\)</span></td>
<td style="text-align: left;"><span class="math inline">\(725\)</span></td>
<td style="text-align: left;"><span class="math inline">\(p_A^2\)</span></td>
</tr>
<tr class="even">
<td style="text-align: left;"><span class="math inline">\(AO\)</span></td>
<td style="text-align: left;"><span class="math inline">\(A\)</span></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"><span class="math inline">\(2 p_A p_O\)</span></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><span class="math inline">\(AB\)</span></td>
<td style="text-align: left;"><span class="math inline">\(AB\)</span></td>
<td style="text-align: left;"><span class="math inline">\(72\)</span></td>
<td style="text-align: left;"><span class="math inline">\(2 p_A p_B\)</span></td>
</tr>
<tr class="even">
<td style="text-align: left;"><span class="math inline">\(BB\)</span></td>
<td style="text-align: left;"><span class="math inline">\(B\)</span></td>
<td style="text-align: left;"><span class="math inline">\(258\)</span></td>
<td style="text-align: left;"><span class="math inline">\(p_B^2\)</span></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><span class="math inline">\(BO\)</span></td>
<td style="text-align: left;"><span class="math inline">\(B\)</span></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"><span class="math inline">\(2 p_B p_O\)</span></td>
</tr>
<tr class="even">
<td style="text-align: left;"><span class="math inline">\(OO\)</span></td>
<td style="text-align: left;"><span class="math inline">\(O\)</span></td>
<td style="text-align: left;"><span class="math inline">\(1073\)</span></td>
<td style="text-align: left;"><span class="math inline">\(p_O^2\)</span></td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="notes">
<p><strong>Script:</strong><br>
“With these sample counts—A = 725, AB = 72, B = 258, O = 1073—we can run EM. I’ll show convergence behavior next, and I want you to notice two things: the <strong>monotone</strong> increase in log‑likelihood and the impact of <strong>starting values</strong>.”</p>
</div>
</section>
<section id="em-starting-values-matter" class="level2">
<h2 class="anchored" data-anchor-id="em-starting-values-matter">EM: Starting Values Matter</h2>
<ul>
<li>Different initial <span class="math inline">\((p_A, p_B, p_O)\)</span> can change speed (not the final MLE)</li>
<li>Strategies:
<ul>
<li>Equal: <span class="math inline">\((1/3, 1/3, 1/3)\)</span></li>
<li>Unbalanced guess: <span class="math inline">\((0.01, 0.98, 0.01)\)</span></li>
<li>Smart (use <span class="math inline">\(p_O^{(0)} = \sqrt{n_O/N}\)</span>, solve for <span class="math inline">\((p_A^{(0)}, p_B^{(0)})\)</span> from <span class="math inline">\(2p_A p_B\)</span>)</li>
</ul></li>
<li>Tolerance of <span class="math inline">\(1 \times 10^{-6}\)</span></li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“The ‘smart start’ uses two near‑identities: <span class="math inline">\(n_O \approx N p_O^2\)</span> gives <span class="math inline">\(p_O^{(0)}\approx\sqrt{n_O/N}\)</span>; then <span class="math inline">\(n_{AB}\approx 2N p_A p_B\)</span> plus <span class="math inline">\(p_A+p_B=1-p_O^{(0)}\)</span> gives a quadratic for <span class="math inline">\((p_A^{(0)},p_B^{(0)})\)</span>. This gets you close to the basin of attraction and trims iterations substantially. Equal and wild unbalanced starts are good stress tests.”</p>
</div>
</section>
<section id="em-starting-values-matter-1" class="level2">
<h2 class="anchored" data-anchor-id="em-starting-values-matter-1">EM: Starting Values Matter</h2>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;">Strategy</th>
<th style="text-align: right;">Iterations</th>
<th style="text-align: right;"><span class="math inline">\(\hat{p}_A\)</span></th>
<th style="text-align: right;"><span class="math inline">\(\hat{p}_B\)</span></th>
<th style="text-align: right;"><span class="math inline">\(\hat{p}_O\)</span></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Equal</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">0.209131</td>
<td style="text-align: right;">0.080801</td>
<td style="text-align: right;">0.710068</td>
</tr>
<tr class="even">
<td style="text-align: left;">Smart</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0.209131</td>
<td style="text-align: right;">0.080801</td>
<td style="text-align: right;">0.710068</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Unbalanced</td>
<td style="text-align: right;">8</td>
<td style="text-align: right;">0.209131</td>
<td style="text-align: right;">0.080801</td>
<td style="text-align: right;">0.710068</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="notes">
<p><strong>Script:</strong><br>
“Read this table left to right. Each row is a starting strategy; the last three columns are the converged allele frequency estimates. The point is not that the answers differ—they don’t—but that the <strong>iterations</strong> column can vary a lot. You can buy free speed by initializing intelligently.”</p>
</div>
</section>
<section id="em-parameter-trajectories" class="level2">
<h2 class="anchored" data-anchor-id="em-parameter-trajectories">EM Parameter Trajectories</h2>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lecture-03_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="notes">
<p><strong>Script:</strong><br>
“Here are the trajectories for <span class="math inline">\(p_A\)</span>. Different starts head toward the same fixed point, and the paths are smooth. If you ever see oscillation or jaggedness here, it’s a warning sign about numerical underflow or a bad update rule.”</p>
</div>
</section>
<section id="when-does-em-fail" class="level2">
<h2 class="anchored" data-anchor-id="when-does-em-fail">When does EM fail?</h2>
<ul>
<li><strong>Non-identifiability / flat likelihood</strong>: Ridges in segregation or linkage models; EM wanders or stalls.</li>
<li><strong>Local maxima &amp; starts</strong>: Multiple modes (e.g., mixture of penetrance classes); poor initialization traps EM.</li>
<li><strong>Boundary degeneracy</strong>: Rare allele/component weight driven to 0; variance or frequency estimates collapse.</li>
<li><strong>Model misspecification</strong>: Violated assumptions (e.g., Hardy-Weinberg, stratification) give misleading “convergence.”</li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“EM is not magic. If your model is <strong>not identifiable</strong>, EM will happily march along a ridge forever. If the likelihood is <strong>multimodal</strong>, bad starts can trap you. Near <strong>boundaries</strong>—think vanishing allele frequency—updates can collapse and stick. And if assumptions like HWE are wrong, EM will converge…to the wrong place. The antidotes are multiple starts, likelihood profiling, and stress‑testing the assumptions.”</p>
</div>
</section>
<section id="linkage-vs-association-testing" class="level2">
<h2 class="anchored" data-anchor-id="linkage-vs-association-testing">Linkage vs Association Testing</h2>
<ul>
<li>Linkage: Tracks co-segregation of markers and traits within families to map disease genes.
<ul>
<li>Null hypothesis: no linkage (independent assortment), <span class="math inline">\(\theta = 0.5\)</span>.</li>
</ul></li>
<li>Association: Tests for correlation between variants and traits in populations to pinpoint causal loci.
<ul>
<li>Null hypothesis: no association (e.g., <span class="math inline">\(\beta=0\)</span> or OR=1).</li>
</ul></li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“Two mapping mindsets. <strong>Linkage</strong> works within families, asks whether a trait and a marker co‑segregate more than chance, and summarizes evidence on a <strong>centimorgan</strong> scale with LOD curves. It’s robust to population structure. <strong>Association</strong> works across unrelateds, tests regression coefficients or odds ratios, and resolves at <strong>kilobase</strong> scale, but is sensitive to confounding. We need both lenses.”</p>
</div>
</section>
<section id="two-point-linkage-terms" class="level2">
<h2 class="anchored" data-anchor-id="two-point-linkage-terms">Two-Point Linkage: Terms</h2>
<ul>
<li>Independent assortment (recall): when loci are unlinked, transmissions are independent and the chance a crossover separates them is <span class="math inline">\(\theta=0.5\)</span> (Mendel’s Second Law).</li>
<li>Informative transmission/meiosis: a parent <span class="math inline">\(\to\)</span> child transmission where the transmitting parent is heterozygous at the loci of interest and the genotypes allow us to tell whether a crossover occurred.
<ul>
<li>Example: parent <span class="math inline">\(Aa/Bb\)</span> transmits <span class="math inline">\(AB\)</span> or <span class="math inline">\(ab\)</span> (nonrecombinant) vs <span class="math inline">\(Ab\)</span> or <span class="math inline">\(aB\)</span> (recombinant).</li>
</ul></li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“An <strong>informative</strong> meiosis is one where you can tell whether a crossover happened. If the transmitting parent is not heterozygous at both loci, you can’t tell—so that meiosis doesn’t contribute information about <span class="math inline">\(\theta\)</span>.”</p>
</div>
</section>
<section id="two-point-linkage-terms-1" class="level2">
<h2 class="anchored" data-anchor-id="two-point-linkage-terms-1">Two-Point Linkage: Terms</h2>
<ul>
<li>Recombinant fraction <span class="math inline">\(\theta\)</span>: the probability that a crossover occurs between two loci in a single meiosis; <span class="math inline">\(0 \le \theta \le 0.5\)</span>.</li>
<li>Nonrecombinant transmission: the child receives an allele combination that matches one of the transmitting parent’s original allele pairs (no crossover between the loci).</li>
<li>Recombinant transmission: the child receives a new combination relative to the transmitting parent’s original pair (a crossover occurred between the loci).</li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“Operationally, <strong>nonrecombinants</strong> copy one of the parental haplotypes; <strong>recombinants</strong> mix them. Our complete‑data likelihoods will be binomial in the counts of these two outcomes.”</p>
</div>
</section>
<section id="two-point-linkage-repulsion-vs-coupling" class="level2">
<h2 class="anchored" data-anchor-id="two-point-linkage-repulsion-vs-coupling">Two-Point Linkage: Repulsion vs Coupling</h2>
<p><img src="images/paste-8.png" class="img-fluid"></p>
<div class="notes">
<p><strong>Script:</strong><br>
“Phase matters. In <strong>coupling</strong>, <span class="math inline">\(AB/ab\)</span>, nonrecombinants are <span class="math inline">\(\{AB,ab\}\)</span> and recombinants are <span class="math inline">\(\{Ab,aB\}\)</span>. In <strong>repulsion</strong>, <span class="math inline">\(Ab/aB\)</span>, those roles flip. When phase is unknown, we’ll treat it as <strong>latent</strong> and estimate a posterior weight on the two configurations.”</p>
</div>
</section>
<section id="two-point-linkage-direct-counting" class="level2">
<h2 class="anchored" data-anchor-id="two-point-linkage-direct-counting">Two-Point Linkage: Direct Counting</h2>
<ul>
<li><p>Count <span class="math inline">\(R\)</span> recombinants and <span class="math inline">\(NR\)</span> nonrecombinants; total informative <span class="math inline">\(I=R+NR\)</span>.</p></li>
<li><p>Point estimate: <span class="math inline">\(\hat{\theta}=R/I\)</span>.</p></li>
<li><p>LOD from counts (vs.&nbsp;independence at <span class="math inline">\(\theta=0.5\)</span>):</p></li>
</ul>
<p><span class="math display">\[
\mathrm{LOD}(\theta) = \log_{10}\!\left\{ \frac{\theta^{R} (1-\theta)^{NR}}{0.5^{I}} \right\}
\]</span></p>
<ul>
<li>This is a likelihood ratio test!</li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“In the simplest case, count recombinants <span class="math inline">\(R\)</span> and informatives <span class="math inline">\(I\)</span>, and compute <span class="math inline">\(\hat\theta=R/I\)</span>. The LOD is a base‑10 log likelihood ratio against <span class="math inline">\(\theta=0.5\)</span>. The curve peaks at <span class="math inline">\(\hat\theta\)</span> and tells you where the evidence lies.”</p>
</div>
</section>
<section id="worked-example-direct-counting-lod" class="level2">
<h2 class="anchored" data-anchor-id="worked-example-direct-counting-lod">Worked Example: Direct Counting LOD</h2>
<ul>
<li>Suppose <span class="math inline">\(I=40\)</span> informative meioses with <span class="math inline">\(R=12\)</span> recombinants.</li>
<li>MLE <span class="math inline">\(\hat{\theta}=R/I=12/40=0.3\)</span>.</li>
<li>For <span class="math inline">\(\theta=0.3\)</span></li>
</ul>
<p><span class="math display">\[
\mathrm{LOD}(0.3) = \log_{10}\!\left\{ \frac{0.3^{12} \cdot 0.7^{28}}{0.5^{40}} \right\}.
\]</span></p>
<div class="notes">
<p><strong>Script:</strong><br>
“With <span class="math inline">\(R=12\)</span> out of <span class="math inline">\(I=40\)</span>, <span class="math inline">\(\hat\theta=0.30\)</span>. Plugging into the LOD gives about <span class="math inline">\(1.43\)</span>, i.e., roughly 27:1 odds for linkage at the MLE. That’s suggestive, but below the classical <span class="math inline">\(LOD\ge 3\)</span> bar. Modern practice often uses permutation‑based thresholds instead of a hard rule.”</p>
</div>
</section>
<section id="worked-example-direct-counting-lod-1" class="level2">
<h2 class="anchored" data-anchor-id="worked-example-direct-counting-lod-1">Worked Example: Direct Counting LOD</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>R <span class="ot">&lt;-</span> <span class="dv">12</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="dv">40</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>thetahat <span class="ot">&lt;-</span> R <span class="sc">/</span> N</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>lod <span class="ot">&lt;-</span> <span class="fu">log10</span>((thetahat<span class="sc">^</span>R <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> thetahat)<span class="sc">^</span>(N <span class="sc">-</span> R)) <span class="sc">/</span> (<span class="fl">0.5</span><span class="sc">^</span>N))</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>lod</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1.4294</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span><span class="sc">^</span>lod</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 26.87819</code></pre>
</div>
</div>
<ul>
<li>Maximize <span class="math inline">\(\mathrm{LOD}(\theta)\)</span> over <span class="math inline">\(\theta \in [0,0.5]\)</span> to estimate the recombination fraction; <span class="math inline">\(LOD \ge 3\)</span> and <span class="math inline">\(\leq -2\)</span> are classic heuristics (context-dependent).</li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“This code reproduces the arithmetic: LOD around 1.43, odds around 27:1. The maximizer of the LOD is exactly <span class="math inline">\(R/I\)</span>. The <span class="math inline">\(\pm\)</span> thresholds are heuristics, not laws—use context and proper genome‑wide calibration.”</p>
</div>
</section>
<section id="worked-example-direct-counting-lod-2" class="level2">
<h2 class="anchored" data-anchor-id="worked-example-direct-counting-lod-2">Worked Example: Direct Counting LOD</h2>
<ul>
<li>Under <span class="math inline">\(H_0:\theta=0.5\)</span> (a boundary point), <span class="math inline">\(D = 2\ln(10)\cdot \mathrm{LOD}(\hat\theta)\)</span> has the mixture limit <span class="math inline">\(\tfrac{1}{2}\chi^2_0 + \tfrac{1}{2}\chi^2_1\)</span>; thus <span class="math inline">\(p = \tfrac{1}{2}\, \Pr(\chi^2_1 \ge D)\)</span>.</li>
</ul>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>D <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="sc">*</span> <span class="fu">log</span>(<span class="dv">10</span>) <span class="sc">*</span> lod</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>pval <span class="ot">&lt;-</span> <span class="fl">0.5</span> <span class="sc">*</span> <span class="fu">pchisq</span>(D, <span class="at">df =</span> <span class="dv">1</span>, <span class="at">lower.tail =</span> <span class="cn">FALSE</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>pval</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.005148931</code></pre>
</div>
</div>
<div class="notes">
<p><strong>Script:</strong><br>
“At the <strong>boundary</strong> <span class="math inline">\(\theta=0.5\)</span>, Wilks’ theorem gives a <strong>mixture</strong> null: half the mass at zero degrees of freedom and half at one. Concretely, compute <span class="math inline">\(D=2\ln(10)\cdot \mathrm{LOD}\)</span> and take <span class="math inline">\(p=\tfrac{1}{2}\Pr(\chi^2_1\ge D)\)</span>. This is the right conversion for two‑point tests against <span class="math inline">\(\theta=0.5\)</span>.”</p>
</div>
</section>
<section id="two-point-linkage-em-with-unknown-phase" class="level2">
<h2 class="anchored" data-anchor-id="two-point-linkage-em-with-unknown-phase">Two-Point Linkage: EM with Unknown Phase</h2>
<ul>
<li>Transmitting parent: <strong>double het</strong> <span class="math inline">\(Aa/Bb\)</span> with <strong>unknown phase</strong>.</li>
<li>Mate: <strong>Aa/bb</strong> (heterozygous at <span class="math inline">\(A\)</span>, homozygous at <span class="math inline">\(B\)</span>).<br>
</li>
<li>Observed per child: two-locus <strong>genotypes</strong> <span class="math inline">\((A\text{-genotype}, B\text{-genotype})\)</span>.</li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“Now a case where EM is <strong>actually necessary</strong>. The mate is <strong>bb</strong>, so the child’s <span class="math inline">\(B\)</span> allele comes entirely from the transmitting parent. That’s helpful. But the mate is <strong>Aa</strong> at <span class="math inline">\(A\)</span>, so whenever a child is <strong>Aa</strong>, we can’t tell whether the transmitting parent contributed <span class="math inline">\(A\)</span> or <span class="math inline">\(a\)</span>. Those are <strong>ambiguous transmissions</strong> between two haplotypes—exactly the kind of soft labeling EM was built for.”</p>
</div>
</section>
<section id="two-point-linkage-em-with-unknown-phase-1" class="level2">
<h2 class="anchored" data-anchor-id="two-point-linkage-em-with-unknown-phase-1">Two-Point Linkage: EM with Unknown Phase</h2>
<ul>
<li>Missing-data view: the parental phase (coupling vs repulsion) is not observed. We make it a complete-data problem by introducing a weight <span class="math inline">\(w\in[0,1]\)</span> that blends the two phase-specific models.</li>
<li>Model (given recombination fraction <span class="math inline">\(\theta\)</span>):
<ul>
<li>Coupling: <span class="math inline">\(P(\text{NR})=1-\theta\)</span>, <span class="math inline">\(P(\text{R})=\theta\)</span> (split equally across the two categories).</li>
<li>Repulsion: <span class="math inline">\(P(\text{NR})=\theta\)</span>, <span class="math inline">\(P(\text{R})=1-\theta\)</span>.</li>
</ul></li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“Think of <strong>phase</strong> as a two‑component latent class: coupling vs repulsion. We’ll carry a single scalar, <span class="math inline">\(w\)</span>, the posterior weight on coupling. Under coupling, nonrecombinants are more likely by <span class="math inline">\(1-\theta\)</span>; under repulsion the roles swap. EM will update <span class="math inline">\(w\)</span> and then update <span class="math inline">\(\theta\)</span> from <strong>weighted</strong> recombinant counts.”</p>
</div>
</section>
<section id="two-point-linkage-em-with-unknown-phase-2" class="level2">
<h2 class="anchored" data-anchor-id="two-point-linkage-em-with-unknown-phase-2">Two-Point Linkage: EM with Unknown Phase</h2>
<p>E-step (update the phase weight)</p>
<p><span class="math display">\[
w = \frac{(1-\theta^{(t)})^{\,n_{\text{NR}}}\,(\theta^{(t)})^{\,n_{\text{R}}}}
{(1-\theta^{(t)})^{\,n_{\text{NR}}}\,(\theta^{(t)})^{\,n_{\text{R}}} + (\theta^{(t)})^{\,n_{\text{NR}}}\,(1-\theta^{(t)})^{\,n_{\text{R}}}}
\]</span></p>
<p>and use <span class="math inline">\(1-w\)</span> as the weight on the repulsion configuration.</p>
<div class="notes">
<p><strong>Script:</strong><br>
“This is the <strong>posterior</strong> weight on coupling: likelihood under coupling divided by the sum of likelihoods under both phases. In code, use logs and a log‑sum‑exp trick to avoid underflow if <span class="math inline">\(n_{\text{NR}}\)</span> or <span class="math inline">\(n_{\text{R}}\)</span> is large.”</p>
</div>
<p>M-step (update <span class="math inline">\(\theta\)</span>)</p>
<p><span class="math display">\[\begin{align*}
\theta^{(t+1)}
&amp;= \frac{\text{(weighted recombinants)}}{N}
= \frac{w\,n_{\text{R}} + (1-w)\,n_{\text{NR}}}{N}.
\end{align*}\]</span></p>
<div class="notes">
<p><strong>Script:</strong><br>
“The M‑step is ‘recombinants over informatives’, but with a twist: we average the recombinant counts under coupling and repulsion using the current <span class="math inline">\(w\)</span>. That keeps the intuition intact.”</p>
</div>
</section>
<section id="from-two-point-em-to-multipoint" class="level2">
<h2 class="anchored" data-anchor-id="from-two-point-em-to-multipoint">From Two-Point EM to Multipoint</h2>
<ul>
<li>Same objective: estimate recombination while marginalizing over latent transmissions/phase under current <span class="math inline">\(\theta\)</span>.</li>
<li>E-step engine: compute required state weights via Elston–Stewart peeling (pedigrees) or Lander–Green forward–backward (marker HMM) when you have multiple markers and missing genotypes.</li>
<li>Quantities needed: <span class="math inline">\(\mathbb E[\#\,\text{recombinants between adjacent markers}]\)</span> and <span class="math inline">\(\mathbb E[\#\,\text{informative transmissions}]\)</span> under current map.</li>
<li>Use in practice: either (a) plug these expectations into an EM-style update, or (b) more commonly, scan positions to build a LOD curve and report peak and 1-LOD interval.</li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“In multipoint linkage, the hidden object is the <strong>inheritance state</strong> along the chromosome. The forward–backward algorithm gives us posterior state probabilities; from those we compute expected <strong>recombinant counts between markers</strong> and assemble LOD curves across the map. It’s the same EM logic, just powered by an HMM.”</p>
</div>
</section>
<section id="multipoint-linkage-hmm" class="level2">
<h2 class="anchored" data-anchor-id="multipoint-linkage-hmm">Multipoint Linkage (HMM)</h2>
<ul>
<li>Markers along a chromosome define an HMM over inheritance states; adjacent recombination rates drive transitions.</li>
<li>Combine penetrance with the marker HMM to compute <span class="math inline">\(L(\theta)\)</span> efficiently as you slide along the map (forward–backward yields the needed state probabilities even with missing phase/genotypes).</li>
<li>Pros: more information than two-point; better localization. Cons: requires a genetic map and error modeling.</li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“Two practical cautions: include a <strong>genotyping‑error model</strong>—even a small error rate flattens LODs if ignored—and consider sex‑specific maps or interference if your study demands it. Otherwise, the HMM machinery scales cleanly.”</p>
</div>
</section>
<section id="missing-data-em-in-linkage" class="level2">
<h2 class="anchored" data-anchor-id="missing-data-em-in-linkage">Missing-Data EM in Linkage</h2>
<ul>
<li>Complete data: informative meioses labeled as recombinant/nonrecombinant; incomplete data: untyped genotypes/phase.</li>
<li>E-step: compute <span class="math inline">\(E[\text{\# recombinants}]\)</span> and <span class="math inline">\(\mathrm E[\text{\# informative meioses}]\)</span> given current <span class="math inline">\(\theta^{(t)}\)</span> via peeling/HMM.</li>
<li>M-step: <span class="math inline">\(\displaystyle \theta^{(t+1)} = \frac{\mathrm E[R\mid\text{data},\theta^{(t)}]}{\mathrm E[I\mid\text{data},\theta^{(t)}]}\)</span>.</li>
<li>Iterate across <span class="math inline">\(\theta\)</span> grid to produce a LOD curve; maximize or report support interval.</li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“This slide distills the recipe: compute <strong>expected</strong> recombinants and informatives under the current map, then update <span class="math inline">\(\theta\)</span> as their ratio. In practice we report the <strong>LOD curve</strong> and a 1‑LOD support interval rather than a single point estimate.”</p>
</div>
</section>
<section id="practical-issues-in-linkage" class="level2">
<h2 class="anchored" data-anchor-id="practical-issues-in-linkage">Practical Issues in Linkage</h2>
<ul>
<li>Model specification: penetrance, phenocopy, allele frequencies; misspecification can distort LOD.</li>
<li>Marker selection: highly polymorphic markers increase information; account for sex-specific recombination if needed.</li>
<li>Computational trade-offs: many markers and large pedigrees require approximations or pruning; consider parametric vs nonparametric linkage.</li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“Linkage analyses are only as good as their <strong>trait model</strong>. Probe sensitivity to penetrance and phenocopy. Use polymorphic markers to maximize information, and thin dense SNPs if LD bloats state spaces. For large pedigrees, Elston–Stewart pruning is your friend.”</p>
</div>
</section>
<section id="linkage-disequilibrium" class="level2">
<h2 class="anchored" data-anchor-id="linkage-disequilibrium">Linkage Disequilibrium</h2>
<ul>
<li>Let alleles at two markers be denoted <span class="math inline">\(A, a\)</span> and <span class="math inline">\(B, b\)</span></li>
<li>Define allele frequencies by <span class="math inline">\(P_A, P_a, P_B, P_b\)</span></li>
<li>Define frequencies of haplotypes as <span class="math inline">\(P_{AB}, P_{Ab}, P_{aB}, P_{ab}\)</span></li>
<li>Define linkage equilibrium as independence in the <span class="math inline">\(2 \times 2\)</span> table
<ul>
<li>For example, <span class="math inline">\(P_{AB} = P_A \times P_B\)</span></li>
</ul></li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“LD is simply <strong>statistical dependence</strong> between alleles on the same haplotype. Think forces: drift and finite population size create dependence, recombination erodes it, selection and admixture can inflate it.”</p>
</div>
</section>
<section id="linkage-disequilibrium-1" class="level2">
<h2 class="anchored" data-anchor-id="linkage-disequilibrium-1">Linkage Disequilibrium</h2>
<ul>
<li>Define the LD coefficient as <span class="math inline">\(D\)</span></li>
</ul>
<p><span class="math display">\[D = p_{AB} - p_A p_B\]</span></p>
<ul>
<li>What are we looking at here?</li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“Algebraically, <span class="math inline">\(D\)</span> is the <strong>covariance</strong> between the allele indicators on a random haplotype. If <span class="math inline">\(D&gt;0\)</span>, <span class="math inline">\(AB\)</span> is over‑represented relative to independence; if <span class="math inline">\(D&lt;0\)</span>, it’s under‑represented.”</p>
</div>
</section>
<section id="linkage-disequilibrium-2" class="level2">
<h2 class="anchored" data-anchor-id="linkage-disequilibrium-2">Linkage Disequilibrium</h2>
<ul>
<li>Define the LD coefficient as <span class="math inline">\(D\)</span></li>
</ul>
<p><span class="math display">\[D = p_{AB} - p_A p_B\]</span></p>
<ul>
<li>Note that
<ul>
<li><span class="math inline">\(\operatorname{E}[A] = p_A, \operatorname{E}[B] = p_B\)</span></li>
<li><span class="math inline">\(\operatorname{E}[AB] = p_{AB}\)</span></li>
<li><span class="math inline">\(\operatorname{Cov}(A, B) = \operatorname{E}[AB] - \operatorname{E}[A]\operatorname{E}[B]\)</span></li>
</ul></li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“Formalizing that intuition: let <span class="math inline">\(A,B\in\{0,1\}\)</span> indicate the presence of alleles on a haplotype. Then <span class="math inline">\(D=\operatorname{Cov}(A,B)\)</span>. This is why correlation <span class="math inline">\(r\)</span> shows up naturally on the next slide.”</p>
</div>
</section>
<section id="linkage-disequilibrium-3" class="level2">
<h2 class="anchored" data-anchor-id="linkage-disequilibrium-3">Linkage Disequilibrium</h2>
<ul>
<li><span class="math inline">\(D\)</span>, or <span class="math inline">\(\operatorname{Cov}(A, B)\)</span>, is scale-dependent</li>
<li>Normalized measures:
<ul>
<li><span class="math inline">\(D' = D / D_{\max}\)</span>, where <span class="math inline">\(D_{\max} = \min(p_A p_b, p_a p_B)\)</span> if <span class="math inline">\(D &gt; 0\)</span></li>
<li>Pearson correlation:</li>
</ul></li>
</ul>
<p><span class="math display">\[\begin{align}
r
&amp;= \frac{\operatorname{Cov}(A, B)}{\sqrt{\operatorname{Var}(A) \operatorname{Var}(B)}} \\
&amp;= \frac{D}{\sqrt{p_A(1 - p_A) p_B(1 - p_B)}}
\end{align}\]</span></p>
<div class="notes">
<p><strong>Script:</strong><br>
“<span class="math inline">\(D\)</span> depends on allele frequencies, so we normalize. Two popular choices: <span class="math inline">\(D'\)</span> and <span class="math inline">\(r\)</span>. For <strong>association power</strong>, <span class="math inline">\(r\)</span>—and really <span class="math inline">\(r^2\)</span>—matters most because it directly scales the noncentrality parameter: effect sizes at a tag SNP are attenuated by about <span class="math inline">\(r\)</span> relative to the causal.”</p>
</div>
</section>
<section id="goal-of-association-ld-link" class="level2">
<h2 class="anchored" data-anchor-id="goal-of-association-ld-link">Goal of Association &amp; LD Link</h2>
<ul>
<li>Goal: test the null of no effect (e.g., <span class="math inline">\(\beta=0\)</span> or OR<span class="math inline">\(=1\)</span>) to identify regions harboring causal variants.</li>
<li>We test markers, not necessarily the causal variant. If a marker has nonzero LD (correlation <span class="math inline">\(r\)</span>) with a causal variant, its association test can detect signal from that causal.</li>
<li>Stronger <span class="math inline">\(|r|\)</span> implies a stronger expected signal at the marker; if <span class="math inline">\(r\approx 0\)</span>, the marker carries no information about that causal.</li>
<li>Sign matters: <span class="math inline">\(r&lt;0\)</span> flips the observed effect direction at the marker relative to the causal.</li>
<li>In practice, nearby markers within LD blocks show clustered p-values and similar effect signs.</li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“Here’s the bridge: association tests fire at <strong>markers</strong>. If a marker correlates with a causal by <span class="math inline">\(r\)</span>, the observed effect is roughly <span class="math inline">\(r\)</span> times the causal effect, and power scales with <span class="math inline">\(r^2\)</span>. That’s why p‑values cluster within LD blocks and why effect <strong>signs</strong> can flip if <span class="math inline">\(r&lt;0\)</span>.”</p>
</div>
</section>
<section id="single-marker-association-core-test-linear" class="level2">
<h2 class="anchored" data-anchor-id="single-marker-association-core-test-linear">Single-Marker Association: Core Test (Linear)</h2>
<ul>
<li>Linear trait: <span class="math inline">\(Y=\alpha+\beta G+\gamma^T C+\varepsilon\)</span>; test <span class="math inline">\(H_0{:}\,\beta=0\)</span>.</li>
<li>Genotype encodings: additive (0/1/2) or genotypic (AA/Aa/aa); include dominance to test non-additivity.</li>
<li>Binary outcomes (case–control) and logistic models will be covered later.</li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“Regression 101 with genetics details. Code genotypes additively unless biology suggests dominance or recessivity; add a dominance term if you want to test non‑additivity. Always include relevant covariates—sex, age, and <strong>PCs</strong> for ancestry—because stratification can fake associations.”</p>
</div>
</section>
<section id="basic-qc-pipeline-variant-level" class="level2">
<h2 class="anchored" data-anchor-id="basic-qc-pipeline-variant-level">Basic QC Pipeline (Variant-level)</h2>
<ul>
<li>Call rate/missingness (e.g., variant missingness &lt; 2%).</li>
<li>Hardy–Weinberg equilibrium in controls (e.g., P &gt; 1e−6), mindful of true deviations.</li>
<li>Minor allele frequency threshold (e.g., MAF ≥ 0.01 unless rare-variant methods used).</li>
<li>Differential missingness across case/control; strand/allele checks.</li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“QC is prevention. Missingness filters remove low‑quality sites; HWE in <strong>controls</strong> flags genotyping problems; MAF thresholds avoid unstable estimates unless you’re using rare‑variant methods. Don’t forget <strong>strand and allele</strong> checks—mismatches there create phantom effects.”</p>
</div>
</section>
<section id="basic-qc-pipeline-sample-level" class="level2">
<h2 class="anchored" data-anchor-id="basic-qc-pipeline-sample-level">Basic QC Pipeline (Sample-level)</h2>
<ul>
<li>Sample call rate; sex checks from X chr; heterozygosity outliers.</li>
<li>Relatedness/duplicates via IBD; ancestry via PCA; remove outliers or adjust with PCs.</li>
<li>Duplicates/cryptic relatedness: retain one per pair or use mixed models.</li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“At the sample level, verify <strong>sex</strong> against X‑chromosome heterozygosity, look for <strong>heterozygosity outliers</strong>, and identify relatives with IBD. Use PCA to quantify ancestry and either exclude outliers or adjust downstream. Mixed models help with relatedness, but don’t use them as an excuse to skip QC.”</p>
</div>
</section>
<section id="summary-key-takeaways" class="level2">
<h2 class="anchored" data-anchor-id="summary-key-takeaways">Summary &amp; Key Takeaways</h2>
<ul>
<li>Likelihood optimization: Newton–Raphson (score/Hessian), Gradient Descent/SGD (scalable), and EM (latent-data) are core tools.</li>
<li>EM for ABO gene frequencies: initialization affects speed; log-likelihood increases monotonically until convergence.</li>
<li>Linkage: two-point LOD and unknown phase via EM; multipoint mapping framed as an HMM for efficient inference with missing data.</li>
<li>LD basics: r and r^2 quantify correlation between markers and explain why nearby markers can show similar association signals when a causal variant is unobserved</li>
</ul>
<div class="notes">
<p><strong>Script:</strong><br>
“Take‑home messages. Newton and first‑order methods are your optimization workhorses. EM turns missingness into a tractable expectation–maximization cycle and increases the log‑likelihood monotonically. Linkage is about family transmissions; association is about population correlation. And <span class="math inline">\(r^2\)</span> is the currency that converts LD into expected association signal.”</p>
</div>
</section>
<section id="hands-on-lab-03-applied" class="level2">
<h2 class="anchored" data-anchor-id="hands-on-lab-03-applied">Hands-On: Lab 03 (Applied)</h2>
<ul>
<li>Implement ABO EM and compare starts; plot log-likelihood gap.</li>
<li>Two-point linkage: LOD grid search + EM with unknown phase.</li>
<li>Two-SNP haplotype EM; compute LD (D, r^2).</li>
<li>Mini association demo: linear regression + HWE QC in controls.</li>
</ul>
<p>See <code>labs/unsolved/lab-03.R</code> for the scaffold (with TODOs) and <code>labs/solved/lab-03.R</code> for a worked solution.</p>
<div class="notes">
<p><strong>Script:</strong><br>
“For the lab: keep a log of starting values and attained log‑likelihoods—this is good hygiene with EM. In the two‑point EM, verify that ambiguous transmissions are being split sensibly as <span class="math inline">\(\theta\)</span> and <span class="math inline">\(w\)</span> update. In the haplotype EM, watch for boundary estimates at very low MAFs. And in the association mini‑demo, confirm that including PCs meaningfully reduces ancestry structure before you trust the p‑values.”</p>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>