<!DOCTYPE html>
<html lang="en"><head>
<link href="../course_logo.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/tabby.min.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dc55a5b9e770e841cd82e46aadbfb9b0.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.8.24">

  <title>Statistical Genetics – Lecture 04: Population Structure and Bayesian Methods in Statistical Genetics</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    html { -webkit-text-size-adjust: 100%; }
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; font-weight: bold; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="../site_libs/revealjs/dist/theme/quarto-fd091bfd51e2c02178447c91813c7d86.css">
  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.css">
  <link href="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Lecture 04: Population Structure and Bayesian Methods in Statistical Genetics</h1>
  <p class="subtitle">PUBH 8878, Statistical Genetics</p>

<div class="quarto-title-authors">
</div>

</section>
<section id="the-bayesian-paradigm" class="slide level2">
<h2>The Bayesian Paradigm</h2>
<style>
    figcaption, p.caption {
        text-align: center;
    }
</style>
<ul>
<li>Let <span class="math inline">\theta</span> be unknown parameters, <span class="math inline">y</span> observed data</li>
<li>While frequentist methods treat <span class="math inline">\theta</span> as fixed but unknown, Bayesian methods treat <span class="math inline">\theta</span> as <strong>random variables</strong></li>
</ul>

<img data-src="images/thomas-bayes.png" class="r-stretch quarto-figure-center"><p class="caption">Sir Thomas Bayes</p></section>
<section id="anatomy-of-bayes-rule" class="slide level2">
<h2>Anatomy of Bayes’ Rule</h2>
<ul>
<li>Uncertainty in <span class="math inline">\theta</span> is modeled via a <strong>prior distribution</strong> <span class="math inline">p(\theta)</span></li>
<li>Observed data <span class="math inline">y</span> are modeled via a <strong>likelihood</strong> <span class="math inline">p(y|\theta)</span></li>
<li>Bayes’ theorem combines these to yield the <strong>posterior distribution</strong> <span class="math inline">p(\theta|y) \propto p(y|\theta)p(\theta)</span></li>
</ul>

<img data-src="images/bayes-dist.png" class="r-stretch"></section>
<section id="frequentist-vs-bayesian-terminology" class="slide level2">
<h2>Frequentist vs Bayesian Terminology</h2>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top" style="width:100%;">
<colgroup>
<col style="width: 22%">
<col style="width: 36%">
<col style="width: 41%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Concept</th>
<th style="text-align: left;">Frequentist</th>
<th style="text-align: left;">Bayesian</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Parameter</td>
<td style="text-align: left;">Fixed</td>
<td style="text-align: left;">Random</td>
</tr>
<tr class="even">
<td style="text-align: left;">Data</td>
<td style="text-align: left;">Random</td>
<td style="text-align: left;">Fixed</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Estimate</td>
<td style="text-align: left;">Point estimate</td>
<td style="text-align: left;">Posterior distribution</td>
</tr>
<tr class="even">
<td style="text-align: left;">Confidence Interval</td>
<td style="text-align: left;">Interval with coverage</td>
<td style="text-align: left;">Credible interval</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Hypothesis Test</td>
<td style="text-align: left;">Reject/Fail to reject</td>
<td style="text-align: left;">Posterior predictive check</td>
</tr>
<tr class="even">
<td style="text-align: left;">P-value</td>
<td style="text-align: left;">Probability of data under null</td>
<td style="text-align: left;">Probability of hypothesis given data</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Likelihood</td>
<td style="text-align: left;">Function of parameter given data</td>
<td style="text-align: left;">Function of data given parameter</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="interpretation-of-intervals" class="slide level2">
<h2>Interpretation of Intervals</h2>
<p>Compare statements on intervals</p>
<ul>
<li>Frequentist: “Over many repeated samples, 95% of such intervals will contain the true parameter.”</li>
<li>Bayesian: “Given the observed data and prior, there is a 95% probability that the parameter lies within this interval.”</li>
</ul>
</section>
<section id="bayesian-pros-and-cons" class="slide level2">
<h2>Bayesian Pros and Cons</h2>
<h3 id="pros">Pros</h3>
<ul>
<li>Intuitive: direct probability statements about parameters</li>
<li>Flexibility: complex models, hierarchical structures, small-n</li>
<li>Incorporate prior knowledge or expert opinion</li>
<li>Full uncertainty quantification via posterior distributions</li>
</ul>
<h3 id="cons">Cons</h3>
<ul>
<li>Computationally intensive: MCMC, variational inference</li>
<li>Sensitivity to prior choice</li>
<li>Not the status quo in many fields</li>
</ul>
</section>
<section id="bayes-vs-frequentist-in-genetic-inference" class="slide level2">
<h2>Bayes vs Frequentist in Genetic Inference</h2>
<ul>
<li>Frequentist pipelines (logistic regression + PCs, LMMs) offer calibrated <span class="math inline">p</span>-values but often ignore uncertainty in latent structure</li>
<li>Bayesian models propagate uncertainty in allele frequencies, effect sizes, and structure into downstream predictions (e.g., polygenic risk scores)</li>
<li>Empirical Bayes shrinkage (e.g., LDpred) blurs the line: priors estimated from data approximate Bayesian decisions with frequentist guarantees</li>
<li>Debate centers on transparency of priors vs.&nbsp;reliance on asymptotics—highlight assumptions explicitly to compare answers, not ideologies</li>
</ul>
</section>
<section id="further-resources" class="slide level2">
<h2>Further Resources</h2>
<div class="quarto-layout-panel" data-layout="[60, 40]">
<div class="quarto-layout-row quarto-layout-valign-bottom">
<div class="quarto-layout-cell" style="flex-basis: 60.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/statistical-rethinking.png"></p>
<figcaption><span class="citation" data-cites="mcelreathStatisticalRethinkingBayesian2020">McElreath (<a href="#/reference" role="doc-biblioref" onclick="">2020</a>)</span></figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 40.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://sites.stat.columbia.edu/gelman/book/bda_cover.png"></p>
<figcaption><span class="citation" data-cites="gelmanBayesianDataAnalysis2013">Gelman (<a href="#/reference" role="doc-biblioref" onclick="">2013</a>)</span></figcaption>
</figure>
</div>
</div>
</div>
</div>
</section>
<section id="conjugacy" class="slide level2">
<h2>Conjugacy</h2>
<ul>
<li>A prior is <strong>conjugate</strong> to a likelihood if the posterior is in the same family as the prior.</li>
<li>Example: Beta prior + Binomial likelihood <span class="math inline">to</span> Beta posterior</li>
</ul>
<h3 id="conjugacy-example">Conjugacy Example</h3>
<p>Consider the allele-frequency estimation setup introduced in Lecture 1.</p>
<ul>
<li>Previous experiment: observed <span class="math inline">x_{\text{prev}} = 3</span> successes (allele A) out of <span class="math inline">n_{\text{prev}} = 20</span> trials.</li>
<li>Current data: <span class="math inline">x = 11</span> successes out of <span class="math inline">n = 27</span> trials.</li>
<li>Parameter of interest: allele (“success”) frequency <span class="math inline">p</span>.</li>
</ul>
</section>
<section id="step-1-prior" class="slide level2">
<h2>Step 1: Prior</h2>
<p>From the previous experiment (3 successes out of 20 trials), we can form a Beta prior:</p>
<p>Interpret <span class="math inline">\alpha</span> and <span class="math inline">\beta</span> as pseudo-counts of successes and failures, respectively. (Note that <span class="math inline">\text{Beta}(1,1)</span> is uniform on [0,1].)</p>
<p><span class="math display">\begin{gather*}
p \sim \text{Beta}(\alpha, \beta) \\
\alpha = x_{\text{prev}} + 1 = 4 \\
\beta = n_{\text{prev}} - x_{\text{prev}} + 1 = 18.
\end{gather*}</span></p>
</section>
<section id="step-2-likelihood" class="slide level2">
<h2>Step 2: Likelihood</h2>
<p>Current data likelihood (up to proportionality in <span class="math inline">p</span>): <span class="math display">L(p; x,n) \propto p^{x} (1-p)^{n-x} = p^{11}(1-p)^{16}.</span></p>
<p>Recognize this kernel as</p>
<p><span class="math display">
\text{Beta}(x+1, n-x+1)=\text{Beta}(12,17)
</span></p>
</section>
<section id="step-3-posterior-conjugacy" class="slide level2">
<h2>Step 3: Posterior (Conjugacy)</h2>
<p>Multiply prior and likelihood kernels: <span class="math display">p^{\alpha-1}(1-p)^{\beta-1}\times p^{x}(1-p)^{n-x} = p^{(\alpha + x) - 1}(1-p)^{(\beta + n - x) - 1}.</span></p>
<p>Thus</p>
<p><span class="math display">\begin{align*}
p | x \sim \text{Beta}(\alpha + x,\ \beta + n - x)
&amp;= \text{Beta}(4+11,\ 18+16) \\
&amp;= \text{Beta}(15,34).
\end{align*}</span></p>
</section>
<section id="posterior-summaries" class="slide level2">
<h2>Posterior Summaries</h2>
<ul>
<li><p>Posterior mean: <span class="math inline">E[p|x] = \frac{15}{15+34} = 0.306</span> (shrinks slightly toward prior mean <span class="math inline">\frac{4}{22}=0.182</span> relative to MLE <span class="math inline">\hat p = 11/27 \approx 0.407</span>).</p></li>
<li><p>Effective sample size intuition: prior contributes <span class="math inline">(\alpha+\beta-2)=20</span> pseudo-trials; data contribute 27 real trials.</p></li>
</ul>
</section>
<section id="visualization" class="slide level2">
<h2>Visualization</h2>

<img data-src="lecture-04_files/figure-revealjs/unnamed-chunk-3-1.png" width="960" class="r-stretch"></section>
<section id="non-conjugate-models" class="slide level2">
<h2>Non-Conjugate Models</h2>
<ul>
<li>Conjugacy is convenient but limited to simple models</li>
<li>Many realistic models (e.g., multinomial/Dirichlet with nonlinear transforms) are non-conjugate</li>
<li>Bayesian inference in these models typically requires computational methods like MCMC or variational inference</li>
<li>Richard McElreath has a great <a href="https://www.youtube.com/watch?v=rZk2FqX2XnY">youtube video</a> on MCMC</li>
</ul>
</section>
<section id="bayesian-revisit-abo-allele-frequencies" class="slide level2">
<h2>Bayesian Revisit: ABO Allele Frequencies</h2>
<ul>
<li>Goal: Infer allele frequencies <code>(p_A, p_B, p_O)</code> given phenotype counts <code>(n_A, n_AB, n_B, n_O)</code>.</li>
<li>Frequentist (Lecture 03): EM treats latent genotypes for A and B phenotypes.</li>
<li>Bayesian: Place a prior on allele frequencies; integrate (average) over uncertainty rather than impute expected counts.</li>
</ul>
<aside class="notes">
<p>The EM algorithm gave us point estimates by iteratively reassigning ambiguous A and B phenotypes. In the Bayesian framing we directly place a prior on the simplex and obtain a posterior over allele frequencies that reflects finite-sample uncertainty.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="data-sufficient-statistics" class="slide level2">
<h2>Data &amp; Sufficient Statistics</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines"><code class="sourceCode r"><span id="cb1-1"><a></a><span class="co"># Phenotype counts (same as EM slide)</span></span>
<span id="cb1-2"><a></a>counts <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">nA =</span> <span class="dv">725</span>, <span class="at">nAB =</span> <span class="dv">72</span>, <span class="at">nB =</span> <span class="dv">258</span>, <span class="at">nO =</span> <span class="dv">1073</span>)</span>
<span id="cb1-3"><a></a>N <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="fu">unlist</span>(counts))</span>
<span id="cb1-4"><a></a>counts</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$nA
[1] 725

$nAB
[1] 72

$nB
[1] 258

$nO
[1] 1073</code></pre>
</div>
</div>
<ul>
<li>Total sample size: 2128 individuals.</li>
<li>Minimal information for this model are the four phenotype counts.</li>
</ul>
<aside class="notes">
<p>Placing the data early clarifies what the likelihood conditions on before diving into priors or Stan code.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="model-specification-recap" class="slide level2">
<h2>Model Specification (Recap)</h2>
<ul>
<li>Allele frequency vector: <span class="math inline">\boldsymbol p=(p_A,p_B,p_O)</span>, <span class="math inline">\boldsymbol p\sim\text{Dirichlet}(\boldsymbol\alpha)</span>.</li>
<li>Under HWE, genotype frequencies: <span class="math inline">p_A^2, 2p_A p_O, p_B^2, 2 p_B p_O, 2 p_A p_B, p_O^2</span>.</li>
<li>Phenotype probabilities (aggregating ambiguous genotypes):
<ul>
<li><span class="math inline">P(\text{A}) = p_A^2 + 2 p_A p_O</span></li>
<li><span class="math inline">P(\text{AB}) = 2 p_A p_B</span></li>
<li><span class="math inline">P(\text{B}) = p_B^2 + 2 p_B p_O</span></li>
<li><span class="math inline">P(\text{O}) = p_O^2</span></li>
</ul></li>
<li>Likelihood: <span class="math inline">(n_A,n_{AB},n_B,n_O) \sim \text{Multinomial}(N, \boldsymbol q)</span> with <span class="math inline">\boldsymbol q</span> above.</li>
</ul>
<aside class="notes">
<p>Choosing Dirichlet priors keeps conjugacy partially: not fully conjugate because of the nonlinear mapping from allele to phenotype probabilities, but the prior still supplies soft pseudo-counts that regularize towards symmetry (if all alphas equal) or prior knowledge.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="prior-families" class="slide level2">
<h2>Prior Families</h2>
<h3 id="weakly-informative">Weakly Informative</h3>
<ul>
<li>Weak: Dirichlet(1,1,1) (uniform over allele simplex)</li>
<li>Mild: Dirichlet(2,2,2) (light shrink toward center)</li>
</ul>
<aside class="notes">
<p>Set expectations for how each prior should behave before showing shapes or fits.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="prior-families-1" class="slide level2">
<h2>Prior Families</h2>
<h3 id="using-historical-data">Using Historical Data</h3>
<ul>
<li>Consider global survey means <span class="math inline">(p_A = 0.26, p_B = 0.09, p_O = 0.65)</span> <span class="citation" data-cites="mourant1976 yamamoto2012abo">(<a href="#/reference" role="doc-biblioref" onclick="">Mourant <em>et al.</em>, 1976</a>; <a href="#/reference" role="doc-biblioref" onclick="">Yamamoto, 2012</a>)</span>.</li>
<li>Dirichlet prior: <span class="math inline">\mathbf{\alpha} = k (0.26, 0.09, 0.65)</span></li>
<li>Effective sample size idea: acts like observing k allele draws before current data (2N alleles in sample).</li>
<li>Relative weight vs data (here total alleles = 2N = 4256):
<ul>
<li>k = 200 ⇒ prior weight ≈ 4.5% of total information.</li>
<li>k = 1000 ⇒ prior weight ≈ 19.0%.</li>
</ul></li>
</ul>
<aside class="notes">
<p>Explain that k trades off prior certainty vs flexibility. If study population differs (e.g., ancestry-specific shifts), large k may bias estimates. Showing prior SD table makes the 1/k scaling concrete.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="prior-shapes-marginal-densities" class="slide level2">
<h2>Prior Shapes (Marginal Densities)</h2>

<img data-src="lecture-04_files/figure-revealjs/unnamed-chunk-5-1.png" width="960" class="r-stretch"><aside class="notes">
<ul>
<li>Weak priors (Dir(1,1,1), Dir(2,2,2)) are diffuse over the simplex.</li>
<li>High-concentration informative priors collapse mass tightly around external means.</li>
<li>Marginal densities illustrate shrinkage strength before seeing data.</li>
</ul>
<p>Showing the priors <em>alone</em> helps students distinguish likelihood-driven updates from prior-imposed structure. The near-degenerate appearance for large k reinforces the concept of effective sample size encoded in the concentration.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="how-to-fit-these-models" class="slide level2">
<h2>How to fit these models?</h2>
<div class="quarto-layout-panel" data-layout="[80, -1, 20]">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="quarto-layout-cell" style="flex-basis: 79.2%;justify-content: flex-start;">
<ul>
<li>So we have our data, likelihood, and priors</li>
<li>We can use MCMC to sample from the posterior distribution of allele frequencies</li>
<li>There exist many software packages to do this! We will use Stan <span class="citation" data-cites="carpenterStanProbabilisticProgramming2017">(<a href="#/reference" role="doc-biblioref" onclick="">Carpenter <em>et al.</em>, 2017</a>)</span> via the <code>cmdstanr</code> R package</li>
</ul>
</div>
<div class="quarto-figure-spacer quarto-layout-cell" style="flex-basis: 1.0%;justify-content: flex-start;">
<p>&nbsp;</p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 19.8%;justify-content: center;">
<p><img data-src="https://mc-stan.org/img/logo_tm.png"></p>
</div>
</div>
</div>
</section>
<section id="stan-model" class="slide level2">
<h2>Stan Model</h2>
<h3 id="step-1-phenotype-probabilities">Step 1: Phenotype Probabilities</h3>
<div class="sourceCode" id="annotated-cell-2"><pre class="sourceCode numberSource stan code-annotation-code number-lines code-annotated"><code class="sourceCode stan"><button class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="1">1</button><span id="annotated-cell-2-1" class="code-annotation-target"><a href="#annotated-cell-2-1"></a><span class="kw">functions</span> {</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="2">2</button><span id="annotated-cell-2-2" class="code-annotation-target"><a href="#annotated-cell-2-2"></a>    <span class="dt">vector</span> abo_pheno_probs(<span class="dt">vector</span> p) {</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="3">3</button><span id="annotated-cell-2-3" class="code-annotation-target"><a href="#annotated-cell-2-3"></a>        <span class="dt">real</span> pA = p[<span class="dv">1</span>];</span>
<span id="annotated-cell-2-4"><a href="#annotated-cell-2-4"></a>        <span class="dt">real</span> pB = p[<span class="dv">2</span>];</span>
<span id="annotated-cell-2-5"><a href="#annotated-cell-2-5"></a>        <span class="dt">real</span> pO = p[<span class="dv">3</span>];</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="4">4</button><span id="annotated-cell-2-6" class="code-annotation-target"><a href="#annotated-cell-2-6"></a>        <span class="dt">vector</span>[<span class="dv">4</span>] q;</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="5">5</button><span id="annotated-cell-2-7" class="code-annotation-target"><a href="#annotated-cell-2-7"></a>        q[<span class="dv">1</span>] = pA * pA + <span class="dv">2</span> * pA * pO; <span class="co">// A phenotype</span></span>
<span id="annotated-cell-2-8"><a href="#annotated-cell-2-8"></a>        q[<span class="dv">2</span>] = <span class="dv">2</span> * pA * pB;            <span class="co">// AB phenotype</span></span>
<span id="annotated-cell-2-9"><a href="#annotated-cell-2-9"></a>        q[<span class="dv">3</span>] = pB * pB + <span class="dv">2</span> * pB * pO; <span class="co">// B phenotype</span></span>
<span id="annotated-cell-2-10"><a href="#annotated-cell-2-10"></a>        q[<span class="dv">4</span>] = pO * pO;               <span class="co">// O phenotype</span></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="6">6</button><span id="annotated-cell-2-11" class="code-annotation-target"><a href="#annotated-cell-2-11"></a>        <span class="cf">return</span> q;</span>
<span id="annotated-cell-2-12"><a href="#annotated-cell-2-12"></a>        }                                            </span>
<span id="annotated-cell-2-13"><a href="#annotated-cell-2-13"></a>}</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div>
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-2" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="1,13" data-code-annotation="1">Declare the <code>functions</code> block (optional in Stan, but lets us encapsulate logic).</span>
</dd>
<dt data-target-cell="annotated-cell-2" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="2" data-code-annotation="2">Define a helper that maps allele frequencies to phenotype probabilities.</span>
</dd>
<dt data-target-cell="annotated-cell-2" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="3,4,5" data-code-annotation="3">Extract components (purely for readability in later expressions).</span>
</dd>
<dt data-target-cell="annotated-cell-2" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="6" data-code-annotation="4">Allocate a length-4 vector <code>q</code> (A, AB, B, O) for phenotype probabilities.</span>
</dd>
<dt data-target-cell="annotated-cell-2" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="7,8,9,10" data-code-annotation="5">Hardy–Weinberg genotype algebra aggregated into phenotype probabilities.</span>
</dd>
<dt data-target-cell="annotated-cell-2" data-target-annotation="6">6</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="11" data-code-annotation="6">Return the vector; braces close the function and block.</span>
</dd>
</dl>
</section>
<section id="stan-model-1" class="slide level2">
<h2>Stan Model</h2>
<h3 id="step-2-data-transforms">Step 2: Data &amp; Transforms</h3>
<div class="sourceCode" id="annotated-cell-3"><pre class="sourceCode numberSource stan code-annotation-code number-lines code-annotated"><code class="sourceCode stan"><button class="code-annotation-anchor" data-target-cell="annotated-cell-3" data-target-annotation="1">1</button><span id="annotated-cell-3-1" class="code-annotation-target"><a href="#annotated-cell-3-1"></a><span class="kw">data</span> {</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-3" data-target-annotation="2">2</button><span id="annotated-cell-3-2" class="code-annotation-target"><a href="#annotated-cell-3-2"></a>    <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; nA;</span>
<span id="annotated-cell-3-3"><a href="#annotated-cell-3-3"></a>    <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; nAB;</span>
<span id="annotated-cell-3-4"><a href="#annotated-cell-3-4"></a>    <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; nB;</span>
<span id="annotated-cell-3-5"><a href="#annotated-cell-3-5"></a>    <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; nO;</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-3" data-target-annotation="3">3</button><span id="annotated-cell-3-6" class="code-annotation-target"><a href="#annotated-cell-3-6"></a>    <span class="dt">vector</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt;[<span class="dv">3</span>] alpha; <span class="co"># Dirichlet hyperparameters</span></span>
<span id="annotated-cell-3-7"><a href="#annotated-cell-3-7"></a>}</span>
<span id="annotated-cell-3-8"><a href="#annotated-cell-3-8"></a></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-3" data-target-annotation="4">4</button><span id="annotated-cell-3-9" class="code-annotation-target"><a href="#annotated-cell-3-9"></a><span class="kw">transformed data</span> {</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-3" data-target-annotation="5">5</button><span id="annotated-cell-3-10" class="code-annotation-target"><a href="#annotated-cell-3-10"></a>    <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; N = nA + nAB + nB + nO;</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-3" data-target-annotation="6">6</button><span id="annotated-cell-3-11" class="code-annotation-target"><a href="#annotated-cell-3-11"></a>    <span class="dt">array</span>[<span class="dv">4</span>] <span class="dt">int</span> y = { nA, nAB, nB, nO };</span>
<span id="annotated-cell-3-12"><a href="#annotated-cell-3-12"></a>}</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div>
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-3" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-3" data-code-lines="1,7" data-code-annotation="1">Raw observed counts and prior hyperparameters enter in the <code>data</code> block.</span>
</dd>
<dt data-target-cell="annotated-cell-3" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-3" data-code-lines="2,3,4,5" data-code-annotation="2">ABO phenotype counts (non-negative integers) per category.</span>
</dd>
<dt data-target-cell="annotated-cell-3" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-3" data-code-lines="6" data-code-annotation="3">Dirichlet prior parameters supplied from R (allow different priors via <code>alpha</code>).</span>
</dd>
<dt data-target-cell="annotated-cell-3" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-3" data-code-lines="9,12" data-code-annotation="4"><code>transformed data</code> pre-computes deterministic quantities once (saves work per draw).</span>
</dd>
<dt data-target-cell="annotated-cell-3" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-3" data-code-lines="10" data-code-annotation="5">Total sample size N used for reference or diagnostics.</span>
</dd>
<dt data-target-cell="annotated-cell-3" data-target-annotation="6">6</dt>
<dd>
<span data-code-cell="annotated-cell-3" data-code-lines="11" data-code-annotation="6">Assemble counts into an array to pass to the multinomial.</span>
</dd>
</dl>
</section>
<section id="stan-model-2" class="slide level2">
<h2>Stan Model</h2>
<h3 id="step-3-parameters-derived-q">Step 3: Parameters &amp; Derived <code>q</code></h3>
<div class="sourceCode" id="annotated-cell-4"><pre class="sourceCode numberSource stan code-annotation-code number-lines code-annotated"><code class="sourceCode stan"><button class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="1">1</button><span id="annotated-cell-4-1" class="code-annotation-target"><a href="#annotated-cell-4-1"></a><span class="kw">parameters</span> {</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="2">2</button><span id="annotated-cell-4-2" class="code-annotation-target"><a href="#annotated-cell-4-2"></a>    <span class="dt">simplex</span>[<span class="dv">3</span>] p;  <span class="co"># (pA, pB, pO) on 2-simplex</span></span>
<span id="annotated-cell-4-3"><a href="#annotated-cell-4-3"></a>}</span>
<span id="annotated-cell-4-4"><a href="#annotated-cell-4-4"></a></span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="3">3</button><span id="annotated-cell-4-5" class="code-annotation-target"><a href="#annotated-cell-4-5"></a><span class="kw">transformed parameters</span> {</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="4">4</button><span id="annotated-cell-4-6" class="code-annotation-target"><a href="#annotated-cell-4-6"></a>    <span class="dt">vector</span>[<span class="dv">4</span>] q = abo_pheno_probs(p);</span>
<span id="annotated-cell-4-7"><a href="#annotated-cell-4-7"></a>}</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div>
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-4" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="1,3" data-code-annotation="1">Declare unknown quantities to infer in <code>parameters</code>.</span>
</dd>
<dt data-target-cell="annotated-cell-4" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="2" data-code-annotation="2"><code>simplex[3]</code> enforces positivity and sum-to-one constraints automatically.</span>
</dd>
<dt data-target-cell="annotated-cell-4" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="5,7" data-code-annotation="3"><code>transformed parameters</code> recomputes per-draw derived values.</span>
</dd>
<dt data-target-cell="annotated-cell-4" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="6" data-code-annotation="4">Reuse helper to obtain phenotype probabilities from allele frequencies.</span>
</dd>
</dl>
</section>
<section id="stan-model-step-4-prior-likelihood" class="slide level2">
<h2>Stan Model (Step 4: Prior &amp; Likelihood)</h2>
<div class="sourceCode" id="annotated-cell-5"><pre class="sourceCode numberSource stan code-annotation-code number-lines code-annotated"><code class="sourceCode stan"><button class="code-annotation-anchor" data-target-cell="annotated-cell-5" data-target-annotation="1">1</button><span id="annotated-cell-5-1" class="code-annotation-target"><a href="#annotated-cell-5-1"></a><span class="kw">model</span> {</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-5" data-target-annotation="2">2</button><span id="annotated-cell-5-2" class="code-annotation-target"><a href="#annotated-cell-5-2"></a>    p ~ dirichlet(alpha);</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-5" data-target-annotation="3">3</button><span id="annotated-cell-5-3" class="code-annotation-target"><a href="#annotated-cell-5-3"></a>    y ~ multinomial(q);</span>
<span id="annotated-cell-5-4"><a href="#annotated-cell-5-4"></a>}</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div>
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-5" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-5" data-code-lines="1,4" data-code-annotation="1">The <code>model</code> block contains all sampling statements contributing to log density.</span>
</dd>
<dt data-target-cell="annotated-cell-5" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-5" data-code-lines="2" data-code-annotation="2">Dirichlet prior supplies pseudo-count style regularization.</span>
</dd>
<dt data-target-cell="annotated-cell-5" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-5" data-code-lines="3" data-code-annotation="3">Multinomial likelihood over phenotype counts with probabilities <code>q</code>.</span>
</dd>
</dl>
</section>
<section id="stan-model-step-5-generated-quantities" class="slide level2">
<h2>Stan Model (Step 5: Generated Quantities)</h2>
<div class="sourceCode" id="annotated-cell-6"><pre class="sourceCode numberSource stan code-annotation-code number-lines code-annotated"><code class="sourceCode stan"><button class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="1">1</button><span id="annotated-cell-6-1" class="code-annotation-target"><a href="#annotated-cell-6-1"></a><span class="kw">generated quantities</span> {</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="2">2</button><span id="annotated-cell-6-2" class="code-annotation-target"><a href="#annotated-cell-6-2"></a>    <span class="dt">real</span> log_lik = multinomial_lpmf(y | q);</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="3">3</button><span id="annotated-cell-6-3" class="code-annotation-target"><a href="#annotated-cell-6-3"></a>    <span class="dt">vector</span>[<span class="dv">6</span>] geno_freq;</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="4">4</button><span id="annotated-cell-6-4" class="code-annotation-target"><a href="#annotated-cell-6-4"></a>    <span class="dt">real</span> pA = p[<span class="dv">1</span>];</span>
<span id="annotated-cell-6-5"><a href="#annotated-cell-6-5"></a>    <span class="dt">real</span> pB = p[<span class="dv">2</span>];</span>
<span id="annotated-cell-6-6"><a href="#annotated-cell-6-6"></a>    <span class="dt">real</span> pO = p[<span class="dv">3</span>];</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="5">5</button><span id="annotated-cell-6-7" class="code-annotation-target"><a href="#annotated-cell-6-7"></a>    geno_freq[<span class="dv">1</span>] = pA * pA;        <span class="co"># AA</span></span>
<span id="annotated-cell-6-8"><a href="#annotated-cell-6-8"></a>    geno_freq[<span class="dv">2</span>] = <span class="dv">2</span> * pA * pO;    <span class="co"># AO</span></span>
<span id="annotated-cell-6-9"><a href="#annotated-cell-6-9"></a>    geno_freq[<span class="dv">3</span>] = pB * pB;        <span class="co"># BB</span></span>
<span id="annotated-cell-6-10"><a href="#annotated-cell-6-10"></a>    geno_freq[<span class="dv">4</span>] = <span class="dv">2</span> * pB * pO;    <span class="co"># BO</span></span>
<span id="annotated-cell-6-11"><a href="#annotated-cell-6-11"></a>    geno_freq[<span class="dv">5</span>] = <span class="dv">2</span> * pA * pB;    <span class="co"># AB</span></span>
<span id="annotated-cell-6-12"><a href="#annotated-cell-6-12"></a>    geno_freq[<span class="dv">6</span>] = pO * pO;        <span class="co"># OO</span></span>
<span id="annotated-cell-6-13"><a href="#annotated-cell-6-13"></a>}</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div>
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-6" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="1,13" data-code-annotation="1">Post-processing: quantities saved per posterior draw (no effect on inference math).</span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="2" data-code-annotation="2">Store log-likelihood for model comparison / LOO / WAIC.</span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="3" data-code-annotation="3">Allocate genotype frequency vector (optional pedagogical output).</span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="4,5,6" data-code-annotation="4">Local aliases improve clarity when computing genotype frequencies.</span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="7,8,9,10,11,12" data-code-annotation="5">Hardy–Weinberg genotype probabilities (could be summarized later in R).</span>
</dd>
</dl>
</section>
<section id="compile-stan-model" class="slide level2">
<h2>Compile Stan Model</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines"><code class="sourceCode r"><span id="cb3-1"><a></a>mod_path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"stan"</span>, <span class="st">"abo_multinomial.stan"</span>)</span>
<span id="cb3-2"><a></a>abo_mod <span class="ot">&lt;-</span> <span class="fu">cmdstan_model</span>(mod_path)</span></code></pre></div>
</div>
<ul>
<li>Stan uses C++ on the backend, so we need to compile the model once before fitting</li>
<li><code>cmdstan_model()</code> handles compilation and returns a model object for sampling</li>
</ul>
</section>
<section id="fit-weak-prior-dirichlet111" class="slide level2">
<h2>Fit: Weak Prior (Dirichlet(1,1,1))</h2>
<div class="cell">
<div class="sourceCode cell-code" id="annotated-cell-8"><pre class="sourceCode numberSource r code-annotation-code number-lines code-annotated"><code class="sourceCode r"><button class="code-annotation-anchor" data-target-cell="annotated-cell-8" data-target-annotation="1">1</button><span id="annotated-cell-8-1" class="code-annotation-target"><a href="#annotated-cell-8-1"></a>fit_weak <span class="ot">&lt;-</span> abo_mod<span class="sc">$</span><span class="fu">sample</span>(</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-8" data-target-annotation="2">2</button><span id="annotated-cell-8-2" class="code-annotation-target"><a href="#annotated-cell-8-2"></a>    <span class="at">data =</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-8-3"><a href="#annotated-cell-8-3"></a>        <span class="at">nA =</span> counts<span class="sc">$</span>nA, <span class="at">nAB =</span> counts<span class="sc">$</span>nAB, <span class="at">nB =</span> counts<span class="sc">$</span>nB, <span class="at">nO =</span> counts<span class="sc">$</span>nO,</span>
<span id="annotated-cell-8-4"><a href="#annotated-cell-8-4"></a>        <span class="at">alpha =</span> <span class="fu">rep</span>(<span class="dv">1</span>, <span class="dv">3</span>)</span>
<span id="annotated-cell-8-5"><a href="#annotated-cell-8-5"></a>    ),</span>
<span id="annotated-cell-8-6"><a href="#annotated-cell-8-6"></a>    <span class="at">seed =</span> <span class="dv">8878</span>,</span>
<button class="code-annotation-anchor" data-target-cell="annotated-cell-8" data-target-annotation="3">3</button><span id="annotated-cell-8-7" class="code-annotation-target"><a href="#annotated-cell-8-7"></a>    <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">parallel_chains =</span> <span class="dv">4</span>,</span>
<span id="annotated-cell-8-8"><a href="#annotated-cell-8-8"></a>    <span class="at">iter_warmup =</span> <span class="dv">1000</span>, <span class="at">iter_sampling =</span> <span class="dv">1000</span>,</span>
<span id="annotated-cell-8-9"><a href="#annotated-cell-8-9"></a>    <span class="at">refresh =</span> <span class="dv">0</span></span>
<span id="annotated-cell-8-10"><a href="#annotated-cell-8-10"></a>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-hidden code-annotation-container-grid">
<dt data-target-cell="annotated-cell-8" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-8" data-code-lines="1" data-code-annotation="1">Execute the MCMC sampler using the <code>sample()</code> method on the compiled model object. This will generate draws from the posterior distribution <span class="math inline">p(\boldsymbol{p} | \boldsymbol{y})</span>.</span>
</dd>
<dt data-target-cell="annotated-cell-8" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-8" data-code-lines="2,3,4,5" data-code-annotation="2">Pass data from R to Stan as a named list. The names must match the variables declared in the Stan <code>data</code> block. Here, <code>alpha = rep(1, 3)</code> sets a <code>Dirichlet(1,1,1)</code> prior, which is uniform over the simplex.</span>
</dd>
<dt data-target-cell="annotated-cell-8" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-8" data-code-lines="7,8,9" data-code-annotation="3">Specify MCMC sampler settings. We run multiple independent <code>chains</code> to diagnose convergence (e.g., via <span class="math inline">\hat{R}</span>). Each chain has a <code>warmup</code> phase (for adaptation, discarded) and a <code>sampling</code> phase (kept for inference).</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Running MCMC with 4 parallel chains...

Chain 1 finished in 0.0 seconds.
Chain 2 finished in 0.0 seconds.
Chain 3 finished in 0.0 seconds.
Chain 4 finished in 0.0 seconds.

All 4 chains finished successfully.
Mean chain execution time: 0.0 seconds.
Total execution time: 0.2 seconds.</code></pre>
</div>
</div>
</section>
<section id="posterior-summaries-intervals-weak-prior" class="slide level2">
<h2>Posterior Summaries &amp; Intervals (Weak Prior)</h2>

<img data-src="lecture-04_files/figure-revealjs/unnamed-chunk-8-1.png" width="960" class="r-stretch"></section>
<section id="em-vs-bayesian-point-estimates" class="slide level2">
<h2>EM vs Bayesian Point Estimates</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 5
  method            pA     pB    pO loglik
  &lt;chr&gt;          &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;
1 EM             0.209 0.0808 0.710 -2304.
2 Posterior Mean 0.209 0.0810 0.710    NA </code></pre>
</div>
</div>
<aside class="notes">
<p>This slide emphasizes calibration: the Bayesian machinery reproduces the frequentist point estimates but adds interval estimates derived from the full posterior. With alternative priors you can incorporate external knowledge or shrink extreme estimates for sparse categories.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="consolidated-posterior-comparison" class="slide level2">
<h2>Consolidated Posterior Comparison</h2>
<p>Posterior histograms only (models as rows, alleles as columns). Dashed line marks posterior mean.</p>

<img data-src="lecture-04_files/figure-revealjs/unnamed-chunk-11-1.png" width="960" class="r-stretch"><aside class="notes">
<p>Removed prior overlay to focus attention on posterior variance contraction across priors without dual-density distraction.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="population-substructure" class="slide level2">
<h2>Population Substructure</h2>
<ul>
<li>Features of a population which result from variation of expected allele frequencies across individuals</li>
<li>Standard allele counting (<span class="math inline">\hat{p} = (2n_{AA} + n_{Aa}/2n)</span>) will still be unbiased</li>
<li>But, not all subjects may have the same probability of being represented in the sample</li>
<li>Variance of estimate will be effected</li>
</ul>
</section>
<section id="population-stratification" class="slide level2">
<h2>Population Stratification</h2>
<ul>
<li>Individuals in a population can be subdivided into mutually exclusive strata</li>
<li>Within each strata the allele frequency is the same for all individuals</li>
<li>Intuitively, we are partitioning a large dataset into multiple smaller datasets</li>
</ul>
</section>
<section id="population-admixture" class="slide level2">
<h2>Population Admixture</h2>
<ul>
<li>When individuals in a population have a mixture of different genetic ancestries due to prior mixing of two or more populations</li>
<li>Often result of migration</li>
</ul>
</section>
<section id="population-admixture-1" class="slide level2">
<h2>Population Admixture</h2>

<img data-src="images/admixture.png" class="r-stretch quarto-figure-center"><p class="caption">From <span class="citation" data-cites="korunesHumanGeneticAdmixture2021">Korunes and Goldberg (<a href="#/reference" role="doc-biblioref" onclick="">2021</a>)</span></p></section>
<section id="population-inbreeding" class="slide level2">
<h2>Population Inbreeding</h2>
<ul>
<li>Occurs when there is a preference for mating among relatives in a population or because geographic isolation of subgroups restricts mating choices</li>
<li>Possibility that an offspring will inherit two copies of the same ancestral allele</li>
<li>Define <span class="math inline">F</span>, the inbreeding coefficient, as the probability that a random individual in the population inherits two copies of the same allele from a common ancestor</li>
</ul>
</section>
<section id="admixture-as-a-confounder" class="slide level2">
<h2>Admixture as a Confounder</h2>
<ul>
<li>Consider the problem of estimating the effect of a SNP on a disease phenotype: <span class="math inline">\beta</span> in <span class="math inline">P(Y=1) = \text{logit}^{-1}(\alpha + \beta G)</span></li>
<li>Recent admixture mixes ancestries within individuals: genotype is a convex combination of source populations</li>
<li>If phenotype prevalence differs by ancestry, local or global ancestry proportions act like hidden covariates</li>
<li>Association tests must separate causal signal from ancestry-driven allele frequency differences</li>
</ul>
</section>
<section id="detecting-structure-diagnostics" class="slide level2">
<h2>Detecting Structure: Diagnostics</h2>
<ul>
<li>Inflation factors (<span class="math inline">\lambda_{GC}</span>), QQ plots, and LD score intercepts flag confounding at the genome-wide scale</li>
<li><span class="math inline">F_{ST}</span>, population differentiation statistics, and heterozygosity contrasts summarize divergence</li>
<li>Eigen-decomposition of the genotype covariance (PCA) reveals axes of ancestry variation</li>
</ul>
</section>
<section id="principal-components-for-structure" class="slide level2">
<h2>Principal Components for Structure</h2>
<ul>
<li>Construct the standardized genotype matrix <span class="math inline">Z</span> and compute <span class="math inline">Z^T Z / M</span> (with <span class="math inline">M</span> markers)</li>
<li>Top eigenvectors capture major ancestry gradients; outliers highlight batch or sample swaps</li>
<li>Use the leading PCs as covariates in association tests or to stratify downstream analyses</li>
<li>Classroom demo: leverage the <code>adegenet::eHGDP</code> microsatellite panel (ships in R) to avoid VCF parsing—convert to CSV and color PCs by continental ancestry</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines"><code class="sourceCode r"><span id="cb6-1"><a></a><span class="co"># Load tidy microsatellite panel (1,350 individuals × 678 loci)</span></span>
<span id="cb6-2"><a></a><span class="fu">library</span>(adegenet)</span>
<span id="cb6-3"><a></a><span class="fu">data</span>(<span class="st">"eHGDP"</span>)</span>
<span id="cb6-4"><a></a>hgdp_df <span class="ot">&lt;-</span> <span class="fu">genind2df</span>(eHGDP, <span class="at">sep =</span> <span class="st">"/"</span>)</span>
<span id="cb6-5"><a></a><span class="fu">write.csv</span>(hgdp_df, <span class="st">"data/hgdp_microsat.csv"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb6-6"><a></a></span>
<span id="cb6-7"><a></a><span class="co"># Convert to allele count matrix, center columns (replace missing with locus means)</span></span>
<span id="cb6-8"><a></a>geno_mat <span class="ot">&lt;-</span> <span class="fu">scaleGen</span>(eHGDP, <span class="at">center =</span> <span class="cn">TRUE</span>, <span class="at">scale =</span> <span class="cn">FALSE</span>, <span class="at">NA.method =</span> <span class="st">"mean"</span>)</span>
<span id="cb6-9"><a></a>pc_fit <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(geno_mat, <span class="at">center =</span> <span class="cn">FALSE</span>, <span class="at">scale. =</span> <span class="cn">FALSE</span>)</span>
<span id="cb6-10"><a></a></span>
<span id="cb6-11"><a></a><span class="co"># Map individuals to geographic regions for coloring</span></span>
<span id="cb6-12"><a></a>pop_info <span class="ot">&lt;-</span> eHGDP<span class="sc">@</span>other<span class="sc">$</span>popInfo</span>
<span id="cb6-13"><a></a>pop_index <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(<span class="fu">pop</span>(eHGDP))</span>
<span id="cb6-14"><a></a>region <span class="ot">&lt;-</span> pop_info<span class="sc">$</span>Region[pop_index]</span>
<span id="cb6-15"><a></a></span>
<span id="cb6-16"><a></a>plot_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">PC1 =</span> pc_fit<span class="sc">$</span>x[, <span class="dv">1</span>], <span class="at">PC2 =</span> pc_fit<span class="sc">$</span>x[, <span class="dv">2</span>], <span class="at">Region =</span> region)</span>
<span id="cb6-17"><a></a><span class="fu">ggplot</span>(plot_df, <span class="fu">aes</span>(PC1, PC2, <span class="at">color =</span> Region)) <span class="sc">+</span></span>
<span id="cb6-18"><a></a>    <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.7</span>, <span class="at">size =</span> <span class="fl">1.5</span>) <span class="sc">+</span></span>
<span id="cb6-19"><a></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Population structure diagnostic"</span>, <span class="at">x =</span> <span class="st">"PC1"</span>, <span class="at">y =</span> <span class="st">"PC2"</span>) <span class="sc">+</span></span>
<span id="cb6-20"><a></a>    <span class="fu">theme_minimal</span>()</span></code></pre></div>
</div>
</section>
<section id="mixed-models-kinship" class="slide level2">
<h2>Mixed Models &amp; Kinship</h2>
<ul>
<li>Genomic Relationship Matrices (GRMs) model polygenic background sharing via linear mixed models (LMMs)</li>
<li>LMMs decompose phenotypic variance into fixed covariates + random genetic effects, soaking up stratification</li>
<li>Software: <code>GCTA</code>, <code>BOLT-LMM</code>, <code>SAIGE</code> scale to biobank settings; Bayesian analogues treat GRM variance components as parameters</li>
</ul>
</section>
<section id="correction-toolbox" class="slide level2">
<h2>Correction Toolbox</h2>
<ul>
<li>Include global ancestry covariates (PCs, admixture proportions)</li>
<li>Calibrate test statistics (genomic control) when residual inflation remains</li>
<li>Within-ancestry, case-only, or family-based designs remove between-group confounding at the design stage</li>
<li>Mixed-model association (REML/BLUP or Bayesian) further mitigates subtle relatedness</li>
</ul>
</section>
<section id="when-pca-is-not-enough" class="slide level2">
<h2>When PCA Is Not Enough</h2>
<ul>
<li>Fine-scale substructure (Ashkenazi vs Sephardi, Yoruba vs Igbo) can survive adjustment on a handful of PCs</li>
<li>Long-range LD, recent relatedness, and batch artefacts violate homoscedastic residual assumptions</li>
<li>Hierarchical models can borrow strength across loci while preserving uncertainty in latent structure</li>
</ul>
</section>
<section id="bayesian-admixture-structure-style" class="slide level2">
<h2>Bayesian Admixture (STRUCTURE-Style)</h2>
<ul>
<li>Latent populations <span class="math inline">k = 1,\ldots,K</span> possess allele frequencies <span class="math inline">\theta_{k\ell}</span> at locus <span class="math inline">\ell</span></li>
<li>Individual ancestry proportions <span class="math inline">\boldsymbol{\pi}_i \sim \text{Dirichlet}(\boldsymbol{\alpha})</span></li>
<li>Genotype <span class="math inline">y_{i\ell} \sim \text{Binomial}\left(2, \sum_{k} \pi_{ik} \theta_{k\ell}\right)</span> assuming HWE within each ancestral population</li>
<li>Posterior draws propagate ancestry/allele-frequency uncertainty into association testing, local ancestry, and polygenic prediction</li>
</ul>
</section>
<section id="stan-model-data-parameters" class="slide level2">
<h2>Stan Model: Data &amp; Parameters</h2>
<div class="sourceCode" id="cb7"><pre class="sourceCode numberSource stan number-lines"><code class="sourceCode stan"><span id="cb7-1"><a></a><span class="kw">data</span> {</span>
<span id="cb7-2"><a></a>    <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">1</span>&gt; N;                <span class="co">// individuals</span></span>
<span id="cb7-3"><a></a>    <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">1</span>&gt; L;                <span class="co">// loci</span></span>
<span id="cb7-4"><a></a>    <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">1</span>&gt; K;                <span class="co">// ancestral pops</span></span>
<span id="cb7-5"><a></a>    <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>, <span class="kw">upper</span>=<span class="dv">2</span>&gt; y[N, L]; <span class="co">// genotypes</span></span>
<span id="cb7-6"><a></a>    <span class="dt">vector</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt;[K] alpha;      <span class="co">// ancestry prior</span></span>
<span id="cb7-7"><a></a>    <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; a_theta;</span>
<span id="cb7-8"><a></a>    <span class="dt">real</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; b_theta;</span>
<span id="cb7-9"><a></a>}</span>
<span id="cb7-10"><a></a></span>
<span id="cb7-11"><a></a><span class="kw">parameters</span> {</span>
<span id="cb7-12"><a></a>    <span class="dt">array</span>[N] <span class="dt">simplex</span>[K] pi;        <span class="co">// admixture proportions</span></span>
<span id="cb7-13"><a></a>    <span class="dt">matrix</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>, <span class="kw">upper</span>=<span class="dv">1</span>&gt;[K, L] theta; <span class="co">// allele freqs</span></span>
<span id="cb7-14"><a></a>}</span>
<span id="cb7-15"><a></a></span>
<span id="cb7-16"><a></a><span class="kw">transformed parameters</span> {</span>
<span id="cb7-17"><a></a>    <span class="dt">matrix</span>[N, L] p_mix;</span>
<span id="cb7-18"><a></a>    <span class="cf">for</span> (n <span class="cf">in</span> <span class="dv">1</span>:N)</span>
<span id="cb7-19"><a></a>        <span class="cf">for</span> (l <span class="cf">in</span> <span class="dv">1</span>:L)</span>
<span id="cb7-20"><a></a>            p_mix[n, l] = dot_product(pi[n], col(theta, l));</span>
<span id="cb7-21"><a></a>}</span></code></pre></div>
</section>
<section id="stan-model-likelihood-priors" class="slide level2">
<h2>Stan Model: Likelihood &amp; Priors</h2>
<div class="sourceCode" id="cb8"><pre class="sourceCode numberSource stan number-lines"><code class="sourceCode stan"><span id="cb8-1"><a></a><span class="kw">model</span> {</span>
<span id="cb8-2"><a></a>    <span class="cf">for</span> (n <span class="cf">in</span> <span class="dv">1</span>:N)</span>
<span id="cb8-3"><a></a>        pi[n] ~ dirichlet(alpha);</span>
<span id="cb8-4"><a></a></span>
<span id="cb8-5"><a></a>    <span class="cf">for</span> (k <span class="cf">in</span> <span class="dv">1</span>:K)</span>
<span id="cb8-6"><a></a>        <span class="cf">for</span> (l <span class="cf">in</span> <span class="dv">1</span>:L)</span>
<span id="cb8-7"><a></a>            theta[k, l] ~ beta(a_theta, b_theta);</span>
<span id="cb8-8"><a></a></span>
<span id="cb8-9"><a></a>    <span class="cf">for</span> (n <span class="cf">in</span> <span class="dv">1</span>:N)</span>
<span id="cb8-10"><a></a>        <span class="cf">for</span> (l <span class="cf">in</span> <span class="dv">1</span>:L)</span>
<span id="cb8-11"><a></a>            y[n, l] ~ binomial(<span class="dv">2</span>, p_mix[n, l]);</span>
<span id="cb8-12"><a></a>}</span>
<span id="cb8-13"><a></a></span>
<span id="cb8-14"><a></a><span class="kw">generated quantities</span> {</span>
<span id="cb8-15"><a></a>    <span class="dt">matrix</span>[N, K] logit_pi;</span>
<span id="cb8-16"><a></a>    <span class="cf">for</span> (n <span class="cf">in</span> <span class="dv">1</span>:N)</span>
<span id="cb8-17"><a></a>        <span class="cf">for</span> (k <span class="cf">in</span> <span class="dv">1</span>:K)</span>
<span id="cb8-18"><a></a>            logit_pi[n, k] = logit(pi[n, k]);</span>
<span id="cb8-19"><a></a>}</span></code></pre></div>
<aside class="notes">
<p>Logit-transformed ancestry proportions support diagnostics (ternary plots) and regression models on the real line.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="fitting-the-admixture-model-in-r" class="slide level2">
<h2>Fitting the Admixture Model in R</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines"><code class="sourceCode r"><span id="cb9-1"><a></a><span class="fu">library</span>(cmdstanr)</span>
<span id="cb9-2"><a></a><span class="fu">set.seed</span>(<span class="dv">887804</span>)</span>
<span id="cb9-3"><a></a></span>
<span id="cb9-4"><a></a>rdirichlet <span class="ot">&lt;-</span> <span class="cf">function</span>(n, alpha) {</span>
<span id="cb9-5"><a></a>    k <span class="ot">&lt;-</span> <span class="fu">length</span>(alpha)</span>
<span id="cb9-6"><a></a>    gamma_draws <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rgamma</span>(n <span class="sc">*</span> k, <span class="at">shape =</span> alpha, <span class="at">rate =</span> <span class="dv">1</span>), <span class="at">ncol =</span> k, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span>
<span id="cb9-7"><a></a>    <span class="fu">sweep</span>(gamma_draws, <span class="dv">1</span>, <span class="fu">rowSums</span>(gamma_draws), <span class="st">"/"</span>)</span>
<span id="cb9-8"><a></a>}</span>
<span id="cb9-9"><a></a></span>
<span id="cb9-10"><a></a>N <span class="ot">&lt;-</span> <span class="dv">180</span>  <span class="co"># individuals</span></span>
<span id="cb9-11"><a></a>L <span class="ot">&lt;-</span> <span class="dv">60</span>   <span class="co"># loci</span></span>
<span id="cb9-12"><a></a>K <span class="ot">&lt;-</span> <span class="dv">3</span>    <span class="co"># ancestral populations</span></span>
<span id="cb9-13"><a></a></span>
<span id="cb9-14"><a></a>theta_true <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rbeta</span>(K <span class="sc">*</span> L, <span class="at">shape1 =</span> <span class="fl">2.5</span>, <span class="at">shape2 =</span> <span class="dv">5</span>), <span class="at">nrow =</span> K, <span class="at">ncol =</span> L)</span>
<span id="cb9-15"><a></a>pi_true <span class="ot">&lt;-</span> <span class="fu">rdirichlet</span>(N, <span class="fu">rep</span>(<span class="dv">3</span>, K))</span>
<span id="cb9-16"><a></a></span>
<span id="cb9-17"><a></a>y <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>L, <span class="at">nrow =</span> N, <span class="at">ncol =</span> L)</span>
<span id="cb9-18"><a></a><span class="cf">for</span> (n <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>N) {</span>
<span id="cb9-19"><a></a>    <span class="cf">for</span> (l <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>L) {</span>
<span id="cb9-20"><a></a>        p_mix <span class="ot">&lt;-</span> <span class="fu">sum</span>(pi_true[n, ] <span class="sc">*</span> theta_true[, l])</span>
<span id="cb9-21"><a></a>        y[n, l] <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">prob =</span> p_mix)</span>
<span id="cb9-22"><a></a>    }</span>
<span id="cb9-23"><a></a>}</span>
<span id="cb9-24"><a></a></span>
<span id="cb9-25"><a></a>stan_data <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb9-26"><a></a>    <span class="at">N =</span> N,</span>
<span id="cb9-27"><a></a>    <span class="at">L =</span> L,</span>
<span id="cb9-28"><a></a>    <span class="at">K =</span> K,</span>
<span id="cb9-29"><a></a>    <span class="at">y =</span> y,</span>
<span id="cb9-30"><a></a>    <span class="at">alpha =</span> <span class="fu">rep</span>(<span class="dv">1</span>, K),</span>
<span id="cb9-31"><a></a>    <span class="at">a_theta =</span> <span class="fl">2.5</span>,</span>
<span id="cb9-32"><a></a>    <span class="at">b_theta =</span> <span class="dv">5</span></span>
<span id="cb9-33"><a></a>)</span>
<span id="cb9-34"><a></a></span>
<span id="cb9-35"><a></a>admix_mod <span class="ot">&lt;-</span> <span class="fu">cmdstan_model</span>(<span class="fu">file.path</span>(<span class="st">"lectures"</span>, <span class="st">"stan"</span>, <span class="st">"structure_admixture.stan"</span>))</span>
<span id="cb9-36"><a></a></span>
<span id="cb9-37"><a></a>admix_fit <span class="ot">&lt;-</span> admix_mod<span class="sc">$</span><span class="fu">sample</span>(</span>
<span id="cb9-38"><a></a>    <span class="at">data =</span> stan_data,</span>
<span id="cb9-39"><a></a>    <span class="at">chains =</span> <span class="dv">4</span>,</span>
<span id="cb9-40"><a></a>    <span class="at">parallel_chains =</span> <span class="dv">4</span>,</span>
<span id="cb9-41"><a></a>    <span class="at">iter_warmup =</span> <span class="dv">500</span>,</span>
<span id="cb9-42"><a></a>    <span class="at">iter_sampling =</span> <span class="dv">500</span>,</span>
<span id="cb9-43"><a></a>    <span class="at">seed =</span> <span class="dv">887804</span>,</span>
<span id="cb9-44"><a></a>    <span class="at">refresh =</span> <span class="dv">0</span></span>
<span id="cb9-45"><a></a>)</span>
<span id="cb9-46"><a></a></span>
<span id="cb9-47"><a></a>admix_pi_draws <span class="ot">&lt;-</span> admix_fit<span class="sc">$</span><span class="fu">draws</span>(<span class="at">variables =</span> <span class="fu">c</span>(<span class="st">"pi"</span>), <span class="at">format =</span> <span class="st">"draws_array"</span>) </span>
<span id="cb9-48"><a></a></span>
<span id="cb9-49"><a></a><span class="fu">mcmc_hist</span>(admix_pi_draws, <span class="at">pars =</span> <span class="fu">c</span>(<span class="st">"pi[1,1]"</span>)) <span class="sc">+</span></span>
<span id="cb9-50"><a></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> pi_true[<span class="dv">2</span>, <span class="dv">1</span>], <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"red"</span>) </span>
<span id="cb9-51"><a></a></span>
<span id="cb9-52"><a></a></span>
<span id="cb9-53"><a></a>pi_rvar    <span class="ot">&lt;-</span> <span class="fu">as_draws_rvars</span>(admix_fit<span class="sc">$</span><span class="fu">draws</span>(<span class="st">"pi"</span>))<span class="sc">$</span>pi      <span class="co"># N × K matrix of rvars</span></span>
<span id="cb9-54"><a></a>theta_rvar <span class="ot">&lt;-</span> <span class="fu">as_draws_rvars</span>(admix_fit<span class="sc">$</span><span class="fu">draws</span>(<span class="st">"theta"</span>))<span class="sc">$</span>theta  <span class="co"># K × L matrix of rvars</span></span>
<span id="cb9-55"><a></a></span>
<span id="cb9-56"><a></a><span class="co"># Posterior means (and SDs if you need them)</span></span>
<span id="cb9-57"><a></a>pi_est_mat    <span class="ot">&lt;-</span> <span class="fu">mean</span>(pi_rvar)        <span class="co"># N × K numeric matrix</span></span>
<span id="cb9-58"><a></a>theta_est_mat <span class="ot">&lt;-</span> <span class="fu">mean</span>(theta_rvar)     <span class="co"># K × L numeric matrix</span></span>
<span id="cb9-59"><a></a></span>
<span id="cb9-60"><a></a><span class="co"># Example diagnostics</span></span>
<span id="cb9-61"><a></a>pi_rmse    <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(<span class="fu">mean</span>((pi_est_mat    <span class="sc">-</span> pi_true)<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb9-62"><a></a>theta_rmse <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(<span class="fu">mean</span>((theta_est_mat <span class="sc">-</span> theta_true)<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb9-63"><a></a>pi_rmse</span>
<span id="cb9-64"><a></a>theta_rmse</span></code></pre></div>
</div>
</section>
<section id="posterior-summaries-admixture" class="slide level2">
<h2>Posterior Summaries (Admixture)</h2>
<ul>
<li>Compare posterior means <span class="math inline">\hat{\pi}_{ik}</span> against the simulated truth to confirm the model recovers ancestry proportions (RMSE reported above)</li>
<li>Posterior draws for <span class="math inline">\theta_{k\ell}</span> provide credible intervals on population-specific allele frequencies for downstream association</li>
<li>Posterior predictive checks or replicated phenotypes can diagnose residual stratification before moving to real data</li>
</ul>
</section>
<section id="posterior-density-plots" class="slide level2">
<h2>Posterior Density Plots</h2>
<ul>
<li>Convert Stan output to <code>posterior::rvar</code> objects to preserve array shapes automatically</li>
<li>Density overlays with dashed true values visually confirm the model recovers simulated parameters</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines"><code class="sourceCode r"><span id="cb10-1"><a></a>pi_rvar <span class="ot">&lt;-</span> <span class="fu">as_draws_rvars</span>(admix_fit<span class="sc">$</span><span class="fu">draws</span>(<span class="st">"pi"</span>))<span class="sc">$</span>pi</span>
<span id="cb10-2"><a></a>theta_rvar <span class="ot">&lt;-</span> <span class="fu">as_draws_rvars</span>(admix_fit<span class="sc">$</span><span class="fu">draws</span>(<span class="st">"theta"</span>))<span class="sc">$</span>theta</span>
<span id="cb10-3"><a></a></span>
<span id="cb10-4"><a></a>rvar_to_long <span class="ot">&lt;-</span> <span class="cf">function</span>(x, truth, idx_names) {</span>
<span id="cb10-5"><a></a>    arr <span class="ot">&lt;-</span> <span class="fu">as.array</span>(x)</span>
<span id="cb10-6"><a></a>    <span class="fu">dimnames</span>(arr) <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb10-7"><a></a>        <span class="at">draw =</span> <span class="fu">seq_len</span>(<span class="fu">dim</span>(arr)[<span class="dv">1</span>]),</span>
<span id="cb10-8"><a></a>        <span class="at">idx1 =</span> <span class="fu">seq_len</span>(<span class="fu">dim</span>(arr)[<span class="dv">2</span>]),</span>
<span id="cb10-9"><a></a>        <span class="at">idx2 =</span> <span class="fu">seq_len</span>(<span class="fu">dim</span>(arr)[<span class="dv">3</span>])</span>
<span id="cb10-10"><a></a>    )</span>
<span id="cb10-11"><a></a>    df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">as.table</span>(arr), <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb10-12"><a></a>    <span class="fu">names</span>(df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"draw"</span>, idx_names, <span class="st">"value"</span>)</span>
<span id="cb10-13"><a></a>    df[[idx_names[<span class="dv">1</span>]]] <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(df[[idx_names[<span class="dv">1</span>]]])</span>
<span id="cb10-14"><a></a>    df[[idx_names[<span class="dv">2</span>]]] <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(df[[idx_names[<span class="dv">2</span>]]])</span>
<span id="cb10-15"><a></a>    df<span class="sc">$</span>true <span class="ot">&lt;-</span> truth[<span class="fu">cbind</span>(df[[idx_names[<span class="dv">1</span>]]], df[[idx_names[<span class="dv">2</span>]]])]</span>
<span id="cb10-16"><a></a>    df</span>
<span id="cb10-17"><a></a>}</span>
<span id="cb10-18"><a></a></span>
<span id="cb10-19"><a></a>pi_long <span class="ot">&lt;-</span> <span class="fu">rvar_to_long</span>(pi_rvar, pi_true, <span class="fu">c</span>(<span class="st">"individual"</span>, <span class="st">"component"</span>))</span>
<span id="cb10-20"><a></a>theta_long <span class="ot">&lt;-</span> <span class="fu">rvar_to_long</span>(theta_rvar, theta_true, <span class="fu">c</span>(<span class="st">"population"</span>, <span class="st">"locus"</span>))</span>
<span id="cb10-21"><a></a></span>
<span id="cb10-22"><a></a><span class="fu">ggplot</span>(<span class="fu">filter</span>(pi_long, individual <span class="sc">&lt;=</span> <span class="dv">6</span>),</span>
<span id="cb10-23"><a></a>       <span class="fu">aes</span>(value, <span class="at">colour =</span> <span class="fu">factor</span>(component), <span class="at">fill =</span> <span class="fu">factor</span>(component))) <span class="sc">+</span></span>
<span id="cb10-24"><a></a>    <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.25</span>) <span class="sc">+</span></span>
<span id="cb10-25"><a></a>    <span class="fu">geom_vline</span>(<span class="fu">aes</span>(<span class="at">xintercept =</span> true, <span class="at">colour =</span> <span class="fu">factor</span>(component)),</span>
<span id="cb10-26"><a></a>               <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">linewidth =</span> <span class="fl">0.4</span>) <span class="sc">+</span></span>
<span id="cb10-27"><a></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span> individual, <span class="at">ncol =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb10-28"><a></a>    <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Ancestry proportion"</span>, <span class="at">y =</span> <span class="st">"Posterior density"</span>,</span>
<span id="cb10-29"><a></a>         <span class="at">colour =</span> <span class="st">"Component"</span>, <span class="at">fill =</span> <span class="st">"Component"</span>,</span>
<span id="cb10-30"><a></a>         <span class="at">title =</span> <span class="st">"Posterior of π (true values dashed)"</span>) <span class="sc">+</span></span>
<span id="cb10-31"><a></a>    <span class="fu">theme_minimal</span>()</span>
<span id="cb10-32"><a></a></span>
<span id="cb10-33"><a></a><span class="fu">ggplot</span>(<span class="fu">filter</span>(theta_long, locus <span class="sc">&lt;=</span> <span class="dv">9</span>),</span>
<span id="cb10-34"><a></a>       <span class="fu">aes</span>(value, <span class="at">colour =</span> <span class="fu">factor</span>(population), <span class="at">fill =</span> <span class="fu">factor</span>(population))) <span class="sc">+</span></span>
<span id="cb10-35"><a></a>    <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.25</span>) <span class="sc">+</span></span>
<span id="cb10-36"><a></a>    <span class="fu">geom_vline</span>(<span class="fu">aes</span>(<span class="at">xintercept =</span> true, <span class="at">colour =</span> <span class="fu">factor</span>(population)),</span>
<span id="cb10-37"><a></a>               <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">linewidth =</span> <span class="fl">0.4</span>) <span class="sc">+</span></span>
<span id="cb10-38"><a></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span> locus, <span class="at">ncol =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb10-39"><a></a>    <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Allele frequency"</span>, <span class="at">y =</span> <span class="st">"Posterior density"</span>,</span>
<span id="cb10-40"><a></a>         <span class="at">colour =</span> <span class="st">"Population"</span>, <span class="at">fill =</span> <span class="st">"Population"</span>,</span>
<span id="cb10-41"><a></a>         <span class="at">title =</span> <span class="st">"Posterior of θ (true values dashed)"</span>) <span class="sc">+</span></span>
<span id="cb10-42"><a></a>    <span class="fu">theme_minimal</span>()</span></code></pre></div>
</div>
</section>
<section id="summary-next-steps" class="slide level2">
<h2>Summary &amp; Next Steps</h2>
<ul>
<li>Detect and correct stratification with diagnostics (λ<sub>GC</sub>, PCA) and mixed-model adjustments before testing</li>
<li>Bayesian framing exposes prior choices, enables posterior uncertainty on structure (ABO example, admixture Stan model)</li>
<li>Next: Bayesian association (regression) and hierarchical shrinkage for polygenic traits, building on latent structure posteriors</li>
</ul>
<aside class="notes">
<p>Close with a segue to upcoming material reinforcing how current ideas generalize.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="reference" class="slide level2 smaller scrollable">
<h2>Reference</h2>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-carpenterStanProbabilisticProgramming2017" class="csl-entry" role="listitem">
Carpenter,B. <em>et al.</em> (2017) <a href="https://doi.org/10.18637/jss.v076.i01">Stan: <span>A Probabilistic Programming Language</span></a>. <em>Journal of Statistical Software</em>, <strong>76</strong>, 1–32.
</div>
<div id="ref-gelmanBayesianDataAnalysis2013" class="csl-entry" role="listitem">
Gelman,A. (2013) Bayesian <span>Data Analysis</span> 3rd ed. CRC Press.
</div>
<div id="ref-korunesHumanGeneticAdmixture2021" class="csl-entry" role="listitem">
Korunes,K.L. and Goldberg,A. (2021) <a href="https://doi.org/10.1371/journal.pgen.1009374">Human genetic admixture</a>. <em>PLOS Genetics</em>, <strong>17</strong>, e1009374.
</div>
<div id="ref-mcelreathStatisticalRethinkingBayesian2020" class="csl-entry" role="listitem">
McElreath,R. (2020) Statistical rethinking: A <span>Bayesian</span> course with examples in <span>R</span> and <span>Stan</span> Second edition. CRC Press, Boca Raton.
</div>
<div id="ref-mourant1976" class="csl-entry" role="listitem">
Mourant,A.E. <em>et al.</em> (1976) The distribution of the human blood groups and other polymorphisms 2nd ed. Oxford University Press, London.
</div>
<div id="ref-yamamoto2012abo" class="csl-entry" role="listitem">
Yamamoto,F. (2012) <a href="https://doi.org/10.1159/000342197">ABO blood groups</a>. <em>Transfusion Medicine and Hemotherapy</em>, <strong>39</strong>, 213–220.
</div>
</div>
</section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">

</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"8\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp('/' + window.location.host + '/');
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
            let selectedAnnoteEl;
            const selectorForAnnotation = ( cell, annotation) => {
              let cellAttr = 'data-code-cell="' + cell + '"';
              let lineAttr = 'data-code-annotation="' +  annotation + '"';
              const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
              return selector;
            }
            const selectCodeLines = (annoteEl) => {
              const doc = window.document;
              const targetCell = annoteEl.getAttribute("data-target-cell");
              const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
              const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
              const lines = annoteSpan.getAttribute("data-code-lines").split(",");
              const lineIds = lines.map((line) => {
                return targetCell + "-" + line;
              })
              let top = null;
              let height = null;
              let parent = null;
              if (lineIds.length > 0) {
                  //compute the position of the single el (top and bottom and make a div)
                  const el = window.document.getElementById(lineIds[0]);
                  top = el.offsetTop;
                  height = el.offsetHeight;
                  parent = el.parentElement.parentElement;
                if (lineIds.length > 1) {
                  const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
                  const bottom = lastEl.offsetTop + lastEl.offsetHeight;
                  height = bottom - top;
                }
                if (top !== null && height !== null && parent !== null) {
                  // cook up a div (if necessary) and position it 
                  let div = window.document.getElementById("code-annotation-line-highlight");
                  if (div === null) {
                    div = window.document.createElement("div");
                    div.setAttribute("id", "code-annotation-line-highlight");
                    div.style.position = 'absolute';
                    parent.appendChild(div);
                  }
                  div.style.top = top - 2 + "px";
                  div.style.height = height + 4 + "px";
                  div.style.left = 0;
                  let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
                  if (gutterDiv === null) {
                    gutterDiv = window.document.createElement("div");
                    gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                    gutterDiv.style.position = 'absolute';
                    const codeCell = window.document.getElementById(targetCell);
                    const gutter = codeCell.querySelector('.code-annotation-gutter');
                    gutter.appendChild(gutterDiv);
                  }
                  gutterDiv.style.top = top - 2 + "px";
                  gutterDiv.style.height = height + 4 + "px";
                }
                selectedAnnoteEl = annoteEl;
              }
            };
            const unselectCodeLines = () => {
              const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
              elementsIds.forEach((elId) => {
                const div = window.document.getElementById(elId);
                if (div) {
                  div.remove();
                }
              });
              selectedAnnoteEl = undefined;
            };
              // Handle positioning of the toggle
          window.addEventListener(
            "resize",
            throttle(() => {
              elRect = undefined;
              if (selectedAnnoteEl) {
                selectCodeLines(selectedAnnoteEl);
              }
            }, 10)
          );
          function throttle(fn, ms) {
          let throttle = false;
          let timer;
            return (...args) => {
              if(!throttle) { // first call gets through
                  fn.apply(this, args);
                  throttle = true;
              } else { // all the others get throttled
                  if(timer) clearTimeout(timer); // cancel #2
                  timer = setTimeout(() => {
                    fn.apply(this, args);
                    timer = throttle = false;
                  }, ms);
              }
            };
          }
            const annoteTargets = window.document.querySelectorAll('.code-annotation-anchor');
            for (let i=0; i<annoteTargets.length; i++) {
              const annoteTarget = annoteTargets[i];
              const targetCell = annoteTarget.getAttribute("data-target-cell");
              const targetAnnotation = annoteTarget.getAttribute("data-target-annotation");
              const contentFn = () => {
                const content = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
                if (content) {
                  const tipContent = content.cloneNode(true);
                  tipContent.classList.add("code-annotation-tip-content");
                  return tipContent.outerHTML;
                }
              }
              const config = {
                allowHTML: true,
                content: contentFn,
                onShow: (instance) => {
                  selectCodeLines(instance.reference);
                  instance.reference.classList.add('code-annotation-active');
                  window.tippy.hideAll();
                },
                onHide: (instance) => {
                  unselectCodeLines();
                  instance.reference.classList.remove('code-annotation-active');
                },
                maxWidth: 300,
                delay: [50, 0],
                duration: [200, 0],
                offset: [5, 10],
                arrow: true,
                trigger: 'click',
                appendTo: function(el) {
                  return el.parentElement.parentElement.parentElement;
                },
                interactive: true,
                interactiveBorder: 10,
                theme: 'light-border',
                placement: 'right',
                positionFixed: true,
                popperOptions: {
                  modifiers: [
                  {
                    name: 'flip',
                    options: {
                      flipVariations: false, // true by default
                      allowedAutoPlacements: ['right'],
                      fallbackPlacements: ['right', 'top', 'top-start', 'top-end', 'bottom', 'bottom-start', 'bottom-end', 'left'],
                    },
                  },
                  {
                    name: 'preventOverflow',
                    options: {
                      mainAxis: false,
                      altAxis: false
                    }
                  }
                  ]        
                }      
              };
              window.tippy(annoteTarget, config); 
            }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    

</body></html>