<!DOCTYPE html>
<html lang="en"><head>
<link href="../course_logo.png" rel="icon" type="image/png">
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-html/tabby.min.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.8.25">

  <title>Statistical Genetics – Lecture 07: False Discovery Rates</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    html { -webkit-text-size-adjust: 100%; }
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; font-weight: bold; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="../site_libs/revealjs/dist/theme/quarto-b35a3bfecdac5fc8de409df89ff20628.css">
  <link rel="stylesheet" href="styles.css">
  <link href="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Lecture 07: False Discovery Rates</h1>
  <p class="subtitle">PUBH 8878, Statistical Genetics</p>

<div class="quarto-title-authors">
</div>

</section>
<section id="obligatory-xkcd" class="slide level2">
<h2>Obligatory XKCD</h2>
<div class="quarto-layout-panel" data-layout="[45, 55]">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 45.0%;justify-content: center;">
<p><img data-src="images/paste-9.png"></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 55.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/paste-10.png"></p>
<figcaption><a href="https://xkcd.com/882">From xkcd</a></figcaption>
</figure>
</div>
</div>
</div>
</div>
</section>
<section id="problem-set-up" class="slide level2">
<h2>Problem Set-up</h2>
<ul>
<li>Consider a GWAS where <span class="math inline">\(m\)</span> SNPs are tested for association with a trait of interest</li>
<li>For SNP <span class="math inline">\(i\)</span>, Let <span class="math inline">\(\delta\)</span> be a unobserved random variable where</li>
</ul>
<p><span class="math display">\[
\delta_i =
\begin{cases}
0 \text{ if the i'th hypothesis is a true null} \\
1 \text{ if the alternative is true}
\end{cases}
\]</span></p>
<ul>
<li>Based on some rule, we observe <span class="math inline">\(r_i\)</span>, where</li>
</ul>
<p><span class="math display">\[
r_i =
\begin{cases}
0 \text{ if we fail to reject the i'th hypothesis} \\
1 \text{ if we reject the i'th hypothesis}
\end{cases}
\]</span></p>
</section>
<section id="problem-set-up-1" class="slide level2">
<h2>Problem Set-up</h2>
<ul>
<li>For each SNP, given our data, we compute a test statistic <span class="math inline">\(z_i\)</span></li>
<li>We will reject hypothesis <span class="math inline">\(i\)</span> if <span class="math inline">\(z_i \in A\)</span>, where <span class="math inline">\(A\)</span> is some rejection region</li>
<li>Let <span class="math inline">\(R\)</span> be the total number of rejections, where <span class="math inline">\(R = \sum_{i=1}^m r_i\)</span></li>
<li>Let <span class="math inline">\(T, V\)</span> be the total number of true and false discoveries respectively, where</li>
</ul>
<p><span class="math display">\[
T = \sum_{i=1}^m r_i \delta_i, \qquad V = \sum_{i=1}^m r_i (1-\delta_i)
\]</span></p>
<ul>
<li>The Family Wise Error Rate (FWER) is the probably that we have at least one false discovery, or <span class="math inline">\(\text{Pr}(V \geq 1)\)</span></li>
<li>The False Discovery Proportion is <span class="math inline">\(Q = \frac{V}{R}\)</span>, where <span class="math inline">\(Q=0\)</span> if <span class="math inline">\(R=0\)</span></li>
</ul>
</section>
<section id="problem-set-up-2" class="slide level2">
<h2>Problem Set-up</h2>
<ul>
<li>Note that <span class="math inline">\(Q\)</span> is not observable!</li>
<li><span class="math inline">\(V\)</span> is unknown as <span class="math inline">\(\delta\)</span> is unobservable</li>
<li>There exist a variety of methods to estimate <span class="math inline">\(Q\)</span></li>
</ul>
</section>
<section id="two-preliminaries" class="slide level2">
<h2>Two Preliminaries</h2>
<ol type="1">
<li>Distribution of <span class="math inline">\(p\)</span>-values</li>
<li>Generating a Random Variable with a desired distribution</li>
</ol>
</section>
<section id="distribution-of-p-values" class="slide level2">
<h2>Distribution of <span class="math inline">\(p\)</span>-values</h2>
<ul>
<li>Let <span class="math inline">\(X \in \mathbb{R}\)</span> be a random variable with pdf <span class="math inline">\(f_X(x) = \frac{d}{dx}F(x)\)</span></li>
<li>Let <span class="math inline">\(y = F(x)\)</span> such that <span class="math inline">\(x = F^{-1}(y)\)</span></li>
<li>Via theory of transformations, we achieve for <span class="math inline">\(y \in (0, 1)\)</span></li>
</ul>
<p><span class="math display">\[
f_Y(y) = f_X\left(F^{-1}(y)\right)\left| \frac{dF^{-1}(y)}{dy}\right|
\]</span></p>
<ul>
<li>Let <span class="math inline">\(x = F^{-1}(y)\)</span>. Then, <span class="math inline">\(y = F(x)\)</span>, and,</li>
</ul>
<p><span class="math display">\[\begin{align*}
\left| \frac{dF^{-1}(y)}{dy}\right| =
\frac{dx}{dF(x)} =
\frac{dx}{f_X(x) dx} =
\frac{1}{f_X(x)} =
\frac{1}{f_X(F^{-1}(y))}
\end{align*}\]</span></p>
<ul>
<li>Which means <span class="math inline">\(f_Y(y) = 1\)</span> for <span class="math inline">\(y \in (0, 1)\)</span>, which is the standard uniform distribution</li>
</ul>
</section>
<section id="generating-random-variables" class="slide level2">
<h2>Generating Random Variables</h2>
<p>To generate a random variable with the desired distribution <span class="math inline">\(F\)</span>, we use the following algorithm:</p>
<ol type="1">
<li>Generate <span class="math inline">\(U \sim \text{Uniform}(0, 1)\)</span></li>
<li>Set <span class="math inline">\(X = F_X^{-1}(U)\)</span></li>
</ol>
</section>
<section id="generating-random-variables-1" class="slide level2">
<h2>Generating Random Variables</h2>
<p>Let our desired random variable have the standard normal distribution.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a></a>U <span class="ot">&lt;-</span> <span class="fu">runif</span>(<span class="dv">10000</span>)</span>
<span id="cb1-2"><a></a>Z <span class="ot">&lt;-</span> <span class="fu">qnorm</span>(U)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>

<img data-src="lecture-07_files/figure-revealjs/unnamed-chunk-3-1.png" width="960" class="r-stretch"></section>
<section id="generating-random-variables-2" class="slide level2">
<h2>Generating Random Variables</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a></a>U <span class="ot">&lt;-</span> <span class="fu">runif</span>(<span class="dv">10000</span>)</span>
<span id="cb2-2"><a></a>Z <span class="ot">&lt;-</span> <span class="fu">qexp</span>(U)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>

<img data-src="lecture-07_files/figure-revealjs/unnamed-chunk-5-1.png" width="960" class="r-stretch"></section>
<section id="benjamini-hochberg-false-discovery-rate-bh" class="slide level2">
<h2>Benjamini-Hochberg False Discovery Rate (BH)</h2>
<ul>
<li>Let FDR-BH be the expected value of FDP</li>
</ul>
<p><span class="math display">\[\begin{align*}
\text{FDR}
&amp;= \text{E}\left[Q\right] \\
&amp;= \text{E}\left[\frac{V}{R} \Big| R &gt; 0\right] \text{Pr}(R &gt; 0)
\end{align*}\]</span></p>
<ul>
<li>Let <span class="math inline">\(r_q\)</span> be a value that controls the FDR at level <span class="math inline">\(q \in (0, 1)\)</span> if</li>
</ul>
<p><span class="math display">\[
\text{FDR}(r_q) &lt; q
\]</span></p>
</section>
<section id="benjamini-hochberg-false-discovery-rate-bh-1" class="slide level2">
<h2>Benjamini-Hochberg False Discovery Rate (BH)</h2>
<p>Decision rule:</p>
<ul>
<li>Consider an ordered set of <span class="math inline">\(p\)</span>-values: <span class="math inline">\(\{p_i | i = 1, 2, ..., m\}\)</span>, where <span class="math inline">\(p_i \leq p_{i+1}\)</span></li>
<li>Let <span class="math inline">\(r_q = \text{max}\{i: p_i \leq (i/m)q\}\)</span></li>
<li>Then, let our rejection set be <span class="math inline">\(\{H_i\ \, | \, i \in (0, r_q)\}\)</span></li>
</ul>
</section>
<section id="benjamini-hochberg-false-discovery-rate-bh-2" class="slide level2">
<h2>Benjamini-Hochberg False Discovery Rate (BH)</h2>
<p>Another method:</p>
<ul>
<li>Let <span class="math inline">\(k = \max\left\{i: p_i \leq \frac{i}{m}q\right\}\)</span></li>
<li>Construct <span class="math inline">\(q_i^* = q(p_i) = \min_{i \leq k}\left(p_k \frac{m}{k}\right)\)</span></li>
<li>Reject <span class="math inline">\(H_i\)</span> if <span class="math inline">\(q_i^* \leq q\)</span></li>
</ul>
</section>
<section id="bayesian-fdr" class="slide level2">
<h2>Bayesian FDR</h2>
<ul>
<li>Let H be a binary random variable that takes the value 0 when a hypothesis is null, and 1 when a hypothesis is non-null</li>
<li>Let <span class="math inline">\(m\)</span> be the number of total hypotheses, <span class="math inline">\(m_0\)</span> be the number of null, and <span class="math inline">\(m - m_0\)</span> be the number of non-null</li>
<li><span class="math inline">\(\text{Pr}(H = 0) = \frac{m_0}{m} := \pi_0\)</span></li>
<li><span class="math inline">\(\text{Pr}(H = 1) = \frac{m - m_0}{m} = 1 - \pi_0 := \pi_1\)</span></li>
<li><span class="math inline">\(\pi_0\)</span> is the a priori probability of null for each hypothesis</li>
<li>In GWAS, <span class="math inline">\(\pi_0 &gt;&gt; \pi_1\)</span></li>
<li>Assume that <span class="math inline">\(H \sim \text{Bernoulli}(\pi_i)\)</span></li>
</ul>
</section>
<section id="bayesian-fdr-1" class="slide level2">
<h2>Bayesian FDR</h2>
<ul>
<li>Our observed test statistics then have the marginal distribution</li>
</ul>
<p><span class="math display">\[
p(z) = \text{Pr}(z | H = 0)\pi_0 + \text{Pr}(z | H = 1)\pi_1
\]</span></p>
<ul>
<li>Given a fixed rejection region <span class="math inline">\(A\)</span>, we denote its probability under the null or marginal marginal distributions as</li>
</ul>
<p><span class="math display">\[
F_0(A) = \text{Pr}(Z \in A | H_0) = \int_A p_0(z)dz
\]</span></p>
<p><span class="math display">\[
F(A) = \text{Pr}(Z \in A) = \int_A p(z)dz
\]</span></p>
</section>
<section id="bayesian-fdr-2" class="slide level2">
<h2>Bayesian FDR</h2>
<ul>
<li>Define the probability of a false discovery as</li>
</ul>
<p><span class="math display">\[\begin{align*}
\text{BFDR}(A)
&amp;= \text{Pr}(H = 0 | z \in A) \\[.5em]
&amp;= \frac{\text{Pr}(z \in A |H = 0) \text{Pr}(H = 0)}{\text{Pr}(Z \in A)} \\[.5em]
&amp;= \frac{F_0(A)\pi_0}{F(A)}
\end{align*}\]</span></p>
<ul>
<li>Expected proportion of the false discoveries among all features considered significant</li>
</ul>
</section>
<section id="how-to-compute-bfdr" class="slide level2">
<h2>How to compute BFDR?</h2>
<ul>
<li>Note that to compute BFDR, we require knowledge of
<ul>
<li><span class="math inline">\(\pi_0\)</span>, the a priori probability of a null hypothesis</li>
<li><span class="math inline">\(F_0\)</span></li>
<li><span class="math inline">\(F\)</span></li>
</ul></li>
<li>Empirical bayes approach: estimate these quantities from the data, and then approximate the BFDR</li>
</ul>
</section>
<section id="empirical-bayes-estimation" class="slide level2">
<h2>Empirical Bayes Estimation</h2>
<ul>
<li>Consider that we have the results of a <span class="math inline">\(t\)</span>-test for each of the <span class="math inline">\(m\)</span> genes</li>
<li>Assume <span class="math inline">\(F_0\)</span> to be the standard normal distribution by transforming the <span class="math inline">\(p\)</span>-values into <span class="math inline">\(z\)</span>-values, set <span class="math inline">\(\pi_0\)</span> to 1, and estimate <span class="math inline">\(F\)</span> via:</li>
</ul>
<p><span class="math display">\[\begin{align*}
F(A) = \text{Pr}(Z \in A) &amp;= \int_A p(z)dz \\
&amp;= \int_{-\infty}^{\infty} I(z \in A)p(z)dz \\
&amp;= \text{E}[I(z \in A)]
\end{align*}\]</span></p>
<p>Which we empirically estimate via</p>
<p><span class="math display">\[
\hat{F}(A) = \hat{\text{Pr}}(Z \in A) = \frac{1}{m}\sum_{i=1}^m I(z_i \in A)
\]</span></p>
</section>
<section id="empirical-bayes-estimation-1" class="slide level2">
<h2>Empirical Bayes Estimation</h2>
<p>Plugging that result in our empirical Bayes estimator, we achieve:</p>
<p><span class="math display">\[
\hat{BFDR}(A) = \frac{\hat{pi}_0 F_0(A)}{\hat{F}(A)}
\]</span></p>
<ul>
<li>Posterior probability of false discoveries among the features whose <span class="math inline">\(z\)</span>-values fall in the rejection region <span class="math inline">\(A\)</span></li>
</ul>
</section>
<section id="using-p-values-as-data" class="slide level2">
<h2>Using <span class="math inline">\(p\)</span>-values as data</h2>
<ul>
<li>If we set a threshold <span class="math inline">\(p_t\)</span>, where <span class="math inline">\(p_i \leq p_t \implies p_i \textsf{ is significant}\)</span>, we can then write</li>
</ul>
<p><span class="math display">\[
F_0(p_t) = \text{Pr}(P_i \leq p_t \, |\, H_i = 0) = \int_0^{p_t} \text{Unif}(0, 1) = p_t
\]</span></p>
<p><span class="math display">\[
F(p_t) = \text{Pr}(P \leq p_t)
\]</span></p>
<p><span class="math display">\[
\hat{F}(p_t) = \hat{\text{Pr}}(P \leq p_t) = \frac{1}{m}\sum_{i=1}^m I(p_i \leq p_t)
\]</span></p>
<p>and then estimate BFDR via</p>
<p><span class="math display">\[
\hat{BFDR}(p_t) = \frac{\hat{\pi}_0 p_t}{\frac{1}{m}\sum_{i=1}^m I(p_i \leq p_t)}
\]</span></p>
</section>
<section id="example" class="slide level2">
<h2>Example</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a></a>snps <span class="ot">&lt;-</span> <span class="dv">10000</span></span>
<span id="cb3-2"><a></a></span>
<span id="cb3-3"><a></a><span class="co"># mixture of null and alternative p-values</span></span>
<span id="cb3-4"><a></a>p_values <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">runif</span>(snps <span class="sc">*</span> <span class="fl">0.9</span>), <span class="fu">runif</span>(snps <span class="sc">*</span> <span class="fl">0.1</span>, <span class="dv">0</span>, <span class="fl">0.01</span>))</span>
<span id="cb3-5"><a></a>p_t <span class="ot">&lt;-</span> <span class="fl">0.05</span></span>
<span id="cb3-6"><a></a>significant <span class="ot">&lt;-</span> p_values <span class="sc">&lt;=</span> p_t</span>
<span id="cb3-7"><a></a>pi_0 <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">length.out =</span> <span class="dv">100</span>)</span>
<span id="cb3-8"><a></a>bfdr <span class="ot">&lt;-</span> (pi_0 <span class="sc">*</span> p_t) <span class="sc">/</span> <span class="fu">mean</span>(significant)</span>
<span id="cb3-9"><a></a></span>
<span id="cb3-10"><a></a><span class="fu">ggplot</span>(<span class="fu">data.frame</span>(pi_0, bfdr), <span class="fu">aes</span>(<span class="at">x =</span> pi_0, <span class="at">y =</span> bfdr)) <span class="sc">+</span></span>
<span id="cb3-11"><a></a>  <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb3-12"><a></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="fu">expression</span>(pi[<span class="dv">0</span>]), <span class="at">y =</span> <span class="st">"BFDR"</span>, <span class="at">title =</span> <span class="st">"Bayesian False Discovery Rate"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>

</div>
<img data-src="lecture-07_files/figure-revealjs/unnamed-chunk-6-1.png" width="960" class="r-stretch"></section>
<section id="local-false-discovery-rates" class="slide level2">
<h2>Local False Discovery Rates</h2>
<ul>
<li>Note that FDR-BH and BFDR are interested in identifying a <em>discovery set</em></li>
<li>Instead, consider the probability of a false positive for a specific SNP <span class="math inline">\(j\)</span> given that we observed <span class="math inline">\(y_j\)</span>:</li>
</ul>
<p><span class="math display">\[\begin{align*}
\text{fdr}(y_j)
&amp;= \text{Pr}(H_j = 0 \, | \, y_j) \\[.5em]
&amp;= \frac{p_0(y_j \, | \, H_j = 0) \pi_0}{p(y_j)} \qquad \textsf{(via Bayes' theorem)}\\[.5em]
&amp;= \frac{p_0(y_j \, | \, H_j = 0) \pi_0}{p_0(y_j \, | \, H_j = 0)\pi_0 + p_1(y_j \, | \, H_j = 1)(1 - \pi_0)}
\end{align*}\]</span></p>
<ul>
<li>Given <span class="math inline">\(z\)</span>-values, <span class="math inline">\(p_0 = \phi(y_j)\)</span>, <span class="math inline">\(\pi_0 := 1\)</span></li>
<li>The denominator can be estimated using MLE or MCMC</li>
</ul>
</section>
<section id="q-values" class="slide level2">
<h2><span class="math inline">\(q\)</span>-values</h2>
<ul>
<li>Formally,</li>
</ul>
<p><span class="math display">\[
q(t) = \min_{\{A:t\in A\}}\left[pBFDR(A)\right]
\]</span></p>
<ul>
<li>More intuitively, the <span class="math inline">\(q\)</span>-value is the smallest pBFDR over all regions that reject <span class="math inline">\(H_i\)</span></li>
<li>Expected proportion of false positives among all features as or more extreme than the one observed</li>
<li>Empirical bayes estimate: <span class="math inline">\(q_i = q_i^* \hat{\pi}_0\)</span></li>
<li>If the prior probability a randomly chosen hypothesis is null is 1, then the two values above are equal</li>
</ul>
</section>
<section id="q-values-vs-alpha-values" class="slide level2">
<h2><span class="math inline">\(q\)</span>-values vs <span class="math inline">\(\alpha\)</span>-values</h2>
<ul>
<li><span class="math inline">\(\alpha\)</span>-value = <span class="math inline">\(\text{Pr}(\textsf{Test Positive} | H_0) =\)</span> False positive rate</li>
<li><span class="math inline">\(q\)</span>-value = <span class="math inline">\(\text{Pr}(H_0 | \textsf{Test Positive}) =\)</span> False discovery rate</li>
<li><span class="math inline">\(q = .05 \implies\)</span> “An estimated 5% of the results in the discovery set are false discoveries”</li>
<li><span class="math inline">\(\alpha = .05 \implies\)</span> “An estimated 5% of all the tests performed are false discoveries”</li>
</ul>
<p><a href="https://bioconductor.org/packages/release/bioc/html/qvalue.html"><code>qvalue</code> package in <code>R</code></a></p>
</section>
<section id="fully-bayesian-fdr" class="slide level2">
<h2>Fully Bayesian FDR</h2>
<p>We observe test statistics <span class="math inline">\(Z_1,\dots,Z_m\)</span> (e.g., GWAS <span class="math inline">\(z\)</span>‑scores) and introduce latent indicators <span class="math inline">\(H_i\in\{0,1\}\)</span> with <span class="math display">\[
H_i=0 \text{ (null)},\quad H_i=1 \text{ (non‑null)}
\]</span> <span class="math display">\[
\Pr(H_i=0)=\pi_0,\ \Pr(H_i=1)=1-\pi_0.
\]</span></p>
</section>
<section id="fully-bayesian-fdr-1" class="slide level2">
<h2>Fully Bayesian FDR</h2>
<p><strong>Likelihood (symmetric, zero‑mean alternative):</strong></p>
<p><span class="math display">\[\begin{align*}
&amp;Z_i \mid H_i=0 \sim \mathcal N(0,\tau^2) \\[.5em]
&amp;Z_i \mid H_i=1 \sim \mathcal N(0,\sigma^2)
\end{align*}\]</span></p>
<p>(Allow <span class="math inline">\(\tau\neq 1\)</span> to capture empirical‑null inflation/deflation.)</p>
</section>
<section id="fully-bayesian-fdr-2" class="slide level2">
<h2>Fully Bayesian FDR</h2>
<p><strong>Priors (GWAS‑motivated):</strong></p>
<p><span class="math display">\[\begin{align*}
&amp;\pi_0 \sim \mathrm{Beta}(20,1) \\
&amp;\sigma \sim \mathrm{LogNormal}(0,0.5^2) \\
&amp;\tau \sim \mathrm{LogNormal}(0,0.05^2)
\end{align*}\]</span></p>
</section>
<section id="fully-bayesian-fdr-3" class="slide level2">
<h2>Fully Bayesian FDR</h2>
<p><strong>Posterior</strong> over parameters <span class="math inline">\(\theta=(\pi_0,\sigma,\tau)\)</span>: <span class="math display">\[
p(\theta\mid \mathbf z)\ \propto\ p(\theta)\ \prod_{i=1}^m\Big[\pi_0\,\phi_{\tau}(z_i) + (1-\pi_0)\,\phi_{\sigma}(z_i)\Big],
\]</span> with <span class="math inline">\(\phi_s\)</span> the <span class="math inline">\(\mathcal N(0,s^2)\)</span> pdf.</p>
</section>
<section id="fully-bayesian-fdr-4" class="slide level2">
<h2>Fully Bayesian FDR</h2>
<p><strong>Local false discovery rate (per feature), integrating parameter uncertainty:</strong> <span class="math display">\[
\operatorname{lfdr}_i
=\Pr(H_i=0\mid z_i,\mathbf z)
=\mathbb E_{\theta\mid\mathbf z}\!\left[
\frac{\pi_0\,\phi_{\tau}(z_i)}
{\pi_0\,\phi_{\tau}(z_i)+(1-\pi_0)\,\phi_{\sigma}(z_i)}
\right].
\]</span></p>
</section>
<section id="fully-bayesian-fdr-5" class="slide level2">
<h2>Fully Bayesian FDR</h2>
<p>Let a rejection rule be <span class="math inline">\(r_i(t)=\mathbf 1\{|z_i|\ge z_t\}\)</span> where <span class="math inline">\(z_t=\Phi^{-1}(1-t/2)\)</span>. Define <span class="math display">\[
R(t)=\sum_i r_i(t),\quad V(t)=\sum_i r_i(t)\,(1-H_i),\quad \mathrm{FDP}(t)=\frac{V(t)}{\max(1,R(t))}.
\]</span></p>
<p>A <strong>fully Bayesian FDR</strong> at threshold <span class="math inline">\(t\)</span> is the posterior expectation: <span class="math display">\[
\boxed{\ \mathrm{FDR}(t)\ :=\ \mathbb E\!\left[\frac{V(t)}{\max(1,R(t))}\ \middle|\ \mathbf z\right]
\ \approx\ \frac{1}{S}\sum_{s=1}^S \frac{\sum_i r_i(t)\,\operatorname{lfdr}_i^{(s)} }{\max(1,\sum_i r_i(t))}\ },
\]</span> where <span class="math inline">\(\operatorname{lfdr}_i^{(s)}\)</span> is the null posterior probability at draw <span class="math inline">\(\theta^{(s)}\)</span>.</p>
</section>
<section id="fully-bayesian-fdr-6" class="slide level2">
<h2>Fully Bayesian FDR</h2>
<p><strong>Bayes‑optimal discovery set at level</strong> <span class="math inline">\(q\)</span>.</p>
<ul>
<li>Rank features by <span class="math inline">\(\widehat{\operatorname{lfdr}}_i=\mathbb E[\Pr(H_i=0\mid z_i,\theta)\mid \mathbf z]\)</span></li>
<li>Let <span class="math inline">\(\bar\ell(k)\)</span> be the running mean of the first <span class="math inline">\(k\)</span> lfdrs.</li>
<li>Choose <span class="math inline">\(\hat k=\max\{k:\bar\ell(k)\le q\}\)</span> and take those <span class="math inline">\(\hat k\)</span> features
<ul>
<li>This minimizes posterior expected false discoveries among all sets of size <span class="math inline">\(\hat k\)</span>.</li>
</ul></li>
</ul>
</section>
<section id="simulate-gwas-like-z-scores" class="slide level2">
<h2>Simulate GWAS-like <span class="math inline">\(z\)</span>-scores</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-4"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-4-1" class="code-annotation-target"><a href="#annotated-cell-4-1"></a>m <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="annotated-cell-4-2"><a href="#annotated-cell-4-2"></a>pi0_true   <span class="ot">&lt;-</span> <span class="fl">0.95</span></span>
<span id="annotated-cell-4-3"><a href="#annotated-cell-4-3"></a>sigma_true <span class="ot">&lt;-</span> <span class="fl">1.8</span></span>
<span id="annotated-cell-4-4"><a href="#annotated-cell-4-4"></a>tau_true   <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="annotated-cell-4-5"><a href="#annotated-cell-4-5"></a></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-4-6" class="code-annotation-target"><a href="#annotated-cell-4-6"></a>H  <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(m, <span class="dv">1</span>, <span class="dv">1</span> <span class="sc">-</span> pi0_true)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-4-7" class="code-annotation-target"><a href="#annotated-cell-4-7"></a>z <span class="ot">&lt;-</span> <span class="fu">numeric</span>(m)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-4-8" class="code-annotation-target"><a href="#annotated-cell-4-8"></a>z[H<span class="sc">==</span><span class="dv">0</span>] <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="fu">sum</span>(H<span class="sc">==</span><span class="dv">0</span>), <span class="dv">0</span>, tau_true)</span>
<span id="annotated-cell-4-9"><a href="#annotated-cell-4-9"></a>z[H<span class="sc">==</span><span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="fu">sum</span>(H<span class="sc">==</span><span class="dv">1</span>), <span class="dv">0</span>, sigma_true)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="5" onclick="event.preventDefault();">5</a><span id="annotated-cell-4-10" class="code-annotation-target"><a href="#annotated-cell-4-10"></a>p <span class="ot">&lt;-</span> <span class="dv">2</span><span class="sc">*</span><span class="fu">pnorm</span>(<span class="sc">-</span><span class="fu">abs</span>(z))</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-4" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="1,2,3,4" data-code-annotation="1">Choose parameters to control sparsity and scales.</span>
</dd>
<dt data-target-cell="annotated-cell-4" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="6" data-code-annotation="2">Encode the unknown truth via latent indicators in a two-groups model: <span class="math inline">\(H_i \sim \mathrm{Bernoulli}(1-\pi_0)\)</span>, so that on average <span class="math inline">\((1-\pi_0)m\)</span> tests are non-null.</span>
</dd>
<dt data-target-cell="annotated-cell-4" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="7" data-code-annotation="3">Impose directional neutrality for two-sided testing by using a zero-centered alternative (no global shift of effects).</span>
</dd>
<dt data-target-cell="annotated-cell-4" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="8,9" data-code-annotation="4">Generate <span class="math inline">\(z\)</span>-scores under the two-groups mixture. Null <span class="math inline">\(z\)</span>-scores are centered at 0 with variance <span class="math inline">\(\tau^2\)</span>; non-null <span class="math inline">\(z\)</span>-scores come from a zero-centered wider normal to induce heavy tails.</span>
</dd>
<dt data-target-cell="annotated-cell-4" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="10" data-code-annotation="5">Map <span class="math inline">\(z\)</span>-scores to two-sided <span class="math inline">\(p\)</span>-values under the standard normal null (<span class="math inline">\(\tau=1\)</span>): <span class="math inline">\(p_i = 2\,\Phi(-|z_i|)\)</span>.</span>
</dd>
</dl>
</div>
</div>
</section>
<section id="simulate-gwas-like-z-scores-1" class="slide level2">
<h2>Simulate GWAS-like <span class="math inline">\(z\)</span>-scores</h2>

<img data-src="lecture-07_files/figure-revealjs/unnamed-chunk-8-1.png" width="960" class="r-stretch"></section>
<section id="bayesian-two-groups-fit-stan" class="slide level2">
<h2>Bayesian Two-Groups Fit (Stan)</h2>
<ul>
<li>Model the marginal <span class="math inline">\(z\)</span>-score density as a two-groups mixture:</li>
</ul>
<p><span class="math display">\[
p(z_i \mid \theta) = \pi_0\, \varphi_{\tau}(z_i) + (1-\pi_0)\, \varphi_{\sigma}(z_i),
\]</span></p>
<p>where <span class="math inline">\(\varphi_{\sigma}(\cdot)\)</span> is the <span class="math inline">\(\mathcal N(0,\sigma^2)\)</span> pdf. Priors encode GWAS sparsity and standard-null assumptions:</p>
<p><span class="math display">\[
\pi_0 \sim \mathrm{Beta}(20,1),\quad \sigma \sim \mathrm{LogNormal}(0,0.5),\quad \tau \sim \mathrm{LogNormal}(0,0.05).
\]</span></p>
<ul>
<li>This matches the lecture: <span class="math inline">\(\pi_0 \gg \pi_1\)</span> (sparse signals) and a tight null around 1 while allowing mild inflation.</li>
</ul>
</section>
<section id="posterior-summary" class="slide level2">
<h2>Posterior Summary</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a></a>fit<span class="sc">$</span><span class="fu">summary</span>(<span class="fu">c</span>(<span class="st">"pi0"</span>,<span class="st">"sigma"</span>,<span class="st">"tau"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb4-2"><a></a>  <span class="fu">select</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>) <span class="sc">|&gt;</span></span>
<span id="cb4-3"><a></a>  knitr<span class="sc">::</span><span class="fu">kable</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<table class="caption-top" style="width:100%;">
<colgroup>
<col style="width: 13%">
<col style="width: 14%">
<col style="width: 14%">
<col style="width: 14%">
<col style="width: 14%">
<col style="width: 14%">
<col style="width: 14%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">variable</th>
<th style="text-align: right;">mean</th>
<th style="text-align: right;">median</th>
<th style="text-align: right;">sd</th>
<th style="text-align: right;">mad</th>
<th style="text-align: right;">q5</th>
<th style="text-align: right;">q95</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">pi0</td>
<td style="text-align: right;">0.9299875</td>
<td style="text-align: right;">0.9345666</td>
<td style="text-align: right;">0.0313251</td>
<td style="text-align: right;">0.0280454</td>
<td style="text-align: right;">0.8676778</td>
<td style="text-align: right;">0.9703593</td>
</tr>
<tr class="even">
<td style="text-align: left;">sigma</td>
<td style="text-align: right;">1.7883926</td>
<td style="text-align: right;">1.7657928</td>
<td style="text-align: right;">0.1873689</td>
<td style="text-align: right;">0.1780550</td>
<td style="text-align: right;">1.5202518</td>
<td style="text-align: right;">2.1319676</td>
</tr>
<tr class="odd">
<td style="text-align: left;">tau</td>
<td style="text-align: right;">0.9907577</td>
<td style="text-align: right;">0.9912889</td>
<td style="text-align: right;">0.0178861</td>
<td style="text-align: right;">0.0172593</td>
<td style="text-align: right;">0.9601886</td>
<td style="text-align: right;">1.0194893</td>
</tr>
</tbody>
</table>
</div>
</div>
<ul>
<li><span class="math inline">\(\pi_0\)</span> is the estimated proportion of nulls</li>
<li><span class="math inline">\(\sigma\)</span> is the estimated standard deviation of the alternative (zero-centered)</li>
<li><span class="math inline">\(\tau\)</span> is the estimated standard deviation of the null</li>
</ul>
</section>
<section id="posterior-predictive-check" class="slide level2">
<h2>Posterior Predictive Check</h2>
<ul>
<li>Posterior predictive density under parameters <span class="math inline">\(\theta=(\pi_0,\sigma,\tau)\)</span> is</li>
</ul>
<p><span class="math display">\[
p_\theta(z) = \pi_0\, \varphi_{\tau}(z) + (1-\pi_0)\, \varphi_{\sigma}(z).
\]</span></p>
<ul>
<li>Overlay the posterior mean predictive density on the <span class="math inline">\(z\)</span> histogram.</li>
</ul>

<img data-src="lecture-07_files/figure-revealjs/unnamed-chunk-11-1.png" width="960" class="r-stretch"></section>
<section id="posterior-local-fdr" class="slide level2">
<h2>Posterior Local fdr</h2>
<ul>
<li>The local false discovery rate for SNP <span class="math inline">\(i\)</span> is the posterior probability of null given <span class="math inline">\(z_i\)</span> and parameters:</li>
</ul>
<p><span class="math display">\[
\operatorname{lfdr}_i(\theta) = \Pr(H_i=0 \mid z_i,\theta) = \frac{\pi_0\,\varphi_{\tau}(z_i)}{\pi_0\,\varphi_{\tau}(z_i) + (1-\pi_0)\,\varphi_{\sigma}(z_i)}.
\]</span></p>
</section>
<section id="posterior-local-fdr-1" class="slide level2">
<h2>Posterior Local fdr</h2>
<ul>
<li>Our Stan model returns <code>lfdr</code> in generated quantities. Compute posterior means <span class="math inline">\(\hat{\operatorname{lfdr}}_i\)</span> and visualize.</li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using method = 'gam' and formula = 'y ~ s(x, bs = "cs")'</code></pre>
</div>

</div>


<img data-src="lecture-07_files/figure-revealjs/unnamed-chunk-12-1.png" width="960" class="r-stretch"></section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">

</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
          const outerScaffold = trigger.parentElement.cloneNode(true);
          const codeEl = outerScaffold.querySelector('code');
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp('/' + window.location.host + '/');
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    

</body></html>