<!DOCTYPE html>
<html lang="en"><head>
<link href="../course_logo.png" rel="icon" type="image/png">
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-html/tabby.min.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.8.25">

  <title>Statistical Genetics – Lecture 05: GWAS in R</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    html { -webkit-text-size-adjust: 100%; }
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; font-weight: bold; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="../site_libs/revealjs/dist/theme/quarto-264c76fdbb68a1a4f90edbec2b556612.css">
  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.css">
  <link href="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Lecture 05: GWAS in R</h1>
  <p class="subtitle">PUBH 8878, Statistical Genetics</p>

<div class="quarto-title-authors">
</div>

</section>
<section id="step-1-setup" class="slide level2">
<h2>Step 1: Setup</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-1"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-1" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-1-1" class="code-annotation-target"><a href="#annotated-cell-1-1"></a><span class="fu">library</span>(SNPRelate)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-1" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-1-2" class="code-annotation-target"><a href="#annotated-cell-1-2"></a><span class="fu">library</span>(GENESIS)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-1" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-1-3" class="code-annotation-target"><a href="#annotated-cell-1-3"></a><span class="fu">library</span>(GWASTools)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-1" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-1-4" class="code-annotation-target"><a href="#annotated-cell-1-4"></a><span class="fu">library</span>(qqman)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-1" data-target-annotation="5" onclick="event.preventDefault();">5</a><span id="annotated-cell-1-5" class="code-annotation-target"><a href="#annotated-cell-1-5"></a><span class="fu">library</span>(data.table)</span>
<span id="annotated-cell-1-6"><a href="#annotated-cell-1-6"></a><span class="fu">library</span>(ggplot2)</span>
<span id="annotated-cell-1-7"><a href="#annotated-cell-1-7"></a><span class="fu">library</span>(dplyr)</span>
<span id="annotated-cell-1-8"><a href="#annotated-cell-1-8"></a><span class="fu">set.seed</span>(<span class="dv">8878</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-1" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-1" data-code-lines="1" data-code-annotation="1">SNPRelate (Zheng et al., 2012) for GDS file handling and genotype data management</span>
</dd>
<dt data-target-cell="annotated-cell-1" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-1" data-code-lines="2" data-code-annotation="2">GENESIS for genome-wide association analysis and relatedness estimation</span>
</dd>
<dt data-target-cell="annotated-cell-1" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-1" data-code-lines="3" data-code-annotation="3">GWASTools for genotype data management and quality control</span>
</dd>
<dt data-target-cell="annotated-cell-1" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-1" data-code-lines="4" data-code-annotation="4">qqman for creating Q-Q and Manhattan plots</span>
</dd>
<dt data-target-cell="annotated-cell-1" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-1" data-code-lines="5" data-code-annotation="5">data.table for efficient data manipulation</span>
</dd>
</dl>
</div>
</div>
</section>
<section id="step-1-setup-data-access" class="slide level2">
<h2>Step 1: Setup &amp; data access</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-2" data-code-line-numbers="1"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-2-1" class="code-annotation-target"><a href="#annotated-cell-2-1"></a>gdsfile <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"HapMap_ASW_MXL_geno.gds"</span>, <span class="at">package =</span> <span class="st">"GENESIS"</span>)</span>
<span id="annotated-cell-2-2"><a href="#annotated-cell-2-2"></a></span>
<span id="annotated-cell-2-3"><a href="#annotated-cell-2-3"></a>g <span class="ot">&lt;-</span> SNPRelate<span class="sc">::</span><span class="fu">snpgdsOpen</span>(gdsfile)</span>
<span id="annotated-cell-2-4"><a href="#annotated-cell-2-4"></a></span>
<span id="annotated-cell-2-5"><a href="#annotated-cell-2-5"></a>samp.id <span class="ot">&lt;-</span> <span class="fu">read.gdsn</span>(<span class="fu">index.gdsn</span>(g, <span class="st">"sample.id"</span>))</span>
<span id="annotated-cell-2-6"><a href="#annotated-cell-2-6"></a>snp.id <span class="ot">&lt;-</span> <span class="fu">read.gdsn</span>(<span class="fu">index.gdsn</span>(g, <span class="st">"snp.id"</span>))</span>
<span id="annotated-cell-2-7"><a href="#annotated-cell-2-7"></a>chr <span class="ot">&lt;-</span> <span class="fu">read.gdsn</span>(<span class="fu">index.gdsn</span>(g, <span class="st">"snp.chromosome"</span>))</span>
<span id="annotated-cell-2-8"><a href="#annotated-cell-2-8"></a>pos <span class="ot">&lt;-</span> <span class="fu">read.gdsn</span>(<span class="fu">index.gdsn</span>(g, <span class="st">"snp.position"</span>))</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-2" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="1" data-code-annotation="1">Locate the example GDS file included with the GENESIS package</span>
</dd>
</dl>
</div>
</div>
</section>
<section id="step-1-setup-data-access-1" class="slide level2">
<h2>Step 1: Setup &amp; data access</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-3" data-code-line-numbers="1-3"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-3" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-3-1" class="code-annotation-target"><a href="#annotated-cell-3-1"></a>gdsfile <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"HapMap_ASW_MXL_geno.gds"</span>, <span class="at">package =</span> <span class="st">"GENESIS"</span>)</span>
<span id="annotated-cell-3-2"><a href="#annotated-cell-3-2"></a></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-3" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-3-3" class="code-annotation-target"><a href="#annotated-cell-3-3"></a>g <span class="ot">&lt;-</span> SNPRelate<span class="sc">::</span><span class="fu">snpgdsOpen</span>(gdsfile)</span>
<span id="annotated-cell-3-4"><a href="#annotated-cell-3-4"></a></span>
<span id="annotated-cell-3-5"><a href="#annotated-cell-3-5"></a>samp.id <span class="ot">&lt;-</span> <span class="fu">read.gdsn</span>(<span class="fu">index.gdsn</span>(g, <span class="st">"sample.id"</span>))</span>
<span id="annotated-cell-3-6"><a href="#annotated-cell-3-6"></a>snp.id <span class="ot">&lt;-</span> <span class="fu">read.gdsn</span>(<span class="fu">index.gdsn</span>(g, <span class="st">"snp.id"</span>))</span>
<span id="annotated-cell-3-7"><a href="#annotated-cell-3-7"></a>chr <span class="ot">&lt;-</span> <span class="fu">read.gdsn</span>(<span class="fu">index.gdsn</span>(g, <span class="st">"snp.chromosome"</span>))</span>
<span id="annotated-cell-3-8"><a href="#annotated-cell-3-8"></a>pos <span class="ot">&lt;-</span> <span class="fu">read.gdsn</span>(<span class="fu">index.gdsn</span>(g, <span class="st">"snp.position"</span>))</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-3" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-3" data-code-lines="1" data-code-annotation="1">Locate the example GDS file included with the GENESIS package</span>
</dd>
<dt data-target-cell="annotated-cell-3" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-3" data-code-lines="3" data-code-annotation="2">Open the GDS file for random access to genotype data</span>
</dd>
</dl>
</div>
</div>
</section>
<section id="step-1-setup-data-access-2" class="slide level2">
<h2>Step 1: Setup &amp; data access</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-4"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-4-1" class="code-annotation-target"><a href="#annotated-cell-4-1"></a>gdsfile <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"HapMap_ASW_MXL_geno.gds"</span>, <span class="at">package =</span> <span class="st">"GENESIS"</span>)</span>
<span id="annotated-cell-4-2"><a href="#annotated-cell-4-2"></a></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-4-3" class="code-annotation-target"><a href="#annotated-cell-4-3"></a>g <span class="ot">&lt;-</span> SNPRelate<span class="sc">::</span><span class="fu">snpgdsOpen</span>(gdsfile)</span>
<span id="annotated-cell-4-4"><a href="#annotated-cell-4-4"></a></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-4" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-4-5" class="code-annotation-target"><a href="#annotated-cell-4-5"></a>samp.id <span class="ot">&lt;-</span> <span class="fu">read.gdsn</span>(<span class="fu">index.gdsn</span>(g, <span class="st">"sample.id"</span>))</span>
<span id="annotated-cell-4-6"><a href="#annotated-cell-4-6"></a>snp.id <span class="ot">&lt;-</span> <span class="fu">read.gdsn</span>(<span class="fu">index.gdsn</span>(g, <span class="st">"snp.id"</span>))</span>
<span id="annotated-cell-4-7"><a href="#annotated-cell-4-7"></a>chr <span class="ot">&lt;-</span> <span class="fu">read.gdsn</span>(<span class="fu">index.gdsn</span>(g, <span class="st">"snp.chromosome"</span>))</span>
<span id="annotated-cell-4-8"><a href="#annotated-cell-4-8"></a>pos <span class="ot">&lt;-</span> <span class="fu">read.gdsn</span>(<span class="fu">index.gdsn</span>(g, <span class="st">"snp.position"</span>))</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-4" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="1" data-code-annotation="1">Locate the example GDS file included with the GENESIS package</span>
</dd>
<dt data-target-cell="annotated-cell-4" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="3" data-code-annotation="2">Open the GDS file for random access to genotype data</span>
</dd>
<dt data-target-cell="annotated-cell-4" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-4" data-code-lines="5,6,7,8" data-code-annotation="3">Read sample IDs, SNP IDs, chromosome numbers, and positions from the GDS file</span>
</dd>
</dl>
</div>
</div>
</section>
<section>
<section id="step-2-pre-imputation-quality-control-qc-filtering" class="title-slide slide level1 center">
<h1>Step 2: Pre-imputation Quality Control (QC) &amp; filtering</h1>

</section>
<section id="genotype-quality-control" class="slide level2">
<h2>Genotype quality control</h2>
<ol type="1">
<li>Filtering for minor allele frequency</li>
<li>Filtering SNPs based on violation of HWE</li>
</ol>
</section>
<section id="filtering-for-minor-allele-frequency-maf" class="slide level2">
<h2>Filtering for minor allele frequency (MAF)</h2>
<ul>
<li>SNP-chips often determine genotypes based on intensities for each allele</li>
<li>When MAF is low, genotype clusters can overlap, leading to genotyping errors</li>
</ul>
</section>
<section id="filtering-for-minor-allele-frequency-maf-1" class="slide level2">
<h2>Filtering for minor allele frequency (MAF)</h2>

<img data-src="lecture-05_files/figure-revealjs/unnamed-chunk-5-1.png" width="960" class="r-stretch"></section>
<section id="filtering-for-minor-allele-frequency-maf-2" class="slide level2">
<h2>Filtering for minor allele frequency (MAF)</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-5"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-5" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-5-1" class="code-annotation-target"><a href="#annotated-cell-5-1"></a>af <span class="ot">&lt;-</span> <span class="fu">snpgdsSNPRateFreq</span>(g)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-5" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-5-2" class="code-annotation-target"><a href="#annotated-cell-5-2"></a>maf <span class="ot">&lt;-</span> af<span class="sc">$</span>MinorFreq</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-5" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-5-3" class="code-annotation-target"><a href="#annotated-cell-5-3"></a>miss_var <span class="ot">&lt;-</span> af<span class="sc">$</span>MissingRate</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-5" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-5" data-code-lines="1" data-code-annotation="1">Compute allele frequencies and missing rates for all SNPs</span>
</dd>
<dt data-target-cell="annotated-cell-5" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-5" data-code-lines="2" data-code-annotation="2">Extract minor allele frequencies (MAF)</span>
</dd>
<dt data-target-cell="annotated-cell-5" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-5" data-code-lines="3" data-code-annotation="3">Extract variant missingness rates</span>
</dd>
</dl>
</div>
</div>
</section>
<section id="filtering-for-minor-allele-frequency-maf-3" class="slide level2">
<h2>Filtering for minor allele frequency (MAF)</h2>

<img data-src="lecture-05_files/figure-revealjs/unnamed-chunk-7-1.png" width="960" class="r-stretch"></section>
<section id="variant-missingness" class="slide level2">
<h2>Variant missingness</h2>
<ul>
<li>Missingness can indicate genotyping errors or batch effects</li>
<li>Missingness can be non-random with respect to phenotype or ancestry</li>
<li>Typically filter variants with &gt;2% missingness</li>
</ul>
</section>
<section id="variant-missingness-1" class="slide level2">
<h2>Variant missingness</h2>

<img data-src="lecture-05_files/figure-revealjs/unnamed-chunk-8-1.png" width="960" class="r-stretch"></section>
<section id="filtering-snps-based-on-violation-of-hwe" class="slide level2">
<h2>Filtering SNPs based on violation of HWE</h2>
<ul>
<li>Deviations (rejections of <span class="math inline">H_0</span>) can indicate genotyping errors, population structure, or selection</li>
<li>Typically tested in controls or unrelateds for case-control studies</li>
</ul>
</section>
<section id="filtering-snps-based-on-violation-of-hwe-1" class="slide level2">
<h2>Filtering SNPs based on violation of HWE</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a></a>hwe_p <span class="ot">&lt;-</span> <span class="fu">snpgdsHWE</span>(g, <span class="at">snp.id =</span> snp.id)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>

<img data-src="lecture-05_files/figure-revealjs/hwe-plot-1.png" width="960" class="r-stretch"></section>
<section id="sample-level-missingness-heterozygosity" class="slide level2">
<h2>Sample-level missingness &amp; heterozygosity</h2>
<ul>
<li>Samples with high missingness may have poor DNA quality or technical issues</li>
<li>Samples with extreme heterozygosity rates may be contaminated or mislabelled</li>
</ul>
</section>
<section id="sample-level-missingness-heterozygosity-1" class="slide level2">
<h2>Sample-level missingness &amp; heterozygosity</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a></a>samp_miss <span class="ot">&lt;-</span> <span class="fu">snpgdsSampMissRate</span>(g)</span>
<span id="cb2-2"><a></a></span>
<span id="cb2-3"><a></a>G <span class="ot">&lt;-</span> <span class="fu">snpgdsGetGeno</span>(g, <span class="at">with.id =</span> <span class="cn">TRUE</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Genotype matrix: 20000 SNPs X 173 samples</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a></a>geno_mat <span class="ot">&lt;-</span> G<span class="sc">$</span>genotype</span>
<span id="cb4-2"><a></a>geno_mat[geno_mat <span class="sc">==</span> <span class="dv">3</span>] <span class="ot">&lt;-</span> <span class="cn">NA_integer_</span>  <span class="co"># SNPRelate encodes missing genotypes as 3</span></span>
<span id="cb4-3"><a></a>het_rate <span class="ot">&lt;-</span> <span class="fu">colMeans</span>(geno_mat <span class="sc">==</span> <span class="dv">1</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="sample-level-missingness-heterozygosity-2" class="slide level2">
<h2>Sample-level missingness &amp; heterozygosity</h2>

<img data-src="lecture-05_files/figure-revealjs/unnamed-chunk-10-1.png" width="960" class="r-stretch"></section>
<section id="thresholding" class="slide level2">
<h2>Thresholding</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a></a>thr_samp_miss <span class="ot">&lt;-</span> <span class="fl">0.02</span></span>
<span id="cb5-2"><a></a>thr_var_miss <span class="ot">&lt;-</span> <span class="fl">0.02</span></span>
<span id="cb5-3"><a></a>thr_maf <span class="ot">&lt;-</span> <span class="fl">0.01</span></span>
<span id="cb5-4"><a></a>thr_hwe <span class="ot">&lt;-</span> <span class="fl">1e-6</span></span>
<span id="cb5-5"><a></a>thr_het_z <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb5-6"><a></a></span>
<span id="cb5-7"><a></a>keep_sample <span class="ot">&lt;-</span> <span class="fu">which</span>(samp_miss <span class="sc">&lt;=</span> thr_samp_miss <span class="sc">&amp;</span> <span class="fu">abs</span>(<span class="fu">scale</span>(het_rate)) <span class="sc">&lt;</span> thr_het_z)</span>
<span id="cb5-8"><a></a>keep_snp <span class="ot">&lt;-</span> <span class="fu">which</span>(miss_var <span class="sc">&lt;=</span> thr_var_miss <span class="sc">&amp;</span> maf <span class="sc">&gt;=</span> thr_maf <span class="sc">&amp;</span> (<span class="fu">is.na</span>(hwe_p) <span class="sc">|</span> hwe_p <span class="sc">&gt;</span> thr_hwe))</span>
<span id="cb5-9"><a></a></span>
<span id="cb5-10"><a></a>samp.keep.id <span class="ot">&lt;-</span> samp.id[keep_sample]</span>
<span id="cb5-11"><a></a>snp.keep.id <span class="ot">&lt;-</span> snp.id[keep_snp]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="snp-correlation-structure" class="slide level2">
<h2>SNP correlation structure</h2>
<ul>
<li>Inspect pairwise SNP correlations (LD) among variants that pass the basic filters before pruning.<br>
</li>
<li>When many SNPs remain, plot the first few hundred to keep the heatmap interpretable.</li>
</ul>
</section>
<section id="snp-correlation-structure-1" class="slide level2">
<h2>SNP correlation structure</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a></a>max_plot_snps <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb6-2"><a></a>ld_snp_ids <span class="ot">&lt;-</span> <span class="cf">if</span> (<span class="fu">length</span>(snp.keep.id) <span class="sc">&gt;</span> max_plot_snps) {</span>
<span id="cb6-3"><a></a>    snp.keep.id[<span class="fu">seq_len</span>(max_plot_snps)]</span>
<span id="cb6-4"><a></a>} <span class="cf">else</span> {</span>
<span id="cb6-5"><a></a>    snp.keep.id</span>
<span id="cb6-6"><a></a>}</span>
<span id="cb6-7"><a></a></span>
<span id="cb6-8"><a></a>ld_mat <span class="ot">&lt;-</span> <span class="fu">snpgdsLDMat</span>(</span>
<span id="cb6-9"><a></a>    g,</span>
<span id="cb6-10"><a></a>    <span class="at">sample.id =</span> samp.keep.id,</span>
<span id="cb6-11"><a></a>    <span class="at">snp.id =</span> ld_snp_ids,</span>
<span id="cb6-12"><a></a>    <span class="at">slide =</span> <span class="sc">-</span><span class="dv">1</span>,</span>
<span id="cb6-13"><a></a>    <span class="at">method =</span> <span class="st">"corr"</span></span>
<span id="cb6-14"><a></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Linkage Disequilibrium (LD) estimation on genotypes:
    # of samples: 167
    # of SNPs: 100
    using 1 thread
    method: correlation
LD matrix:    the sum of all selected genotypes (0,1,2) = 9659</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a></a>ld_vals <span class="ot">&lt;-</span> ld_mat<span class="sc">$</span>LD</span>
<span id="cb8-2"><a></a><span class="fu">dimnames</span>(ld_vals) <span class="ot">&lt;-</span> <span class="fu">list</span>(ld_snp_ids, ld_snp_ids)</span>
<span id="cb8-3"><a></a></span>
<span id="cb8-4"><a></a>ld_long <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">as.table</span>(ld_vals))</span>
<span id="cb8-5"><a></a><span class="fu">colnames</span>(ld_long) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"snp_i"</span>, <span class="st">"snp_j"</span>, <span class="st">"corr"</span>)</span>
<span id="cb8-6"><a></a></span>
<span id="cb8-7"><a></a><span class="fu">ggplot</span>(ld_long, <span class="fu">aes</span>(<span class="at">x =</span> snp_i, <span class="at">y =</span> snp_j, <span class="at">fill =</span> corr)) <span class="sc">+</span></span>
<span id="cb8-8"><a></a>    <span class="fu">geom_tile</span>() <span class="sc">+</span></span>
<span id="cb8-9"><a></a>    <span class="fu">coord_equal</span>() <span class="sc">+</span></span>
<span id="cb8-10"><a></a>    <span class="fu">scale_fill_gradient2</span>(</span>
<span id="cb8-11"><a></a>        <span class="at">low =</span> <span class="st">"#313695"</span>, <span class="at">mid =</span> <span class="st">"#ffffbf"</span>, <span class="at">high =</span> <span class="st">"#a50026"</span>,</span>
<span id="cb8-12"><a></a>        <span class="at">limits =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="dv">1</span>), <span class="at">midpoint =</span> <span class="dv">0</span>, <span class="at">oob =</span> scales<span class="sc">::</span>squish</span>
<span id="cb8-13"><a></a>    ) <span class="sc">+</span></span>
<span id="cb8-14"><a></a>    <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb8-15"><a></a>    <span class="fu">theme</span>(</span>
<span id="cb8-16"><a></a>        <span class="at">axis.text.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb8-17"><a></a>        <span class="at">axis.text.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb8-18"><a></a>        <span class="at">axis.ticks =</span> <span class="fu">element_blank</span>()</span>
<span id="cb8-19"><a></a>    ) <span class="sc">+</span></span>
<span id="cb8-20"><a></a>    <span class="fu">labs</span>(</span>
<span id="cb8-21"><a></a>        <span class="at">title =</span> <span class="fu">sprintf</span>(<span class="st">"Pairwise SNP correlation (first %d passing variants)"</span>, <span class="fu">length</span>(ld_snp_ids)),</span>
<span id="cb8-22"><a></a>        <span class="at">x =</span> <span class="st">"SNP1"</span>,</span>
<span id="cb8-23"><a></a>        <span class="at">y =</span> <span class="st">"SNP2"</span>,</span>
<span id="cb8-24"><a></a>        <span class="at">fill =</span> <span class="st">"r"</span></span>
<span id="cb8-25"><a></a>    )</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>

</div>
<img data-src="lecture-05_files/figure-revealjs/unnamed-chunk-12-1.png" width="960" class="r-stretch"></section>
<section id="ld-pruning" class="slide level2">
<h2>LD pruning</h2>
<ol type="1">
<li>Randomly select a starting position i (‘start.pos=“random”’), i=1 if ‘start.pos=“first”’, or i=last if ‘start.pos=“last”’; and let the current SNP set S={ i };</li>
<li>For each right position j from i+1 to n: if any LD between j and k is greater than ‘ld.threshold’, where k belongs to S, and both of j and k are in the sliding window, then skip j; otherwise, let S be S + { j };</li>
<li>For each left position j from i-1 to 1: if any LD between j and k is greater than ‘ld.threshold’, where k belongs to S, and both of j and k are in the sliding window, then skip j; otherwise, let S be S + { j };</li>
<li>Output S, the final selection of SNPs.</li>
</ol>
</section>
<section id="ld-pruning-1" class="slide level2">
<h2>LD pruning</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb9-2"><a></a>ld_prune <span class="ot">&lt;-</span> <span class="fu">snpgdsLDpruning</span>(g,</span>
<span id="cb9-3"><a></a>    <span class="at">sample.id =</span> samp.keep.id, </span>
<span id="cb9-4"><a></a>    <span class="at">snp.id =</span> snp.keep.id,</span>
<span id="cb9-5"><a></a>    <span class="at">method =</span> <span class="st">"r"</span>,</span>
<span id="cb9-6"><a></a>    <span class="at">ld.threshold =</span> <span class="fu">sqrt</span>(.<span class="dv">1</span>),  </span>
<span id="cb9-7"><a></a>    <span class="at">autosome.only =</span> <span class="cn">TRUE</span></span>
<span id="cb9-8"><a></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>SNP pruning based on LD:
Excluding 532 SNPs (non-autosomes or non-selection)
Excluding 0 SNP (monomorphic: TRUE, MAF: 0.005, missing rate: 0.05)
    # of samples: 167
    # of SNPs: 19,468
    using 1 thread
    sliding window: 500,000 basepairs, Inf SNPs
    |LD| threshold: 0.316228
    method: R
Chrom 1: |====================|====================|
    21.59%, 4,317 / 20,000 (Thu Oct  2 11:27:25 2025)
4,317 markers are selected in total.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a></a>snps_pruned <span class="ot">&lt;-</span> <span class="fu">unlist</span>(ld_prune, <span class="at">use.names =</span> <span class="cn">FALSE</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="relatedness" class="slide level2">
<h2>Relatedness</h2>
<ul>
<li>We need to account for relatedness to avoid confounding in PCA and association testing</li>
<li>KING (Kinship-based INference for GWAS) is method for estimating pairwise kinship robust to population structure (<span class="citation" data-cites="manichaikulRobustRelationshipInference2010">(<a href="#/wrapup-what-to-record" role="doc-biblioref" onclick="">Manichaikul <em>et al.</em>, 2010</a>)</span></li>
</ul>
</section>
<section id="relatedness-1" class="slide level2">
<h2>Relatedness</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-14"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-14" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-14-1" class="code-annotation-target"><a href="#annotated-cell-14-1"></a>king <span class="ot">&lt;-</span> <span class="fu">snpgdsIBDKING</span>(g,</span>
<span id="annotated-cell-14-2"><a href="#annotated-cell-14-2"></a>                      <span class="at">sample.id =</span> samp.keep.id,</span>
<span id="annotated-cell-14-3"><a href="#annotated-cell-14-3"></a>                      <span class="at">snp.id =</span> snps_pruned,</span>
<span id="annotated-cell-14-4"><a href="#annotated-cell-14-4"></a>                      <span class="at">type =</span> <span class="st">"KING-robust"</span>)</span>
<span id="annotated-cell-14-5"><a href="#annotated-cell-14-5"></a></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-14" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-14-6" class="code-annotation-target"><a href="#annotated-cell-14-6"></a>KINGmat <span class="ot">&lt;-</span> GENESIS<span class="sc">::</span><span class="fu">kingToMatrix</span>(king)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-14" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-14" data-code-lines="1,2,3,4" data-code-annotation="1">Run KING on our dataset with the filtered samples and SNPs</span>
</dd>
<dt data-target-cell="annotated-cell-14" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-14" data-code-lines="6" data-code-annotation="2">Convert into a kinship matrix we can use for downstream tasks</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>IBD analysis (KING method of moment) on genotypes:
Excluding 15,683 SNPs (non-autosomes or non-selection)
Excluding 0 SNP (monomorphic: TRUE, MAF: NaN, missing rate: NaN)
    # of samples: 167
    # of SNPs: 4,317
    using 1 thread
No family is specified, and all individuals are treated as singletons.
Relationship inference in the presence of population stratification.
KING IBD:    the sum of all selected genotypes (0,1,2) = 366279
CPU capabilities:
Thu Oct  2 11:27:25 2025    (internal increment: 65536)

[..................................................]  0%, ETC: ---        
[==================================================] 100%, completed, 0s
Thu Oct  2 11:27:25 2025    Done.</code></pre>
</div>
</div>
</section>
<section id="estimated-family-clusters" class="slide level2">
<h2>Estimated family clusters</h2>
<ul>
<li>Build a kinship graph using a threshold (default: 3rd-degree or closer).</li>
<li>Connected components approximate families/households; singletons have no relatives above the threshold.</li>
</ul>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a></a>kinship_threshold <span class="ot">&lt;-</span> <span class="fl">0.0625</span></span>
<span id="cb13-2"><a></a></span>
<span id="cb13-3"><a></a>king_adj <span class="ot">&lt;-</span> KINGmat</span>
<span id="cb13-4"><a></a><span class="fu">diag</span>(king_adj) <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb13-5"><a></a>king_adj[king_adj <span class="sc">&lt;</span> kinship_threshold] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb13-6"><a></a></span>
<span id="cb13-7"><a></a>family_graph <span class="ot">&lt;-</span> igraph<span class="sc">::</span><span class="fu">graph_from_adjacency_matrix</span>(king_adj, <span class="at">mode =</span> <span class="st">"undirected"</span>, <span class="at">weighted =</span> <span class="cn">TRUE</span>, <span class="at">diag =</span> <span class="cn">FALSE</span>)</span>
<span id="cb13-8"><a></a>comp <span class="ot">&lt;-</span> igraph<span class="sc">::</span><span class="fu">components</span>(family_graph)</span>
<span id="cb13-9"><a></a></span>
<span id="cb13-10"><a></a>family_map <span class="ot">&lt;-</span> tibble<span class="sc">::</span><span class="fu">tibble</span>(</span>
<span id="cb13-11"><a></a>    <span class="at">scanID =</span> <span class="fu">names</span>(comp<span class="sc">$</span>membership),</span>
<span id="cb13-12"><a></a>    <span class="at">family_id =</span> <span class="fu">paste0</span>(<span class="st">"fam_"</span>, <span class="fu">sprintf</span>(<span class="st">"%03d"</span>, comp<span class="sc">$</span>membership))</span>
<span id="cb13-13"><a></a>)</span>
<span id="cb13-14"><a></a></span>
<span id="cb13-15"><a></a>family_summary <span class="ot">&lt;-</span> family_map <span class="sc">|&gt;</span></span>
<span id="cb13-16"><a></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>(family_id) <span class="sc">|&gt;</span></span>
<span id="cb13-17"><a></a>    dplyr<span class="sc">::</span><span class="fu">summarise</span>(</span>
<span id="cb13-18"><a></a>        <span class="at">n_members =</span> dplyr<span class="sc">::</span><span class="fu">n</span>(),</span>
<span id="cb13-19"><a></a>        <span class="at">members =</span> <span class="fu">paste</span>(scanID, <span class="at">collapse =</span> <span class="st">", "</span>),</span>
<span id="cb13-20"><a></a>        <span class="at">.groups =</span> <span class="st">"drop"</span></span>
<span id="cb13-21"><a></a>    ) <span class="sc">|&gt;</span></span>
<span id="cb13-22"><a></a>    dplyr<span class="sc">::</span><span class="fu">arrange</span>(dplyr<span class="sc">::</span><span class="fu">desc</span>(n_members), family_id)</span>
<span id="cb13-23"><a></a></span>
<span id="cb13-24"><a></a><span class="fu">head</span>(family_summary, <span class="dv">5</span>) <span class="sc">|&gt;</span></span>
<span id="cb13-25"><a></a>    knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="at">col.names =</span> <span class="fu">c</span>(<span class="st">"Family"</span>, <span class="st">"n"</span>, <span class="st">"Sample IDs"</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="estimated-family-clusters-1" class="slide level2">
<h2>Estimated family clusters</h2>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top">
<colgroup>
<col style="width: 8%">
<col style="width: 3%">
<col style="width: 87%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Family</th>
<th style="text-align: right;">n</th>
<th style="text-align: left;">Sample IDs</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">fam_002</td>
<td style="text-align: right;">9</td>
<td style="text-align: left;">NA19663, NA19664, NA19665, NA19686, NA19684, NA19661, NA19660, NA19662, NA19685</td>
</tr>
<tr class="even">
<td style="text-align: left;">fam_017</td>
<td style="text-align: right;">6</td>
<td style="text-align: left;">NA19675, NA19676, NA19677, NA19678, NA19680, NA19679</td>
</tr>
<tr class="odd">
<td style="text-align: left;">fam_024</td>
<td style="text-align: right;">6</td>
<td style="text-align: left;">NA20346, NA20359, NA20363, NA20360, NA20347, NA20364</td>
</tr>
<tr class="even">
<td style="text-align: left;">fam_022</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">NA20340, NA20349, NA20344, NA20350, NA20345</td>
</tr>
<tr class="odd">
<td style="text-align: left;">fam_033</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">NA19713, NA19982, NA19983, NA19714, NA19985</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="ancestry-pcs" class="slide level2">
<h2>Ancestry PCs</h2>
<ul>
<li>So we have our kinship matrix</li>
<li>We now want ancestry PCs for covariates</li>
<li><strong>PC‑AiR</strong> (Conomos et al., 2015) computes PCs on unrelateds, then projects onto relateds</li>
<li>Relateds are identified using the KING kinship matrix</li>
</ul>
</section>
<section id="ancestry-pcs-1" class="slide level2">
<h2>Ancestry PCs</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-16"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-16" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-16-1" class="code-annotation-target"><a href="#annotated-cell-16-1"></a><span class="fu">snpgdsClose</span>(g)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-16" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-16-2" class="code-annotation-target"><a href="#annotated-cell-16-2"></a>geno_reader <span class="ot">&lt;-</span> GWASTools<span class="sc">::</span><span class="fu">GdsGenotypeReader</span>(gdsfile)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-16" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-16-3" class="code-annotation-target"><a href="#annotated-cell-16-3"></a>geno_data <span class="ot">&lt;-</span> <span class="fu">GenotypeData</span>(geno_reader)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-16" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-16-4" class="code-annotation-target"><a href="#annotated-cell-16-4"></a>pcair_out <span class="ot">&lt;-</span> <span class="fu">pcair</span>(geno_data,</span>
<span id="annotated-cell-16-5"><a href="#annotated-cell-16-5"></a>                   <span class="at">kinobj =</span> KINGmat,</span>
<span id="annotated-cell-16-6"><a href="#annotated-cell-16-6"></a>                   <span class="at">divobj =</span> KINGmat,</span>
<span id="annotated-cell-16-7"><a href="#annotated-cell-16-7"></a>                   <span class="at">snp.include =</span> snps_pruned,</span>
<span id="annotated-cell-16-8"><a href="#annotated-cell-16-8"></a>                   <span class="at">sample.include =</span> samp.keep.id)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-16" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-16" data-code-lines="1" data-code-annotation="1">We need to close the SNPRelate connection before opening with GWASTools</span>
</dd>
<dt data-target-cell="annotated-cell-16" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-16" data-code-lines="2" data-code-annotation="2">Open the GDS file with GWASTools</span>
</dd>
<dt data-target-cell="annotated-cell-16" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-16" data-code-lines="3" data-code-annotation="3">Create a GenotypeData object for analysis</span>
</dd>
<dt data-target-cell="annotated-cell-16" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-16" data-code-lines="4,5,6,7,8" data-code-annotation="4">Run PC-AiR using the KING kinship matrix to identify unrelateds and compute PCs</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Principal Component Analysis (PCA) on genotypes:
Excluding 15,683 SNPs (non-autosomes or non-selection)
Excluding 0 SNP (monomorphic: TRUE, MAF: NaN, missing rate: NaN)
    # of samples: 93
    # of SNPs: 4,317
    using 1 thread
    # of principal components: 32
PCA:    the sum of all selected genotypes (0,1,2) = 204226
CPU capabilities:
Thu Oct  2 11:27:25 2025    (internal increment: 5284)

[..................................................]  0%, ETC: ---        
[==================================================] 100%, completed, 0s
Thu Oct  2 11:27:25 2025    Begin (eigenvalues and eigenvectors)
Thu Oct  2 11:27:25 2025    Done.
SNP Loading:
    # of samples: 93
    # of SNPs: 4,317
    using 1 thread
    using the top 32 eigenvectors
SNP Loading:    the sum of all selected genotypes (0,1,2) = 204226
Thu Oct  2 11:27:25 2025    (internal increment: 42280)

[..................................................]  0%, ETC: ---        
[==================================================] 100%, completed, 0s
Thu Oct  2 11:27:25 2025    Done.
Sample Loading:
    # of samples: 74
    # of SNPs: 4,317
    using 1 thread
    using the top 32 eigenvectors
Sample Loading:    the sum of all selected genotypes (0,1,2) = 162053
Thu Oct  2 11:27:25 2025    (internal increment: 53136)

[..................................................]  0%, ETC: ---        
[==================================================] 100%, completed, 0s
Thu Oct  2 11:27:25 2025    Done.</code></pre>
</div>
</div>
</section>
<section id="ancestry-pcs-2" class="slide level2">
<h2>Ancestry PCs</h2>

<img data-src="lecture-05_files/figure-revealjs/unnamed-chunk-13-1.png" width="960" class="r-stretch"></section>
<section id="ancestry-pcs-3" class="slide level2">
<h2>Ancestry PCs</h2>
</section>
<section id="step-6-phenotype-covariates-code" class="slide level2">
<h2>Step 6 — Phenotype &amp; covariates (code)</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a></a><span class="co"># Simulate simple covariates</span></span>
<span id="cb15-2"><a></a><span class="fu">set.seed</span>(<span class="dv">99</span>)</span>
<span id="cb15-3"><a></a>N <span class="ot">&lt;-</span> <span class="fu">nrow</span>(pcs)</span>
<span id="cb15-4"><a></a>sex <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">c</span>(<span class="st">"M"</span>, <span class="st">"F"</span>), N, <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb15-5"><a></a>age <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">rnorm</span>(N, <span class="dv">50</span>, <span class="dv">10</span>))</span>
<span id="cb15-6"><a></a></span>
<span id="cb15-7"><a></a><span class="co"># Choose one moderately frequent SNP and simulate a quantitative trait with a planted effect</span></span>
<span id="cb15-8"><a></a>maf_full <span class="ot">&lt;-</span> maf</span>
<span id="cb15-9"><a></a>ix <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="sc">!</span><span class="fu">is.na</span>(maf_full) <span class="sc">&amp;</span> maf_full <span class="sc">&gt;</span> <span class="fl">0.2</span> <span class="sc">&amp;</span> maf_full <span class="sc">&lt;</span> <span class="fl">0.3</span>)[<span class="dv">1</span>]</span>
<span id="cb15-10"><a></a></span>
<span id="cb15-11"><a></a>G1 <span class="ot">&lt;-</span> GWASTools<span class="sc">::</span><span class="fu">getGenotypeSelection</span>(geno, <span class="at">snpID =</span> snp.id[ix], <span class="at">scanID =</span> samp.keep.id)</span>
<span id="cb15-12"><a></a>G1 <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(G1) <span class="co"># 0,1,2 with NAs</span></span>
<span id="cb15-13"><a></a>G1[<span class="fu">is.na</span>(G1)] <span class="ot">&lt;-</span> <span class="fu">mean</span>(G1, <span class="at">na.rm =</span> <span class="cn">TRUE</span>) <span class="co"># quick mean-impute for demo</span></span>
<span id="cb15-14"><a></a></span>
<span id="cb15-15"><a></a>y <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fl">0.30</span> <span class="sc">*</span> <span class="fu">scale</span>(G1) <span class="sc">+</span> <span class="fl">0.02</span> <span class="sc">*</span> <span class="fu">scale</span>(age) <span class="sc">+</span> <span class="fl">0.05</span> <span class="sc">*</span> (sex <span class="sc">==</span> <span class="st">"M"</span>) <span class="sc">+</span> <span class="fu">rnorm</span>(N))</span>
<span id="cb15-16"><a></a></span>
<span id="cb15-17"><a></a><span class="co"># Assemble scan annotation with covariates and PCs</span></span>
<span id="cb15-18"><a></a>scanDF <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">scanID =</span> samp.keep.id, <span class="at">y =</span> y, <span class="at">age =</span> age, <span class="at">sex =</span> sex, pcs[, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>])</span>
<span id="cb15-19"><a></a><span class="fu">colnames</span>(scanDF)[<span class="dv">6</span><span class="sc">:</span><span class="dv">9</span>] <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"PC"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>)</span>
<span id="cb15-20"><a></a>scanAnnot2 <span class="ot">&lt;-</span> GWASTools<span class="sc">::</span><span class="fu">ScanAnnotationDataFrame</span>(scanDF)</span>
<span id="cb15-21"><a></a><span class="co"># Update 'geno' to carry covariates</span></span>
<span id="cb15-22"><a></a>GWASTools<span class="sc">::</span><span class="fu">close</span>(geno)</span>
<span id="cb15-23"><a></a>geno <span class="ot">&lt;-</span> GWASTools<span class="sc">::</span><span class="fu">GenotypeData</span>(GWASTools<span class="sc">::</span><span class="fu">GdsGenotypeReader</span>(gdsfile), <span class="at">scanAnnot =</span> scanAnnot2)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="step-6-phenotype-covariates-explain" class="slide level2">
<h2>Step 6 — Phenotype &amp; covariates (explain)</h2>
<p><strong>Why simulate?</strong><br>
HapMap doesn’t ship with rich traits. We plant a small genotype effect so that the mixed‑model GWAS produces a recognizable signal.</p>
<p><strong>Model intent</strong> - Covariates: <code>age</code>, <code>sex</code>, <code>PC1–PC4</code> (ancestry).<br>
- Genetic effect: ~0.30 SD per minor allele on a standardized SNP.</p>
<p><strong>Deliverable</strong> - <code>geno</code> now contains sample annotations with (y, X), and PCs.</p>
</section>
<section id="step-6-phenotype-covariates-output" class="slide level2">
<h2>Step 6 — Phenotype &amp; covariates (output)</h2>
</section>
<section id="step-7-null-model-lmm-code" class="slide level2">
<h2>Step 7 — Null model (LMM) (code)</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a></a>nullmod <span class="ot">&lt;-</span> GENESIS<span class="sc">::</span><span class="fu">fitNullModel</span>(scanAnnot2,</span>
<span id="cb16-2"><a></a>    <span class="at">outcome =</span> <span class="st">"y"</span>,</span>
<span id="cb16-3"><a></a>    <span class="at">covars =</span> <span class="fu">c</span>(<span class="st">"age"</span>, <span class="st">"sex"</span>, <span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span>, <span class="st">"PC4"</span>),</span>
<span id="cb16-4"><a></a>    <span class="at">cov.mat =</span> K, <span class="at">family =</span> <span class="st">"gaussian"</span></span>
<span id="cb16-5"><a></a>)</span>
<span id="cb16-6"><a></a></span>
<span id="cb16-7"><a></a><span class="co"># Genotype iterator (block-wise scan of all SNPs kept by QC)</span></span>
<span id="cb16-8"><a></a>it <span class="ot">&lt;-</span> GWASTools<span class="sc">::</span><span class="fu">GenotypeBlockIterator</span>(geno, <span class="at">snpInclude =</span> snp.keep.id, <span class="at">snpBlock =</span> <span class="dv">5000</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="step-7-null-model-explain" class="slide level2">
<h2>Step 7 — Null model (explain)</h2>
<p><strong>LMM</strong><br>
[ y = X+ G + u + ,u (0,_g^2 K),(0,_e^2 I). ]</p>
<ul>
<li>We fit variance components under (H_0:) (the <strong>null model</strong>) once.<br>
</li>
<li>Association uses a <strong>Score test</strong>, reusing the null for each SNP.<br>
</li>
<li>PCs adjust for residual structure; (K) captures realized relatedness.</li>
</ul>
<p><strong>Deliverables</strong> - <code>nullmod</code>: object with (_g^2,_e^2), fixed effects, and residuals.<br>
- <code>it</code>: a block iterator to stream SNPs without loading them all at once.</p>
</section>
<section id="step-7-null-model-output" class="slide level2">
<h2>Step 7 — Null model (output)</h2>
</section>
<section id="step-8-association-testing-code" class="slide level2">
<h2>Step 8 — Association testing (code)</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a></a>assoc <span class="ot">&lt;-</span> GENESIS<span class="sc">::</span><span class="fu">assocTestSingle</span>(it, <span class="at">null.model =</span> nullmod, <span class="at">test =</span> <span class="st">"Score"</span>)</span>
<span id="cb17-2"><a></a><span class="co"># Prepare results for plotting</span></span>
<span id="cb17-3"><a></a>res <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb17-4"><a></a>    <span class="at">SNP =</span> assoc<span class="sc">$</span>snpID,</span>
<span id="cb17-5"><a></a>    <span class="at">CHR =</span> <span class="fu">as.integer</span>(assoc<span class="sc">$</span>chr),</span>
<span id="cb17-6"><a></a>    <span class="at">BP =</span> assoc<span class="sc">$</span>pos,</span>
<span id="cb17-7"><a></a>    <span class="at">P =</span> assoc<span class="sc">$</span>Score.pval,</span>
<span id="cb17-8"><a></a>    <span class="at">BETA =</span> assoc<span class="sc">$</span>Est,</span>
<span id="cb17-9"><a></a>    <span class="at">SE =</span> assoc<span class="sc">$</span>Est.SE</span>
<span id="cb17-10"><a></a>)</span>
<span id="cb17-11"><a></a><span class="co"># Genomic control</span></span>
<span id="cb17-12"><a></a>lambda_gc <span class="ot">&lt;-</span> <span class="fu">median</span>(stats<span class="sc">::</span><span class="fu">qchisq</span>(<span class="dv">1</span> <span class="sc">-</span> res<span class="sc">$</span>P, <span class="at">df =</span> <span class="dv">1</span>), <span class="at">na.rm =</span> <span class="cn">TRUE</span>) <span class="sc">/</span> <span class="fl">0.456</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="step-8-association-testing-explain" class="slide level2">
<h2>Step 8 — Association testing (explain)</h2>
<ul>
<li><strong>Score test</strong>: efficient per‑SNP test using derivatives at the null; numerically stable at scale.<br>
</li>
<li>With our simulated effect, we expect one clear peak near the planted SNP, modest elsewhere.<br>
</li>
<li><strong>Calibration</strong>: ( _{GC} ) near 1 indicates good control; large values suggest stratification, relatedness, or batch effects.</li>
</ul>
<p><strong>Deliverables</strong> - <code>res</code>: per‑SNP estimates and p‑values.<br>
- <code>lambda_gc</code>: quick inflation diagnostic.</p>
</section>
<section id="step-8-association-testing-output" class="slide level2">
<h2>Step 8 — Association testing (output)</h2>
</section>
<section id="step-9-figures-manhattan-qq-code" class="slide level2">
<h2>Step 9 — Figures: Manhattan &amp; QQ (code)</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a></a><span class="fu">library</span>(qqman)</span>
<span id="cb18-2"><a></a><span class="co"># Clean CHR for qqman (drop non-autosomes for the plot)</span></span>
<span id="cb18-3"><a></a>plotdat <span class="ot">&lt;-</span> <span class="fu">subset</span>(res, CHR <span class="sc">%in%</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">22</span> <span class="sc">&amp;</span> <span class="fu">is.finite</span>(P))</span>
<span id="cb18-4"><a></a><span class="fu">manhattan</span>(plotdat,</span>
<span id="cb18-5"><a></a>    <span class="at">chr =</span> <span class="st">"CHR"</span>, <span class="at">bp =</span> <span class="st">"BP"</span>, <span class="at">p =</span> <span class="st">"P"</span>, <span class="at">snp =</span> <span class="st">"SNP"</span>,</span>
<span id="cb18-6"><a></a>    <span class="at">suggestiveline =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">1e-5</span>), <span class="at">genomewideline =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">5e-8</span>),</span>
<span id="cb18-7"><a></a>    <span class="at">main =</span> <span class="st">"GWAS (LMM) — Manhattan"</span></span>
<span id="cb18-8"><a></a>)</span>
<span id="cb18-9"><a></a><span class="fu">qq</span>(plotdat<span class="sc">$</span>P, <span class="at">main =</span> <span class="st">"QQ plot"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="step-9-figures-manhattan-qq-explain" class="slide level2">
<h2>Step 9 — Figures: Manhattan &amp; QQ (explain)</h2>
<p><strong>How to read the plots</strong> - <strong>Manhattan</strong>: look for peaks spanning multiple SNPs in LD; single‑SNP spikes can be artifacts.<br>
- <strong>QQ</strong>: early alignment with the diagonal + tail deviation at top hits is typical of a polygenic trait.<br>
- Show ( _{GC} ) on the figure or caption.</p>
<p><strong>Caveats</strong> - Our dataset is <strong>small</strong> and <strong>simulated</strong>; the figures illustrate <strong>method behavior</strong>, not discovery.</p>
</section>
<section id="step-9-figures-manhattan-qq-output" class="slide level2">
<h2>Step 9 — Figures: Manhattan &amp; QQ (output)</h2>
</section>
<section id="optional-step-10-binary-trait-spa-code" class="slide level2">
<h2>(Optional) Step 10 — Binary trait &amp; SPA (code)</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a></a><span class="co"># Simulate a low-prevalence binary trait using the same planted SNP</span></span>
<span id="cb19-2"><a></a>logit_p <span class="ot">&lt;-</span> <span class="sc">-</span><span class="fl">2.3</span> <span class="sc">+</span> <span class="fl">0.8</span> <span class="sc">*</span> <span class="fu">scale</span>(G1) <span class="sc">+</span> <span class="fl">0.1</span> <span class="sc">*</span> <span class="fu">scale</span>(age) <span class="sc">+</span> <span class="fl">0.1</span> <span class="sc">*</span> (sex <span class="sc">==</span> <span class="st">"M"</span>) <span class="sc">+</span> <span class="fu">as.numeric</span>(pcs[, <span class="dv">1</span>]) <span class="sc">*</span> <span class="fl">0.0</span></span>
<span id="cb19-3"><a></a>p_case <span class="ot">&lt;-</span> <span class="fu">plogis</span>(logit_p)</span>
<span id="cb19-4"><a></a>case <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="fu">length</span>(p_case), <span class="dv">1</span>, p_case)</span>
<span id="cb19-5"><a></a></span>
<span id="cb19-6"><a></a>scanDF2 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">scanID =</span> samp.keep.id, <span class="at">case =</span> case, <span class="at">age =</span> age, <span class="at">sex =</span> sex, pcs[, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>])</span>
<span id="cb19-7"><a></a><span class="fu">colnames</span>(scanDF2)[<span class="dv">4</span><span class="sc">:</span><span class="dv">7</span>] <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"PC"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>)</span>
<span id="cb19-8"><a></a>scanAnnot3 <span class="ot">&lt;-</span> GWASTools<span class="sc">::</span><span class="fu">ScanAnnotationDataFrame</span>(scanDF2)</span>
<span id="cb19-9"><a></a></span>
<span id="cb19-10"><a></a>null_cc <span class="ot">&lt;-</span> GENESIS<span class="sc">::</span><span class="fu">fitNullModel</span>(scanAnnot3,</span>
<span id="cb19-11"><a></a>    <span class="at">outcome =</span> <span class="st">"case"</span>,</span>
<span id="cb19-12"><a></a>    <span class="at">covars =</span> <span class="fu">c</span>(<span class="st">"age"</span>, <span class="st">"sex"</span>, <span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span>, <span class="st">"PC4"</span>),</span>
<span id="cb19-13"><a></a>    <span class="at">cov.mat =</span> K, <span class="at">family =</span> <span class="st">"binomial"</span></span>
<span id="cb19-14"><a></a>)</span>
<span id="cb19-15"><a></a></span>
<span id="cb19-16"><a></a>it2 <span class="ot">&lt;-</span> GWASTools<span class="sc">::</span><span class="fu">GenotypeBlockIterator</span>(geno, <span class="at">snpInclude =</span> snp.keep.id, <span class="at">snpBlock =</span> <span class="dv">5000</span>)</span>
<span id="cb19-17"><a></a>assoc_cc <span class="ot">&lt;-</span> GENESIS<span class="sc">::</span><span class="fu">assocTestSingle</span>(it2, <span class="at">null.model =</span> null_cc, <span class="at">test =</span> <span class="st">"Score.SPA"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="optional-step-10-binary-trait-spa-explain" class="slide level2">
<h2>(Optional) Step 10 — Binary trait &amp; SPA (explain)</h2>
<ul>
<li>Severe <strong>case:control imbalance</strong> and low MAC inflate type‑I error for naive tests.<br>
</li>
<li><strong>SPA (saddlepoint approximation)</strong> yields accurate p‑values in the tails by correcting score distributions.<br>
</li>
<li><code>assocTestSingle(..., test="Score.SPA")</code> invokes SPA within GENESIS.</li>
</ul>
</section>
<section id="wrapup-what-to-record" class="slide level2 smaller scrollable">
<h2>Wrap‑up &amp; what to record</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a></a>GWASTools<span class="sc">::</span><span class="fu">close</span>(geno_reader)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<ul>
<li><strong>Parameters</strong>: thresholds (QC), LD‑prune settings, PCs used, null model family, random effect (K), LOCO (used? no).<br>
</li>
<li><strong>Software</strong>: package versions; genome build; random seeds.<br>
</li>
<li><strong>Deliverables</strong>: annotated code, QQ/Manhattan, top hits table, (_{GC}).</li>
</ul>
<p><strong>Homework (separate)</strong><br>
Install PLINK and reproduce QC + PCs on a small chr22 dataset; compare with R results.</p>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-manichaikulRobustRelationshipInference2010" class="csl-entry" role="listitem">
Manichaikul,A. <em>et al.</em> (2010) <a href="https://doi.org/10.1093/bioinformatics/btq559">Robust relationship inference in genome-wide association studies</a>. <em>Bioinformatics</em>, <strong>26</strong>, 2867–2873.
</div>
</div>
</section></section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">

</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
          const outerScaffold = trigger.parentElement.cloneNode(true);
          const codeEl = outerScaffold.querySelector('code');
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp('/' + window.location.host + '/');
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    

</body></html>